<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Glossary &mdash; RNAlysis 2.1.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="History" href="history.html" />
    <link rel="prev" title="User guide" href="user_guide.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html">
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Glossary</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Glossary</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/glossary.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="section" id="glossary">
<span id="id1"></span><h1>Glossary<a class="headerlink" href="#glossary" title="Permalink to this headline"></a></h1>
<dl class="glossary simple">
<dt id="term-Attribute-Reference-Table">Attribute Reference Table<a class="headerlink" href="#term-Attribute-Reference-Table" title="Permalink to this term"></a></dt><dd><p>A <cite>csv</cite> table that contains user-defined information (‘attributes’, such as ‘genes expressed in intestine’, ‘epigenetic genes’ or ‘genes that have paralogs’)
abut genes/genomic features.
Every row in the file is a gene/genomic feature, and every column is an attribute.
You can define such table to be your default Attribute Reference Table using the function <cite>general.set_attr_ref_table_path()</cite>.
The attributes in the table can be used both to filter your data with functions like <cite>Filter.filter_by_attribute’,
and to perform enrichment analysis with various functions from the `enrichment</cite> module.</p>
</dd>
<dt id="term-Biotype-Reference-Table">Biotype Reference Table<a class="headerlink" href="#term-Biotype-Reference-Table" title="Permalink to this term"></a></dt><dd><p>A <cite>csv</cite> table that contains information about the biotype (protein coding, pseudogene, lncRNA, etc…) of genomic features.
You can define such table to be your default Biotype Reference Table using the function <cite>general.set_biotype_ref_table_path()</cite>.
Various functions in RNAlysis, such <cite>Filter.biotypes()</cite> and <cite>Filter.filter_biotype()</cite> will use the information in the Biotype Reference Table
to filter data based on biotype, or display information about the biotype of the genomic features in your Filter objects.</p>
</dd>
<dt id="term-CountFilter">CountFilter<a class="headerlink" href="#term-CountFilter" title="Permalink to this term"></a></dt><dd><p>A <a class="reference internal" href="#term-Filter-object"><span class="xref std std-term">Filter object</span></a> that supports count matrices, where each row represents a gene or genomic feature,
and every column represents a condition or replicate.</p>
</dd>
<dt id="term-DESeqFilter">DESeqFilter<a class="headerlink" href="#term-DESeqFilter" title="Permalink to this term"></a></dt><dd><p>A <a class="reference internal" href="#term-Filter-object"><span class="xref std std-term">Filter object</span></a> that supports differential expression data, formatted like the output files of DESeq2.
See the user guide for more details.</p>
</dd>
<dt id="term-FeatureSet">FeatureSet<a class="headerlink" href="#term-FeatureSet" title="Permalink to this term"></a></dt><dd><p>A container for a set of gene/feature IDs. A FeatureSet can optionally be named.
Using the Enrichmemt module, you can run various enrichment analyses on FeatureSet objects.</p>
</dd>
<dt id="term-Filter">Filter<a class="headerlink" href="#term-Filter" title="Permalink to this term"></a></dt><dd><p>The most basic type of <a class="reference internal" href="#term-Filter-object"><span class="xref std std-term">Filter object</span></a>.
Can support any form of tabular data, and contains only the basic filtering functions.</p>
</dd>
<dt id="term-Filter-object">Filter object<a class="headerlink" href="#term-Filter-object" title="Permalink to this term"></a></dt><dt id="term-Filter-objects">Filter objects<a class="headerlink" href="#term-Filter-objects" title="Permalink to this term"></a></dt><dd><p>An object that stores your tabular data and its filename.
You can apply filtering/normalizing functions to your data using the Filter object.
Each type of Filter object supports different data structures and offers different utilities.
For more details, see <a class="reference internal" href="#term-Filter"><span class="xref std std-term">Filter</span></a>, :term:’CountFilter’, :term:<a href="#id2"><span class="problematic" id="id3">`</span></a>DESeqFilter’ and :term:<a href="#id4"><span class="problematic" id="id5">`</span></a>FoldChangeFilter’.</p>
</dd>
<dt id="term-FoldChangeFilter">FoldChangeFilter<a class="headerlink" href="#term-FoldChangeFilter" title="Permalink to this term"></a></dt><dd><p>A <a class="reference internal" href="#term-Filter-object"><span class="xref std std-term">Filter object</span></a> that supports fold change values, where each row represents a gene or genomic feature,
and the <strong>only</strong> other column contains fold-change values between two conditions.
Can be loaded from a <cite>csv</cite> file, or generated from a <a class="reference internal" href="#term-CountFilter"><span class="xref std std-term">CountFilter</span></a>.</p>
</dd>
<dt id="term-Pipeline">Pipeline<a class="headerlink" href="#term-Pipeline" title="Permalink to this term"></a></dt><dd><p>An object that stores a series of functions and their corresponding arguments.
You can then use the Pipeline object to apply this series of functions, with the same order and same arguments, to any number of <a class="reference internal" href="#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a>.
Pipelines can contain filtering functions, normalization functions, splitting functions and visualization functions.</p>
</dd>
<dt id="term-RankedSet">RankedSet<a class="headerlink" href="#term-RankedSet" title="Permalink to this term"></a></dt><dd><p>A subtype of <a class="reference internal" href="#term-FeatureSet"><span class="xref std std-term">FeatureSet</span></a> that, instead of storing an unsorted set of gene/feature IDs,
also stores their order. You can run enrichment analyses on RankedSet objects as you would on a FeatureSet object.
However, you can also perform single-list enrichment (enrichment analysis without a background set, based on the ranking/order of the gene IDs)
on RankedSet objects.</p>
</dd>
</dl>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="user_guide.html" class="btn btn-neutral float-left" title="User guide" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="history.html" class="btn btn-neutral float-right" title="History" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>