<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>History &mdash; RNAlysis 3.5.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.fastq" href="rnalysis.fastq.html" />
    <link rel="prev" title="Glossary - programmatic interface" href="glossary.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html">
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">3.5.2 (2023-02-23)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#changed">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fixed">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id2">3.5.1 (2023-02-22)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id4">3.5.0 (2023-02-08)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#added">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id6">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id7">3.4.2 (2023-02-01)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id8">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id9">3.4.0 (2023-02-01)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id10">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id11">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#removed">Removed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#new-contributors">New Contributors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id15">3.3.0 (2022-12-02)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id16">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id19">3.2.2 (2022-11-25)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id20">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id21">3.2.1 (2022-11-25)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id22">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id24">3.2.0 (2022-11-23)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id25">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id26">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id28">3.1.0 (2022-10-16)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id29">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id30">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id32">3.0.1 (2022-10-12)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id33">3.0.0 (2022-10-10)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id34">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id35">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id37">2.1.1 (2022-07-05)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id38">Changed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id39">2.1.0 (2022-04-16)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id40">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id41">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id42">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id43">2.0.1 (2022-04-02)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id44">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id45">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id46">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id47">2.0.0 (2021-12-05)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id48">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id49">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id50">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id51">1.3.5 (2020-05-27)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id52">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id53">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id54">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id55">1.3.4 (2020-04-07)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id56">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id57">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id58">1.3.3 (2020-03-28)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id59">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id60">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id61">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id62">Removed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id63">1.3.2 (2019-12-11)</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>History</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/history.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="history">
<h1>History<a class="headerlink" href="#history" title="Permalink to this heading"></a></h1>
<section id="id1">
<h2>3.5.2 (2023-02-23)<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h2>
<p>This version includes bug fixes for a few quality-of-life issues which were introduced in version RNAlysis 3.5.0.</p>
<section id="changed">
<h3>Changed<a class="headerlink" href="#changed" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>It is now possible to change the parallel backend of performance-intensive functions such as clustering an enrichment analysis in non-standalone versions of RNAlysis.</p></li>
<li><p>Expanded the Frequently Asked Questions page.</p></li>
<li><p>Added Perl as a dependency for Windows users on the How to Install page.</p></li>
</ul>
</section>
<section id="fixed">
<h3>Fixed<a class="headerlink" href="#fixed" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed bug where occasionally newly-created tabs would open twice instead of once.</p></li>
<li><p>Fixed bug where the ‘Add Function’ button of the Pipeline window would appear in the wrong location.</p></li>
<li><p>Fixed bug where RNAlysis sessions saved after version 3.5.0 which contain gene sets would raise an error when loaded.</p></li>
<li><p>Fixed typos in the RNAlysis tutorial.</p></li>
</ul>
</section>
</section>
<section id="id2">
<h2>3.5.1 (2023-02-22)<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h2>
<p>This version introduces minor bug fixes.</p>
<section id="id3">
<h3>Fixed<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed bug where the <em>RNAlysis</em> stand-alone app would sometimes fail to run CutAdapt (thanks to Matthias Wilm).</p></li>
<li><p>Fixed bug where the User Guide action in the graphical interface would point to the Tutorial, and vice versa.</p></li>
<li><p>The X and Y axis labels on volcano plots should now format the ‘log’ in the label correctly.</p></li>
</ul>
</section>
</section>
<section id="id4">
<h2>3.5.0 (2023-02-08)<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h2>
<p>This version introduces new features such as differential expression analysis with the Limma-Voom pipeline,
customizable databases for the quick-search function, basic filtering and procrssing functions for gene sets,
improved progammatic API for FeatureSet and RankedSet objects, and RPKM and TPM read count normalization.
Several changes have been made to improve the user experience, including updated documentation,
improved clarity of function tooltips, clearer output formats, and faster download speeds for tutorial videos.</p>
<section id="added">
<h3>Added<a class="headerlink" href="#added" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added differential expression analysis with the Limma-Voom pipelines (CountFilter.differential_expression_limma_voom)</p></li>
<li><p>You can now select which databases to display in the right-click quick-search menu, using the settings menu.</p></li>
<li><p>Gene sets now support some basic operations (filtering, gene ID translating, etc) through the graphical interface.</p></li>
<li><p>enrichment.FeatureSet and enrichment.RankedSet now support some filtering operations from the filtering module (such as filtering by user-defined attributes, GO terms, or KEGG pathways).</p></li>
<li><p>Added reads-per-kilobase-million (RPKM) and transcripts-per-million (TPM) normalization methods (CountFilter.normalize_to_rpkm() and CountFilter.normalize_to_tpm()).</p></li>
</ul>
</section>
<section id="id5">
<h3>Changed<a class="headerlink" href="#id5" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Classes enrichment.FeatureSet and enrichment.RankedSet now inherit from Python base-class set, and can be interacted with like other Python sets. The old API and attributes of these classes were maintained as they were.</p></li>
<li><p>Improved documentation for some functions.</p></li>
<li><p>Function selection tooltips should now display information more clearly.</p></li>
<li><p>Pipelines that contain consecutive clustering/splitting functions will now return their outputs in a clearer format.</p></li>
<li><p>Enrichment bar-plots should now adjust the x-axis limits more tightly to fit the displayed data.</p></li>
<li><p>Improved clarity of automatic titles in enrichment plots.</p></li>
<li><p>Download/update speed of tutorial videos has improved significantly.</p></li>
</ul>
</section>
<section id="id6">
<h3>Fixed<a class="headerlink" href="#id6" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed bug where Pipelines would not always properly run ‘translate_gene_ids’</p></li>
</ul>
</section>
</section>
<section id="id7">
<h2>3.4.2 (2023-02-01)<a class="headerlink" href="#id7" title="Permalink to this heading"></a></h2>
<p>This version introduces minor bug fixes.</p>
<section id="id8">
<h3>Fixed<a class="headerlink" href="#id8" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed bug where updating RNAlysis through the graphical interface would not update some of the optional dependencies.</p></li>
<li><p>Fixed typos in the documentation.</p></li>
</ul>
</section>
</section>
<section id="id9">
<h2>3.4.0 (2023-02-01)<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h2>
<p>From this release forward, <em>RNAlysis</em> is made available as a stand-alone app for Windows and MacOS. You can download these stand-alone versions from the GitHub Releases page.
In addition, new features were added, including new plots, filtering functions, integration of the external tools bowtie2 and featureCounts, and the ability to generate Gene Ontology Graphs and KEGG Pathway Graphs without running enrichment analysis from scratch.</p>
<section id="id10">
<h3>Added<a class="headerlink" href="#id10" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added a Scree Plot (explained variance per PC) to Principle Component Analysis</p></li>
<li><p>Added CountFilter.split_by_principal_component(), allowing users to filter genes based on their contributions (loadings) to PCA Principal Components.</p></li>
<li><p>Added Filter.drop_columns</p></li>
<li><p>Added support for the Sharpened Cosine distance metric in clustering analyses</p></li>
<li><p>KEGG enrichment can now generate KEGG pathway graphs for pathways that were found to be statistically significant</p></li>
<li><p>Added functions to the enrichment module that can generate KEGG Pathway or Gene Ontology plots based on previously-generated enrichment results</p></li>
<li><p>You can now clear the <em>RNAlysis</em> cache from the <em>RNAlysis</em> GUI, or through the general module.</p></li>
<li><p>Added bowtie2 alignment to the fastq module.</p></li>
<li><p>Added FeatureCounts feature-counting to the fastq module.</p></li>
<li><p>You can now choose whether or not to discard genes from enrichment analysis if they have no annotations associated with them.</p></li>
<li><p>When right-clicking on a specific cell in a table or line in a gene set view, a context menu will open, allowing you to copy the associated value, or look it up in one of few common biology databases.</p></li>
<li><p>Added sections to the programmatic user guide about the <cite>fastq</cite> module.</p></li>
</ul>
</section>
<section id="id11">
<h3>Changed<a class="headerlink" href="#id11" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Replaced the ‘parallel’ parameter in enrichment functions with the ‘parallel_backend’ parameter, allowing users to choose which parallel backend (if any) will be used in the function.</p></li>
<li><p>Added ‘parallel_backend’ parameter to all clustering functions under the filtering module.</p></li>
<li><p>When generating Gene Ontology/KEGG Pathway graphs, users can choose whether or not to generate the figure in an additional separate file.</p></li>
<li><p>Updated type annotations of some functions to be more precise and helpful (for example, setting a lower bound on some int function parameters).</p></li>
<li><p>The colorbar ticks in enrichment bar plots now match the axis ticks on the main axis.</p></li>
<li><p>Slight improvements in GUI performance, stability, and looks.</p></li>
<li><p>Slight improvements in performance of enrichment analysis when examining a small number of attributes.</p></li>
<li><p>enrichment.plot_enrichment() was replaced by enrichment.enrichment_bar_plot().</p></li>
<li><p>CountFilter.differential_expression() has new optional parameter <cite>output_folder</cite>, which allows users to save the generated data tables and the R script that generated them into a specified folder.</p></li>
</ul>
</section>
<section id="id12">
<h3>Fixed<a class="headerlink" href="#id12" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>In CountFilter.differential_expression_deseq2(), fixed a bug where design matrix files with non-comma delimiters would cause an error (thanks to <a class="reference external" href="https://github.com/Mintxoklet">Mintxoklet</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/7">#7</a>)</p></li>
<li><p>Fixed bug where setup.py would install a directory named tests into site-packages folder (thanks to <a class="reference external" href="https://github.com/kbipinkumar">Bipin Kumar</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/9">#9</a>)</p></li>
<li><p>Fixed bug where the windows of some functions (differential expression, adapter trimming, etc) did not show a link to the function’s documentation page.</p></li>
<li><p>Fixed typos in some parts of the <em>RNAlysis</em> documentation</p></li>
<li><p>When filtering a table by a single user-defined attribute, the automatic table name will now be more informative about the operation applied.</p></li>
<li><p>Fixed bug where occasionally a Pipeline or Function would generate multiple tables of the same name, but only one of them will appear in the GUI.</p></li>
<li><p>Fixed bug where occasionally significance asterisks on enrichment bar-plots would appear in the wrong location.</p></li>
<li><p>Fixed bug where fastq.create_kallisto_index() (Create Kallisto Index) would not make use of the <cite>make_unique</cite> parameter (thanks to Matthias Wilm)</p></li>
</ul>
</section>
<section id="removed">
<h3>Removed<a class="headerlink" href="#removed" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Removed the previously-deprecated functions <cite>enrichment.enrich_randomization()</cite> and <cite>enrichment.enrich_hypergeometric()</cite>.</p></li>
</ul>
</section>
<section id="new-contributors">
<h3>New Contributors<a class="headerlink" href="#new-contributors" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/Mintxoklet">Mintxoklet</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/7">#7</a></p></li>
<li><p><a class="reference external" href="https://github.com/kbipinkumar">Bipin Kumar</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/9">#9</a></p></li>
<li><p>Matthias Wilm</p></li>
</ul>
</section>
</section>
<section id="id15">
<h2>3.3.0 (2022-12-02)<a class="headerlink" href="#id15" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduced quality-of-life improvements to the graphical user interface.</p></li>
</ul>
<section id="id16">
<h3>Added<a class="headerlink" href="#id16" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added a Frequently Asked Questions page, and linked all <em>RNAlysis</em> help material inside the graphical interface Help menu.</p></li>
<li><p>Pipelines can now be edited and deleted through the Pipeline menu of the graphical interface.</p></li>
<li><p>Added Contributing page to the documentation, with basic guidelines on how you can contribute to the <em>RNAlysis</em> project!</p></li>
</ul>
</section>
<section id="id17">
<h3>Changed<a class="headerlink" href="#id17" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>All open tabs are now always visible in the main menu screen. Tab names are now shortened with ellipsis if nessecary.</p></li>
<li><p>The right-click context menu of the main menu tabs now allows users to open a new tab at a specific position, or close a specific tab/tabs to the right/tabs to the left/all other tabs.</p></li>
<li><p><em>RNAlysis</em> documentation is now split into GUI documentation (quick-start video guide, tutorial, GUI user guide), and programmatic documentation (programmatic user guide)</p></li>
<li><p>Improved readability of <em>RNAlysis</em> logs</p></li>
<li><p>Pipelines are now exported with additional metadata - the version of <em>RNAlysis</em> they were exported from, and the date and time it was exported. This metadata should not affect Pipelines that were created in older versions, and does not affect the way Pipelines are applied to data tables.</p></li>
</ul>
</section>
<section id="id18">
<h3>Fixed<a class="headerlink" href="#id18" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><em>RNAlysis</em> now warns users if they attempt to overwrite an existing Pipeline.</p></li>
<li><p>Fixed an incorrect keyboard shortcut for Export Pipeline action</p></li>
</ul>
</section>
</section>
<section id="id19">
<h2>3.2.2 (2022-11-25)<a class="headerlink" href="#id19" title="Permalink to this heading"></a></h2>
<section id="id20">
<h3>Fixed<a class="headerlink" href="#id20" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed bug with DESeq2 automatic installation on Windows computers.</p></li>
</ul>
</section>
</section>
<section id="id21">
<h2>3.2.1 (2022-11-25)<a class="headerlink" href="#id21" title="Permalink to this heading"></a></h2>
<section id="id22">
<h3>Changed<a class="headerlink" href="#id22" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Updated citation information for <em>RNAlysis</em></p></li>
</ul>
</section>
<section id="id23">
<h3>Fixed<a class="headerlink" href="#id23" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed typos in the <em>RNAlysis</em> tutorial</p></li>
</ul>
</section>
</section>
<section id="id24">
<h2>3.2.0 (2022-11-23)<a class="headerlink" href="#id24" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces quality-of-life changes to the graphical user interface, functions for translating gene IDs and running differential expression analysis, and extends RNAlysis to support Python versions 3.9 and 3.10.</p></li>
</ul>
<section id="id25">
<h3>Added<a class="headerlink" href="#id25" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added Filter.translate_gene_ids()</p></li>
<li><p>Added CountFilter.differential_expression_deseq2()</p></li>
<li><p>Added Filter.filter_by_kegg_annotations()</p></li>
<li><p>Added Filter.filter_by_go_annotations()</p></li>
<li><p>Added CountFilter.average_replicate_samples()</p></li>
<li><p>Added fastq module that contains adapter-trimming functions utilizing CutAdapt, and mRNA-sequencing quantification using kallisto.</p></li>
</ul>
</section>
<section id="id26">
<h3>Changed<a class="headerlink" href="#id26" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added additional plotting parameters to visualization functions.</p></li>
<li><p>Improved performance of some aspects of the graphical user interface.</p></li>
<li><p>RNAlysis’ basic features are now supported on Python versions 3.9 and 3.10.</p></li>
<li><p>CountFilter.pca() now generates a plot for <em>every</em> pair of Principal Components requested by the user.</p></li>
<li><p>CountFilter.split_clicom() now supports clustering each batch of replicates separately, using the ‘replicates_grouping’ parameter</p></li>
<li><p>Biotype-based filtering and summary can now be done based on GTF annotation files instead of a Biotype Reference Table.</p></li>
<li><p>Filter.biotypes() was refactored into Filter.biotypes_from_ref_table()</p></li>
<li><p>Filter.filter_biotype() was refactored into Filter.filter_biotype_from_ref_table()</p></li>
</ul>
</section>
<section id="id27">
<h3>Fixed<a class="headerlink" href="#id27" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Users can now queue multiple computationally-intense enrichment/clustering tasks while another task is running.</p></li>
<li><p>Fixed a bug where sometimes some function parameters would disappear from the graphical user interface.</p></li>
<li><p>Fixed a bug where exceptions during computationally-intense tasks would cause <em>RNAlysis</em> to crash.</p></li>
<li><p>Auxillary windows are now properly minimized when analysis starts, and restored when analysis ends or encounters an error.</p></li>
</ul>
</section>
</section>
<section id="id28">
<h2>3.1.0 (2022-10-16)<a class="headerlink" href="#id28" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces new count matrix normalization methods, as well as MA plots and minor bug fixes.</p></li>
</ul>
<section id="id29">
<h3>Added<a class="headerlink" href="#id29" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added the visualization function ma_plot() for CountFilter</p></li>
<li><p>Added functions for the normalization functions Relative Log Ratio (RLE), Trimmed Mean of M-values (TMM), Median of Ratios (MRN), Quantile normalization (quantile)</p></li>
</ul>
</section>
<section id="id30">
<h3>Changed<a class="headerlink" href="#id30" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>CountFilter.normalize_to_rpm() was renamed to CountFilter.normalize_to_rpm_htseqcount(), and was supplemented by the more general function for normalizing to Reads Per Million CountFilter.normalize_to_rpm()</p></li>
</ul>
</section>
<section id="id31">
<h3>Fixed<a class="headerlink" href="#id31" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed a bug where some elements of the graphical user interface would not display correctly</p></li>
</ul>
</section>
</section>
<section id="id32">
<h2>3.0.1 (2022-10-12)<a class="headerlink" href="#id32" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version fixes a bug with displaying the tutorial videos in the graphical user interface.</p></li>
</ul>
</section>
<section id="id33">
<h2>3.0.0 (2022-10-10)<a class="headerlink" href="#id33" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces a graphical user interface for RNAlysis, as well as new functions for KEGG Pathways enrichment analysis.</p></li>
</ul>
<section id="id34">
<h3>Added<a class="headerlink" href="#id34" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>RNAlysis now includes a graphical user interface</p></li>
<li><p>Pipelines can now be imported and exported</p></li>
<li><p>Enrichment and single-set-enrichment for KEGG pathway data</p></li>
</ul>
</section>
<section id="id35">
<h3>Changed<a class="headerlink" href="#id35" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added function FeatureSet.user_defined_enrichment(), which will replace FeatureSet.enrich_hypergeometric() and FeatureSet.enrich_randomization()</p></li>
<li><p>Updated signature of enrichment.venn_diagram</p></li>
<li><p>enrichment.venn_diagram and enrichment.upset_plot can now be generated on a user-supplied FIgure</p></li>
<li><p>Clustering functions now apply a power transform to count data prior to clustering by default</p></li>
<li><p>Non-deprecated enrichment functions no longer filter the background set by biotype by default</p></li>
<li><p>Changed signature of CountFilter.pca, CountFilter.box_plot, CountFilter.enhanced_box_plot, CountFilter.clustergram, and CountFilter.pairplot to ensure consistency among visualization functions.</p></li>
</ul>
</section>
<section id="id36">
<h3>Fixed<a class="headerlink" href="#id36" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>enrichment.venn_diagram can now be plotted with outlines when the circles are unweighted</p></li>
<li><p>Fixed bug in Pipeline.apply_to() where a Filter object would be returned even when the Pipeline was applied inplace</p></li>
</ul>
</section>
</section>
<section id="id37">
<h2>2.1.1 (2022-07-05)<a class="headerlink" href="#id37" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version fixes issues with running GO enrichment that resulted from recent changes to UniProt’s API.  Moreover, this version slightly improves the performance of some functions.</p></li>
</ul>
<section id="id38">
<h3>Changed<a class="headerlink" href="#id38" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed issues with running GO enrichment that resulted from changes to UniProt’s API.</p></li>
<li><p>Some functions that fetch annotations now cache their results, leading to improved runtimes.</p></li>
<li><p>Updated the documentation of some functions to better reflect their usage and input parameters.</p></li>
</ul>
</section>
</section>
<section id="id39">
<h2>2.1.0 (2022-04-16)<a class="headerlink" href="#id39" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces multiple new features, as well as generally improved graphs and quality-of-life changes.</p></li>
</ul>
<section id="id40">
<h3>Added<a class="headerlink" href="#id40" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>GO enrichment can now generate Ontology Graphs for the statistically significant GO terms.</p></li>
<li><p>Added CountFilter.split_clicom(), an implementation of the CLICOM ensemble-based clustering method (Mimaroglu and Yagci 2012).</p></li>
<li><p>Added Filter.transform(), a method that can transform your data tables with either predefined or user-defined transformations.</p></li>
</ul>
</section>
<section id="id41">
<h3>Changed<a class="headerlink" href="#id41" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>CountFilter.pairplot() now uses a logarithmic scale by default.</p></li>
<li><p>Visually improved the graphs generated by many functions, including CountFilter.pairplot() and CountFilter.plot_expression().</p></li>
<li><p>The clusters resulting from all clustering functions are now sorted by size instead of being sorted randomly.</p></li>
</ul>
</section>
<section id="id42">
<h3>Fixed<a class="headerlink" href="#id42" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Minor bug fixes.</p></li>
</ul>
</section>
</section>
<section id="id43">
<h2>2.0.1 (2022-04-02)<a class="headerlink" href="#id43" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces small bug fixes, as well as a new function in the Filtering module.</p></li>
</ul>
<section id="id44">
<h3>Added<a class="headerlink" href="#id44" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added Filter.majority_vote_intersection(), which returns a set/string of the features that appear in at least (majority_threhold * 100)% of the given Filter objects/sets.</p></li>
</ul>
</section>
<section id="id45">
<h3>Changed<a class="headerlink" href="#id45" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>When mapping/inferring taxon IDs during GO enrichment analysis, organisms will now be prioritized based on their taxon ID values (numerically lower IDs will be considered to be more relevant).</p></li>
</ul>
</section>
<section id="id46">
<h3>Fixed<a class="headerlink" href="#id46" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed bug that occured when mapping/inferring taxon IDs during GO enrichment analysis, where integer taxon IDs would be matched by name similarity before trying an exact ID match, leading to spurious matches.</p></li>
<li><p>Fixed bug that occursed when plotting clustering results with style=’all’ on Python 3.8.</p></li>
</ul>
</section>
</section>
<section id="id47">
<h2>2.0.0 (2021-12-05)<a class="headerlink" href="#id47" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces new method to cluster your read count matrices, including K-Means/Medoids clustering, Hierarchical clustering, and HDBSCAN.</p></li>
<li><p>This version introduces many new ways to perform enrichment analysis and to visualize your results, including highly customizable GO Enrichment, enrichment based on ranked lists of genes, and enrichment for non-categorical attributes.</p></li>
<li><p>This version introduces Pipelines - a quicker and more convenient way to apply a particular analysis pipeline to multiple Filter objects.</p></li>
<li><p>This version improves the performance of many functions in RNAlysis, and in particular the performance of randomization tests.</p></li>
<li><p>This version includes changes to names and signatures of some functions in the module, as elaborated below.</p></li>
</ul>
<section id="id48">
<h3>Added<a class="headerlink" href="#id48" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added class Pipeline to filtering module, which applies a series of filter functions to specified Filter objects.</p></li>
<li><p>Added CountFilter.split_kmeans(), CountFilter.split_kmedoids(), CountFilter.split_hierarchical() and CountFilter.split_hdbscan(), which split your read count matrices into clusters with similar expression patterns.</p></li>
<li><p>Added class RankedSet to enrichment module, which accepts a ranked list of genes/features, and can perform single-list enrichment analysis</p></li>
<li><p>Added RankedSet.single_set_enrichment(), which can perfofm single-list enrichment analysis of user-defined attributes using XL-mHG test (see <a class="reference external" href="https://dx.doi.org/10.1371/journal.pcbi.0030039">Eden et al. (PLoS Comput Biol, 2007)</a>  and <a class="reference external" href="https://dx.doi.org/10.1371/journal.pone.0143196">Wagner (PLoS One, 2015)</a> ).</p></li>
<li><p>Added FeatureSet.go_enrichment() and RankedSet.single_set_go_enrichment(), which let you compute Gene Ontology enrichment for any organism of your choice, and filter the GO annotations used according to your preferences.</p></li>
<li><p>Added FeatureSet.enrich_hypergeometric(), which can perform enrichment analysis using the Hypergeometric Test.</p></li>
<li><p>Added more visualization functions, such CountFilter.enhanced_box_plot().</p></li>
<li><p>Added FeatureSet.change_set_name(), to give a new ‘set_name’ to a FeatureSet object.</p></li>
</ul>
</section>
<section id="id49">
<h3>Changed<a class="headerlink" href="#id49" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>FeatureSet.enrich_randomization_parallel() was deprecated. Instead, you can compute your enrichment analysis with parallel computing by calling FeatureSet.enrich_randomization() with the argument ‘parallel_processing=True’. Moreover, parallel session will now start automatically if one was not already active.</p></li>
<li><p>Improved running time of enrich_randomization() about six-fold.</p></li>
<li><p>Filter objects can be created from any delimiter-separated file format (.csv, .tsv, .txt, etc).</p></li>
<li><p>CountFilter.pca() can now be plotted without labeled points.</p></li>
<li><p>Filter.index_string is now sorted by the current order of indices in the Filter object, instead of by alphabetical order.</p></li>
<li><p>CountFilter.violin_plot() now accepts a y_title argument.</p></li>
<li><p>Added more optional arguments to visualization functions such as CountFilter.violin_plot() and CountFilter.clustergram().</p></li>
<li><p>Automatic filenames for Filter objects should now reflect more clearly the operations that were performed.</p></li>
<li><p>The DataFrame returned by enrich_randomization() and enrich_randomization_parallel() now contains the additional column ‘data_scale’, determined by the new optional argument ‘data_scale’.</p></li>
<li><p>The columns ‘n obs’ and ‘n exp’ in the DataFrame returned by enrich_randomization() and enrich_randomization_parallel() were renamed to ‘obs’ and ‘exp’ respectively.</p></li>
<li><p>FeatureSets no longer support in-place set operations (intersection, union, difference, symmetric difference). Instead, these functions return a new FeatureSet.</p></li>
<li><p>Filter.biotypes_from_ref_table() now accepts the boolean parameter ‘long_format’ instead of the str parameter ‘format’.</p></li>
<li><p>Filter.biotypes_from_ref_table() and FeatureSet.biotypes_from_ref_table() now count features which do not appear in the Biotype Reference Table as ‘_missing_from_biotype_reference’ instead of ‘not_in_biotype_reference’.</p></li>
</ul>
</section>
<section id="id50">
<h3>Fixed<a class="headerlink" href="#id50" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Updated type-hinting of specific functions.</p></li>
<li><p>Filter.biotypes_from_ref_table() and FeatureSet.biotypes_from_ref_table() now support Biotype Reference Tables with different column names.</p></li>
<li><p>Generally improved performance of RNAlysis.</p></li>
<li><p>Fixed bug in Filter.filter_percentile() where the value at the exact percentile speficied (e.g. the median for percentile=0.5) would be removed from the Filter object.</p></li>
<li><p>Fixed bug in enrichment.FeatureSet, where creating a FeatureSet from input string would result in an empty set.</p></li>
<li><p>Various minor bug fixes.</p></li>
</ul>
</section>
</section>
<section id="id51">
<h2>1.3.5 (2020-05-27)<a class="headerlink" href="#id51" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version introduces minor bug fixes and a few more visualization options.</p></li>
</ul>
<section id="id52">
<h3>Added<a class="headerlink" href="#id52" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added Filter.filter_missing_values(), which can remove rows with NaN values in some (or all) columns.</p></li>
<li><p>Added the visualization function CountFilter.box_plot().</p></li>
</ul>
</section>
<section id="id53">
<h3>Changed<a class="headerlink" href="#id53" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Updated docstrings and printouts of several functions.</p></li>
<li><p>Slightly improved speed and performance across the board.</p></li>
<li><p>Filter.feature_string() is now sorted alphabetically.</p></li>
<li><p>Enrichment randomization functions in the enrichment module now accept a ‘random_seed’ argument, to be able to generate consistent results over multiple sessions.</p></li>
<li><p>Enrichment randomization functions can now return the matplotlib Figure object, in addition to the results table.</p></li>
</ul>
</section>
<section id="id54">
<h3>Fixed<a class="headerlink" href="#id54" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed DepracationWarning on parsing functions from the general module.</p></li>
<li><p>Fixed bug where saving csv files on Linux systems would save the files under the wrong directory.</p></li>
<li><p>Fixed a bug where UTF-8-encoded Reference Tables won’t be loaded correctly</p></li>
<li><p>Fixed a bug where enrichment.upsetplot() and enrichment.venn_diagram() would sometimes modify the user dict input ‘objs’.</p></li>
<li><p>Fixed a bug in CountFilter.pairplot where log2 would be calculated without a pseudocount added, leading to division by 0.</p></li>
</ul>
</section>
</section>
<section id="id55">
<h2>1.3.4 (2020-04-07)<a class="headerlink" href="#id55" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>This version fixed a bug that prevented installation of the package.</p></li>
</ul>
<section id="id56">
<h3>Changed<a class="headerlink" href="#id56" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Updated docstrings and printouts of several functions</p></li>
</ul>
</section>
<section id="id57">
<h3>Fixed<a class="headerlink" href="#id57" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Fixed a bug with installation of the previous version</p></li>
</ul>
</section>
</section>
<section id="id58">
<h2>1.3.3 (2020-03-28)<a class="headerlink" href="#id58" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>First stable release on PyPI.</p></li>
</ul>
<section id="id59">
<h3>Added<a class="headerlink" href="#id59" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Added Filter.sort(), and upgraded the functionality of Filter.filter_top_n().</p></li>
<li><p>Added UpSet plots and Venn diagrams to enrichment module.</p></li>
<li><p>User-defined biotype reference tables can now be used.</p></li>
<li><p>Filter operations now print out the result of the operation.</p></li>
<li><p>Enrichment randomization tests now also support non-WBGene indexing.</p></li>
<li><p>Filter.biotypes_from_ref_table() and FeatureSet.biotypes_from_ref_table() now report genes that don’t appear in the biotype reference table.</p></li>
<li><p>Filter.biotypes_from_ref_table() can now give a long-form report with descriptive statistics of all columns, grouped by biotype.</p></li>
<li><p>Added code examples to the user guide and to the docstrings of most functions.</p></li>
</ul>
</section>
<section id="id60">
<h3>Changed<a class="headerlink" href="#id60" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Changed argument order and default values in filtering.CountFilter.from_folder().</p></li>
<li><p>Changed default title in scatter_sample_vs_sample().</p></li>
<li><p>Changed default filename in CountFilter.fold_change().</p></li>
<li><p>Settings are now saved in a .yaml format. Reading and writing of settings have been modified.</p></li>
<li><p>Changed argument name ‘deseq_highlight’ to ‘highlight’ in scatter_sample_vs_sample(). It can now accept any Filter object.</p></li>
<li><p>Updated documentation and default ‘mode’ value for FeatureSet.go_enrichment().</p></li>
<li><p>Updated the signature and function of general.load_csv() to be clearer and more predictable.</p></li>
<li><p>Changed argument names in CountFilter.from_folder().</p></li>
<li><p>Modified names and signatures of .csv test files functions to make them more comprehensible.</p></li>
<li><p>Renamed ‘Filter.filter_by_ref_table_attr()’ to ‘Filter.filter_by_attribute()’.</p></li>
<li><p>Renamed ‘Filter.split_by_ref_table_attr()’ to ‘Filter.split_by_attribute()’.</p></li>
<li><p>Renamed ‘Filter.norm_reads_with_size_factor()’ to ‘Filter.normalize_with_scaling_factors()’. It can now use any set of scaling factors to normalize libraries.</p></li>
<li><p>Renamed ‘Filter.norm_reads_to_rpm()’ to ‘Filter.normalize_to_rpm()’.</p></li>
<li><p>Made some functions in the general module hidden.</p></li>
</ul>
</section>
<section id="id61">
<h3>Fixed<a class="headerlink" href="#id61" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Various bug fixes</p></li>
</ul>
</section>
<section id="id62">
<h3>Removed<a class="headerlink" href="#id62" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p>Removed the ‘feature_name_to_wbgene’ module from RNAlysis.</p></li>
</ul>
</section>
</section>
<section id="id63">
<h2>1.3.2 (2019-12-11)<a class="headerlink" href="#id63" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>First beta release on PyPI.</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="glossary.html" class="btn btn-neutral float-left" title="Glossary - programmatic interface" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.fastq.html" class="btn btn-neutral float-right" title="rnalysis.fastq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>