<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>History &mdash; RNAlysis 4.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=3304f9e4"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="_static/copybutton.js?v=5b738548"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.fastq" href="rnalysis.fastq.html" />
    <link rel="prev" title="Glossary - programmatic interface" href="glossary.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">üß¨ What is <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#see-rnalysis-in-action">üé• See <em>RNAlysis</em> in Action</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#key-features">üöÄ Key Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">üõ†Ô∏è How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">üìä How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_report.html">How to use the <em>RNAlysis</em> analysis reports</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">History</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#id1">4.0.0 (2024-06-29)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#added">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#changed">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#fixed">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#backwards-incompatible-changes">Backwards-incompatible changes</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id2">3.12.0 (2024-05-14)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id3">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id4">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#new-contributors">New Contributors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id8">3.11.0 (2024-01-05)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id9">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id10">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id11">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id12">3.10.1 (2023-11-22)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id13">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id14">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id15">3.10.0 (2023-10-31)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id16">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id17">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id18">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id19">3.9.2 (2023-06-23)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id20">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id21">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id22">3.9.1 (2023-06-19)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id23">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id25">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id26">3.9.0 (2023-06-09)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id27">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id28">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id29">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">New Contributors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id32">3.8.0 (2023-05-07)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id33">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id34">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id35">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id37">New Contributors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id38">3.7.0 (2023-04-07)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id39">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id40">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id41">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id42">3.6.2 (2023-03-25)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id43">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id44">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id45">3.6.1 (2023-03-22)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id46">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id47">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id48">3.6.0 (2023-03-07)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id49">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id50">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id51">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id53">New Contributors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id54">3.5.2 (2023-02-23)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id55">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id56">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id57">3.5.1 (2023-02-22)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id58">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id59">3.5.0 (2023-02-08)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id60">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id61">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id62">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id63">3.4.2 (2023-02-01)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id64">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id65">3.4.0 (2023-02-01)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id66">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id67">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id68">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#removed">Removed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id71">New Contributors</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id72">3.3.0 (2022-12-02)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id73">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id74">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id75">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id76">3.2.2 (2022-11-25)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id77">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id78">3.2.1 (2022-11-25)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id79">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id80">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id81">3.2.0 (2022-11-23)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id82">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id83">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id84">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id85">3.1.0 (2022-10-16)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id86">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id87">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id88">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id89">3.0.1 (2022-10-12)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id90">3.0.0 (2022-10-10)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id91">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id92">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id93">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id94">2.1.1 (2022-07-05)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id95">Changed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id96">2.1.0 (2022-04-16)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id97">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id98">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id99">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id100">2.0.1 (2022-04-02)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id101">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id102">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id103">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id104">2.0.0 (2021-12-05)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id105">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id106">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id107">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id108">1.3.5 (2020-05-27)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id109">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id110">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id111">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id112">1.3.4 (2020-04-07)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id113">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id114">Fixed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id115">1.3.3 (2020-03-28)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id116">Added</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id117">Changed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id118">Fixed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id119">Removed</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#id120">1.3.2 (2019-12-11)</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">History</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/history.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="history">
<h1>History<a class="headerlink" href="#history" title="Link to this heading">ÔÉÅ</a></h1>
<section id="id1">
<h2>4.0.0 (2024-06-29)<a class="headerlink" href="#id1" title="Link to this heading">ÔÉÅ</a></h2>
<p>Version 4.0.0 of RNAlysis introduces two major improvements:
First, the software has switched from Pandas to Polars, which should substantially improve overall performance, particularly when generating automatic analysis reports.</p>
<p>Second, significant enhancements have been made to the automatic analysis reports.
These include new interactive features such as the ability to highlight analysis paths and filter the report by node types,
improved hierarchical layout for better readability, performance optimizations, and additional customization options.
The reports now offer a more comprehensive and user-friendly way to navigate and understand analysis workflows.</p>
<p>This update also includes various bug fixes and minor improvements to other functions within the software.</p>
<p>Happy analysis!</p>
<section id="added">
<h3>Added<a class="headerlink" href="#added" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added additional optional parameters to automatic report generation.</p></li>
<li><p>Added additional parameters to ‚ÄòHierarchical clustergram plot‚Äô (CountFilter.clustergram).</p></li>
</ul>
</section>
<section id="changed">
<h3>Changed<a class="headerlink" href="#changed" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis now uses Polars instead of Pandas. This change should improve the performance of RNAlysis by an order of magnitude, especially when generating automatic analysis reports.</p></li>
<li><p>Automatic analysis reports are now laid out in a hierarchical structure, making it easier to navigate through the report.</p></li>
<li><p>When clicking on a node in an automatic analysis report, the analysis path leading to that node will be highlighted.</p></li>
<li><p>When clicking on a legend node in an automatic analysis report, all nodes of that type will be highlighted.</p></li>
<li><p>The ‚ÄòHierarchical clustergram plot‚Äô function should now run faster on large datasets.</p></li>
<li><p>Added support for running tasks using ‚Äòsequential‚Äô backend on standalone releases of RNAlysis.</p></li>
</ul>
</section>
<section id="fixed">
<h3>Fixed<a class="headerlink" href="#fixed" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where the visualization functions ‚ÄòPlot histogram of p-values‚Äô and ‚ÄòHistogram of a column‚Äô would not display a graph immediately when using the stand-alone app.</p></li>
<li><p>Fixed bug re-loading a saved session report would sometimes lead to missing connections in the analysis flow graph.</p></li>
<li><p>Fixed bug where applying not-inplace operations to differential expression tables would change the expected name of the p-value column.</p></li>
<li><p>Fixed bug where table previews in automatic analysis reports would display an extra empty line.</p></li>
</ul>
</section>
<section id="backwards-incompatible-changes">
<h3>Backwards-incompatible changes<a class="headerlink" href="#backwards-incompatible-changes" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Since RNAlysis now uses Polars instead of Pandas, all functions that previously returned Pandas DataFrames will now return Polars DataFrames. To keep your code compatible with the new version, you may need to adjust the way you interact with the returned DataFrames, or convert them back into Pandas DataFrames.</p></li>
<li><p>Sessions saved with previous versions of RNAlysis are mostly compatible with RNAlysis 4.0.0, but some of the new features of automatic analysis reports (such as highlighting the report by node type) may not work in older sessions.</p></li>
</ul>
</section>
</section>
<section id="id2">
<h2>3.12.0 (2024-05-14)<a class="headerlink" href="#id2" title="Link to this heading">ÔÉÅ</a></h2>
<p>I‚Äôm happy to announce the latest release of RNAlysis, which brings a variety of new features and improvements.</p>
<p>One of the key additions in this version is expanded support for advanced differential expression analysis using DESeq2 and Limma-Voom.
You can now test continuous covariates, perform likelihood ratio tests for factors, interactions, and polynomials, and take advantage of sample quality weights in Limma-Voom analysis.
We‚Äôve also added several new visualization options, such as a p-value histogram plot and more customization choices for gene expression plots.</p>
<p>This release also includes various usability enhancements and bug fixes to provide a smoother analysis experience. We‚Äôve clarified error messages, improved parameter annotations, and resolved issues that could lead to crashes or incorrect results in certain scenarios. RNAlysis now integrates with the latest versions of the OrthoInspector and ShortStack APIs as well.</p>
<p>As always, your support and feedback are greatly appreciated.
If you have any questions, encounter issues, or have suggestions for future development, please don‚Äôt hesitate to reach out.
Happy analysis!
Guy</p>
<section id="id3">
<h3>Added<a class="headerlink" href="#id3" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added support for advanced differential expression analysis with DESeq2/Limma-Voom, including testing continuous covariates, as well as likelihood ratio tests for factors, interactions, and polynomials.</p></li>
<li><p>Added support for sample quality weights in Limma-Voom differential expression analysis.</p></li>
<li><p>Added support for the ‚ÄòcooksCutoff‚Äô parameter in DESq2 differential expression analysis.</p></li>
<li><p>Added support for user-provided scaling factors in DESeq2 differential expression analysis.</p></li>
<li><p>bowtie2 and kallisto paired-end modes now support the new file naming method ‚Äòsmart‚Äô, that attempts to automatically determine the common name of each pair of paired-end fastq files.</p></li>
<li><p>Added a p-value histogram plot (DESeqFilter.pval_histogram) that displays the distribution of p-values in a differential expression table.</p></li>
<li><p>Added new visualization options for ‚Äòplot expression of specific genes‚Äô function (CountFilter.plot_expression).</p></li>
<li><p>Added new function ‚ÄòHistogram of a column‚Äô (Filter.histogram) that generates a histogram of a column in a table.</p></li>
<li><p>Added new function ‚ÄòConcatenate two tables‚Äô (Filter.concatenate) that concatenates two tables along their rows.</p></li>
<li><p>The filtering function ‚ÄòFilter with a number filter‚Äô now supports filtering based on absolute values.</p></li>
</ul>
</section>
<section id="id4">
<h3>Changed<a class="headerlink" href="#id4" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>When running differential expression analysis, RNAlysis will automatically ensure that the order of samples in your design matrix matches the order of samples in your count matrix, avoiding erronious results.</p></li>
<li><p>The ‚ÄòFilter genes with low expression in all columns‚Äô function (CountFilter.filter_low_reads) now supports the ‚Äòn_samples‚Äô parameter, allowing users to filter genes with a minimal expression threshold in a specific number of samples.</p></li>
<li><p>The ‚ÄòPlot expressino of specific genes‚Äô function (CountFilter.plot_expression) now supports the ‚Äòjitter‚Äô, ‚Äògroup_names‚Äô and ‚Äòlog_scale‚Äô parameters, allowing users to further customize the plot.</p></li>
<li><p>The ‚ÄòScatter plot - sample VS sample‚Äô function (CountFilter.scatter_sample_vs_sample) now always displays highlighted points on top of the plot, making it easier to see which points are highlighted.</p></li>
<li><p>Kallisto quantification (kallisto_quantify_single_end and kallisto_quantify_paired_end) now supports the ‚Äòsummation_method‚Äô parameter, allowing users to choose between ‚Äòraw‚Äô and ‚Äòscaled_tpm‚Äô transcript summation methods. The default behavior of the functions did not change (it corresponds to ‚Äòscaled_tpm‚Äô).</p></li>
<li><p>Enrichment bar plots now have optional parameters that control font sizes for titles and labels.</p></li>
<li><p>Moved the enrichment analysis and enrichment graphs actions to the ‚ÄúEnrichment‚Äù menu in the graphical interface to make the actions easier to find.</p></li>
<li><p>Improved the clarity of error messages when attempting to read an invalid GTF file.</p></li>
<li><p>RNAlysis now supports the latest version of OrthoInspector API.</p></li>
<li><p>RNAlysis now supports the latest version of Ensembl Ortholog API.</p></li>
<li><p>Improved annotation for the ‚Äòmetric‚Äô parameter of the ‚ÄòHierarchical clustergram plot‚Äô function (CountFilter.clustergram).</p></li>
<li><p>Improved performance of RNAlysis when generating automatic session reports.</p></li>
<li><p>RNAlysis now offers default values for differential expression tables‚Äô column names.</p></li>
<li><p>Functions that average replicates now display clearer group names by default.</p></li>
<li><p>The RNAlysis interface to ShortStack now uses the most recent API (replaced ‚ÄòknownRNAs‚Äô with ‚Äòknown_miRNAs‚Äô).</p></li>
<li><p>When running differential expression, RNAlysis session reports will automatically include the auto-generated R script, as well as the sanitized design matrix used.</p></li>
<li><p>Added optional parameters to all differential expression functions, allowing users to return a path to the auto-generated R script and data sanitized design matrix used.</p></li>
</ul>
</section>
<section id="id5">
<h3>Fixed<a class="headerlink" href="#id5" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where some FASTQ/SAM functions could not be added to a FASTQ pipeline.</p></li>
<li><p>Fixed bug where bowtie2 could not be run in/from directories with spaces in their names.</p></li>
<li><p>Fixed bug where RNAlysis would crash when launched without an internet connection.</p></li>
<li><p>Fixed bug that cause ID-mapping functions to raise an error when called from the MacOS stand-alone app (thanks to <a class="reference external" href="https://github.com/Mitchzw">Mitchzw</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/34">#34</a>).</p></li>
<li><p>Fixed bug that caused R package installations (DESeq2, limma, etc) to fail on some computers (thanks to <a class="reference external" href="https://github.com/Celine-075">Celine-075</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/35">#35</a>).</p></li>
<li><p>Fixed bug where Limma-Voom differential expression would fit numerical covariates as categorical factors.</p></li>
<li><p>Fixed bug where generating enrichment bar plots with ylim=‚Äôauto‚Äô would cause bars with 100% depletion (log2FC=-inf) to disappear.</p></li>
<li><p>Fixed bug where defining 10 or more sample groups in the graphical interface would cause the groups to be ordered incorrectly in graphs.</p></li>
<li><p>Fixed bug where the ‚Äòreturn_scaling_factors‚Äô argument would not return the normalization scaling factors on the graphical interface.</p></li>
<li><p>Fixed various visual issues in the graphical interface</p></li>
<li><p>Fixed bug where the ‚Äòfilter_by_row_sum‚Äô function would raise an error when the table contains non-numerical columns</p></li>
<li><p>Fixed bug where running enrichment on an empty gene set would raise an error.</p></li>
<li><p>Fixed bug where RNAlysis would suggest resuming an auto-report from loaded session even when auto-report is turned off.</p></li>
<li><p>Fixed bug where disabling auto-report in the middle of the session would raise errors when trying to create new graphs.</p></li>
<li><p>Fixed bug where generating multiple gene expression plots (split_plots=True) with auto-generated report would only add the last graph to the session report.</p></li>
<li><p>Fixed bug where the function ‚Äònormalize_to_quantile‚Äô generated unclear table names.</p></li>
<li><p>Fixed bug where sometimes the first operation performed in a session would not display correctly in the automatic session report.</p></li>
</ul>
</section>
<section id="new-contributors">
<h3>New Contributors<a class="headerlink" href="#new-contributors" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/Mitchzw">Mitchzw</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/34">#34</a></p></li>
<li><p><a class="reference external" href="https://github.com/Celine-075">Celine-075</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/35">#35</a></p></li>
</ul>
</section>
</section>
<section id="id8">
<h2>3.11.0 (2024-01-05)<a class="headerlink" href="#id8" title="Link to this heading">ÔÉÅ</a></h2>
<p>This release brings several exciting new features.
Notably, these inclue the ability to run functions from the Picardtools suite, and the ability to automatically save interactive session reports and later resume them from any saved session.
In addition, this release includes several visual upgrades, bug fixes, and quality-of-life improvements.
Happy analysis!</p>
<section id="id9">
<h3>Added<a class="headerlink" href="#id9" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><em>RNAlysis</em> can now run functions from the Picardtools suite, including conversion functions (BAM to SAM, SAM to FASTQ, FASTQ to SAM, etc), quality control (validate SAM), and post-processing functions (remove PCR duplicates, sort SAM, create BAM index).</p></li>
<li><p><em>RNAlysis</em> interactive session reports can now be resumed from any saved session, instead of having to start a new report from scratch. When loading a session created by <em>RNAlysis</em> 3.11 and beyond, you will have the option to resume the interactive report from the last saved state.</p></li>
</ul>
</section>
<section id="id10">
<h3>Changed<a class="headerlink" href="#id10" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>CutAdapt adapter trimming functions can now receive an optional ‚Äúnew_filenames‚Äù parameter, which allows users to specify the names of the output files.</p></li>
<li><p>Hierarchical clustergram plot (CountFilter.clustergram) now supports the ‚Äòcolormap‚Äô parameter, which allows users to specify a custom color map for the plot.</p></li>
<li><p>Hierarchical clustergram plot (CountFilter.clustergram) now displays continuous values on the color bar, instead of discrete values.</p></li>
<li><p>Generally improved the looks of Hierarchical clustergram plot (CountFilter.clustergram).</p></li>
<li><p>Previously-added functions (such as ortholog/paralog mapping, biotype summary by GTF file, etc) can now be applied to gene sets. Previously, some of these functions could only be applied to data tables.</p></li>
</ul>
</section>
<section id="id11">
<h3>Fixed<a class="headerlink" href="#id11" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Function ‚ÄúSummarize feature biotypes (based on a reference table)‚Äù (biotypes_from_ref_table) now treats rows with missing values as ‚Äú_missing_from_biotype_reference‚Äù instead of ignoring them entirely.</p></li>
<li><p>Fixed bug where the Ensembl paralog-finding function would appear under the wrong tab in the graphical interface.</p></li>
<li><p>Fixed bug where the description of the MA Plot function and parameters would not display correctly in the graphical interface.</p></li>
</ul>
</section>
</section>
<section id="id12">
<h2>3.10.1 (2023-11-22)<a class="headerlink" href="#id12" title="Link to this heading">ÔÉÅ</a></h2>
<p>Version 3.10.1 introduces several bug fixes, as well as well as support for random effect analysis in Limma-Voom differential expression.</p>
<section id="id13">
<h3>Added<a class="headerlink" href="#id13" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Limma-Voom differential expression can now fit mixed linear models containing a random effect (e.g. nested design).</p></li>
</ul>
</section>
<section id="id14">
<h3>Fixed<a class="headerlink" href="#id14" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where trying to load sessions created with RNAlysis version 3.10.0 would result in an error.</p></li>
<li><p>Fixed bug where using the OrthoInspector ortholog mapping function with database=‚Äôauto‚Äô would sometimes fail to find an appropriate mapping database, even when one exists.</p></li>
<li><p>Fixed bug where kallisto paired-end quantification window would display the ‚Äòread1‚Äô and ‚Äòread2‚Äô parameters twice.</p></li>
<li><p>Fixed bug where empty sub-menus would appear under the FASTQ menu. These sub-menus will be implemented in future versions of RNAlysis.</p></li>
</ul>
</section>
</section>
<section id="id15">
<h2>3.10.0 (2023-10-31)<a class="headerlink" href="#id15" title="Link to this heading">ÔÉÅ</a></h2>
<p>I‚Äôm thrilled to introduce RNAlysis version 3.10.0.
This version includes features that were requested by users for a while, alongside quality-of-life improvements and bug fixes.
Here is a brief highlight of the most important additions:</p>
<p><strong>Ortholog Mapping:</strong> <em>RNAlysis</em> can now map genes to their closest orthologs in different organisms.
You can map genes to their orthologs using four different databases - Ensembl, Panther, PhylomeDB, and OrthoInspector - extracting both one-to-one and one-to-many ortholog relationships and filtering them based on their reliability.</p>
<p><strong>Discovering Paralogs:</strong> In the same vein, <em>RNAlysis</em> now facilitates the discovery of paralogs within a specific organism, using either the Ensembl or Panther databases.</p>
<p><strong>New visualization and analysis options for Principal Component Analysis (PCA):</strong> We‚Äôve introduced new functions and parameters to allow you to get more out of your principal component analysis.</p>
<p>I would also like to extend my personal apology for the delay in bringing you this update.
Due to personal reasons, this release, originally scheduled for the end of August, took longer than expected.
Your patience and support have been invaluable, and I‚Äôm eager to share these exciting additions with you.
Thank you for being a part of the RNAlysis community, and stay tuned for more updates in the near future!</p>
<section id="id16">
<h3>Added<a class="headerlink" href="#id16" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added new functions to the filtering module that map genes to their closest orthologs in a different organism, using four different databases: Ensembl, Panther, PhylomeDB, and OrthoInspector.</p></li>
<li><p>Added new functions to the filtering module that find paralogs of genes in a given organism, using two different databases: Ensembl and Panther.</p></li>
<li><p>Added new function ‚ÄòSort table by contribution to a Principal Component (PCA)‚Äô (CountFilter.sort_by_principal_component), which allows sorting of genes in a count matrix by their contribution (gene loadings) to a principal component.</p></li>
<li><p>Added a new parameter called ‚Äòlegend‚Äô to ‚ÄòPrincipal Component Analysis (PCA) plot‚Äô (CountFilter.pca), which allows users to display a legend on the PCA plot with a name for each sample group/color.</p></li>
</ul>
</section>
<section id="id17">
<h3>Changed<a class="headerlink" href="#id17" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis now issues a warning when users run PCA or PCA-based functions on an unnormalized count matrix.</p></li>
<li><p>The ‚Äòseek_fusion_genes‚Äô and ‚Äòlearn_bias‚Äô arguments for kallisto quantification (fastq.kallisto_quant_single_end and fastq.kallisto_quant_paired_end), which were depracated in kallisto versions &gt;0.48, are no longer displayed on the graphical interface. Old Pipelines that contain these arguments will still run, but new Pipelines will not contain them.</p></li>
<li><p>Long-running functions now run in the background even when the ‚Äòinplace‚Äô parameter is set to True, instead of freezing the entire graphical interface.</p></li>
</ul>
</section>
<section id="id18">
<h3>Fixed<a class="headerlink" href="#id18" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where functions would sometimes fail to run without displaying an error message.</p></li>
<li><p>Fixed bug where progress bars in the graphical interface would sometimes not disappear after reaching 100% completion.</p></li>
<li><p>RNAlysis should no longer display warning messages about graph layout when graphs are scaled down.</p></li>
<li><p>Fixed bug where the clustergram function (CountFilter.clustergram) would raise an error with specific sets of dependecy versions.</p></li>
<li><p>Loading tables no longer raises a depracation warning when using newer versions of Pandas.</p></li>
</ul>
</section>
</section>
<section id="id19">
<h2>3.9.2 (2023-06-23)<a class="headerlink" href="#id19" title="Link to this heading">ÔÉÅ</a></h2>
<p>This patch contains bug fixes and improved functionality for enrichment lollipop plots,
as well as bug fixes for issues with the stand-alone version.</p>
<section id="id20">
<h3>Changed<a class="headerlink" href="#id20" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Single-set enrichment result tables now contain observed/expected values based on the XL-mHG test cutoff.</p></li>
<li><p>When loading a differential expression design matrix, RNAlysis now issues an error if the design matrix column names contain invalid characters.</p></li>
<li><p>Updated the scaling of enrichment lollipop plots to make small ‚Äòobserved‚Äô values easier to discern.</p></li>
</ul>
</section>
<section id="id21">
<h3>Fixed<a class="headerlink" href="#id21" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where an error message would sometimes appear after RNAlysis finishes generating an automatic session report on the stand-alone app.</p></li>
<li><p>Fixed bug where enrichment lollipop plots in horizontal mode would display the observed/expected values in reverse order.</p></li>
<li><p>Fixed bug where enrichment lollipop plots and the ‚Äòshow_exp‚Äô parameter would not work on single-set enrichment data.</p></li>
<li><p>Fixed bug where sometimes the tab label of clustering/differential expression output tables would not match the name of the generated table.</p></li>
</ul>
</section>
</section>
<section id="id22">
<h2>3.9.1 (2023-06-19)<a class="headerlink" href="#id22" title="Link to this heading">ÔÉÅ</a></h2>
<p>Version 3.9.1 of RNAlysis introduces several improvements and fixes to further improve your analysis experience.
The release includes new optional parameters for single-set enrichment functions, compatibility improvements with newer Python versions,
improved error messaging for R scripts, and adresses minor issues related to enrichment analysis, documentation, plotting parameters, and Pipeline saving.</p>
<section id="id23">
<h3>Added<a class="headerlink" href="#id23" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added new optional parameters to single-set enrichment functions, allowing users to determine the top and bottom cutoffs for the XL-mHG test (‚ÄúX‚Äù and ‚ÄúL‚Äù).</p></li>
</ul>
</section>
<section id="id24">
<h3>Changed<a class="headerlink" href="#id24" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis single-set enrichment analysis using the XL-mHG test now supports Python versions &gt;= 3.8.</p></li>
<li><p>RNAlysis stand-alone app is now built on Python 3.11, improving overall performance.</p></li>
<li><p>Error messages caused by running R tools such as DESeq2, Limma-Voom, and FeatureCounts will now clearly state the reason the script failed, making it easier to understand what went wrong.</p></li>
</ul>
</section>
<section id="id25">
<h3>Fixed<a class="headerlink" href="#id25" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where enrichment analysis would raise an error when running enrichment analysis on a gene set with no relevant annotations, or a gene set that does not intersect at all with the background gene set.</p></li>
<li><p>Added missing documentation for plotting parameters in some enrichment functions.</p></li>
<li><p>Depracation Warning should no longer appear when generating a box-plot or enhanced box-plot with scatter=True (CountFilter.box_plot, CountFilter.enhanced_box_plot)</p></li>
<li><p>Fixed bug in featureCounts single-end mode where the ‚Äòoutput_folder‚Äô parameter could appear as disabled.</p></li>
<li><p>Fixed bug where RNAlysis would raise an error message after saving a Pipeline, even when the Pipeline was saved successfully.</p></li>
</ul>
</section>
</section>
<section id="id26">
<h2>3.9.0 (2023-06-09)<a class="headerlink" href="#id26" title="Link to this heading">ÔÉÅ</a></h2>
<p>Version 3.9.0 of <em>RNAlysis</em> introduces several enhancements and fixes to improve your experience.
The release includes additional enrichment plot styles, a new option for PCA plots,
the ability to load and save data tables in Parquet format, and new actions in the Help menu for reporting issues and suggesting improvements.
The update also improves the performance of various functions, ensures consistency in font and theme settings,
and addresses multiple bug fixes, including issues with automatic session reports and visualization functions.</p>
<section id="id27">
<h3>Added<a class="headerlink" href="#id27" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added additional parameters to enrichment bar plots (enrichment.enrichment_bar_plot), including a new plot style (‚Äòlollipop‚Äô) and observed/expected labels on the graph.</p></li>
<li><p>Added a new parameter to Principal Component Analysis plots (CountFilter.pca) ‚Äòplot_grid‚Äô, which can enable or disable adding a grid to PCA plots.</p></li>
<li><p>RNAlysis can now load and save data tables in Parquet format (.parquet)</p></li>
<li><p>Added new actions to the Help menu, allowing users to report issues, suggest issues, or open discussions.</p></li>
</ul>
</section>
<section id="id28">
<h3>Changed<a class="headerlink" href="#id28" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Functions in the FASTQ model are now added to automatic session reports.</p></li>
<li><p>Many of the functions in RNAlysis should now run faster.</p></li>
<li><p>Font type, size, and color for help tooltips should now match the global font settings.</p></li>
<li><p>True/False toggle switches now scale with font size.</p></li>
<li><p>When loading data tables into RNAlysis, you will now see only supported file formats by default.</p></li>
<li><p>Clustering PCA plots are now plotted in proportion to the % variance explained by each PC.</p></li>
<li><p>The legend in clustering PCA plots is now draggable.</p></li>
</ul>
</section>
<section id="id29">
<h3>Fixed<a class="headerlink" href="#id29" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where data tables generated through the FASTQ model would not display properly in automatic session reports.</p></li>
<li><p>Fixed bug where graphs generated through the Visualize Gene Sets window would not be added to automatic analysis report.</p></li>
<li><p>When saving a file through the graphical interface, automatically-suggested filenames no longer contain illegal characters.</p></li>
<li><p>Improved clarity of error message when R installation folder is not found.</p></li>
<li><p>Fixed bug where some input parameter widgets in the RNAlysis graphical interface would not display properly.</p></li>
<li><p>RNAlysis now provides a clearer warning message when attempting to run HDBSCAN clustering, if the hdbscan package is not installed.</p></li>
<li><p>Label text in PCA plots and hierarchical clustergrams should no longer be cropped outside of the visible region of the plot.</p></li>
<li><p>Fixed bug where some visualization functions, such as pair-plot (CountFilter.pairplot) would not display properly due to version mismatches between pandas and seaborn.</p></li>
<li><p>Improved clarity of error messages when external apps‚Äô (kallisto, bowtie2, etc) installation folders are not found.</p></li>
<li><p>Fixed bug where running the RNAlysis graphical interface on a new computer would sometimes raise an error (thanks to <a class="reference external" href="https://github.com/NeuroRookie">NeuroRookie</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/25">#25</a>).</p></li>
<li><p>Fixed a bug where the ‚Äòmin_samples‚Äô parameter in HDBSCAN clustering could not be disabled.</p></li>
<li><p>Fixed a bug where applying a function to a gene set with inplace=False would cause the new gene set to be called ‚ÄòNew Table‚Äô.</p></li>
<li><p>Fixed a bug where RNAlysis would display the message ‚ÄúPipeline saved successfully‚Äù, even when the user cancels the save operation.</p></li>
</ul>
</section>
<section id="id31">
<h3>New Contributors<a class="headerlink" href="#id31" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/NeuroRookie">NeuroRookie</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/25">#25</a></p></li>
</ul>
</section>
</section>
<section id="id32">
<h2>3.8.0 (2023-05-07)<a class="headerlink" href="#id32" title="Link to this heading">ÔÉÅ</a></h2>
<p>Version 3.8.0 of <em>RNAlysis</em> comes with several exciting new features, including the ability to generate interactive analysis reports automatically.
This feature allows users to create an interactive graph of all the datasets they loaded into RNAlysis, the functions applied, and the outputs generated during the analysis.
You can read more about this feature in the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/tutorial.html#create-analysis-report">Tutorial chapter</a> and the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide_gui.html#rnalysis-interactive-analysis-reports">User Guide chapter</a>.</p>
<p>The new release also includes Pipelines for FASTQ functions, the ability to export normalization scaling factors, and other changes to improve the software‚Äôs performance.
RNAlysis now supports Python 3.11, and many functions should now run faster. The software‚Äôs graphic interface has also improved significantly, and users will now see a clearer error message when attempting to load unsupported file formats.
Lastly, the release also fixes several bugs.</p>
<p>Please note that, since Python 3.7 will be reaching end of life as of June 2023, new versions of <em>RNAlysis</em> will no longer support it.</p>
<section id="id33">
<h3>Added<a class="headerlink" href="#id33" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>You can now generate interactive analysis reports automatically using the RNAlysis graphical interface. Read more about this feature <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide_gui.html">here</a>.</p></li>
<li><p>Added Pipelines for the FASTQ module (SingleEndPipeline and PairedEndPipeline), allowing you to apply a series of functions (adapter trimming, alignment, quantification, etc) to a batch of sequence/alignment files.</p></li>
<li><p>Added new parameter ‚Äòreturn_scaling_factors‚Äô to normalization functions, that allows you to access the scaling factors calculated by RNAlysis.</p></li>
<li><p>Added new parameter ‚Äògzip_output‚Äô to CutAdapt adapter trimming (fastq.trim_adapters_single_end and fastq.trim_adapters_paired_end), allowing users to determine whether or not the output files will be gzipped.</p></li>
</ul>
</section>
<section id="id34">
<h3>Changed<a class="headerlink" href="#id34" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis now supports Python 3.11, and no longer supports Python 3.7.</p></li>
<li><p>Many of the functions in RNAlysis should now run faster.</p></li>
<li><p>The RNAlysis graphic interface should now boot up significantly faster.</p></li>
<li><p>RNAlysis now shows an easier to understand error message when users attempt to load a table in an unsupported format (e.g. Excel files).</p></li>
<li><p>CutAdapt adapter trimming (fastq.trim_adapters_single_end and fastq.trim_adapters_paired_end) now outputs non-gzipped files by default.</p></li>
<li><p>Standardized all plotting functions in the filtering module to return a matplotlib Figure object, which can be further modified by users.</p></li>
</ul>
</section>
<section id="id35">
<h3>Fixed<a class="headerlink" href="#id35" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis failing to map gene IDs during GO enrichment analysis should no longer raise an error (thanks to <a class="reference external" href="https://github.com/clockgene">clockgene</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/16">#16</a>).</p></li>
<li><p>Fixed bug where the Command History window would not display history of the current tab immediately after clearing the current session.</p></li>
<li><p>Fixed bug where adapter trimming would fail to run when using CutAdapt version &gt;= 4.4.0.</p></li>
<li><p>Fixed bug where ‚ÄòFilter rows with duplicate names/IDs‚Äô (Filter.filter_duplicate_ids) would raise an error when applied to some tables.</p></li>
</ul>
</section>
<section id="id37">
<h3>New Contributors<a class="headerlink" href="#id37" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/clockgene">clockgene</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/16">#16</a></p></li>
</ul>
</section>
</section>
<section id="id38">
<h2>3.7.0 (2023-04-07)<a class="headerlink" href="#id38" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces small RNA read alignment using ShortStack, new filtering functions, a new optional parameter for Principal Component Analysis, improvements to the graphical interface, and bug fixes.</p>
<section id="id39">
<h3>Added<a class="headerlink" href="#id39" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added small RNA read alignment using ShortStack (fastq.shortstack_align_smallrna).</p></li>
<li><p>Added new filtering function ‚ÄòFilter specific rows by name‚Äô (Filter.filter_by_row_name).</p></li>
<li><p>Added new filtering function ‚ÄòFilter rows with duplicate names/IDs‚Äô (Filter.filter_duplicate_ids).</p></li>
<li><p>Function parameters in pop-up windows in the graphical interface can now be imported and exported.</p></li>
<li><p>Added new parameter to Principal Component Analysis (CountFilter.pca) ‚Äòproportional_axes‚Äô, that allows you to make the PCA projection axes proportional to the percentage of variance each PC explains.</p></li>
<li><p>Improved clarity of error messages in the graphical interface.</p></li>
</ul>
</section>
<section id="id40">
<h3>Changed<a class="headerlink" href="#id40" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Tables loaded into RNAlysis that use integer-indices will now be converted to use string-indices.</p></li>
<li><p>Refactored CountFilter.from_folder into CountFilter.from_folder_htseqcount, and added a new CountFilter.from_folder method that accepts a folder of count files in any format.</p></li>
<li><p>In the RNAlysis graphical interface, optional parameters that can be disabled will now display the hint ‚Äúdisable this parammeter?‚Äù instead of ‚Äúdisable input?‚Äù.</p></li>
<li><p>Added optional parameter ‚Äòylim‚Äô to ‚Äòcreate enrichment bar-plot‚Äô function (enrichment.enrichment_bar_plot), allowing users to determine the Y-axis limits of the bar plot.</p></li>
<li><p>Updated function signatures of ‚ÄòVisualize gene ontology‚Äô and ‚ÄòVisualize KEGG pathway‚Äô (enrichment.gene_ontology_graph and enrichment.kegg_pathway_graph) to make more sense.</p></li>
<li><p>Removed parameter ‚Äòreport_read_assignment_path‚Äô from featureCounts feature counting (fastq.featurecounts_single_end and fastq.featurecounts_paired_end).</p></li>
<li><p>The RNAlysis graphical interface should now load more quickly.</p></li>
<li><p>Progress bars in the graphical interface should now reflect elapsed/remaining time for tasks more accurately.</p></li>
</ul>
</section>
<section id="id41">
<h3>Fixed<a class="headerlink" href="#id41" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug in the function ‚ÄòSplit into Higly and Lowly expressed genes‚Äô (Filter.split_by_reads) where the two resulting tables would be named incorrectly (highly-expressed genes would be labeled ‚ÄòbelowXreads‚Äô and vice-versa).</p></li>
<li><p>Fixed bug where the ‚Äòcolumn‚Äô parameter of some functions (‚ÄòFilter by percentile‚Äô, ‚ÄòSplit by percentile‚Äô, ‚ÄòFilter with a number filter‚Äô, ‚ÄòFilter with a text filter‚Äô) would not automatically detect column names in the graphical interface.</p></li>
<li><p>Fixed bug where the ‚Äònumerator‚Äô and ‚Äòdenominator‚Äô parameters of of the function ‚ÄòCalculate fold change‚Äô would not automatically detect column names in the graphical interface.</p></li>
<li><p>Fixed bug where tables with integer-indices could not be visualized properly through the graphical interface.</p></li>
<li><p>Fixed bug in the function ‚ÄòfeatureCounts single-end‚Äô (fastq.featurecounts_single_end) where setting parameter ‚Äòreport_read_assignment‚Äô to any value other than None would raise an error.</p></li>
<li><p>Functions that take column name/names as input (transform, filter_missing_values, filter_percentile, split_percentile) can now be applied to fold change tables (FoldCangeFilter objects).</p></li>
<li><p>Fixed bug where the description for the ‚Äòn_bars‚Äô parameter of the ‚Äòcreate enrichment bar-plot‚Äô function (enrichment.enrichment_bar_plot) would not display properly.</p></li>
<li><p>‚ÄòVisualize gene ontology‚Äô and ‚ÄòVisualize KEGG pathway‚Äô (enrichment.gene_ontology_graph and enrichment.kegg_pathway_graph) now have proper parameter descriptions.</p></li>
<li><p>Fixed bug where in-place intersection and difference in the filtering module would fail when using recent versions of pandas.</p></li>
<li><p>Fixed bug where graphs generated through the Visualize Gene Sets window would not immediately display when using the RNAlysis stand-alone app.</p></li>
</ul>
</section>
</section>
<section id="id42">
<h2>3.6.2 (2023-03-25)<a class="headerlink" href="#id42" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces quality-of-life changes to the graphical interface, as well as bug fixes.</p>
<section id="id43">
<h3>Added<a class="headerlink" href="#id43" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Sample groupings in functions such as PCA, Pair-plot, etc., can now be imported and exported through the graphical interface.</p></li>
</ul>
</section>
<section id="id44">
<h3>Fixed<a class="headerlink" href="#id44" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where the stand-alone Mac version of RNAlysis would sometimes fail to map gene IDs (directly or in enrichment analysis).</p></li>
</ul>
</section>
</section>
<section id="id45">
<h2>3.6.1 (2023-03-22)<a class="headerlink" href="#id45" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces minor bug fixes.</p>
<section id="id46">
<h3>Changed<a class="headerlink" href="#id46" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>DESeq2 automatic installation should now work more reliably.</p></li>
</ul>
</section>
<section id="id47">
<h3>Fixed<a class="headerlink" href="#id47" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where PCA plots would not display on the stand-alone app unless another visualization function was applied afterwards.</p></li>
<li><p>Fixed bug where Pipelines that contain specific functions (such as translating gene IDs/filtering biotypes from GTF file) would fail to run through the graphical interface.</p></li>
<li><p>GO Annotations annotated by ComplexPortal are now supported by RNAlysis.</p></li>
</ul>
</section>
</section>
<section id="id48">
<h2>3.6.0 (2023-03-07)<a class="headerlink" href="#id48" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces improvements to the usability and clarity of the graphic interface,
new methods for automatic estimation of the number of clusters in a dataset,
and various bug fixes.</p>
<section id="id49">
<h3>Added<a class="headerlink" href="#id49" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added three new methods for automatic estimation of the number of clusters in a dataset: Callinski-Harabasz, Davies-Bouldin, and Bayesian Information Criterieon.</p></li>
<li><p>Added a ‚ÄòClose all Figures‚Äô actions to the ‚ÄòView‚Äô menu of the <em>RNAlysis</em> graphic interface.</p></li>
<li><p>Added an ‚Äòinteractive‚Äô parameter to Volcano Plots (DESeqFilter.volcano_plot) and ‚ÄòScatter Sample Vs Sample‚Äô (CountFilter.scatter_sample_vs_sample), allowing user to label data points interactively by clicking on them.</p></li>
<li><p>Added more optional plotting parameters to Volcano Plots (DESeqFilter.volcano_plot) and ‚ÄòScatter Sample Vs Sample‚Äô (CountFilter.scatter_sample_vs_sample).</p></li>
</ul>
</section>
<section id="id50">
<h3>Changed<a class="headerlink" href="#id50" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Progress bars are now integrated into the main <em>RNAlysis</em> window instead of opening as a dialog box.</p></li>
<li><p>Information about running proccesses and functions is now displayed in the main <em>RNAlysis</em> window.</p></li>
<li><p>It is now possible to cancel queued jobs through the <em>RNAlysis</em> graphic interface.</p></li>
<li><p>When loading multiple data tables at the same time, it is now possible to change the table type of all data tables at once, instead of one-by-one.</p></li>
</ul>
</section>
<section id="id51">
<h3>Fixed<a class="headerlink" href="#id51" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis KEGG enrichment should now match the new KEGG annotation format from March 1st 2023.</p></li>
<li><p>Fixed bug where importing <em>RNAlysis</em> would raise ImportError when cutadapt is not installed.</p></li>
<li><p>Fixed bug where the ‚ÄòRun‚Äô button in the Enrichment Analysis window would grey out whenever the enrichment dataset is changed.</p></li>
<li><p>Fixed bug where the <em>RNAlysis</em> stand-alone versions were unable to export Figures in specific formats (e.g. PDF, SVG).</p></li>
<li><p>Fixed bug where functions that depend on R scripts (such as DESeq2 and limma) would sometimes fail to run on MacOS (thanks to Matthias Wilm and <a class="reference external" href="https://github.com/sandyl27">sandyl27</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/12">#12</a>).</p></li>
<li><p>Fixed bug where running limma-voom with a design matrix whose column names contained spaces or special characterse would raise an error.</p></li>
<li><p>Fixed bug where the ‚Äòhighlight‚Äô parameter of CountFilter.scatter_sample_vs_sample would not work when used through the graphic interface.</p></li>
<li><p>Fixed bug where enrichment analysis would sometimes fail to run when ‚Äòexclude_unannotated_genes‚Äô is set to False.</p></li>
<li><p>Fixed bug where translate_gene_ids() would fail for RankedSet objects.</p></li>
<li><p>Fixed bug where filtering gene sets by user-defined attributes (FeatureSet.filter_by_attribute()) would occasionally fail to run.</p></li>
</ul>
</section>
<section id="id53">
<h3>New Contributors<a class="headerlink" href="#id53" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/sandyl27">sandyl27</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/12">#12</a></p></li>
</ul>
</section>
</section>
<section id="id54">
<h2>3.5.2 (2023-02-23)<a class="headerlink" href="#id54" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version includes bug fixes for a few quality-of-life issues which were introduced in version RNAlysis 3.5.0.</p>
<section id="id55">
<h3>Changed<a class="headerlink" href="#id55" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>It is now possible to change the parallel backend of performance-intensive functions such as clustering an enrichment analysis in non-standalone versions of RNAlysis.</p></li>
<li><p>Expanded the Frequently Asked Questions page.</p></li>
<li><p>Added Perl as a dependency for Windows users on the How to Install page.</p></li>
<li><p>Automatic row colours in column-picking tables should no longer mismatch with font colors in a way that obscures visibility.</p></li>
</ul>
</section>
<section id="id56">
<h3>Fixed<a class="headerlink" href="#id56" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where occasionally newly-created tabs would open twice instead of once.</p></li>
<li><p>Fixed bug where the ‚ÄòAdd Function‚Äô button of the Pipeline window would appear in the wrong location.</p></li>
<li><p>Fixed bug where RNAlysis sessions saved after version 3.5.0 which contain gene sets would raise an error when loaded.</p></li>
<li><p>Fixed typos in the RNAlysis tutorial.</p></li>
</ul>
</section>
</section>
<section id="id57">
<h2>3.5.1 (2023-02-22)<a class="headerlink" href="#id57" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces minor bug fixes.</p>
<section id="id58">
<h3>Fixed<a class="headerlink" href="#id58" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where the <em>RNAlysis</em> stand-alone app would sometimes fail to run CutAdapt (thanks to Matthias Wilm).</p></li>
<li><p>Fixed bug where the User Guide action in the graphical interface would point to the Tutorial, and vice versa.</p></li>
<li><p>The X and Y axis labels on volcano plots should now format the ‚Äòlog‚Äô in the label correctly.</p></li>
</ul>
</section>
</section>
<section id="id59">
<h2>3.5.0 (2023-02-08)<a class="headerlink" href="#id59" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces new features such as differential expression analysis with the Limma-Voom pipeline,
customizable databases for the quick-search function, basic filtering and procrssing functions for gene sets,
improved progammatic API for FeatureSet and RankedSet objects, and RPKM and TPM read count normalization.
Several changes have been made to improve the user experience, including updated documentation,
improved clarity of function tooltips, clearer output formats, and faster download speeds for tutorial videos.</p>
<section id="id60">
<h3>Added<a class="headerlink" href="#id60" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added differential expression analysis with the Limma-Voom pipelines (CountFilter.differential_expression_limma_voom)</p></li>
<li><p>You can now select which databases to display in the right-click quick-search menu, using the settings menu.</p></li>
<li><p>Gene sets now support some basic operations (filtering, gene ID translating, etc) through the graphical interface.</p></li>
<li><p>enrichment.FeatureSet and enrichment.RankedSet now support some filtering operations from the filtering module (such as filtering by user-defined attributes, GO terms, or KEGG pathways).</p></li>
<li><p>Added reads-per-kilobase-million (RPKM) and transcripts-per-million (TPM) normalization methods (CountFilter.normalize_to_rpkm() and CountFilter.normalize_to_tpm()).</p></li>
</ul>
</section>
<section id="id61">
<h3>Changed<a class="headerlink" href="#id61" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Classes enrichment.FeatureSet and enrichment.RankedSet now inherit from Python base-class set, and can be interacted with like other Python sets. The old API and attributes of these classes were maintained as they were.</p></li>
<li><p>Improved documentation for some functions.</p></li>
<li><p>Function selection tooltips should now display information more clearly.</p></li>
<li><p>Pipelines that contain consecutive clustering/splitting functions will now return their outputs in a clearer format.</p></li>
<li><p>Enrichment bar-plots should now adjust the x-axis limits more tightly to fit the displayed data.</p></li>
<li><p>Improved clarity of automatic titles in enrichment plots.</p></li>
<li><p>Download/update speed of tutorial videos has improved significantly.</p></li>
</ul>
</section>
<section id="id62">
<h3>Fixed<a class="headerlink" href="#id62" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where Pipelines would not always properly run ‚Äòtranslate_gene_ids‚Äô</p></li>
</ul>
</section>
</section>
<section id="id63">
<h2>3.4.2 (2023-02-01)<a class="headerlink" href="#id63" title="Link to this heading">ÔÉÅ</a></h2>
<p>This version introduces minor bug fixes.</p>
<section id="id64">
<h3>Fixed<a class="headerlink" href="#id64" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug where updating RNAlysis through the graphical interface would not update some of the optional dependencies.</p></li>
<li><p>Fixed typos in the documentation.</p></li>
</ul>
</section>
</section>
<section id="id65">
<h2>3.4.0 (2023-02-01)<a class="headerlink" href="#id65" title="Link to this heading">ÔÉÅ</a></h2>
<p>From this release forward, <em>RNAlysis</em> is made available as a stand-alone app for Windows and MacOS. You can download these stand-alone versions from the GitHub Releases page.
In addition, new features were added, including new plots, filtering functions, integration of the external tools bowtie2 and featureCounts, and the ability to generate Gene Ontology Graphs and KEGG Pathway Graphs without running enrichment analysis from scratch.</p>
<section id="id66">
<h3>Added<a class="headerlink" href="#id66" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added a Scree Plot (explained variance per PC) to Principle Component Analysis</p></li>
<li><p>Added CountFilter.split_by_principal_component(), allowing users to filter genes based on their contributions (loadings) to PCA Principal Components.</p></li>
<li><p>Added Filter.drop_columns</p></li>
<li><p>Added support for the Sharpened Cosine distance metric in clustering analyses</p></li>
<li><p>KEGG enrichment can now generate KEGG pathway graphs for pathways that were found to be statistically significant</p></li>
<li><p>Added functions to the enrichment module that can generate KEGG Pathway or Gene Ontology plots based on previously-generated enrichment results</p></li>
<li><p>You can now clear the <em>RNAlysis</em> cache from the <em>RNAlysis</em> GUI, or through the general module.</p></li>
<li><p>Added bowtie2 alignment to the fastq module.</p></li>
<li><p>Added FeatureCounts feature-counting to the fastq module.</p></li>
<li><p>You can now choose whether or not to discard genes from enrichment analysis if they have no annotations associated with them.</p></li>
<li><p>When right-clicking on a specific cell in a table or line in a gene set view, a context menu will open, allowing you to copy the associated value, or look it up in one of few common biology databases.</p></li>
<li><p>Added sections to the programmatic user guide about the <cite>fastq</cite> module.</p></li>
</ul>
</section>
<section id="id67">
<h3>Changed<a class="headerlink" href="#id67" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Replaced the ‚Äòparallel‚Äô parameter in enrichment functions with the ‚Äòparallel_backend‚Äô parameter, allowing users to choose which parallel backend (if any) will be used in the function.</p></li>
<li><p>Added ‚Äòparallel_backend‚Äô parameter to all clustering functions under the filtering module.</p></li>
<li><p>When generating Gene Ontology/KEGG Pathway graphs, users can choose whether or not to generate the figure in an additional separate file.</p></li>
<li><p>Updated type annotations of some functions to be more precise and helpful (for example, setting a lower bound on some int function parameters).</p></li>
<li><p>The colorbar ticks in enrichment bar plots now match the axis ticks on the main axis.</p></li>
<li><p>Slight improvements in GUI performance, stability, and looks.</p></li>
<li><p>Slight improvements in performance of enrichment analysis when examining a small number of attributes.</p></li>
<li><p>enrichment.plot_enrichment() was replaced by enrichment.enrichment_bar_plot().</p></li>
<li><p>CountFilter.differential_expression() has new optional parameter <cite>output_folder</cite>, which allows users to save the generated data tables and the R script that generated them into a specified folder.</p></li>
</ul>
</section>
<section id="id68">
<h3>Fixed<a class="headerlink" href="#id68" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>In CountFilter.differential_expression_deseq2(), fixed a bug where design matrix files with non-comma delimiters would cause an error (thanks to <a class="reference external" href="https://github.com/Mintxoklet">Mintxoklet</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/7">#7</a>)</p></li>
<li><p>Fixed bug where setup.py would install a directory named tests into site-packages folder (thanks to <a class="reference external" href="https://github.com/kbipinkumar">Bipin Kumar</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/9">#9</a>)</p></li>
<li><p>Fixed bug where the windows of some functions (differential expression, adapter trimming, etc) did not show a link to the function‚Äôs documentation page.</p></li>
<li><p>Fixed typos in some parts of the <em>RNAlysis</em> documentation</p></li>
<li><p>When filtering a table by a single user-defined attribute, the automatic table name will now be more informative about the operation applied.</p></li>
<li><p>Fixed bug where occasionally a Pipeline or Function would generate multiple tables of the same name, but only one of them will appear in the GUI.</p></li>
<li><p>Fixed bug where occasionally significance asterisks on enrichment bar-plots would appear in the wrong location.</p></li>
<li><p>Fixed bug where fastq.create_kallisto_index() (Create Kallisto Index) would not make use of the <cite>make_unique</cite> parameter (thanks to Matthias Wilm)</p></li>
</ul>
</section>
<section id="removed">
<h3>Removed<a class="headerlink" href="#removed" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Removed the previously-deprecated functions <cite>enrichment.enrich_randomization()</cite> and <cite>enrichment.enrich_hypergeometric()</cite>.</p></li>
</ul>
</section>
<section id="id71">
<h3>New Contributors<a class="headerlink" href="#id71" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><a class="reference external" href="https://github.com/Mintxoklet">Mintxoklet</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/7">#7</a></p></li>
<li><p><a class="reference external" href="https://github.com/kbipinkumar">Bipin Kumar</a> in <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues/9">#9</a></p></li>
<li><p>Matthias Wilm</p></li>
</ul>
</section>
</section>
<section id="id72">
<h2>3.3.0 (2022-12-02)<a class="headerlink" href="#id72" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduced quality-of-life improvements to the graphical user interface.</p></li>
</ul>
<section id="id73">
<h3>Added<a class="headerlink" href="#id73" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added a Frequently Asked Questions page, and linked all <em>RNAlysis</em> help material inside the graphical interface Help menu.</p></li>
<li><p>Pipelines can now be edited and deleted through the Pipeline menu of the graphical interface.</p></li>
<li><p>Added Contributing page to the documentation, with basic guidelines on how you can contribute to the <em>RNAlysis</em> project!</p></li>
</ul>
</section>
<section id="id74">
<h3>Changed<a class="headerlink" href="#id74" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>All open tabs are now always visible in the main menu screen. Tab names are now shortened with ellipsis if nessecary.</p></li>
<li><p>The right-click context menu of the main menu tabs now allows users to open a new tab at a specific position, or close a specific tab/tabs to the right/tabs to the left/all other tabs.</p></li>
<li><p><em>RNAlysis</em> documentation is now split into GUI documentation (quick-start video guide, tutorial, GUI user guide), and programmatic documentation (programmatic user guide)</p></li>
<li><p>Improved readability of <em>RNAlysis</em> logs</p></li>
<li><p>Pipelines are now exported with additional metadata - the version of <em>RNAlysis</em> they were exported from, and the date and time it was exported. This metadata should not affect Pipelines that were created in older versions, and does not affect the way Pipelines are applied to data tables.</p></li>
</ul>
</section>
<section id="id75">
<h3>Fixed<a class="headerlink" href="#id75" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p><em>RNAlysis</em> now warns users if they attempt to overwrite an existing Pipeline.</p></li>
<li><p>Fixed an incorrect keyboard shortcut for Export Pipeline action</p></li>
</ul>
</section>
</section>
<section id="id76">
<h2>3.2.2 (2022-11-25)<a class="headerlink" href="#id76" title="Link to this heading">ÔÉÅ</a></h2>
<section id="id77">
<h3>Fixed<a class="headerlink" href="#id77" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug with DESeq2 automatic installation on Windows computers.</p></li>
</ul>
</section>
</section>
<section id="id78">
<h2>3.2.1 (2022-11-25)<a class="headerlink" href="#id78" title="Link to this heading">ÔÉÅ</a></h2>
<section id="id79">
<h3>Changed<a class="headerlink" href="#id79" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Updated citation information for <em>RNAlysis</em></p></li>
</ul>
</section>
<section id="id80">
<h3>Fixed<a class="headerlink" href="#id80" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed typos in the <em>RNAlysis</em> tutorial</p></li>
</ul>
</section>
</section>
<section id="id81">
<h2>3.2.0 (2022-11-23)<a class="headerlink" href="#id81" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces quality-of-life changes to the graphical user interface, functions for translating gene IDs and running differential expression analysis, and extends RNAlysis to support Python versions 3.9 and 3.10.</p></li>
</ul>
<section id="id82">
<h3>Added<a class="headerlink" href="#id82" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added Filter.translate_gene_ids()</p></li>
<li><p>Added CountFilter.differential_expression_deseq2()</p></li>
<li><p>Added Filter.filter_by_kegg_annotations()</p></li>
<li><p>Added Filter.filter_by_go_annotations()</p></li>
<li><p>Added CountFilter.average_replicate_samples()</p></li>
<li><p>Added fastq module that contains adapter-trimming functions utilizing CutAdapt, and mRNA-sequencing quantification using kallisto.</p></li>
</ul>
</section>
<section id="id83">
<h3>Changed<a class="headerlink" href="#id83" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added additional plotting parameters to visualization functions.</p></li>
<li><p>Improved performance of some aspects of the graphical user interface.</p></li>
<li><p>RNAlysis‚Äô basic features are now supported on Python versions 3.9 and 3.10.</p></li>
<li><p>CountFilter.pca() now generates a plot for <em>every</em> pair of Principal Components requested by the user.</p></li>
<li><p>CountFilter.split_clicom() now supports clustering each batch of replicates separately, using the ‚Äòreplicates_grouping‚Äô parameter</p></li>
<li><p>Biotype-based filtering and summary can now be done based on GTF annotation files instead of a Biotype Reference Table.</p></li>
<li><p>Filter.biotypes() was refactored into Filter.biotypes_from_ref_table()</p></li>
<li><p>Filter.filter_biotype() was refactored into Filter.filter_biotype_from_ref_table()</p></li>
</ul>
</section>
<section id="id84">
<h3>Fixed<a class="headerlink" href="#id84" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Users can now queue multiple computationally-intense enrichment/clustering tasks while another task is running.</p></li>
<li><p>Fixed a bug where sometimes some function parameters would disappear from the graphical user interface.</p></li>
<li><p>Fixed a bug where exceptions during computationally-intense tasks would cause <em>RNAlysis</em> to crash.</p></li>
<li><p>Auxillary windows are now properly minimized when analysis starts, and restored when analysis ends or encounters an error.</p></li>
</ul>
</section>
</section>
<section id="id85">
<h2>3.1.0 (2022-10-16)<a class="headerlink" href="#id85" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces new count matrix normalization methods, as well as MA plots and minor bug fixes.</p></li>
</ul>
<section id="id86">
<h3>Added<a class="headerlink" href="#id86" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added the visualization function ma_plot() for CountFilter</p></li>
<li><p>Added functions for the normalization functions Relative Log Ratio (RLE), Trimmed Mean of M-values (TMM), Median of Ratios (MRN), Quantile normalization (quantile)</p></li>
</ul>
</section>
<section id="id87">
<h3>Changed<a class="headerlink" href="#id87" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>CountFilter.normalize_to_rpm() was renamed to CountFilter.normalize_to_rpm_htseqcount(), and was supplemented by the more general function for normalizing to Reads Per Million CountFilter.normalize_to_rpm()</p></li>
</ul>
</section>
<section id="id88">
<h3>Fixed<a class="headerlink" href="#id88" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed a bug where some elements of the graphical user interface would not display correctly</p></li>
</ul>
</section>
</section>
<section id="id89">
<h2>3.0.1 (2022-10-12)<a class="headerlink" href="#id89" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version fixes a bug with displaying the tutorial videos in the graphical user interface.</p></li>
</ul>
</section>
<section id="id90">
<h2>3.0.0 (2022-10-10)<a class="headerlink" href="#id90" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces a graphical user interface for RNAlysis, as well as new functions for KEGG Pathways enrichment analysis.</p></li>
</ul>
<section id="id91">
<h3>Added<a class="headerlink" href="#id91" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>RNAlysis now includes a graphical user interface</p></li>
<li><p>Pipelines can now be imported and exported</p></li>
<li><p>Enrichment and single-set-enrichment for KEGG pathway data</p></li>
</ul>
</section>
<section id="id92">
<h3>Changed<a class="headerlink" href="#id92" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added function FeatureSet.user_defined_enrichment(), which will replace FeatureSet.enrich_hypergeometric() and FeatureSet.enrich_randomization()</p></li>
<li><p>Updated signature of enrichment.venn_diagram</p></li>
<li><p>enrichment.venn_diagram and enrichment.upset_plot can now be generated on a user-supplied FIgure</p></li>
<li><p>Clustering functions now apply a power transform to count data prior to clustering by default</p></li>
<li><p>Non-deprecated enrichment functions no longer filter the background set by biotype by default</p></li>
<li><p>Changed signature of CountFilter.pca, CountFilter.box_plot, CountFilter.enhanced_box_plot, CountFilter.clustergram, and CountFilter.pairplot to ensure consistency among visualization functions.</p></li>
</ul>
</section>
<section id="id93">
<h3>Fixed<a class="headerlink" href="#id93" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>enrichment.venn_diagram can now be plotted with outlines when the circles are unweighted</p></li>
<li><p>Fixed bug in Pipeline.apply_to() where a Filter object would be returned even when the Pipeline was applied inplace</p></li>
</ul>
</section>
</section>
<section id="id94">
<h2>2.1.1 (2022-07-05)<a class="headerlink" href="#id94" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version fixes issues with running GO enrichment that resulted from recent changes to UniProt‚Äôs API.  Moreover, this version slightly improves the performance of some functions.</p></li>
</ul>
<section id="id95">
<h3>Changed<a class="headerlink" href="#id95" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed issues with running GO enrichment that resulted from changes to UniProt‚Äôs API.</p></li>
<li><p>Some functions that fetch annotations now cache their results, leading to improved runtimes.</p></li>
<li><p>Updated the documentation of some functions to better reflect their usage and input parameters.</p></li>
</ul>
</section>
</section>
<section id="id96">
<h2>2.1.0 (2022-04-16)<a class="headerlink" href="#id96" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces multiple new features, as well as generally improved graphs and quality-of-life changes.</p></li>
</ul>
<section id="id97">
<h3>Added<a class="headerlink" href="#id97" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>GO enrichment can now generate Ontology Graphs for the statistically significant GO terms.</p></li>
<li><p>Added CountFilter.split_clicom(), an implementation of the CLICOM ensemble-based clustering method (Mimaroglu and Yagci 2012).</p></li>
<li><p>Added Filter.transform(), a method that can transform your data tables with either predefined or user-defined transformations.</p></li>
</ul>
</section>
<section id="id98">
<h3>Changed<a class="headerlink" href="#id98" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>CountFilter.pairplot() now uses a logarithmic scale by default.</p></li>
<li><p>Visually improved the graphs generated by many functions, including CountFilter.pairplot() and CountFilter.plot_expression().</p></li>
<li><p>The clusters resulting from all clustering functions are now sorted by size instead of being sorted randomly.</p></li>
</ul>
</section>
<section id="id99">
<h3>Fixed<a class="headerlink" href="#id99" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Minor bug fixes.</p></li>
</ul>
</section>
</section>
<section id="id100">
<h2>2.0.1 (2022-04-02)<a class="headerlink" href="#id100" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces small bug fixes, as well as a new function in the Filtering module.</p></li>
</ul>
<section id="id101">
<h3>Added<a class="headerlink" href="#id101" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added Filter.majority_vote_intersection(), which returns a set/string of the features that appear in at least (majority_threhold * 100)% of the given Filter objects/sets.</p></li>
</ul>
</section>
<section id="id102">
<h3>Changed<a class="headerlink" href="#id102" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>When mapping/inferring taxon IDs during GO enrichment analysis, organisms will now be prioritized based on their taxon ID values (numerically lower IDs will be considered to be more relevant).</p></li>
</ul>
</section>
<section id="id103">
<h3>Fixed<a class="headerlink" href="#id103" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed bug that occured when mapping/inferring taxon IDs during GO enrichment analysis, where integer taxon IDs would be matched by name similarity before trying an exact ID match, leading to spurious matches.</p></li>
<li><p>Fixed bug that occursed when plotting clustering results with style=‚Äôall‚Äô on Python 3.8.</p></li>
</ul>
</section>
</section>
<section id="id104">
<h2>2.0.0 (2021-12-05)<a class="headerlink" href="#id104" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces new method to cluster your read count matrices, including K-Means/Medoids clustering, Hierarchical clustering, and HDBSCAN.</p></li>
<li><p>This version introduces many new ways to perform enrichment analysis and to visualize your results, including highly customizable GO Enrichment, enrichment based on ranked lists of genes, and enrichment for non-categorical attributes.</p></li>
<li><p>This version introduces Pipelines - a quicker and more convenient way to apply a particular analysis pipeline to multiple Filter objects.</p></li>
<li><p>This version improves the performance of many functions in RNAlysis, and in particular the performance of randomization tests.</p></li>
<li><p>This version includes changes to names and signatures of some functions in the module, as elaborated below.</p></li>
</ul>
<section id="id105">
<h3>Added<a class="headerlink" href="#id105" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added class Pipeline to filtering module, which applies a series of filter functions to specified Filter objects.</p></li>
<li><p>Added CountFilter.split_kmeans(), CountFilter.split_kmedoids(), CountFilter.split_hierarchical() and CountFilter.split_hdbscan(), which split your read count matrices into clusters with similar expression patterns.</p></li>
<li><p>Added class RankedSet to enrichment module, which accepts a ranked list of genes/features, and can perform single-list enrichment analysis</p></li>
<li><p>Added RankedSet.single_set_enrichment(), which can perfofm single-list enrichment analysis of user-defined attributes using XL-mHG test (see <a class="reference external" href="https://dx.doi.org/10.1371/journal.pcbi.0030039">Eden et al. (PLoS Comput Biol, 2007)</a>  and <a class="reference external" href="https://dx.doi.org/10.1371/journal.pone.0143196">Wagner (PLoS One, 2015)</a> ).</p></li>
<li><p>Added FeatureSet.go_enrichment() and RankedSet.single_set_go_enrichment(), which let you compute Gene Ontology enrichment for any organism of your choice, and filter the GO annotations used according to your preferences.</p></li>
<li><p>Added FeatureSet.enrich_hypergeometric(), which can perform enrichment analysis using the Hypergeometric Test.</p></li>
<li><p>Added more visualization functions, such CountFilter.enhanced_box_plot().</p></li>
<li><p>Added FeatureSet.change_set_name(), to give a new ‚Äòset_name‚Äô to a FeatureSet object.</p></li>
</ul>
</section>
<section id="id106">
<h3>Changed<a class="headerlink" href="#id106" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>FeatureSet.enrich_randomization_parallel() was deprecated. Instead, you can compute your enrichment analysis with parallel computing by calling FeatureSet.enrich_randomization() with the argument ‚Äòparallel_processing=True‚Äô. Moreover, parallel session will now start automatically if one was not already active.</p></li>
<li><p>Improved running time of enrich_randomization() about six-fold.</p></li>
<li><p>Filter objects can be created from any delimiter-separated file format (.csv, .tsv, .txt, etc).</p></li>
<li><p>CountFilter.pca() can now be plotted without labeled points.</p></li>
<li><p>Filter.index_string is now sorted by the current order of indices in the Filter object, instead of by alphabetical order.</p></li>
<li><p>CountFilter.violin_plot() now accepts a y_title argument.</p></li>
<li><p>Added more optional arguments to visualization functions such as CountFilter.violin_plot() and CountFilter.clustergram().</p></li>
<li><p>Automatic filenames for Filter objects should now reflect more clearly the operations that were performed.</p></li>
<li><p>The DataFrame returned by enrich_randomization() and enrich_randomization_parallel() now contains the additional column ‚Äòdata_scale‚Äô, determined by the new optional argument ‚Äòdata_scale‚Äô.</p></li>
<li><p>The columns ‚Äòn obs‚Äô and ‚Äòn exp‚Äô in the DataFrame returned by enrich_randomization() and enrich_randomization_parallel() were renamed to ‚Äòobs‚Äô and ‚Äòexp‚Äô respectively.</p></li>
<li><p>FeatureSets no longer support in-place set operations (intersection, union, difference, symmetric difference). Instead, these functions return a new FeatureSet.</p></li>
<li><p>Filter.biotypes_from_ref_table() now accepts the boolean parameter ‚Äòlong_format‚Äô instead of the str parameter ‚Äòformat‚Äô.</p></li>
<li><p>Filter.biotypes_from_ref_table() and FeatureSet.biotypes_from_ref_table() now count features which do not appear in the Biotype Reference Table as ‚Äò_missing_from_biotype_reference‚Äô instead of ‚Äònot_in_biotype_reference‚Äô.</p></li>
</ul>
</section>
<section id="id107">
<h3>Fixed<a class="headerlink" href="#id107" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Updated type-hinting of specific functions.</p></li>
<li><p>Filter.biotypes_from_ref_table() and FeatureSet.biotypes_from_ref_table() now support Biotype Reference Tables with different column names.</p></li>
<li><p>Generally improved performance of RNAlysis.</p></li>
<li><p>Fixed bug in Filter.filter_percentile() where the value at the exact percentile speficied (e.g. the median for percentile=0.5) would be removed from the Filter object.</p></li>
<li><p>Fixed bug in enrichment.FeatureSet, where creating a FeatureSet from input string would result in an empty set.</p></li>
<li><p>Various minor bug fixes.</p></li>
</ul>
</section>
</section>
<section id="id108">
<h2>1.3.5 (2020-05-27)<a class="headerlink" href="#id108" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version introduces minor bug fixes and a few more visualization options.</p></li>
</ul>
<section id="id109">
<h3>Added<a class="headerlink" href="#id109" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added Filter.filter_missing_values(), which can remove rows with NaN values in some (or all) columns.</p></li>
<li><p>Added the visualization function CountFilter.box_plot().</p></li>
</ul>
</section>
<section id="id110">
<h3>Changed<a class="headerlink" href="#id110" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Updated docstrings and printouts of several functions.</p></li>
<li><p>Slightly improved speed and performance across the board.</p></li>
<li><p>Filter.feature_string() is now sorted alphabetically.</p></li>
<li><p>Enrichment randomization functions in the enrichment module now accept a ‚Äòrandom_seed‚Äô argument, to be able to generate consistent results over multiple sessions.</p></li>
<li><p>Enrichment randomization functions can now return the matplotlib Figure object, in addition to the results table.</p></li>
</ul>
</section>
<section id="id111">
<h3>Fixed<a class="headerlink" href="#id111" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed DepracationWarning on parsing functions from the general module.</p></li>
<li><p>Fixed bug where saving csv files on Linux systems would save the files under the wrong directory.</p></li>
<li><p>Fixed a bug where UTF-8-encoded Reference Tables won‚Äôt be loaded correctly</p></li>
<li><p>Fixed a bug where enrichment.upsetplot() and enrichment.venn_diagram() would sometimes modify the user dict input ‚Äòobjs‚Äô.</p></li>
<li><p>Fixed a bug in CountFilter.pairplot where log2 would be calculated without a pseudocount added, leading to division by 0.</p></li>
</ul>
</section>
</section>
<section id="id112">
<h2>1.3.4 (2020-04-07)<a class="headerlink" href="#id112" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>This version fixed a bug that prevented installation of the package.</p></li>
</ul>
<section id="id113">
<h3>Changed<a class="headerlink" href="#id113" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Updated docstrings and printouts of several functions</p></li>
</ul>
</section>
<section id="id114">
<h3>Fixed<a class="headerlink" href="#id114" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Fixed a bug with installation of the previous version</p></li>
</ul>
</section>
</section>
<section id="id115">
<h2>1.3.3 (2020-03-28)<a class="headerlink" href="#id115" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>First stable release on PyPI.</p></li>
</ul>
<section id="id116">
<h3>Added<a class="headerlink" href="#id116" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Added Filter.sort(), and upgraded the functionality of Filter.filter_top_n().</p></li>
<li><p>Added UpSet plots and Venn diagrams to enrichment module.</p></li>
<li><p>User-defined biotype reference tables can now be used.</p></li>
<li><p>Filter operations now print out the result of the operation.</p></li>
<li><p>Enrichment randomization tests now also support non-WBGene indexing.</p></li>
<li><p>Filter.biotypes_from_ref_table() and FeatureSet.biotypes_from_ref_table() now report genes that don‚Äôt appear in the biotype reference table.</p></li>
<li><p>Filter.biotypes_from_ref_table() can now give a long-form report with descriptive statistics of all columns, grouped by biotype.</p></li>
<li><p>Added code examples to the user guide and to the docstrings of most functions.</p></li>
</ul>
</section>
<section id="id117">
<h3>Changed<a class="headerlink" href="#id117" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Changed argument order and default values in filtering.CountFilter.from_folder_htseqcount().</p></li>
<li><p>Changed default title in scatter_sample_vs_sample().</p></li>
<li><p>Changed default filename in CountFilter.fold_change().</p></li>
<li><p>Settings are now saved in a .yaml format. Reading and writing of settings have been modified.</p></li>
<li><p>Changed argument name ‚Äòdeseq_highlight‚Äô to ‚Äòhighlight‚Äô in scatter_sample_vs_sample(). It can now accept any Filter object.</p></li>
<li><p>Updated documentation and default ‚Äòmode‚Äô value for FeatureSet.go_enrichment().</p></li>
<li><p>Updated the signature and function of general.load_csv() to be clearer and more predictable.</p></li>
<li><p>Changed argument names in CountFilter.from_folder_htseqcount().</p></li>
<li><p>Modified names and signatures of .csv test files functions to make them more comprehensible.</p></li>
<li><p>Renamed ‚ÄòFilter.filter_by_ref_table_attr()‚Äô to ‚ÄòFilter.filter_by_attribute()‚Äô.</p></li>
<li><p>Renamed ‚ÄòFilter.split_by_ref_table_attr()‚Äô to ‚ÄòFilter.split_by_attribute()‚Äô.</p></li>
<li><p>Renamed ‚ÄòFilter.norm_reads_with_size_factor()‚Äô to ‚ÄòFilter.normalize_with_scaling_factors()‚Äô. It can now use any set of scaling factors to normalize libraries.</p></li>
<li><p>Renamed ‚ÄòFilter.norm_reads_to_rpm()‚Äô to ‚ÄòFilter.normalize_to_rpm()‚Äô.</p></li>
<li><p>Made some functions in the general module hidden.</p></li>
</ul>
</section>
<section id="id118">
<h3>Fixed<a class="headerlink" href="#id118" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Various bug fixes</p></li>
</ul>
</section>
<section id="id119">
<h3>Removed<a class="headerlink" href="#id119" title="Link to this heading">ÔÉÅ</a></h3>
<ul class="simple">
<li><p>Removed the ‚Äòfeature_name_to_wbgene‚Äô module from RNAlysis.</p></li>
</ul>
</section>
</section>
<section id="id120">
<h2>1.3.2 (2019-12-11)<a class="headerlink" href="#id120" title="Link to this heading">ÔÉÅ</a></h2>
<ul class="simple">
<li><p>First beta release on PyPI.</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="glossary.html" class="btn btn-neutral float-left" title="Glossary - programmatic interface" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.fastq.html" class="btn btn-neutral float-right" title="rnalysis.fastq" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>