<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>🧬 What is RNAlysis? &mdash; RNAlysis 4.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=3304f9e4"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="_static/copybutton.js?v=5b738548"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="How can I contribute to RNAlysis?" href="contributing.html" />
    <link rel="prev" title="RNAlysis 4.0.0 documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">🧬 What is <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="#see-rnalysis-in-action">🎥 See <em>RNAlysis</em> in Action</a></li>
<li class="toctree-l1"><a class="reference internal" href="#key-features">🚀 Key Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="#how-do-i-install-it">🛠️ How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="#how-do-i-use-it">📊 How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="#credits">Credits</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#how-do-i-cite-rnalysis">How do I cite <em>RNAlysis</em>?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#development-lead">Development Lead</a></li>
<li class="toctree-l2"><a class="reference internal" href="#contributors">Contributors</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_report.html">How to use the <em>RNAlysis</em> analysis reports</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">🧬 What is <em>RNAlysis</em>?</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/readme.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <a class="reference external image-reference" href="https://guyteichman.github.io/RNAlysis"><img alt="logo" class="align-center" src="https://raw.githubusercontent.com/GuyTeichman/RNAlysis/master/docs/source/logo.png" style="width: 250px;" /></a>
<p><a class="reference external" href="https://guyteichman.github.io/RNAlysis">Documentation</a> |
<a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">Source code</a> |
<a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/issues">Bug reports</a> |
<a class="reference external" href="https://pypi.python.org/pypi/rnalysis"><img alt="PyPI version" src="https://img.shields.io/pypi/v/rnalysis.svg" /></a> | <a class="reference external" href="https://pypi.python.org/pypi/rnalysis"><img alt="Python versions supported" src="https://img.shields.io/pypi/pyversions/RNAlysis.svg" /></a> |
<a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/actions/workflows/build_ci.yml"><img alt="Build status" src="https://github.com/guyteichman/RNAlysis/actions/workflows/build_ci.yml/badge.svg" /></a> | <a class="reference external" href="https://coveralls.io/github/GuyTeichman/RNAlysis?branch=master"><img alt="Coverage" class="align-middle" src="https://coveralls.io/repos/github/GuyTeichman/RNAlysis/badge.svg?branch=master" /></a> | <a class="reference external" href="https://pepy.tech/project/rnalysis"><img alt="Downloads" src="https://pepy.tech/badge/rnalysis" /></a></p>
<hr class="docutils" />
<section id="what-is-rnalysis">
<h1>🧬 What is <em>RNAlysis</em>?<a class="headerlink" href="#what-is-rnalysis" title="Link to this heading"></a></h1>
<p><em>RNAlysis</em> is a powerful, user-friendly software that allows you to analyze your RNA sequencing data without writing a single line of code.
This Python-based tool offers a complete solution for your RNA-seq analysis needs, from raw data processing to advanced statistical analyses and beautiful visualizations, all through an intuitive graphical interface.</p>
</section>
<hr class="docutils" />
<section id="see-rnalysis-in-action">
<h1>🎥 See <em>RNAlysis</em> in Action<a class="headerlink" href="#see-rnalysis-in-action" title="Link to this heading"></a></h1>
<p><a class="reference internal" href="https://github.com/GuyTeichman/RNAlysis/raw/master/rnalysis/gui/videos/quick_search.webp"><img alt="Rapid Gene Information Lookup" src="https://github.com/GuyTeichman/RNAlysis/raw/master/rnalysis/gui/videos/quick_search.webp" style="width: 600px;" /></a></p>
<p><em>Instantly access gene information from various biological databases with a simple right-click.</em></p>
<p><a class="reference internal" href="https://github.com/GuyTeichman/RNAlysis/raw/master/rnalysis/gui/videos/set_operations.webp"><img alt="Set Operations" src="https://github.com/GuyTeichman/RNAlysis/raw/master/rnalysis/gui/videos/set_operations.webp" style="width: 600px;" /></a></p>
<p><em>Perform advanced set operations to extract and analyze specific subsets of your data.</em></p>
<p><a class="reference internal" href="https://github.com/GuyTeichman/RNAlysis/raw/master/rnalysis/gui/videos/report_overview.webp"><img alt="Interactive Analysis Report" src="https://github.com/GuyTeichman/RNAlysis/raw/master/rnalysis/gui/videos/report_overview.webp" style="width: 600px;" /></a></p>
<p><em>Easily generate comprehensive and intuitive analysis reports to promote reproducibility. Track the entire analysis path with just a click!</em></p>
</section>
<hr class="docutils" />
<section id="key-features">
<h1>🚀 Key Features<a class="headerlink" href="#key-features" title="Link to this heading"></a></h1>
<ul class="simple">
<li><p><strong>Code-Free Analysis</strong>: Perform complex analyses with just a few clicks.</p></li>
<li><p><strong>Comprehensive Analysis Pipeline</strong>: From data import to enrichment analysis, all in one place.</p></li>
<li><p><strong>Interactive Visualizations</strong>: Explore your data with dynamic, publication-ready graphs.</p></li>
<li><p><strong>Customizable Workflows</strong>: Build and share analysis pipelines tailored to your research questions.</p></li>
<li><p><strong>Integration with Popular Tools</strong>: Seamless compatibility with DESeq2, kallisto, bowtie2, and more.</p></li>
<li><p><strong>Rapid Gene Information Lookup</strong>: Instantly access gene information from various biological databases.</p></li>
<li><p><strong>Advanced Set Operations</strong>: Easily extract and analyze specific subsets of your data.</p></li>
<li><p><strong>Reproducible Research</strong>: Generate comprehensive, interactive analysis reports with a single click.</p></li>
</ul>
<p>To get an overview of what <em>RNAlysis</em> can do, read the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/tutorial.html">tutorial</a> and the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide.html">user guide</a>.</p>
</section>
<hr class="docutils" />
<section id="how-do-i-install-it">
<h1>🛠️ How do I install it?<a class="headerlink" href="#how-do-i-install-it" title="Link to this heading"></a></h1>
<p>You can either install <em>RNAlysis</em> as a stand-alone app, or via PyPI.
To learn how to install <em>RNAlysis</em>, visit the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/installation.html">Installation page</a>.</p>
</section>
<hr class="docutils" />
<section id="how-do-i-use-it">
<h1>📊 How do I use it?<a class="headerlink" href="#how-do-i-use-it" title="Link to this heading"></a></h1>
<p>If you installed <em>RNAlysis</em> as a stand-alone app, simply open the app (“RNAlysis.exe” on Windows, “RNAlysis.dmg” on MacOS) and wait for it to load (it may take a minute or two, so be patient!).</p>
<p>If you installed <em>RNAlysis</em> from PyPi, you can launch <em>RNAlysis</em> by typing the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">rnalysis</span><span class="o">-</span><span class="n">gui</span>
</pre></div>
</div>
<p>Or through a python console:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">gui</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gui</span><span class="o">.</span><span class="n">run_gui</span><span class="p">()</span>
</pre></div>
</div>
<p>In addition, you can write Python code that uses <em>RNAlysis</em> functions as described in the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide.html">programmatic interface user guide</a>.</p>
</section>
<hr class="docutils" />
<section id="dependencies">
<h1>Dependencies<a class="headerlink" href="#dependencies" title="Link to this heading"></a></h1>
<p>All of <em>RNAlysis</em>’s dependencies can be installed automatically via PyPI.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://numpy.org/">numpy</a></p></li>
<li><p><a class="reference external" href="https://pola.rs/">polars</a></p></li>
<li><p><a class="reference external" href="https://pandas.pydata.org/">pandas</a></p></li>
<li><p><a class="reference external" href="https://www.scipy.org/">scipy</a></p></li>
<li><p><a class="reference external" href="https://matplotlib.org/">matplotlib</a></p></li>
<li><p><a class="reference external" href="http://numba.pydata.org/">numba</a></p></li>
<li><p><a class="reference external" href="https://github.com/psf/requests/">requests</a></p></li>
<li><p><a class="reference external" href="https://scikit-learn.org/">scikit-learn</a></p></li>
<li><p><a class="reference external" href="https://github.com/scikit-learn-contrib/scikit-learn-extra">scikit-learn-extra</a></p></li>
<li><p><a class="reference external" href="https://github.com/scikit-learn-contrib/hdbscan">hdbscan</a></p></li>
<li><p><a class="reference external" href="https://seaborn.pydata.org/">seaborn</a></p></li>
<li><p><a class="reference external" href="https://www.statsmodels.org/">statsmodels</a></p></li>
<li><p><a class="reference external" href="https://joblib.readthedocs.io/en/latest/">joblib</a></p></li>
<li><p><a class="reference external" href="https://github.com/tqdm/tqdm">tqdm</a></p></li>
<li><p><a class="reference external" href="https://github.com/ActiveState/appdirs">appdirs</a></p></li>
<li><p><a class="reference external" href="https://github.com/matplotlib/grid-strategy">grid_strategy</a></p></li>
<li><p><a class="reference external" href="https://github.com/yaml/pyyaml">pyyaml</a></p></li>
<li><p><a class="reference external" href="https://github.com/jnothman/UpSetPlot">UpSetPlot</a></p></li>
<li><p><a class="reference external" href="https://github.com/konstantint/matplotlib-venn">matplotlib-venn</a></p></li>
<li><p><a class="reference external" href="https://github.com/GuyTeichman/xlmhglite">xlmhglite</a></p></li>
<li><p><a class="reference external" href="https://github.com/GuyTeichman/pairwisedist/">pairwisedist</a></p></li>
<li><p><a class="reference external" href="https://github.com/python/typing_extensions">typing_extensions</a></p></li>
<li><p><a class="reference external" href="https://www.riverbankcomputing.com/software/pyqt/">PyQt5</a></p></li>
<li><p><a class="reference external" href="https://github.com/ColinDuquesnoy/QDarkStyleSheet">qdarkstyle</a></p></li>
<li><p><a class="reference external" href="https://https://github.com/tiran/defusedxml">defusedxml</a></p></li>
<li><p><a class="reference external" href="https://github.com/marcelm/cutadapt">cutadapt</a></p></li>
<li><p><a class="reference external" href="https://docs.aiohttp.org/">aiohttp</a></p></li>
<li><p><a class="reference external" href="https://github.com/saghul/aiodns">aiodns</a></p></li>
<li><p><a class="reference external" href="https://aiolimiter.readthedocs.io/">aiolimiter</a></p></li>
<li><p><a class="reference external" href="https://github.com/google/brotli">Brotli</a></p></li>
<li><p><a class="reference external" href="https://networkx.org">networkx</a></p></li>
<li><p><a class="reference external" href="https://github.com/WestHealth/pyvis">pyvis</a></p></li>
<li><p><a class="reference external" href="https://github.com/jd/tenacity">tenacity</a></p></li>
</ul>
</section>
<hr class="docutils" />
<section id="credits">
<h1>Credits<a class="headerlink" href="#credits" title="Link to this heading"></a></h1>
<section id="how-do-i-cite-rnalysis">
<h2>How do I cite <em>RNAlysis</em>?<a class="headerlink" href="#how-do-i-cite-rnalysis" title="Link to this heading"></a></h2>
<p>If you use <em>RNAlysis</em> in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Teichman</span><span class="p">,</span> <span class="n">G</span><span class="o">.</span><span class="p">,</span> <span class="n">Cohen</span><span class="p">,</span> <span class="n">D</span><span class="o">.</span><span class="p">,</span> <span class="n">Ganon</span><span class="p">,</span> <span class="n">O</span><span class="o">.</span><span class="p">,</span> <span class="n">Dunsky</span><span class="p">,</span> <span class="n">N</span><span class="o">.</span><span class="p">,</span> <span class="n">Shani</span><span class="p">,</span> <span class="n">S</span><span class="o">.</span><span class="p">,</span> <span class="n">Gingold</span><span class="p">,</span> <span class="n">H</span><span class="o">.</span><span class="p">,</span> <span class="ow">and</span> <span class="n">Rechavi</span><span class="p">,</span> <span class="n">O</span><span class="o">.</span> <span class="p">(</span><span class="mi">2023</span><span class="p">)</span><span class="o">.</span>
<span class="n">RNAlysis</span><span class="p">:</span> <span class="n">analyze</span> <span class="n">your</span> <span class="n">RNA</span> <span class="n">sequencing</span> <span class="n">data</span> <span class="n">without</span> <span class="n">writing</span> <span class="n">a</span> <span class="n">single</span> <span class="n">line</span> <span class="n">of</span> <span class="n">code</span><span class="o">.</span> <span class="n">BMC</span> <span class="n">Biology</span><span class="p">,</span> <span class="mi">21</span><span class="p">,</span> <span class="mf">74.</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.1186</span><span class="o">/</span><span class="n">s12915</span><span class="o">-</span><span class="mi">023</span><span class="o">-</span><span class="mi">01574</span><span class="o">-</span><span class="mi">6</span>
</pre></div>
</div>
<p>If you use the <em>CutAdapt</em> adapter trimming tool in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Martin</span><span class="p">,</span> <span class="n">M</span><span class="o">.</span> <span class="p">(</span><span class="mi">2011</span><span class="p">)</span><span class="o">.</span> <span class="n">Cutadapt</span> <span class="n">removes</span> <span class="n">adapter</span> <span class="n">sequences</span> <span class="kn">from</span> <span class="nn">high</span><span class="o">-</span><span class="n">throughput</span> <span class="n">sequencing</span> <span class="n">reads</span><span class="o">.</span>
<span class="n">EMBnet</span><span class="o">.</span><span class="n">journal</span><span class="p">,</span> <span class="mi">17</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">pp</span><span class="o">.</span> <span class="mi">10</span><span class="o">-</span><span class="mf">12.</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.14806</span><span class="o">/</span><span class="n">ej</span><span class="mf">.17.1.200</span>
</pre></div>
</div>
<p>If you use the <em>kallisto</em> RNA sequencing quantification tool in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Bray, N., Pimentel, H., Melsted, P. et al.
Near-optimal probabilistic RNA-seq quantification.
Nat Biotechnol 34, 525–527 (2016).
https://doi.org/10.1038/nbt.3519
</pre></div>
</div>
<p>If you use the bowtie2 aligner in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Langmead, B., and Salzberg, S.L. (2012).
Fast gapped-read alignment with Bowtie 2. Nat. Methods 2012 94 9, 357–359.
https://doi.org/10.1038/nmeth.1923
</pre></div>
</div>
<p>If you use the ShortStack aligner in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Axtell</span><span class="p">,</span> <span class="n">MJ</span><span class="o">.</span> <span class="p">(</span><span class="mi">2013</span><span class="p">)</span><span class="o">.</span>
<span class="n">ShortStack</span><span class="p">:</span> <span class="n">Comprehensive</span> <span class="n">annotation</span> <span class="ow">and</span> <span class="n">quantification</span> <span class="n">of</span> <span class="n">small</span> <span class="n">RNA</span> <span class="n">genes</span><span class="o">.</span> <span class="n">RNA</span> <span class="mi">19</span><span class="p">:</span><span class="mi">740</span><span class="o">-</span><span class="mf">751.</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.1261</span><span class="o">/</span><span class="n">rna</span><span class="mf">.035279.112</span>
</pre></div>
</div>
<p>If you use the <em>DESeq2</em> differential expression tool in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Love MI, Huber W, Anders S (2014).
“Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2.”
Genome Biology, 15, 550.
https://doi.org/10.1186/s13059-014-0550-8
</pre></div>
</div>
<p>If you use the <em>Limma-Voom</em> differential expression pipeline in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015).
limma powers differential expression analyses for RNA-sequencing and microarray studies.
Nucleic Acids Res. 43, e47–e47.
https://doi.org/10.1093/nar/gkv007

Law, C.W., Chen, Y., Shi, W., and Smyth, G.K. (2014).
Voom: Precision weights unlock linear model analysis tools for RNA-seq read counts.
Genome Biol. 15, 1–17.
https://doi.org/10.1186/gb-2014-15-2-r29
</pre></div>
</div>
<p>If you use the <em>HDBSCAN</em> clustering feature in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="n">L</span><span class="o">.</span> <span class="n">McInnes</span><span class="p">,</span> <span class="n">J</span><span class="o">.</span> <span class="n">Healy</span><span class="p">,</span> <span class="n">S</span><span class="o">.</span> <span class="n">Astels</span><span class="p">,</span> <span class="n">hdbscan</span><span class="p">:</span> <span class="n">Hierarchical</span> <span class="n">density</span> <span class="n">based</span> <span class="n">clustering</span>
<span class="n">Journal</span> <span class="n">of</span> <span class="n">Open</span> <span class="n">Source</span> <span class="n">Software</span><span class="p">,</span> <span class="n">The</span> <span class="n">Open</span> <span class="n">Journal</span><span class="p">,</span> <span class="n">volume</span> <span class="mi">2</span><span class="p">,</span> <span class="n">number</span> <span class="mf">11.</span> <span class="mi">2017</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.1371</span><span class="o">/</span><span class="n">journal</span><span class="o">.</span><span class="n">pcbi</span><span class="mf">.0030039</span>
</pre></div>
</div>
<p>If you use the <em>XL-mHG</em> single-set enrichment test in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Eden</span><span class="p">,</span> <span class="n">E</span><span class="o">.</span><span class="p">,</span> <span class="n">Lipson</span><span class="p">,</span> <span class="n">D</span><span class="o">.</span><span class="p">,</span> <span class="n">Yogev</span><span class="p">,</span> <span class="n">S</span><span class="o">.</span><span class="p">,</span> <span class="ow">and</span> <span class="n">Yakhini</span><span class="p">,</span> <span class="n">Z</span><span class="o">.</span> <span class="p">(</span><span class="mi">2007</span><span class="p">)</span><span class="o">.</span>
 <span class="n">Discovering</span> <span class="n">Motifs</span> <span class="ow">in</span> <span class="n">Ranked</span> <span class="n">Lists</span> <span class="n">of</span> <span class="n">DNA</span> <span class="n">Sequences</span><span class="o">.</span> <span class="n">PLOS</span> <span class="n">Comput</span><span class="o">.</span> <span class="n">Biol</span><span class="o">.</span> <span class="mi">3</span><span class="p">,</span> <span class="n">e39</span><span class="o">.</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.1371</span><span class="o">/</span><span class="n">journal</span><span class="o">.</span><span class="n">pcbi</span><span class="mf">.0030039</span><span class="o">&gt;</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.1371</span><span class="o">/</span><span class="n">journal</span><span class="o">.</span><span class="n">pcbi</span><span class="mf">.0030039</span><span class="o">&lt;/</span><span class="n">a</span><span class="o">&gt;</span>

<span class="n">Wagner</span><span class="p">,</span> <span class="n">F</span><span class="o">.</span> <span class="p">(</span><span class="mi">2017</span><span class="p">)</span><span class="o">.</span> <span class="n">The</span> <span class="n">XL</span><span class="o">-</span><span class="n">mHG</span> <span class="n">test</span> <span class="k">for</span> <span class="n">gene</span> <span class="nb">set</span> <span class="n">enrichment</span><span class="o">.</span> <span class="n">ArXiv</span><span class="o">.</span>
<span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">doi</span><span class="o">.</span><span class="n">org</span><span class="o">/</span><span class="mf">10.48550</span><span class="o">/</span><span class="n">arXiv</span><span class="mf">.1507.07905</span>
</pre></div>
</div>
<p>If you use the <em>Ensembl</em> database in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Martin FJ, Amode MR, Aneja A, Austine-Orimoloye O, Azov AG, Barnes I, et al.
Ensembl 2023. Nucleic Acids Res [Internet]. 2023 Jan 6;51(D1):D933–41.
doi.org/10.1093/nar/gkac958
</pre></div>
</div>
<p>If you use the <em>PANTHER</em> database in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Thomas PD, Ebert D, Muruganujan A, Mushayahama T, Albou L-P, Mi H.
PANTHER: Making genome-scale phylogenetics accessible to all. Protein Sci [Internet]. 2022 Jan 1;31(1):8–22.
doi.org/10.1002/pro.4218
</pre></div>
</div>
<p>If you use the <em>OrthoInspector</em> database in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Nevers Y, Kress A, Defosset A, Ripp R, Linard B, Thompson JD, et al.
OrthoInspector 3.0: open portal for comparative genomics. Nucleic Acids Res [Internet]. 2019 Jan 8;47(D1):D411–8.
doi.org/10.1093/nar/gky1068
</pre></div>
</div>
<p>If you use the <em>PhylomeDB</em> database in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Fuentes D, Molina M, Chorostecki U, Capella-Gutiérrez S, Marcet-Houben M, Gabaldón T.
PhylomeDB V5: an expanding repository for genome-wide catalogues of annotated gene phylogenies. Nucleic Acids Res [Internet]. 2022 Jan 7;50(D1):D1062–8.
doi.org/10.1093/nar/gkab966
</pre></div>
</div>
<p>If you use the <em>UniProt</em> gene ID mapping service in your research, please cite:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>The UniProt Consortium.
UniProt: the Universal Protein Knowledgebase in 2023. Nucleic Acids Res [Internet]. 2023 Jan 6;51(D1):D523–31.
doi.org/10.1093/nar/gkac1052
</pre></div>
</div>
</section>
<section id="development-lead">
<h2>Development Lead<a class="headerlink" href="#development-lead" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Guy Teichman: <a class="reference external" href="mailto:guyteichman&#37;&#52;&#48;gmail&#46;com">guyteichman<span>&#64;</span>gmail<span>&#46;</span>com</a></p></li>
</ul>
</section>
<section id="contributors">
<h2>Contributors<a class="headerlink" href="#contributors" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Dror Cohen</p></li>
<li><p>Or Ganon</p></li>
<li><p>Netta Dunsky</p></li>
<li><p>Shachar Shani</p></li>
<li><p><a class="reference external" href="https://github.com/Mintxoklet">Mintxoklet</a></p></li>
<li><p><a class="reference external" href="https://github.com/kbipinkumar">Bipin Kumar</a></p></li>
<li><p>Matthias Wilm</p></li>
<li><p><a class="reference external" href="https://github.com/sandyl27">sandyl27</a></p></li>
<li><p><a class="reference external" href="https://github.com/clockgene">clockgene</a></p></li>
<li><p><a class="reference external" href="https://github.com/NeuroRookie">NeuroRookie</a></p></li>
</ul>
<hr class="docutils" />
<p>This package was created with <a class="reference external" href="https://github.com/audreyr/cookiecutter">Cookiecutter</a> and the <a class="reference external" href="https://github.com/audreyr/cookiecutter-pypackage">audreyr/cookiecutter-pypackage</a> project template.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="RNAlysis 4.0.0 documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contributing.html" class="btn btn-neutral float-right" title="How can I contribute to RNAlysis?" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>