<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>rnalysis.enrichment.RankedSet.user_defined_enrichment &mdash; RNAlysis 3.10.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.enrichment.enrichment_bar_plot" href="rnalysis.enrichment.enrichment_bar_plot.html" />
    <link rel="prev" title="rnalysis.enrichment.RankedSet.update" href="rnalysis.enrichment.RankedSet.update.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.html">rnalysis.enrichment.FeatureSet</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="rnalysis.enrichment.RankedSet.html">rnalysis.enrichment.RankedSet</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.html#rnalysis.enrichment.RankedSet"><code class="docutils literal notranslate"><span class="pre">RankedSet</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.add.html">rnalysis.enrichment.RankedSet.add</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.biotypes_from_ref_table.html">rnalysis.enrichment.RankedSet.biotypes_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.change_set_name.html">rnalysis.enrichment.RankedSet.change_set_name</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.clear.html">rnalysis.enrichment.RankedSet.clear</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.copy.html">rnalysis.enrichment.RankedSet.copy</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.difference.html">rnalysis.enrichment.RankedSet.difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.difference_update.html">rnalysis.enrichment.RankedSet.difference_update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.discard.html">rnalysis.enrichment.RankedSet.discard</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.filter_by_attribute.html">rnalysis.enrichment.RankedSet.filter_by_attribute</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.filter_by_go_annotations.html">rnalysis.enrichment.RankedSet.filter_by_go_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.filter_by_kegg_annotations.html">rnalysis.enrichment.RankedSet.filter_by_kegg_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.go_enrichment.html">rnalysis.enrichment.RankedSet.go_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.intersection.html">rnalysis.enrichment.RankedSet.intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.intersection_update.html">rnalysis.enrichment.RankedSet.intersection_update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.isdisjoint.html">rnalysis.enrichment.RankedSet.isdisjoint</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.issubset.html">rnalysis.enrichment.RankedSet.issubset</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.issuperset.html">rnalysis.enrichment.RankedSet.issuperset</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.kegg_enrichment.html">rnalysis.enrichment.RankedSet.kegg_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.non_categorical_enrichment.html">rnalysis.enrichment.RankedSet.non_categorical_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.pop.html">rnalysis.enrichment.RankedSet.pop</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.remove.html">rnalysis.enrichment.RankedSet.remove</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.save_txt.html">rnalysis.enrichment.RankedSet.save_txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.single_set_enrichment.html">rnalysis.enrichment.RankedSet.single_set_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.single_set_go_enrichment.html">rnalysis.enrichment.RankedSet.single_set_go_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.single_set_kegg_enrichment.html">rnalysis.enrichment.RankedSet.single_set_kegg_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.symmetric_difference.html">rnalysis.enrichment.RankedSet.symmetric_difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.symmetric_difference_update.html">rnalysis.enrichment.RankedSet.symmetric_difference_update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.translate_gene_ids.html">rnalysis.enrichment.RankedSet.translate_gene_ids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.union.html">rnalysis.enrichment.RankedSet.union</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.RankedSet.update.html">rnalysis.enrichment.RankedSet.update</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">rnalysis.enrichment.RankedSet.user_defined_enrichment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#rnalysis.enrichment.RankedSet.user_defined_enrichment"><code class="docutils literal notranslate"><span class="pre">RankedSet.user_defined_enrichment()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.enrichment_bar_plot.html">rnalysis.enrichment.enrichment_bar_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.gene_ontology_graph.html">rnalysis.enrichment.gene_ontology_graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.kegg_pathway_graph.html">rnalysis.enrichment.kegg_pathway_graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.upset_plot.html">rnalysis.enrichment.upset_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.venn_diagram.html">rnalysis.enrichment.venn_diagram</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
          <li class="breadcrumb-item"><a href="rnalysis.enrichment.RankedSet.html">rnalysis.enrichment.RankedSet</a></li>
      <li class="breadcrumb-item active">rnalysis.enrichment.RankedSet.user_defined_enrichment</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/rnalysis.enrichment.RankedSet.user_defined_enrichment.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnalysis-enrichment-rankedset-user-defined-enrichment">
<h1>rnalysis.enrichment.RankedSet.user_defined_enrichment<a class="headerlink" href="#rnalysis-enrichment-rankedset-user-defined-enrichment" title="Permalink to this heading"></a></h1>
<dl class="py method">
<dt class="sig sig-object py" id="rnalysis.enrichment.RankedSet.user_defined_enrichment">
<span class="sig-prename descclassname"><span class="pre">RankedSet.</span></span><span class="sig-name descname"><span class="pre">user_defined_enrichment</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attributes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'all'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">statistical_test</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'fisher'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'hypergeometric'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'randomization'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'fisher'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">alpha</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Fraction</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'all'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'all'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><span class="pre">Filter</span></a><span class="p"><span class="pre">,</span></span><span class="w"> </span><a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><span class="pre">FeatureSet</span></a><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attr_ref_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Path</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'predefined'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'predefined'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">biotype_ref_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Path</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'predefined'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'predefined'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_unannotated_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_nonsignificant</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_csv</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_fig</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_horizontal</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_expected</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plot_style</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'bar'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'lollipop'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'bar'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">randomization_reps</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PositiveInt</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">random_seed</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parallel_backend</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'multiprocessing'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'loky'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'threading'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'sequential'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'loky'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gui_mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">DataFrame</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Figure</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></span><a class="headerlink" href="#rnalysis.enrichment.RankedSet.user_defined_enrichment" title="Permalink to this definition"></a></dt>
<dd><p>Calculates enrichment and depletion of the FeatureSet for user-defined attributes against a background set.        The attributes are drawn from an Attribute Reference Table.         The background set is determined by either the input variable ‘background_genes’,         or by the input variable ‘biotype’ and a Biotype Reference Table.         P-values are corrected for multiple comparisons using the Benjamini–Hochberg step-up procedure         (original FDR method). In plots, for the clarity of display, complete depletion (linear enrichment score = 0)         appears with the smallest value in the scale.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>str</em><em>, </em><em>int</em><em>, </em><em>iterable</em><em> (</em><em>list</em><em>, </em><em>tuple</em><em>, </em><em>set</em><em>, </em><em>etc</em><em>) of </em><em>str/int</em><em>, or </em><em>'all'.</em>) – An iterable of attribute names or attribute numbers         (according to their order in the Attribute Reference Table).         If ‘all’, all of the attributes in the Attribute Reference Table will be used.         If None, a manual input prompt will be raised.</p></li>
<li><p><strong>statistical_test</strong> (<em>'fisher'</em><em>, </em><em>'hypergeometric'</em><em> or </em><em>'randomization'</em><em> (</em><em>default='fisher'</em><em>)</em>) – determines the statistical test to be used for enrichment analysis.         Note that some propagation methods support only some of the available statistical tests.</p></li>
<li><p><strong>alpha</strong> (<em>float between 0 and 1</em><em> (</em><em>default=0.05</em><em>)</em>) – Indicates the FDR threshold for significance.</p></li>
<li><p><strong>attr_ref_path</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default='predefined'</em><em>)</em>) – the path of the Attribute Reference Table from which user-defined attributes will be drawn.</p></li>
<li><p><strong>biotype_ref_path</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default='predefined'</em><em>)</em>) – the path of the Biotype Reference Table.         Will be used to generate background set if ‘biotype’ is specified.</p></li>
<li><p><strong>biotype</strong> (<em>str specifying a specific biotype</em><em>, </em><em>list/set</em><em> of </em><em>strings each specifying a biotype</em><em>, or </em><em>'all'</em><em>         (</em><em>default='protein_coding'</em><em>)</em>) – determines the background genes by their biotype. Requires specifying a Biotype Reference Table.         ‘all’ will include all genomic features in the reference table,         ‘protein_coding’ will include only protein-coding genes from the reference table, etc.         Cannot be specified together with ‘background_genes’.</p></li>
<li><p><strong>background_genes</strong> (<em>set</em><em> of </em><em>feature indices</em><em>, </em><em>filtering.Filter object</em><em>, or </em><em>enrichment.FeatureSet object</em><em>         (</em><em>default=None</em><em>)</em>) – a set of specific feature indices to be used as background genes.         Cannot be specified together with ‘biotype’.</p></li>
<li><p><strong>exclude_unannotated_genes</strong> (<em>bool</em><em> (</em><em>deafult=True</em><em>)</em>) – if True, genes that have no annotation associated with them will be         excluded from the enrichment analysis. This is the recommended practice for enrichment analysis, since         keeping unannotated genes in the analysis increases the chance of discovering spurious enrichment results.</p></li>
<li><p><strong>return_nonsignificant</strong> (<em>bool</em><em> (</em><em>default=True</em><em>)</em>) – if True (default), the results DataFrame will include all tested attributes -         both significant and non-significant ones. If False, only significant attributes will be returned.</p></li>
<li><p><strong>save_csv</strong> (<em>bool</em><em> (</em><em>default=False</em><em>)</em>) – If True, will save the results to a .csv file, under the name specified in ‘fname’.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default=None</em><em>)</em>) – The full path and name of the file to which to save the results. For example:         ‘C:/dir/file’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</p></li>
<li><p><strong>return_fig</strong> (<em>bool</em><em> (</em><em>default=False</em><em>)</em>) – if True, returns a matplotlib Figure object in addition to the results DataFrame.</p></li>
<li><p><strong>plot_horizontal</strong> (<em>bool</em><em> (</em><em>default=True</em><em>)</em>) – if True, results will be plotted with a horizontal bar plot. Otherwise, results         will be plotted with a vertical plot.</p></li>
<li><p><strong>show_expected</strong> (<em>bool</em><em> (</em><em>default=False</em><em>)</em>) – if True, the observed/expected values will be shown on the plot.</p></li>
<li><p><strong>plot_style</strong> (<em>'bar'</em><em> or </em><em>'lollipop'</em><em> (</em><em>default='bar'</em><em>)</em>) – style for the plot. Either ‘bar’ for a bar plot or ‘lollipop’ for a lollipop plot         in which the lollipop size indicates the size of the observed gene set.</p></li>
<li><p><strong>randomization_reps</strong> (<em>int larger than 0</em><em> (</em><em>default=10000</em><em>)</em>) – if using a randomization test, determine how many randomization repititions to run.         Otherwise, this parameter will not affect the analysis.</p></li>
<li><p><strong>parallel_backend</strong> (<em>Literal</em><em>[</em><em>PARALLEL_BACKENDS</em><em>] </em><em>(</em><em>default='loky'</em><em>)</em>) – Determines the babckend used to run the analysis.         if parallel_backend not ‘sequential’, will calculate the statistical tests using parallel processing.         In most cases parallel processing will lead to shorter computation time, but does not affect the results of         the analysis otherwise.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pd.DataFrame (default) or Tuple[pd.DataFrame, matplotlib.figure.Figure]</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a pandas DataFrame with the indicated attribute names as rows/index;         and a matplotlib Figure, if ‘return_figure’ is set to True.</p>
</dd>
</dl>
<figure class="align-center" id="id1">
<a class="reference internal image-reference" href="_images/plot_enrichment_results.png"><img alt="_images/plot_enrichment_results.png" src="_images/plot_enrichment_results.png" style="width: 336.0px; height: 391.2px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of user_defined_enrichment()</span><a class="headerlink" href="#id1" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-center" id="id2">
<a class="reference internal image-reference" href="_images/plot_enrichment_results_vertical.png"><img alt="_images/plot_enrichment_results_vertical.png" src="_images/plot_enrichment_results_vertical.png" style="width: 491.4px; height: 333.59999999999997px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of user_defined_enrichment(plot_horizontal = False)</span><a class="headerlink" href="#id2" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="rnalysis.enrichment.RankedSet.update.html" class="btn btn-neutral float-left" title="rnalysis.enrichment.RankedSet.update" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.enrichment.enrichment_bar_plot.html" class="btn btn-neutral float-right" title="rnalysis.enrichment.enrichment_bar_plot" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>