<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>rnalysis.filtering.CountFilter.map_orthologs_phylomedb &mdash; RNAlysis 4.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.filtering.CountFilter.normalize_median_of_ratios" href="rnalysis.filtering.CountFilter.normalize_median_of_ratios.html" />
    <link rel="prev" title="rnalysis.filtering.CountFilter.map_orthologs_panther" href="rnalysis.filtering.CountFilter.map_orthologs_panther.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">🧬 What is <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#see-rnalysis-in-action">🎥 See <em>RNAlysis</em> in Action</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#key-features">🚀 Key Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">🛠️ How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">📊 How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_report.html">How to use the <em>RNAlysis</em> analysis reports</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="rnalysis.filtering.CountFilter.html">rnalysis.filtering.CountFilter</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.average_replicate_samples.html">rnalysis.filtering.CountFilter.average_replicate_samples</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.biotypes_from_gtf.html">rnalysis.filtering.CountFilter.biotypes_from_gtf</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.biotypes_from_ref_table.html">rnalysis.filtering.CountFilter.biotypes_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.box_plot.html">rnalysis.filtering.CountFilter.box_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.clustergram.html">rnalysis.filtering.CountFilter.clustergram</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.concatenate.html">rnalysis.filtering.CountFilter.concatenate</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.describe.html">rnalysis.filtering.CountFilter.describe</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.difference.html">rnalysis.filtering.CountFilter.difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.differential_expression_deseq2.html">rnalysis.filtering.CountFilter.differential_expression_deseq2</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.differential_expression_deseq2_simplified.html">rnalysis.filtering.CountFilter.differential_expression_deseq2_simplified</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.differential_expression_limma_voom.html">rnalysis.filtering.CountFilter.differential_expression_limma_voom</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.differential_expression_limma_voom_simplified.html">rnalysis.filtering.CountFilter.differential_expression_limma_voom_simplified</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.drop_columns.html">rnalysis.filtering.CountFilter.drop_columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.enhanced_box_plot.html">rnalysis.filtering.CountFilter.enhanced_box_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_biotype_from_gtf.html">rnalysis.filtering.CountFilter.filter_biotype_from_gtf</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_biotype_from_ref_table.html">rnalysis.filtering.CountFilter.filter_biotype_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_by_attribute.html">rnalysis.filtering.CountFilter.filter_by_attribute</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_by_go_annotations.html">rnalysis.filtering.CountFilter.filter_by_go_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_by_kegg_annotations.html">rnalysis.filtering.CountFilter.filter_by_kegg_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_by_row_name.html">rnalysis.filtering.CountFilter.filter_by_row_name</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_by_row_sum.html">rnalysis.filtering.CountFilter.filter_by_row_sum</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_duplicate_ids.html">rnalysis.filtering.CountFilter.filter_duplicate_ids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_low_reads.html">rnalysis.filtering.CountFilter.filter_low_reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_missing_values.html">rnalysis.filtering.CountFilter.filter_missing_values</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_percentile.html">rnalysis.filtering.CountFilter.filter_percentile</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.filter_top_n.html">rnalysis.filtering.CountFilter.filter_top_n</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.find_paralogs_ensembl.html">rnalysis.filtering.CountFilter.find_paralogs_ensembl</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.find_paralogs_panther.html">rnalysis.filtering.CountFilter.find_paralogs_panther</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.fold_change.html">rnalysis.filtering.CountFilter.fold_change</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.from_dataframe.html">rnalysis.filtering.CountFilter.from_dataframe</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.from_folder.html">rnalysis.filtering.CountFilter.from_folder</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.from_folder_htseqcount.html">rnalysis.filtering.CountFilter.from_folder_htseqcount</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.head.html">rnalysis.filtering.CountFilter.head</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.histogram.html">rnalysis.filtering.CountFilter.histogram</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.intersection.html">rnalysis.filtering.CountFilter.intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.ma_plot.html">rnalysis.filtering.CountFilter.ma_plot</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.majority_vote_intersection.html">rnalysis.filtering.CountFilter.majority_vote_intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.map_orthologs_ensembl.html">rnalysis.filtering.CountFilter.map_orthologs_ensembl</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.map_orthologs_orthoinspector.html">rnalysis.filtering.CountFilter.map_orthologs_orthoinspector</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.map_orthologs_panther.html">rnalysis.filtering.CountFilter.map_orthologs_panther</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">rnalysis.filtering.CountFilter.map_orthologs_phylomedb</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_median_of_ratios.html">rnalysis.filtering.CountFilter.normalize_median_of_ratios</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_rle.html">rnalysis.filtering.CountFilter.normalize_rle</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_tmm.html">rnalysis.filtering.CountFilter.normalize_tmm</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_to_quantile.html">rnalysis.filtering.CountFilter.normalize_to_quantile</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_to_rpkm.html">rnalysis.filtering.CountFilter.normalize_to_rpkm</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_to_rpm.html">rnalysis.filtering.CountFilter.normalize_to_rpm</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_to_rpm_htseqcount.html">rnalysis.filtering.CountFilter.normalize_to_rpm_htseqcount</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_to_tpm.html">rnalysis.filtering.CountFilter.normalize_to_tpm</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.normalize_with_scaling_factors.html">rnalysis.filtering.CountFilter.normalize_with_scaling_factors</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.number_filters.html">rnalysis.filtering.CountFilter.number_filters</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.pairplot.html">rnalysis.filtering.CountFilter.pairplot</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.pca.html">rnalysis.filtering.CountFilter.pca</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.plot_expression.html">rnalysis.filtering.CountFilter.plot_expression</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.print_features.html">rnalysis.filtering.CountFilter.print_features</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.save_csv.html">rnalysis.filtering.CountFilter.save_csv</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.save_parquet.html">rnalysis.filtering.CountFilter.save_parquet</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.save_table.html">rnalysis.filtering.CountFilter.save_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.scatter_sample_vs_sample.html">rnalysis.filtering.CountFilter.scatter_sample_vs_sample</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.sort.html">rnalysis.filtering.CountFilter.sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.sort_by_principal_component.html">rnalysis.filtering.CountFilter.sort_by_principal_component</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_by_attribute.html">rnalysis.filtering.CountFilter.split_by_attribute</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_by_percentile.html">rnalysis.filtering.CountFilter.split_by_percentile</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_by_principal_components.html">rnalysis.filtering.CountFilter.split_by_principal_components</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_by_reads.html">rnalysis.filtering.CountFilter.split_by_reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_clicom.html">rnalysis.filtering.CountFilter.split_clicom</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_hdbscan.html">rnalysis.filtering.CountFilter.split_hdbscan</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_hierarchical.html">rnalysis.filtering.CountFilter.split_hierarchical</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_kmeans.html">rnalysis.filtering.CountFilter.split_kmeans</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.split_kmedoids.html">rnalysis.filtering.CountFilter.split_kmedoids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.symmetric_difference.html">rnalysis.filtering.CountFilter.symmetric_difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.tail.html">rnalysis.filtering.CountFilter.tail</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.text_filters.html">rnalysis.filtering.CountFilter.text_filters</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.transform.html">rnalysis.filtering.CountFilter.transform</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.translate_gene_ids.html">rnalysis.filtering.CountFilter.translate_gene_ids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.union.html">rnalysis.filtering.CountFilter.union</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.CountFilter.violin_plot.html">rnalysis.filtering.CountFilter.violin_plot</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.html">rnalysis.filtering.DESeqFilter</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.Filter.html">rnalysis.filtering.Filter</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.FoldChangeFilter.html">rnalysis.filtering.FoldChangeFilter</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.Pipeline.html">rnalysis.filtering.Pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="rnalysis.filtering.html">rnalysis.filtering</a></li>
          <li class="breadcrumb-item"><a href="rnalysis.filtering.CountFilter.html">rnalysis.filtering.CountFilter</a></li>
      <li class="breadcrumb-item active">rnalysis.filtering.CountFilter.map_orthologs_phylomedb</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/rnalysis.filtering.CountFilter.map_orthologs_phylomedb.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnalysis-filtering-countfilter-map-orthologs-phylomedb">
<h1>rnalysis.filtering.CountFilter.map_orthologs_phylomedb<a class="headerlink" href="#rnalysis-filtering-countfilter-map-orthologs-phylomedb" title="Permalink to this heading"></a></h1>
<dl class="py method">
<dt class="sig sig-object py" id="rnalysis.filtering.CountFilter.map_orthologs_phylomedb">
<span class="sig-prename descclassname"><span class="pre">CountFilter.</span></span><span class="sig-name descname"><span class="pre">map_orthologs_phylomedb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">map_to_organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">map_from_organism</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'auto'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_id_type</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'auto'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'UniProtKB</span> <span class="pre">AC/ID'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UniParc'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UniRef50'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UniRef90'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UniRef100'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Gene</span> <span class="pre">Name'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'CRC64'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Ensembl'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Ensembl</span> <span class="pre">Genomes'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Ensembl</span> <span class="pre">Genomes</span> <span class="pre">Protein'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Ensembl</span> <span class="pre">Genomes</span> <span class="pre">Transcript'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Ensembl</span> <span class="pre">Protein'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Ensembl</span> <span class="pre">Transcript'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GeneID'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'KEGG'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PATRIC'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UCSC'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'WBParaSite'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'WBParaSite</span> <span class="pre">Transcript/Protein'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ArachnoServer'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Araport'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'CGD'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ConoServer'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'dictyBase'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'EchoBASE'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'euHCVdb'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'FlyBase'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GeneCards'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GeneReviews'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'HGNC'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'LegioList'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Leproma'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'MaizeGDB'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'MGI'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'MIM'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'neXtProt'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'OpenTargets'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Orphanet'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PharmGKB'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PomBase'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PseudoCAP'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'RGD'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'SGD'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'TubercuList'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'VEuPathDB'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'VGNC'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'WormBase'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'WormBase</span> <span class="pre">Protein'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'WormBase</span> <span class="pre">Transcript'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Xenbase'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ZFIN'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'eggNOG'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GeneTree'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'HOGENOM'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'OMA'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'OrthoDB'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'TreeFam'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'CCDS'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'EMBL/GenBank/DDBJ'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'EMBL/GenBank/DDBJ</span> <span class="pre">CDS'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GI</span> <span class="pre">number'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PIR'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'RefSeq</span> <span class="pre">Nucleotide'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'RefSeq</span> <span class="pre">Protein'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ChiTaRS'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GeneWiki'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'GenomeRNAi'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PHI-base'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'CollecTF'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'BioCyc'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'PlantReactome'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'Reactome'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'UniPathway'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'CPTAC'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'ProteomicsDB'</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'auto'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">consistency_score_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Fraction</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_consistency_score</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_unique_mode</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'first'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'last'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'random'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'none'</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'first'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_unmapped_genes</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#rnalysis.filtering.CountFilter.map_orthologs_phylomedb" title="Permalink to this definition"></a></dt>
<dd><blockquote>
<div><p>Map genes to their nearest orthologs in a different species using the PhylomeDB database.         This function generates a table describing all matching discovered ortholog pairs (both unique and non-unique)         and returns it, and can also translate the genes in this data table into their nearest ortholog,         as well as remove unmapped genes.</p>
<dl class="field-list simple">
<dt class="field-odd">param map_to_organism<span class="colon">:</span></dt>
<dd class="field-odd"><p>organism name or NCBI taxon ID of the target species for ortholog mapping.</p>
</dd>
<dt class="field-even">type map_to_organism<span class="colon">:</span></dt>
<dd class="field-even"><p>str or int</p>
</dd>
<dt class="field-odd">param map_from_organism<span class="colon">:</span></dt>
<dd class="field-odd"><p>organism name or NCBI taxon ID of the input genes’ source species.</p>
</dd>
<dt class="field-even">type map_from_organism<span class="colon">:</span></dt>
<dd class="field-even"><p>str or int</p>
</dd>
<dt class="field-odd">param gene_id_type<span class="colon">:</span></dt>
<dd class="field-odd"><p>the identifier type of the genes/features in the FeatureSet object         (for example: ‘UniProtKB’, ‘WormBase’, ‘RNACentral’, ‘Entrez Gene ID’).         If the annotations fetched from the KEGG server do not match your gene_id_type, RNAlysis will attempt to map         the annotations’ gene IDs to your identifier type.         For a full list of legal ‘gene_id_type’ names, see the UniProt website:         <a class="reference external" href="https://www.uniprot.org/help/api_idmapping">https://www.uniprot.org/help/api_idmapping</a></p>
</dd>
<dt class="field-even">type gene_id_type<span class="colon">:</span></dt>
<dd class="field-even"><p>str or ‘auto’ (default=’auto’)</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>`   :param consistency_score_threshold: the minimum consistency score required for an ortholog mapping to be         considered valid. Consistency scores are calculated by PhylomeDB and represent the confidence of the         ortholog mapping. setting consistency_score_threshold to 0 will keep all mappings. You can read more about         PhylomeDB consistency score on the PhylomeDB website: orthology.phylomedb.org/help</dt><dd><dl class="field-list simple">
<dt class="field-odd">type consistency_score_threshold<span class="colon">:</span></dt>
<dd class="field-odd"><p>float between 0 and 1 (default=0.5)</p>
</dd>
<dt class="field-even">param filter_consistency_score<span class="colon">:</span></dt>
<dd class="field-even"><p>if True (default), when encountering non-unique ortholog mappings,         <em>RNAlysis</em> will only keep the mappings with the highest consistency score.</p>
</dd>
<dt class="field-odd">type filter_consistency_score<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool (default=True)</p>
</dd>
<dt class="field-even">param non_unique_mode<span class="colon">:</span></dt>
<dd class="field-even"><p>How to handle non-unique mappings. ‘first’ will keep the first mapping found for each         gene; ‘last’ will keep the last; ‘random’ will keep a random mapping;         and ‘none’ will discard all non-unique mappings.</p>
</dd>
<dt class="field-odd">type non_unique_mode<span class="colon">:</span></dt>
<dd class="field-odd"><p>‘first’, ‘last’, ‘random’, or ‘none’ (default=’first’)</p>
</dd>
<dt class="field-even">param remove_unmapped_genes<span class="colon">:</span></dt>
<dd class="field-even"><p>if True, rows with gene names/IDs that could not be mapped to an ortholog         will be dropped from the table.         Otherwise, they will remain in the table with their original gene name/ID.</p>
</dd>
<dt class="field-odd">type remove_unmapped_genes<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool (default=False)</p>
</dd>
<dt class="field-even">type inplace<span class="colon">:</span></dt>
<dd class="field-even"><p>bool (default=True)</p>
</dd>
<dt class="field-odd">param inplace<span class="colon">:</span></dt>
<dd class="field-odd"><p>If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p>
</dd>
<dt class="field-even">return<span class="colon">:</span></dt>
<dd class="field-even"><p>DataFrame describing all discovered mappings (unique and otherwise). If inplace=True,         returns a filtered instance of the Filter object as well.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="rnalysis.filtering.CountFilter.map_orthologs_panther.html" class="btn btn-neutral float-left" title="rnalysis.filtering.CountFilter.map_orthologs_panther" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.filtering.CountFilter.normalize_median_of_ratios.html" class="btn btn-neutral float-right" title="rnalysis.filtering.CountFilter.normalize_median_of_ratios" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>