<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>rnalysis.filtering.DESeqFilter &mdash; RNAlysis 3.8.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->

        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.filtering.DESeqFilter.biotypes_from_gtf" href="rnalysis.filtering.DESeqFilter.biotypes_from_gtf.html" />
    <link rel="prev" title="rnalysis.filtering.CountFilter.violin_plot" href="rnalysis.filtering.CountFilter.violin_plot.html" />
</head>

<body class="wy-body-for-nav">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html">
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.CountFilter.html">rnalysis.filtering.CountFilter</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">rnalysis.filtering.DESeqFilter</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#rnalysis.filtering.DESeqFilter"><code class="docutils literal notranslate"><span class="pre">DESeqFilter</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#rnalysis.filtering.DESeqFilter.__init__"><code class="docutils literal notranslate"><span class="pre">DESeqFilter.__init__()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.biotypes_from_gtf.html">rnalysis.filtering.DESeqFilter.biotypes_from_gtf</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.biotypes_from_ref_table.html">rnalysis.filtering.DESeqFilter.biotypes_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.describe.html">rnalysis.filtering.DESeqFilter.describe</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.difference.html">rnalysis.filtering.DESeqFilter.difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.drop_columns.html">rnalysis.filtering.DESeqFilter.drop_columns</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_abs_log2_fold_change.html">rnalysis.filtering.DESeqFilter.filter_abs_log2_fold_change</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_biotype_from_gtf.html">rnalysis.filtering.DESeqFilter.filter_biotype_from_gtf</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_biotype_from_ref_table.html">rnalysis.filtering.DESeqFilter.filter_biotype_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_attribute.html">rnalysis.filtering.DESeqFilter.filter_by_attribute</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_go_annotations.html">rnalysis.filtering.DESeqFilter.filter_by_go_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_kegg_annotations.html">rnalysis.filtering.DESeqFilter.filter_by_kegg_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_row_name.html">rnalysis.filtering.DESeqFilter.filter_by_row_name</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_duplicate_ids.html">rnalysis.filtering.DESeqFilter.filter_duplicate_ids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_fold_change_direction.html">rnalysis.filtering.DESeqFilter.filter_fold_change_direction</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_missing_values.html">rnalysis.filtering.DESeqFilter.filter_missing_values</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_percentile.html">rnalysis.filtering.DESeqFilter.filter_percentile</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_significant.html">rnalysis.filtering.DESeqFilter.filter_significant</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_top_n.html">rnalysis.filtering.DESeqFilter.filter_top_n</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.from_dataframe.html">rnalysis.filtering.DESeqFilter.from_dataframe</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.head.html">rnalysis.filtering.DESeqFilter.head</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.intersection.html">rnalysis.filtering.DESeqFilter.intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.majority_vote_intersection.html">rnalysis.filtering.DESeqFilter.majority_vote_intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.number_filters.html">rnalysis.filtering.DESeqFilter.number_filters</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.print_features.html">rnalysis.filtering.DESeqFilter.print_features</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.save_csv.html">rnalysis.filtering.DESeqFilter.save_csv</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.sort.html">rnalysis.filtering.DESeqFilter.sort</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.split_by_attribute.html">rnalysis.filtering.DESeqFilter.split_by_attribute</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.split_by_percentile.html">rnalysis.filtering.DESeqFilter.split_by_percentile</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.split_fold_change_direction.html">rnalysis.filtering.DESeqFilter.split_fold_change_direction</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.symmetric_difference.html">rnalysis.filtering.DESeqFilter.symmetric_difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.tail.html">rnalysis.filtering.DESeqFilter.tail</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.text_filters.html">rnalysis.filtering.DESeqFilter.text_filters</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.transform.html">rnalysis.filtering.DESeqFilter.transform</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.translate_gene_ids.html">rnalysis.filtering.DESeqFilter.translate_gene_ids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.union.html">rnalysis.filtering.DESeqFilter.union</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.filtering.DESeqFilter.volcano_plot.html">rnalysis.filtering.DESeqFilter.volcano_plot</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.Filter.html">rnalysis.filtering.Filter</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.FoldChangeFilter.html">rnalysis.filtering.FoldChangeFilter</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.filtering.Pipeline.html">rnalysis.filtering.Pipeline</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="rnalysis.filtering.html">rnalysis.filtering</a> &raquo;</li>
      <li>rnalysis.filtering.DESeqFilter</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/rnalysis.filtering.DESeqFilter.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">

  <section id="rnalysis-filtering-deseqfilter">
<h1>rnalysis.filtering.DESeqFilter<a class="headerlink" href="#rnalysis-filtering-deseqfilter" title="Permalink to this heading"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="rnalysis.filtering.DESeqFilter">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">rnalysis.filtering.</span></span><span class="sig-name descname"><span class="pre">DESeqFilter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Path</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2fc_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'log2FoldChange'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padj_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'padj'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suppress_warnings</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#rnalysis.filtering.DESeqFilter" title="Permalink to this definition"></a></dt>
<dd><p>A class that receives a DESeq output file and can filter it according to various characteristics.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the DESeq output file contents.         The DataFrame is modified upon usage of filter operations.</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
<dt>index_set: set</dt><dd><p>All of the indices in the current DataFrame (which were not removed by previously used filter methods)         as a set.</p>
</dd>
<dt>index_string: string</dt><dd><p>A string of all feature indices in the current DataFrame separated by newline.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="rnalysis.filtering.DESeqFilter.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Path</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">tuple</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">drop_columns</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2fc_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'log2FoldChange'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">padj_col</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'padj'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suppress_warnings</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#rnalysis.filtering.DESeqFilter.__init__" title="Permalink to this definition"></a></dt>
<dd><p>Load a differential expression table. A valid differential expression table should have         a column containing log2(fold change) values for each gene, and another column containing         adjusted p-values for each gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<em>Union</em><em>[</em><em>str</em><em>, </em><em>Path</em><em>]</em>) – full path/filename of the .csv file to be loaded into the Filter object</p></li>
<li><p><strong>drop_columns</strong> (<em>str</em><em>, </em><em>list</em><em> of </em><em>str</em><em>, or </em><em>None</em><em> (</em><em>default=None</em><em>)</em>) – if a string or list of strings are specified,         the columns of the same name/s will be dropped from the loaded table.</p></li>
<li><p><strong>log2fc_col</strong> (<em>str</em><em> (</em><em>default='Log2FoldChange'</em><em>)</em>) – name of the table column containing log2(fold change) values.</p></li>
<li><p><strong>padj_col</strong> (<em>str</em><em> (</em><em>default='padj'</em><em>)</em>) – name of the table column containing adjusted p-values.</p></li>
<li><p><strong>suppress_warnings</strong> (<em>bool</em><em> (</em><em>default=False</em><em>)</em>) – if True, RNAlysis will not issue warnings about the loaded table’s         structure or content.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.biotypes_from_gtf.html#rnalysis.filtering.DESeqFilter.biotypes_from_gtf" title="rnalysis.filtering.DESeqFilter.biotypes_from_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.biotypes_from_gtf</span></code></a>(gtf_path[, ...])</p></td>
<td><p>Returns a DataFrame describing the biotypes in the table and their count.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.biotypes_from_ref_table.html#rnalysis.filtering.DESeqFilter.biotypes_from_ref_table" title="rnalysis.filtering.DESeqFilter.biotypes_from_ref_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.biotypes_from_ref_table</span></code></a>([...])</p></td>
<td><p>Returns a DataFrame describing the biotypes in the table and their count.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.describe.html#rnalysis.filtering.DESeqFilter.describe" title="rnalysis.filtering.DESeqFilter.describe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.describe</span></code></a>([percentiles])</p></td>
<td><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.difference.html#rnalysis.filtering.DESeqFilter.difference" title="rnalysis.filtering.DESeqFilter.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.difference</span></code></a>(*others[, ...])</p></td>
<td><p>Keep only the features that exist in the first Filter object/set but NOT in the others.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.drop_columns.html#rnalysis.filtering.DESeqFilter.drop_columns" title="rnalysis.filtering.DESeqFilter.drop_columns"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.drop_columns</span></code></a>(columns[, inplace])</p></td>
<td><p>Drop specific columns from the table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_abs_log2_fold_change.html#rnalysis.filtering.DESeqFilter.filter_abs_log2_fold_change" title="rnalysis.filtering.DESeqFilter.filter_abs_log2_fold_change"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_abs_log2_fold_change</span></code></a>([...])</p></td>
<td><p>Filters out all features whose absolute log2 fold change is below the indicated threshold.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_biotype_from_gtf.html#rnalysis.filtering.DESeqFilter.filter_biotype_from_gtf" title="rnalysis.filtering.DESeqFilter.filter_biotype_from_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_biotype_from_gtf</span></code></a>(gtf_path)</p></td>
<td><p>Filters out all features that do not match the indicated biotype/biotypes         (for example: 'protein_coding', 'ncRNA', etc).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_biotype_from_ref_table.html#rnalysis.filtering.DESeqFilter.filter_biotype_from_ref_table" title="rnalysis.filtering.DESeqFilter.filter_biotype_from_ref_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_biotype_from_ref_table</span></code></a>([...])</p></td>
<td><p>Filters out all features that do not match the indicated biotype/biotypes         (for example: 'protein_coding', 'ncRNA', etc).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_attribute.html#rnalysis.filtering.DESeqFilter.filter_by_attribute" title="rnalysis.filtering.DESeqFilter.filter_by_attribute"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_by_attribute</span></code></a>([...])</p></td>
<td><p>Filters features according to user-defined attributes from an Attribute Reference Table.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_go_annotations.html#rnalysis.filtering.DESeqFilter.filter_by_go_annotations" title="rnalysis.filtering.DESeqFilter.filter_by_go_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_by_go_annotations</span></code></a>(go_ids)</p></td>
<td><p>Filters genes according to GO annotations, keeping only genes that are annotated with a specific GO term.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_kegg_annotations.html#rnalysis.filtering.DESeqFilter.filter_by_kegg_annotations" title="rnalysis.filtering.DESeqFilter.filter_by_kegg_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_by_kegg_annotations</span></code></a>(kegg_ids)</p></td>
<td><p>Filters genes according to KEGG pathways, keeping only genes that belong to specific KEGG pathway.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_by_row_name.html#rnalysis.filtering.DESeqFilter.filter_by_row_name" title="rnalysis.filtering.DESeqFilter.filter_by_row_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_by_row_name</span></code></a>(row_names[, ...])</p></td>
<td><p>Filter out specific rows from the table by their name (index).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_duplicate_ids.html#rnalysis.filtering.DESeqFilter.filter_duplicate_ids" title="rnalysis.filtering.DESeqFilter.filter_duplicate_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_duplicate_ids</span></code></a>([keep, ...])</p></td>
<td><p>Filter out rows with duplicate names/IDs (index).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_fold_change_direction.html#rnalysis.filtering.DESeqFilter.filter_fold_change_direction" title="rnalysis.filtering.DESeqFilter.filter_fold_change_direction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_fold_change_direction</span></code></a>([...])</p></td>
<td><p>Filters out features according to the direction in which they changed between the two conditions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_missing_values.html#rnalysis.filtering.DESeqFilter.filter_missing_values" title="rnalysis.filtering.DESeqFilter.filter_missing_values"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_missing_values</span></code></a>([columns, ...])</p></td>
<td><p>Remove all rows whose values in the specified columns are missing (NaN).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_percentile.html#rnalysis.filtering.DESeqFilter.filter_percentile" title="rnalysis.filtering.DESeqFilter.filter_percentile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_percentile</span></code></a>(percentile, column)</p></td>
<td><p>Removes all entries above the specified percentile in the specified column.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_significant.html#rnalysis.filtering.DESeqFilter.filter_significant" title="rnalysis.filtering.DESeqFilter.filter_significant"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_significant</span></code></a>([alpha, ...])</p></td>
<td><p>Removes all features which did not change significantly, according to the provided alpha.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.filter_top_n.html#rnalysis.filtering.DESeqFilter.filter_top_n" title="rnalysis.filtering.DESeqFilter.filter_top_n"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.filter_top_n</span></code></a>(by[, n, ascending, ...])</p></td>
<td><p>Sort the rows by the values of specified column or columns, then keep only the top 'n' rows.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.from_dataframe.html#rnalysis.filtering.DESeqFilter.from_dataframe" title="rnalysis.filtering.DESeqFilter.from_dataframe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.from_dataframe</span></code></a>(df, name[, ...])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.head.html#rnalysis.filtering.DESeqFilter.head" title="rnalysis.filtering.DESeqFilter.head"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.head</span></code></a>([n])</p></td>
<td><p>Return the first n rows of the Filter object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.intersection.html#rnalysis.filtering.DESeqFilter.intersection" title="rnalysis.filtering.DESeqFilter.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.intersection</span></code></a>(*others[, ...])</p></td>
<td><p>Keep only the features that exist in ALL of the given Filter objects/sets.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.majority_vote_intersection.html#rnalysis.filtering.DESeqFilter.majority_vote_intersection" title="rnalysis.filtering.DESeqFilter.majority_vote_intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.majority_vote_intersection</span></code></a>(*others)</p></td>
<td><p>Returns a set/string of the features that appear in at least         (majority_threhold * 100)% of the given Filter objects/sets.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.number_filters.html#rnalysis.filtering.DESeqFilter.number_filters" title="rnalysis.filtering.DESeqFilter.number_filters"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.number_filters</span></code></a>(column, operator, ...)</p></td>
<td><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.print_features.html#rnalysis.filtering.DESeqFilter.print_features" title="rnalysis.filtering.DESeqFilter.print_features"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.print_features</span></code></a>()</p></td>
<td><p>Print the feature indices in the Filter object, sorted by their current order in the FIlter object,         and separated by newline.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.save_csv.html#rnalysis.filtering.DESeqFilter.save_csv" title="rnalysis.filtering.DESeqFilter.save_csv"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.save_csv</span></code></a>([alt_filename])</p></td>
<td><p>Saves the current filtered data to a .csv file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.sort.html#rnalysis.filtering.DESeqFilter.sort" title="rnalysis.filtering.DESeqFilter.sort"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.sort</span></code></a>(by[, ascending, ...])</p></td>
<td><p>Sort the rows by the values of specified column or columns.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.split_by_attribute.html#rnalysis.filtering.DESeqFilter.split_by_attribute" title="rnalysis.filtering.DESeqFilter.split_by_attribute"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.split_by_attribute</span></code></a>(attributes[, ref])</p></td>
<td><p>Splits the features in the Filter object into multiple Filter objects,         each corresponding to one of the specified Attribute Reference Table attributes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.split_by_percentile.html#rnalysis.filtering.DESeqFilter.split_by_percentile" title="rnalysis.filtering.DESeqFilter.split_by_percentile"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.split_by_percentile</span></code></a>(percentile, ...)</p></td>
<td><p>Splits the features in the Filter object into two non-overlapping Filter objects:         one containing features below the specified percentile in the specfieid column,         and the other containing features about the specified percentile in the specified column.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.split_fold_change_direction.html#rnalysis.filtering.DESeqFilter.split_fold_change_direction" title="rnalysis.filtering.DESeqFilter.split_fold_change_direction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.split_fold_change_direction</span></code></a>()</p></td>
<td><p>Splits the features in the DESeqFilter object into two non-overlapping DESeqFilter         objects, based on the direction of their log2foldchange.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.symmetric_difference.html#rnalysis.filtering.DESeqFilter.symmetric_difference" title="rnalysis.filtering.DESeqFilter.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.symmetric_difference</span></code></a>(other[, ...])</p></td>
<td><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.tail.html#rnalysis.filtering.DESeqFilter.tail" title="rnalysis.filtering.DESeqFilter.tail"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.tail</span></code></a>([n])</p></td>
<td><p>Return the last n rows of the Filter object.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.text_filters.html#rnalysis.filtering.DESeqFilter.text_filters" title="rnalysis.filtering.DESeqFilter.text_filters"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.text_filters</span></code></a>(column, operator, value)</p></td>
<td><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.transform.html#rnalysis.filtering.DESeqFilter.transform" title="rnalysis.filtering.DESeqFilter.transform"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.transform</span></code></a>(function[, columns, ...])</p></td>
<td><p>Transform the values in the Filter object with the specified function.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.translate_gene_ids.html#rnalysis.filtering.DESeqFilter.translate_gene_ids" title="rnalysis.filtering.DESeqFilter.translate_gene_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.translate_gene_ids</span></code></a>(translate_to)</p></td>
<td><p>Translates gene names/IDs from one type to another.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.union.html#rnalysis.filtering.DESeqFilter.union" title="rnalysis.filtering.DESeqFilter.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.union</span></code></a>(*others[, return_type])</p></td>
<td><p>Returns a set/string of the union of features between multiple Filter objects/sets         (the features that exist in at least one of the Filter objects/sets).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.filtering.DESeqFilter.volcano_plot.html#rnalysis.filtering.DESeqFilter.volcano_plot" title="rnalysis.filtering.DESeqFilter.volcano_plot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">DESeqFilter.volcano_plot</span></code></a>([alpha, ...])</p></td>
<td><p>Plots a volcano plot (log2(fold change) vs -log10(adj.</p></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="rnalysis.filtering.CountFilter.violin_plot.html" class="btn btn-neutral float-left" title="rnalysis.filtering.CountFilter.violin_plot" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.filtering.DESeqFilter.biotypes_from_gtf.html" class="btn btn-neutral float-right" title="rnalysis.filtering.DESeqFilter.biotypes_from_gtf" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

</body>
</html>
