<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>rnalysis.enrichment.FeatureSet &mdash; RNAlysis 4.1.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />

  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=e6a937a4"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="_static/copybutton.js?v=5b738548"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.enrichment.FeatureSet.add" href="rnalysis.enrichment.FeatureSet.add.html" />
    <link rel="prev" title="rnalysis.enrichment" href="rnalysis.enrichment.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">üß¨ What is <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#see-rnalysis-in-action">üé• See <em>RNAlysis</em> in Action</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#key-features">üöÄ Key Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">üõ†Ô∏è How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">üìä How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="analysis_report.html">How to use the <em>RNAlysis</em> analysis reports</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">rnalysis.enrichment.FeatureSet</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#rnalysis.enrichment.FeatureSet"><code class="docutils literal notranslate"><span class="pre">FeatureSet</span></code></a><ul>
<li class="toctree-l4"><a class="reference internal" href="#rnalysis.enrichment.FeatureSet.__init__"><code class="docutils literal notranslate"><span class="pre">FeatureSet.__init__()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.add.html">rnalysis.enrichment.FeatureSet.add</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.biotypes_from_gtf.html">rnalysis.enrichment.FeatureSet.biotypes_from_gtf</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.biotypes_from_ref_table.html">rnalysis.enrichment.FeatureSet.biotypes_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.change_set_name.html">rnalysis.enrichment.FeatureSet.change_set_name</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.clear.html">rnalysis.enrichment.FeatureSet.clear</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.copy.html">rnalysis.enrichment.FeatureSet.copy</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.difference.html">rnalysis.enrichment.FeatureSet.difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.difference_update.html">rnalysis.enrichment.FeatureSet.difference_update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.discard.html">rnalysis.enrichment.FeatureSet.discard</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_biotype_from_gtf.html">rnalysis.enrichment.FeatureSet.filter_biotype_from_gtf</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_biotype_from_ref_table.html">rnalysis.enrichment.FeatureSet.filter_biotype_from_ref_table</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_by_attribute.html">rnalysis.enrichment.FeatureSet.filter_by_attribute</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_by_go_annotations.html">rnalysis.enrichment.FeatureSet.filter_by_go_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_by_kegg_annotations.html">rnalysis.enrichment.FeatureSet.filter_by_kegg_annotations</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.find_paralogs_ensembl.html">rnalysis.enrichment.FeatureSet.find_paralogs_ensembl</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.find_paralogs_panther.html">rnalysis.enrichment.FeatureSet.find_paralogs_panther</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.go_enrichment.html">rnalysis.enrichment.FeatureSet.go_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.intersection.html">rnalysis.enrichment.FeatureSet.intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.intersection_update.html">rnalysis.enrichment.FeatureSet.intersection_update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.isdisjoint.html">rnalysis.enrichment.FeatureSet.isdisjoint</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.issubset.html">rnalysis.enrichment.FeatureSet.issubset</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.issuperset.html">rnalysis.enrichment.FeatureSet.issuperset</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.kegg_enrichment.html">rnalysis.enrichment.FeatureSet.kegg_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_ensembl.html">rnalysis.enrichment.FeatureSet.map_orthologs_ensembl</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_orthoinspector.html">rnalysis.enrichment.FeatureSet.map_orthologs_orthoinspector</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_panther.html">rnalysis.enrichment.FeatureSet.map_orthologs_panther</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_phylomedb.html">rnalysis.enrichment.FeatureSet.map_orthologs_phylomedb</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.non_categorical_enrichment.html">rnalysis.enrichment.FeatureSet.non_categorical_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.pop.html">rnalysis.enrichment.FeatureSet.pop</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.remove.html">rnalysis.enrichment.FeatureSet.remove</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.save_txt.html">rnalysis.enrichment.FeatureSet.save_txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.symmetric_difference.html">rnalysis.enrichment.FeatureSet.symmetric_difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.symmetric_difference_update.html">rnalysis.enrichment.FeatureSet.symmetric_difference_update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.translate_gene_ids.html">rnalysis.enrichment.FeatureSet.translate_gene_ids</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.union.html">rnalysis.enrichment.FeatureSet.union</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.update.html">rnalysis.enrichment.FeatureSet.update</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.user_defined_enrichment.html">rnalysis.enrichment.FeatureSet.user_defined_enrichment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.RankedSet.html">rnalysis.enrichment.RankedSet</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.enrichment_bar_plot.html">rnalysis.enrichment.enrichment_bar_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.gene_ontology_graph.html">rnalysis.enrichment.gene_ontology_graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.kegg_pathway_graph.html">rnalysis.enrichment.kegg_pathway_graph</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.upset_plot.html">rnalysis.enrichment.upset_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.venn_diagram.html">rnalysis.enrichment.venn_diagram</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
      <li class="breadcrumb-item active">rnalysis.enrichment.FeatureSet</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/rnalysis.enrichment.FeatureSet.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnalysis-enrichment-featureset">
<h1>rnalysis.enrichment.FeatureSet<a class="headerlink" href="#rnalysis-enrichment-featureset" title="Link to this heading">ÔÉÅ</a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="rnalysis.enrichment.FeatureSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">rnalysis.enrichment.</span></span><span class="sig-name descname"><span class="pre">FeatureSet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><span class="pre">Filter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#rnalysis.enrichment.FeatureSet" title="Link to this definition">ÔÉÅ</a></dt>
<dd><p>Receives a filtered gene set and the set‚Äôs name (optional) and preforms various enrichment analyses on them.</p>
<dl class="py method">
<dt class="sig sig-object py" id="rnalysis.enrichment.FeatureSet.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">Set</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><span class="pre">Filter</span></a></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">set_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#rnalysis.enrichment.FeatureSet.__init__" title="Link to this definition">ÔÉÅ</a></dt>
<dd><dl class="field-list">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>gene_set</strong> (<em>filtering.Filter object</em><em>, </em><em>set</em><em> of </em><em>strings</em><em> or </em><em>list</em><em> of </em><em>strings</em>) ‚Äì the set of genomic features to be used in downstream analyses</p></li>
<li><p><strong>set_name</strong> (<em>str</em>) ‚Äì name of the FeatureSet</p></li>
</ul>
</dd>
<dt class="field-even">Examples<span class="colon">:</span></dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">enrichment</span><span class="p">,</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;gene1&#39;</span><span class="p">,</span><span class="s1">&#39;gene2&#39;</span><span class="p">,</span><span class="s1">&#39;gene2&#39;</span><span class="p">},</span> <span class="s1">&#39;name of my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filter_obj</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_other_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">filter_obj</span><span class="p">,</span> <span class="s1">&#39;name of my other set&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.add.html#rnalysis.enrichment.FeatureSet.add" title="rnalysis.enrichment.FeatureSet.add"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.add</span></code></a></p></td>
<td><p>Add an element to a set.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.biotypes_from_gtf.html#rnalysis.enrichment.FeatureSet.biotypes_from_gtf" title="rnalysis.enrichment.FeatureSet.biotypes_from_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.biotypes_from_gtf</span></code></a>(gtf_path[,¬†...])</p></td>
<td><p>Returns a DataFrame describing the biotypes in the table and their count.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.biotypes_from_ref_table.html#rnalysis.enrichment.FeatureSet.biotypes_from_ref_table" title="rnalysis.enrichment.FeatureSet.biotypes_from_ref_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.biotypes_from_ref_table</span></code></a>([ref])</p></td>
<td><p>Returns a DataFrame describing the biotypes in the gene set and their count.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.change_set_name.html#rnalysis.enrichment.FeatureSet.change_set_name" title="rnalysis.enrichment.FeatureSet.change_set_name"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.change_set_name</span></code></a>(new_name)</p></td>
<td><p>Change the 'set_name' of a FeatureSet to a new name.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.clear.html#rnalysis.enrichment.FeatureSet.clear" title="rnalysis.enrichment.FeatureSet.clear"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.clear</span></code></a></p></td>
<td><p>Remove all elements from this set.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.copy.html#rnalysis.enrichment.FeatureSet.copy" title="rnalysis.enrichment.FeatureSet.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.copy</span></code></a></p></td>
<td><p>Return a shallow copy of a set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.difference.html#rnalysis.enrichment.FeatureSet.difference" title="rnalysis.enrichment.FeatureSet.difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.difference</span></code></a>(*others)</p></td>
<td><p>Calculates the set difference of the indices from multiple FeatureSet objects         (the indices that appear in the first FeatureSet object but NOT in the other objects).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.difference_update.html#rnalysis.enrichment.FeatureSet.difference_update" title="rnalysis.enrichment.FeatureSet.difference_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.difference_update</span></code></a></p></td>
<td><p>Remove all elements of another set from this set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.discard.html#rnalysis.enrichment.FeatureSet.discard" title="rnalysis.enrichment.FeatureSet.discard"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.discard</span></code></a></p></td>
<td><p>Remove an element from a set if it is a member.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_biotype_from_gtf.html#rnalysis.enrichment.FeatureSet.filter_biotype_from_gtf" title="rnalysis.enrichment.FeatureSet.filter_biotype_from_gtf"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.filter_biotype_from_gtf</span></code></a>(gtf_path)</p></td>
<td><p>Filters out all features that do not match the indicated biotype/biotypes         (for example: 'protein_coding', 'ncRNA', etc).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_biotype_from_ref_table.html#rnalysis.enrichment.FeatureSet.filter_biotype_from_ref_table" title="rnalysis.enrichment.FeatureSet.filter_biotype_from_ref_table"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.filter_biotype_from_ref_table</span></code></a>([...])</p></td>
<td><p>Filters out all features that do not match the indicated biotype/biotypes         (for example: 'protein_coding', 'ncRNA', etc).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_by_attribute.html#rnalysis.enrichment.FeatureSet.filter_by_attribute" title="rnalysis.enrichment.FeatureSet.filter_by_attribute"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.filter_by_attribute</span></code></a>(attributes[,¬†...])</p></td>
<td><p>Filters features according to user-defined attributes from an Attribute Reference Table.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_by_go_annotations.html#rnalysis.enrichment.FeatureSet.filter_by_go_annotations" title="rnalysis.enrichment.FeatureSet.filter_by_go_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.filter_by_go_annotations</span></code></a>(go_ids)</p></td>
<td><p>Filters genes according to GO annotations, keeping only genes that are annotated with a specific GO term.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.filter_by_kegg_annotations.html#rnalysis.enrichment.FeatureSet.filter_by_kegg_annotations" title="rnalysis.enrichment.FeatureSet.filter_by_kegg_annotations"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.filter_by_kegg_annotations</span></code></a>(kegg_ids)</p></td>
<td><p>Filters genes according to KEGG pathways, keeping only genes that belong to specific KEGG pathway.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.find_paralogs_ensembl.html#rnalysis.enrichment.FeatureSet.find_paralogs_ensembl" title="rnalysis.enrichment.FeatureSet.find_paralogs_ensembl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.find_paralogs_ensembl</span></code></a>([organism,¬†...])</p></td>
<td><p>Find paralogs within the same species using the Ensembl database.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.find_paralogs_panther.html#rnalysis.enrichment.FeatureSet.find_paralogs_panther" title="rnalysis.enrichment.FeatureSet.find_paralogs_panther"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.find_paralogs_panther</span></code></a>([organism,¬†...])</p></td>
<td><p>Find paralogs within the same species using the PantherDB database.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.go_enrichment.html#rnalysis.enrichment.FeatureSet.go_enrichment" title="rnalysis.enrichment.FeatureSet.go_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.go_enrichment</span></code></a>(background_genes[,¬†...])</p></td>
<td><p>Calculates enrichment and depletion of the FeatureSet for Gene Ontology (GO) terms against a background set.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.intersection.html#rnalysis.enrichment.FeatureSet.intersection" title="rnalysis.enrichment.FeatureSet.intersection"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.intersection</span></code></a>(*others)</p></td>
<td><p>Calculates the set intersection of the indices from multiple FeatureSet objects         (the indices that exist in ALL of the FeatureSet objects).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.intersection_update.html#rnalysis.enrichment.FeatureSet.intersection_update" title="rnalysis.enrichment.FeatureSet.intersection_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.intersection_update</span></code></a></p></td>
<td><p>Update a set with the intersection of itself and another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.isdisjoint.html#rnalysis.enrichment.FeatureSet.isdisjoint" title="rnalysis.enrichment.FeatureSet.isdisjoint"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.isdisjoint</span></code></a></p></td>
<td><p>Return True if two sets have a null intersection.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.issubset.html#rnalysis.enrichment.FeatureSet.issubset" title="rnalysis.enrichment.FeatureSet.issubset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.issubset</span></code></a>(other,¬†/)</p></td>
<td><p>Test whether every element in the set is in other.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.issuperset.html#rnalysis.enrichment.FeatureSet.issuperset" title="rnalysis.enrichment.FeatureSet.issuperset"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.issuperset</span></code></a>(other,¬†/)</p></td>
<td><p>Test whether every element in other is in the set.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.kegg_enrichment.html#rnalysis.enrichment.FeatureSet.kegg_enrichment" title="rnalysis.enrichment.FeatureSet.kegg_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.kegg_enrichment</span></code></a>(background_genes)</p></td>
<td><p>Calculates enrichment and depletion of the FeatureSet for Kyoto Encyclopedia of Genes and Genomes (KEGG)         curated pathways against a background set.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_ensembl.html#rnalysis.enrichment.FeatureSet.map_orthologs_ensembl" title="rnalysis.enrichment.FeatureSet.map_orthologs_ensembl"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.map_orthologs_ensembl</span></code></a>(map_to_organism)</p></td>
<td><p>Map genes to their nearest orthologs in a different species using the Ensembl database.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_orthoinspector.html#rnalysis.enrichment.FeatureSet.map_orthologs_orthoinspector" title="rnalysis.enrichment.FeatureSet.map_orthologs_orthoinspector"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.map_orthologs_orthoinspector</span></code></a>(...)</p></td>
<td><p>Map genes to their nearest orthologs in a different species using the OrthoInspector database.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_panther.html#rnalysis.enrichment.FeatureSet.map_orthologs_panther" title="rnalysis.enrichment.FeatureSet.map_orthologs_panther"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.map_orthologs_panther</span></code></a>(map_to_organism)</p></td>
<td><p>Map genes to their nearest orthologs in a different species using the PantherDB database.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.map_orthologs_phylomedb.html#rnalysis.enrichment.FeatureSet.map_orthologs_phylomedb" title="rnalysis.enrichment.FeatureSet.map_orthologs_phylomedb"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.map_orthologs_phylomedb</span></code></a>(...[,¬†...])</p></td>
<td><p>Map genes to their nearest orthologs in a different species using the PhylomeDB database.         This function generates a table describing all matching discovered ortholog pairs (both unique and non-unique)         and returns it, and can also translate the genes in this data table into their nearest ortholog,         as well as remove unmapped genes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.non_categorical_enrichment.html#rnalysis.enrichment.FeatureSet.non_categorical_enrichment" title="rnalysis.enrichment.FeatureSet.non_categorical_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.non_categorical_enrichment</span></code></a>(...[,¬†...])</p></td>
<td><p>Calculates enrichment and depletion of the FeatureSet for user-defined non-categorical attributes         against a background set using either a one-sample T-test or Sign test.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.pop.html#rnalysis.enrichment.FeatureSet.pop" title="rnalysis.enrichment.FeatureSet.pop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.pop</span></code></a></p></td>
<td><p>Remove and return an arbitrary set element.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.remove.html#rnalysis.enrichment.FeatureSet.remove" title="rnalysis.enrichment.FeatureSet.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.remove</span></code></a></p></td>
<td><p>Remove an element from a set; it must be a member.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.save_txt.html#rnalysis.enrichment.FeatureSet.save_txt" title="rnalysis.enrichment.FeatureSet.save_txt"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.save_txt</span></code></a>(fname)</p></td>
<td><p>Save the list of features in the FeatureSet object under the specified filename and path.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.symmetric_difference.html#rnalysis.enrichment.FeatureSet.symmetric_difference" title="rnalysis.enrichment.FeatureSet.symmetric_difference"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.symmetric_difference</span></code></a>(other)</p></td>
<td><p>Calculates the set symmetric difference of the indices from two FeatureSet objects         (the indices that appear in EXACTLY ONE of the FeatureSet objects, and not both/neither).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.symmetric_difference_update.html#rnalysis.enrichment.FeatureSet.symmetric_difference_update" title="rnalysis.enrichment.FeatureSet.symmetric_difference_update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.symmetric_difference_update</span></code></a></p></td>
<td><p>Update a set with the symmetric difference of itself and another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.translate_gene_ids.html#rnalysis.enrichment.FeatureSet.translate_gene_ids" title="rnalysis.enrichment.FeatureSet.translate_gene_ids"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.translate_gene_ids</span></code></a>(translate_to)</p></td>
<td><p>Translates gene names/IDs from one type to another.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.union.html#rnalysis.enrichment.FeatureSet.union" title="rnalysis.enrichment.FeatureSet.union"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.union</span></code></a>(*others)</p></td>
<td><p>Calculates the set union of the indices from multiple FeatureSet objects         (the indices that exist in at least one of the FeatureSet objects).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.update.html#rnalysis.enrichment.FeatureSet.update" title="rnalysis.enrichment.FeatureSet.update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.update</span></code></a></p></td>
<td><p>Update a set with the union of itself and others.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.enrichment.FeatureSet.user_defined_enrichment.html#rnalysis.enrichment.FeatureSet.user_defined_enrichment" title="rnalysis.enrichment.FeatureSet.user_defined_enrichment"><code class="xref py py-obj docutils literal notranslate"><span class="pre">FeatureSet.user_defined_enrichment</span></code></a>(...[,¬†...])</p></td>
<td><p>Calculates enrichment and depletion of the FeatureSet for user-defined attributes against a background set.</p></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="rnalysis.enrichment.html" class="btn btn-neutral float-left" title="rnalysis.enrichment" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.enrichment.FeatureSet.add.html" class="btn btn-neutral float-right" title="rnalysis.enrichment.FeatureSet.add" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>