

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>rnalysis.enrichment.FeatureSet.enrich_randomization &mdash; RNAlysis 1.3.3 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.enrichment.FeatureSet.enrich_randomization_parallel" href="rnalysis.enrichment.FeatureSet.enrich_randomization_parallel.html" />
    <link rel="prev" title="rnalysis.enrichment.FeatureSet.difference" href="rnalysis.enrichment.FeatureSet.difference.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> RNAlysis
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">RNAlysis - RNA sequencing data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis_user_guide.html">RNAlysis user guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a><ul class="current">
<li class="toctree-l2 current"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.html">rnalysis.enrichment.FeatureSet</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.__init__.html">rnalysis.enrichment.FeatureSet.__init__</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.biotypes.html">rnalysis.enrichment.FeatureSet.biotypes</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.difference.html">rnalysis.enrichment.FeatureSet.difference</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">rnalysis.enrichment.FeatureSet.enrich_randomization</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.enrich_randomization_parallel.html">rnalysis.enrichment.FeatureSet.enrich_randomization_parallel</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.go_enrichment.html">rnalysis.enrichment.FeatureSet.go_enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.intersection.html">rnalysis.enrichment.FeatureSet.intersection</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.save_txt.html">rnalysis.enrichment.FeatureSet.save_txt</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.symmetric_difference.html">rnalysis.enrichment.FeatureSet.symmetric_difference</a></li>
<li class="toctree-l3"><a class="reference internal" href="rnalysis.enrichment.FeatureSet.union.html">rnalysis.enrichment.FeatureSet.union</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.upset_plot.html">rnalysis.enrichment.upset_plot</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.enrichment.venn_diagram.html">rnalysis.enrichment.venn_diagram</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.html">rnalysis package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="rnalysis.enrichment.html">rnalysis.enrichment</a> &raquo;</li>
        
          <li><a href="rnalysis.enrichment.FeatureSet.html">rnalysis.enrichment.FeatureSet</a> &raquo;</li>
        
      <li>rnalysis.enrichment.FeatureSet.enrich_randomization</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/rnalysis.enrichment.FeatureSet.enrich_randomization.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rnalysis-enrichment-featureset-enrich-randomization">
<h1>rnalysis.enrichment.FeatureSet.enrich_randomization<a class="headerlink" href="#rnalysis-enrichment-featureset-enrich-randomization" title="Permalink to this headline">¶</a></h1>
<dl class="method">
<dt id="rnalysis.enrichment.FeatureSet.enrich_randomization">
<code class="sig-prename descclassname">FeatureSet.</code><code class="sig-name descname">enrich_randomization</code><span class="sig-paren">(</span><em class="sig-param">attributes: Union[Iterable[str], str, Iterable[int], int] = None, fdr: float = 0.05, reps: int = 10000, biotype: str = 'protein_coding', background_genes=None, attr_ref_path: str = 'predefined', biotype_ref_path: str = 'predefined', save_csv: bool = False, fname=None</em><span class="sig-paren">)</span><a class="headerlink" href="#rnalysis.enrichment.FeatureSet.enrich_randomization" title="Permalink to this definition">¶</a></dt>
<dd><blockquote>
<div><p>Calculates enrichment scores, p-values and q-values         for enrichment and depletion of selected attributes from an Attribute Reference Table.         Background set is determined by either the input variable ‘background_genes’,         or by the input variable ‘biotype’ and a Biotype Reference Table.         P-values are calculated using a randomization test with the formula p = (successes + 1)/(repeats + 1).         P-values are corrected for multiple comparisons using         the Benjamini–Hochberg step-up procedure (original FDR method).         Enrichment/depletion is determined automatically by the calculated enrichment score:         if log2(enrichment score) is positive then enrichment is assumed,         and if log2(enrichment score) is negative then depletion is assumed.         In plots, for the clarity of display, complete depletion (linear enrichment = 0)         appears with the smallest value in the scale.</p>
<dl class="field-list simple">
<dt class="field-odd">type attributes</dt>
<dd class="field-odd"><p>str, int, iterable (list, tuple, set, etc) of str/int, or ‘all’.</p>
</dd>
<dt class="field-even">param attributes</dt>
<dd class="field-even"><p>An iterable of attribute names or attribute numbers         (according to their order in the Attribute Reference Table).         If ‘all’, all of the attributes in the Attribute Reference Table will be used.         If None, a manual input prompt will be raised.</p>
</dd>
<dt class="field-odd">type fdr</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param fdr</dt>
<dd class="field-even"><p>Indicates the FDR threshold for significance.</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.         10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">type attr_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param attr_ref_path</dt>
<dd class="field-even"><p>the path of the Attribute Reference Table from which user-defined attributes will be drawn.</p>
</dd>
<dt class="field-odd">type biotype_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param biotype_ref_path</dt>
<dd class="field-even"><p>the path of the Biotype Reference Table.         Will be used to generate background set if ‘biotype’ is specified.</p>
</dd>
<dt class="field-odd">type biotype</dt>
<dd class="field-odd"><p>str specifying a specific biotype, list/set of strings each specifying a biotype, or ‘all’.         Default ‘protein_coding’.</p>
</dd>
<dt class="field-even">param biotype</dt>
<dd class="field-even"><p>determines the background genes by their biotype. Requires specifying a Biotype Reference Table.         ‘all’ will include all genomic features in the reference table,         ‘protein_coding’ will include only protein-coding genes from the reference table, etc.         Cannot be specified together with ‘background_genes’.</p>
</dd>
<dt class="field-odd">type background_genes</dt>
<dd class="field-odd"><p>set of feature indices, filtering.Filter object, or enrichment.FeatureSet object</p>
</dd>
<dt class="field-even">param background_genes</dt>
<dd class="field-even"><p>a set of specific feature indices to be used as background genes.         Cannot be specified together with ‘biotype’.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:         r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list simple">
<dt class="field-odd">return</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<p>a pandas DataFrame with the indicated attribute names as rows/index, and the columns ‘log2_fold_enrichment’
and ‘pvalue’.</p>
<div class="figure align-center" id="id1">
<a class="reference internal image-reference" href="_images/enrichment_randomization.png"><img alt="_images/enrichment_randomization.png" src="_images/enrichment_randomization.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of enrich_randomization()</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="rnalysis.enrichment.FeatureSet.enrich_randomization_parallel.html" class="btn btn-neutral float-right" title="rnalysis.enrichment.FeatureSet.enrich_randomization_parallel" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="rnalysis.enrichment.FeatureSet.difference.html" class="btn btn-neutral float-left" title="rnalysis.enrichment.FeatureSet.difference" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Guy Teichman

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>