<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>rnalysis.fastq &mdash; RNAlysis 3.11.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="rnalysis.fastq.PairedEndPipeline" href="rnalysis.fastq.PairedEndPipeline.html" />
    <link rel="prev" title="History" href="history.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">rnalysis.fastq</a><ul>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.PairedEndPipeline.html">rnalysis.fastq.PairedEndPipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.SingleEndPipeline.html">rnalysis.fastq.SingleEndPipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.bowtie2_align_paired_end.html">rnalysis.fastq.bowtie2_align_paired_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.bowtie2_align_single_end.html">rnalysis.fastq.bowtie2_align_single_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.bowtie2_create_index.html">rnalysis.fastq.bowtie2_create_index</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.convert_sam_format.html">rnalysis.fastq.convert_sam_format</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.create_bam_index.html">rnalysis.fastq.create_bam_index</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.fastq_to_sam_paired.html">rnalysis.fastq.fastq_to_sam_paired</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.fastq_to_sam_single.html">rnalysis.fastq.fastq_to_sam_single</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.featurecounts_paired_end.html">rnalysis.fastq.featurecounts_paired_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.featurecounts_single_end.html">rnalysis.fastq.featurecounts_single_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.find_duplicates.html">rnalysis.fastq.find_duplicates</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.kallisto_create_index.html">rnalysis.fastq.kallisto_create_index</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.kallisto_quantify_paired_end.html">rnalysis.fastq.kallisto_quantify_paired_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.kallisto_quantify_single_end.html">rnalysis.fastq.kallisto_quantify_single_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.sam_to_fastq_paired.html">rnalysis.fastq.sam_to_fastq_paired</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.sam_to_fastq_single.html">rnalysis.fastq.sam_to_fastq_single</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.shortstack_align_smallrna.html">rnalysis.fastq.shortstack_align_smallrna</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.sort_sam.html">rnalysis.fastq.sort_sam</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.trim_adapters_paired_end.html">rnalysis.fastq.trim_adapters_paired_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.trim_adapters_single_end.html">rnalysis.fastq.trim_adapters_single_end</a></li>
<li class="toctree-l2"><a class="reference internal" href="rnalysis.fastq.validate_sam.html">rnalysis.fastq.validate_sam</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">rnalysis.fastq</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/rnalysis.fastq.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="rnalysis-fastq">
<h1>rnalysis.fastq<a class="headerlink" href="#rnalysis-fastq" title="Permalink to this heading">ÔÉÅ</a></h1>
<p class="rubric">Description</p>
<span class="target" id="module-rnalysis.fastq"></span><p>The <em>fastq</em> module provides a unified programmatic interface to external tools that process FASTQ files.
Those currently include the <em>CutAdapt</em> adapter-trimming tool, the <em>kallisto</em> RNA-sequencing quantification tool,
the <em>bowtie2</em> alignment tool, and the <em>featureCounts</em> feature counting tool.</p>
<p class="rubric">Classes</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.PairedEndPipeline.html#rnalysis.fastq.PairedEndPipeline" title="rnalysis.fastq.PairedEndPipeline"><code class="xref py py-obj docutils literal notranslate"><span class="pre">PairedEndPipeline</span></code></a>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.SingleEndPipeline.html#rnalysis.fastq.SingleEndPipeline" title="rnalysis.fastq.SingleEndPipeline"><code class="xref py py-obj docutils literal notranslate"><span class="pre">SingleEndPipeline</span></code></a>()</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p class="rubric">Functions</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.bowtie2_align_paired_end.html#rnalysis.fastq.bowtie2_align_paired_end" title="rnalysis.fastq.bowtie2_align_paired_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bowtie2_align_paired_end</span></code></a>(r1_files,¬†r2_files,¬†...)</p></td>
<td><p>Align paired-end reads from FASTQ files to a reference sequence using the     <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2">bowtie2</a> aligner.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.bowtie2_align_single_end.html#rnalysis.fastq.bowtie2_align_single_end" title="rnalysis.fastq.bowtie2_align_single_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bowtie2_align_single_end</span></code></a>(fastq_folder,¬†...)</p></td>
<td><p><p>Align single-end reads from FASTQ files to a reference sequence using the     <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2">bowtie2</a> aligner.</p>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.bowtie2_create_index.html#rnalysis.fastq.bowtie2_create_index" title="rnalysis.fastq.bowtie2_create_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">bowtie2_create_index</span></code></a>(genome_fastas,¬†...[,¬†...])</p></td>
<td><p>builds a bowtie index from FASTA formatted files of target sequences (genome).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.convert_sam_format.html#rnalysis.fastq.convert_sam_format" title="rnalysis.fastq.convert_sam_format"><code class="xref py py-obj docutils literal notranslate"><span class="pre">convert_sam_format</span></code></a>(input_folder,¬†output_folder)</p></td>
<td><p>Convert SAM files to BAM files or vice versa using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#SamFormatConverter">Picard SamFormatConverter</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.create_bam_index.html#rnalysis.fastq.create_bam_index" title="rnalysis.fastq.create_bam_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">create_bam_index</span></code></a>(input_folder,¬†output_folder)</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.fastq_to_sam_paired.html#rnalysis.fastq.fastq_to_sam_paired" title="rnalysis.fastq.fastq_to_sam_paired"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fastq_to_sam_paired</span></code></a>(r1_files,¬†r2_files,¬†...)</p></td>
<td><p>Convert SAM/BAM files to FASTQ files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#SamToFastq">Picard SamToFastq</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.fastq_to_sam_single.html#rnalysis.fastq.fastq_to_sam_single" title="rnalysis.fastq.fastq_to_sam_single"><code class="xref py py-obj docutils literal notranslate"><span class="pre">fastq_to_sam_single</span></code></a>(input_folder,¬†output_folder)</p></td>
<td><p><p>Convert SAM/BAM files to FASTQ files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#SamToFastq">Picard SamToFastq</a>.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.featurecounts_paired_end.html#rnalysis.fastq.featurecounts_paired_end" title="rnalysis.fastq.featurecounts_paired_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">featurecounts_paired_end</span></code></a>(input_folder,¬†...)</p></td>
<td><p>Assign mapped paired-end sequencing reads to specified genomic features using     <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btt656">RSubread featureCounts</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.featurecounts_single_end.html#rnalysis.fastq.featurecounts_single_end" title="rnalysis.fastq.featurecounts_single_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">featurecounts_single_end</span></code></a>(input_folder,¬†...)</p></td>
<td><p><p>Assign mapped single-end sequencing reads to specified genomic features using     <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btt656">RSubread featureCounts</a>.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.find_duplicates.html#rnalysis.fastq.find_duplicates" title="rnalysis.fastq.find_duplicates"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_duplicates</span></code></a>(input_folder,¬†output_folder)</p></td>
<td><p>Find duplicate reads in SAM/BAM files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#MarkDuplicates">Picard MarkDuplicates</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.kallisto_create_index.html#rnalysis.fastq.kallisto_create_index" title="rnalysis.fastq.kallisto_create_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kallisto_create_index</span></code></a>(transcriptome_fasta[,¬†...])</p></td>
<td><p>builds a kallisto index from a FASTA formatted file of target sequences (transcriptome).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.kallisto_quantify_paired_end.html#rnalysis.fastq.kallisto_quantify_paired_end" title="rnalysis.fastq.kallisto_quantify_paired_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kallisto_quantify_paired_end</span></code></a>(r1_files,¬†...)</p></td>
<td><p>Quantify transcript abundance in paired-end mRNA sequencing data using     <a class="reference external" href="https://pachterlab.github.io/kallisto/">kallisto</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.kallisto_quantify_single_end.html#rnalysis.fastq.kallisto_quantify_single_end" title="rnalysis.fastq.kallisto_quantify_single_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">kallisto_quantify_single_end</span></code></a>(fastq_folder,¬†...)</p></td>
<td><p><p>Quantify transcript abundance in single-end mRNA sequencing data using     <a class="reference external" href="https://pachterlab.github.io/kallisto/">kallisto</a>.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.sam_to_fastq_paired.html#rnalysis.fastq.sam_to_fastq_paired" title="rnalysis.fastq.sam_to_fastq_paired"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sam_to_fastq_paired</span></code></a>(input_folder,¬†output_folder)</p></td>
<td><p><p>Convert SAM/BAM files to FASTQ files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#SamToFastq">Picard SamToFastq</a>.</p>
</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.sam_to_fastq_single.html#rnalysis.fastq.sam_to_fastq_single" title="rnalysis.fastq.sam_to_fastq_single"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sam_to_fastq_single</span></code></a>(input_folder,¬†output_folder)</p></td>
<td><p><p>Convert SAM/BAM files to FASTQ files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#SamToFastq">Picard SamToFastq</a>.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.shortstack_align_smallrna.html#rnalysis.fastq.shortstack_align_smallrna" title="rnalysis.fastq.shortstack_align_smallrna"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shortstack_align_smallrna</span></code></a>(fastq_folder,¬†...)</p></td>
<td><p>Align small RNA single-end reads from FASTQ files to a reference sequence using the     <a class="reference external" href="https://github.com/MikeAxtell/ShortStack">ShortStack</a> aligner (version 4).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.sort_sam.html#rnalysis.fastq.sort_sam" title="rnalysis.fastq.sort_sam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort_sam</span></code></a>(input_folder,¬†output_folder[,¬†...])</p></td>
<td><p>Sort SAM/BAM files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#SortSam">Picard SortSam</a>.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.trim_adapters_paired_end.html#rnalysis.fastq.trim_adapters_paired_end" title="rnalysis.fastq.trim_adapters_paired_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim_adapters_paired_end</span></code></a>(r1_files,¬†r2_files,¬†...)</p></td>
<td><p>Trim adapters from paired-end reads using <a class="reference external" href="https://cutadapt.readthedocs.io/">CutAdapt</a>.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="rnalysis.fastq.trim_adapters_single_end.html#rnalysis.fastq.trim_adapters_single_end" title="rnalysis.fastq.trim_adapters_single_end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">trim_adapters_single_end</span></code></a>(fastq_folder,¬†...)</p></td>
<td><p><p>Trim adapters from single-end reads using <a class="reference external" href="https://cutadapt.readthedocs.io/">CutAdapt</a>.</p>
</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="rnalysis.fastq.validate_sam.html#rnalysis.fastq.validate_sam" title="rnalysis.fastq.validate_sam"><code class="xref py py-obj docutils literal notranslate"><span class="pre">validate_sam</span></code></a>(input_folder,¬†output_folder[,¬†...])</p></td>
<td><p>Validate SAM/BAM files using     <a class="reference external" href="https://broadinstitute.github.io/picard/command-line-overview.html#ValidateSamFile">Picard ValidateSamFile</a>.</p></td>
</tr>
</tbody>
</table>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="history.html" class="btn btn-neutral float-left" title="History" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="rnalysis.fastq.PairedEndPipeline.html" class="btn btn-neutral float-right" title="rnalysis.fastq.PairedEndPipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>