

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>rnalysis package &mdash; RNAlysis 1.3.3 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="rnalysis.general.update_settings" href="rnalysis.general.update_settings.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> RNAlysis
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">RNAlysis - RNA sequencing data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis_user_guide.html">RNAlysis user guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">rnalysis package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-filtering-module">rnalysis.filtering module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-enrichment-module">rnalysis.enrichment module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-general-module">rnalysis.general module</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>rnalysis package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/rnalysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rnalysis-package">
<h1>rnalysis package<a class="headerlink" href="#rnalysis-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rnalysis-filtering-module">
<h2>rnalysis.filtering module<a class="headerlink" href="#rnalysis-filtering-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="rnalysis-enrichment-module">
<h2>rnalysis.enrichment module<a class="headerlink" href="#rnalysis-enrichment-module" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="rnalysis-general-module">
<h2>rnalysis.general module<a class="headerlink" href="#rnalysis-general-module" title="Permalink to this headline">¶</a></h2>
<p>This module contains general-purpose functions, such as loading and saving files, removing table rows with no WBGene  index, identifying type of input variables, etc. This module is used mainly by other modules.</p>
<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">check_is_df</code><span class="sig-paren">(</span><em class="sig-param">inp</em><span class="sig-paren">)</span></dt>
<dd><p>checks whether an input file is a pandas DataFrame, a string that represent a path of a .csv file, a Path object     of a .csv file, or an invalid input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>inp</strong> – the input we wish to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if pandas DataFrame, False if a string/Path object that leads to a .csv file.    Raises ValueError otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">filter_low_rpm</code><span class="sig-paren">(</span><em class="sig-param">df: pandas.core.frame.DataFrame</em>, <em class="sig-param">threshold: float = 5</em><span class="sig-paren">)</span></dt>
<dd><p>remove all features which have less then ‘threshold’ reads per million in all conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas DataFrame to be filtered</p></li>
<li><p><strong>threshold</strong> – float. The minimal rpm a feature needs to have in at least one sample in order not to be filtered out.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A filtered DataFrame</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">get_settings_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd></dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">load_csv</code><span class="sig-paren">(</span><em class="sig-param">filename: str</em>, <em class="sig-param">idx_col: int = None</em>, <em class="sig-param">drop_columns: bool = False</em>, <em class="sig-param">squeeze=False</em>, <em class="sig-param">comment: str = None</em><span class="sig-paren">)</span></dt>
<dd><p>loads a csv df into a pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – name of the csv file to be loaded</p></li>
<li><p><strong>idx_col</strong> (<em>int</em><em>, </em><em>default None</em>) – number of column to be used as index. default is None, meaning no column will be used as index.</p></li>
<li><p><strong>drop_columns</strong> (<em>str</em><em>, </em><em>list of str</em><em>, or </em><em>False</em><em> (</em><em>default False</em><em>)</em>) – if a string or list of strings are specified,     the columns of the same name/s will be dropped from the loaded DataFrame.</p></li>
<li><p><strong>squeeze</strong> (<em>bool</em><em>, </em><em>default False</em>) – If the parsed data only contains one column then return a Series.</p></li>
<li><p><strong>comment</strong> (<em>str</em><em> (</em><em>optional</em><em>)</em>) – Indicates remainder of line should not be parsed.     If found at the beginning of a line, the line will be ignored altogether. This parameter must be a single character.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a pandas dataframe of the csv file</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">load_settings</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd></dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_gene_name_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains gene names (like ‘daf-2’ or ‘lin15B’).     Parses the string into a set of gene names.     The format of a gene name is a sequence consisting of the expression     ‘[a-z]{3,4}’, the character ‘-‘, and the expression ‘[A-Z,0-9]{1,4}’.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_sequence_name_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains sequence names (such as ‘Y55D5A.5’).     Parses the string into a set of WBGene indices.     The format of a sequence name is a sequence consisting of the expression ‘[A-Z,0-9]{5,6}’,     the character ‘.’, and a digit.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_wbgene_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains WBGene indices. Parses the string into a set of WBGene indices.     The format of a WBGene index is ‘WBGene’ and exactly 8 digits.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">read_attr_ref_table_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd></dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">read_biotype_ref_table_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd></dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">read_path_from_settings</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span></dt>
<dd><p>Attempt to read the reference table path from settings.yaml.     If the path was not previously defined, will prompt user to define it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>key</strong> (<em>str</em>) – the key in the settings file whose path to read.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>The path of the reference table.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">remove_unindexed_rows</code><span class="sig-paren">(</span><em class="sig-param">df: pandas.core.frame.DataFrame</em><span class="sig-paren">)</span></dt>
<dd><p>removes rows which have no WBGene index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>df</strong> – a DataFrame with WBGene indices</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a new DataFrame in which all rows with no WBGene index are removed</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">save_to_csv</code><span class="sig-paren">(</span><em class="sig-param">df: pandas.core.frame.DataFrame</em>, <em class="sig-param">filename: str</em>, <em class="sig-param">suffix: str = None</em>, <em class="sig-param">index: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>save a pandas DataFrame to csv.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas DataFrame to be saved</p></li>
<li><p><strong>filename</strong> – a string or pathlib.Path object stating the original name of the file</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>default None</em>) – A suffix to be added to the original name of the file. If None, no suffix will be added.</p></li>
<li><p><strong>index</strong> – if True, saves the DataFrame with the indices. If false, ignores the index.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">start_ipcluster</code><span class="sig-paren">(</span><em class="sig-param">n_engines: int = 'default'</em><span class="sig-paren">)</span></dt>
<dd><p>Start an ipyparallel ipcluster in order to perform parallelized computation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_engines</strong> (<em>int</em><em> or </em><em>'default'</em>) – if ‘default’, will initiate the default amount of engines.     Otherwise, will initiate n_engines engines.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">start_parallel_session</code><span class="sig-paren">(</span><em class="sig-param">n_engines: int = 'default'</em><span class="sig-paren">)</span></dt>
<dd><p>Stop previous ipyparallel ipcluster and start a new one in order to perform parallelized computation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_engines</strong> (<em>int</em><em> or </em><em>'default'</em>) – if ‘default’, will initiate the default amount of engines.     Otherwise, will initiate n_engines engines.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">stop_ipcluster</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Stop a previously started ipyparallel ipcluster.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">update_settings</code><span class="sig-paren">(</span><em class="sig-param">path: str</em>, <em class="sig-param">key</em><span class="sig-paren">)</span></dt>
<dd></dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="rnalysis.general.update_settings.html" class="btn btn-neutral float-left" title="rnalysis.general.update_settings" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Guy Teichman

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>