

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">

  <meta name="viewport" content="width=device-width, initial-scale=1.0">

  <title>rnalysis package &mdash; RNAlysis 1.3.1 documentation</title>








  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>


      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>

    <script type="text/javascript" src="_static/js/theme.js"></script>




  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="rnalysis.feature_name_to_wbgene.GeneNameTranslator.run" href="rnalysis.feature_name_to_wbgene.GeneNameTranslator.run.html" />
</head>

<body class="wy-body-for-nav">


  <div class="wy-grid-for-nav">

    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >



            <a href="index.html" class="icon icon-home"> RNAlysis



          </a>







<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>


        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">






              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">RNAlysis - <em>C. elegans</em> RNA sequencing data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis_user_guide.html">RNAlysis user guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.feature_name_to_wbgene.html">rnalysis.feature_name_to_wbgene</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">rnalysis package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-filtering-module">rnalysis.filtering module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-enrichment-module">rnalysis.enrichment module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-general-module">rnalysis.general module</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-feature-name-to-wbgene-module">rnalysis.feature_name_to_wbgene module</a></li>
</ul>
</li>
</ul>



        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">


      <nav class="wy-nav-top" aria-label="top navigation">

          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>

      </nav>


      <div class="wy-nav-content">

        <div class="rst-content">

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">

      <li><a href="index.html">Docs</a> &raquo;</li>

      <li>rnalysis package</li>


      <li class="wy-breadcrumbs-aside">


            <a href="_sources/rnalysis.rst.txt" rel="nofollow"> View page source</a>


      </li>

  </ul>


  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">

  <div class="section" id="rnalysis-package">
<h1>rnalysis package<a class="headerlink" href="#rnalysis-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rnalysis-filtering-module">
<h2>rnalysis.filtering module<a class="headerlink" href="#rnalysis-filtering-module" title="Permalink to this headline">¶</a></h2>
<p>This module can import HTCount and DESeq .csv files, perform various filtering operations on them,  perform set operations (union, intersection, etc), run basic exploratory analyses and plots (such as PCA, clustergram, violin plots, scatter, etc), save the filtered files to the computer, and return set of features that appear in an imported DESeq file. When a filtered/modified DESeq/HTCount is saved back to disk, its new name will include by default  all of the operations performed on it, in the order they were performed, to allow easy traceback of analyses.</p>
<dl class="class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">CountFilter</code><span class="sig-paren">(</span><em class="sig-param">fname: str</em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">rnalysis.filtering.Filter</span></code></a></p>
<p>A class that receives a count matrix and can filter it according to various characteristics.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the count matrix contents.         The DataFrame is modified upon usage of filter operations.</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
</dl>
<dl class="method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">clustergram</code><span class="sig-paren">(</span><em class="sig-param">sample_names: list = 'all'</em>, <em class="sig-param">metric: str = 'euclidean'</em>, <em class="sig-param">linkage: str = 'average'</em><span class="sig-paren">)</span></dt>
<dd><p>Runs and plots a clustergram on the base-2 log of a given set of samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_names</strong> (<em>'all'</em><em> or </em><em>list.</em>) – the names of the relevant samples in a list.         Example input: [“1A_N2_25”, “1B_N2_25”, “1C_N2_25”, “2A_rde4_25”, “2B_rde4_25”, “2C_rde4_25”]</p></li>
<li><p><strong>metric</strong> – the distance metric to use in the clustergram.         Example inputs: ‘euclidean’, ‘hamming’, ‘correlation’.         For all possible inputs see scipy.spatial.distance.pdist documentation online.</p></li>
<li><p><strong>linkage</strong> – the linkage method to use in the clustergram.         Example inputs: ‘single’, ‘average’, ‘complete’, ‘ward’.         For all possible inputs see scipy.cluster.hierarchy.linkage documentation online.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A seaborn clustermap object.</p>
<div class="figure align-center" id="id1">
<a class="reference internal image-reference" href="_images/clustergram.png"><img alt="_images/clustergram.png" src="_images/clustergram.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of clustergram()</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param">percentiles: list = [0.01, 0.25, 0.5, 0.75, 0.99]</em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>Summary statistics of the dataset.
:rtype: Series or DataFrame</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em>, <em class="sig-param">inplace: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param">biotype='protein_coding'</em>, <em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>str</em><em> or </em><em>list</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_by_ref_table_attr</code><span class="sig-paren">(</span><em class="sig-param">attributes: list = None</em>, <em class="sig-param">mode='union'</em>, <em class="sig-param">inclusion: bool = True</em>, <em class="sig-param">ref: str = 'predefined'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters features by inclusion in or exclusion from a Big Table attribute, or multiple Big Table attributes.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> – list of Big Table attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – ‘union’ or ‘intersection’. If ‘union’, filters out every feature that does not match at least one         of the indicated Big Table attributes. If ‘intersection’,         filters out every feature that does not match all of the indicated Big Table attributes.</p></li>
<li><p><strong>inclusion</strong> – boolean. If True (default), features will be filtered by inclusion in the specified criteria         (meaning features that  belong to the attributes are kept         and features that don’t belong to the attributes are filtered out).         If False, features will be filtered out by exclusion from the specified criteria         (meaning features that DON’T belong to the attributes         are kept, and features that do belong to the attributes are filtered out).</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_by_row_sum</code><span class="sig-paren">(</span><em class="sig-param">threshold: float = 5</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>remove all features which have less then ‘threshold’ reads per million in all conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>threshold</strong> (<em>float</em>) – The minimal sum a row should have in order not to be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of CountFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_low_reads</code><span class="sig-paren">(</span><em class="sig-param">threshold: float = 5</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>remove all features which have less then ‘threshold’ reads per million in all conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>threshold</strong> (<em>float</em>) – The minimal number of reads (counts, rpm, rpkm, tpm, etc) a feature should have         in at least one sample in order not to be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of CountFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param">percentile: float</em>, <em class="sig-param">column: str</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param">col: str</em>, <em class="sig-param">n: int = 100</em>, <em class="sig-param">ascending: bool = True</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all features except the ‘n’ most significantly-changed features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col</strong> (<em>str</em>) – The column to filter by</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – sort ascending (keep lowest values) or descending (keep highest values).</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">fold_change</code><span class="sig-paren">(</span><em class="sig-param">numerator</em>, <em class="sig-param">denominator</em>, <em class="sig-param">numer_name: str = 'default'</em>, <em class="sig-param">denom_name: str = 'default'</em><span class="sig-paren">)</span></dt>
<dd><p>Calculate the fold change between the numerator condition and the denominator condition,         and return it as a FoldChangeFilter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>numerator</strong> (<em>str</em><em>, or </em><em>list of strs</em>) – the CountFilter columns to be used as the numerator. If multiple arguments are given         in a list, they will be averaged.</p></li>
<li><p><strong>denominator</strong> (<em>str</em><em>, or </em><em>list of strs</em>) – the CountFilter columns to be used as the denominator. If multiple arguments are given         in a list, they will be averaged.</p></li>
<li><p><strong>numer_name</strong> (<em>str</em><em> or </em><em>'default'</em>) – name to give the numerator condition. If ‘default’, the name will be generarated         automatically from the names of numerator columns.</p></li>
<li><p><strong>denom_name</strong> (<em>str</em><em> or </em><em>'default'</em>) – name to give the denominator condition. If ‘default’, the name will be generarated         automatically from the names of denominator columns.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="rnalysis.filtering.FoldChangeFilter.html#rnalysis.filtering.FoldChangeFilter" title="rnalysis.filtering.FoldChangeFilter">FoldChangeFilter</a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A new instance of FoldChangeFilter</p>
</dd></dl>

<dl class="method">
<dt>
<em class="property">static </em><code class="sig-name descname">from_folder</code><span class="sig-paren">(</span><em class="sig-param">folder_path: str</em>, <em class="sig-param">save_csv: bool = True</em>, <em class="sig-param">norm_to_rpm: bool = False</em>, <em class="sig-param">save_reads_fname: str = None</em>, <em class="sig-param">save_not_counted_fname: str = None</em>, <em class="sig-param">input_format: str = '.txt'</em><span class="sig-paren">)</span></dt>
<dd><p>Iterates over HTSeq count .txt files in a given folder and combines them into a single CountFilter table.         Can also save the count data table and the uncounted data table to .csv files, and normalize the CountFilter         table to reads per million (RPM). Note that the saved data will always be count data, and not normalized data,         regardless if the CountFilter table was normalized or not.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>folder_path</strong> – str or pathlib.Path. Full path of the folder that contains individual htcount .txt files.</p></li>
<li><p><strong>save_csv</strong> – bool. If True (default), the joint DataFrame of count data and uncounted data will be saved         to two separate .csv files. The files will be saved in ‘folder_path’, and named according to the parameters         ‘save_reads_fname’ for the count data, and ‘save_not_counted_fname’ for the uncounted data (unaligned,         alignment not unique, etc).</p></li>
<li><p><strong>norm_to_rpm</strong> – bool. If True, the CountFilter table will be automatically normalized to reads per         million (RPM). If False (defualt), the CountFilter object will not be normalized, and will instead contain         absolute count data (as in the original htcount .txt files).         Note that if save_csv is True, the saved .csv fill will contain ABSOLUTE COUNT DATA, as in the original         htcount .txt files, and NOT normalized data.</p></li>
<li><p><strong>save_reads_fname</strong> – str. Name under which to save the combined count data table. Does not need to include         the ‘.csv’ suffix.</p></li>
<li><p><strong>save_not_counted_fname</strong> – save_reads_fname: str. Name under which to save the combined uncounted data.         Does not need to include the ‘.csv’ suffix.</p></li>
<li><p><strong>input_format</strong> – the file format of the input files. Default is ‘.txt’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>an CountFilter object containing the combined count data from all individual htcount .txt files in the         specified folder.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param">n=5</em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>returns the first n rows of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">index_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; set</dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         Warning: if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A set of WBGene names.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">index_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str</dt>
<dd><p>Returns a string of all feature indices in the current DataFrame separated by newline. </p>
<p>This includes all of the feature indices which were not filtered out by previously-used filter methods. Warning: if any duplicate features exist in the filter object (same index appears more than once), the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A string of WBGene indices separated by newlines (\n). For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em>, <em class="sig-param">inplace: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">norm_reads_to_rpm</code><span class="sig-paren">(</span><em class="sig-param">special_counter_fname: str</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Normalizes the reads in the CountFilter to reads per million (RPM).         Uses a table of feature counts (ambiguous, no feature, not aligned, etc) from HTSeq’s output.         Divides each column in the CountFilter object by (total reads + ambiguous + no feature)*10^-6 .</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>special_counter_fname</strong> – the .csv file which contains feature information about the RNA library         (ambiguous, no feature, not aligned, etc).</p></li>
<li><p><strong>inplace</strong> – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If inplace is False, returns a new instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">norm_reads_with_size_factor</code><span class="sig-paren">(</span><em class="sig-param">size_factor_fname: str</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Normalizes the reads in the CountFilter using pre-calculated size factors.         Such size factors can be calculated using DESeq2’s median of ratios method.         Receives a table of sample names and their corresponding size factors,         and divides each column in the CountFilter object dataframe by the corresponding size factor.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>size_factor_fname</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – the .csv file which contains size factors for the different libraries.</p></li>
<li><p><strong>inplace</strong> – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If inplace is False, returns a new instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param">column: str</em>, <em class="sig-param">operator: str</em>, <em class="sig-param">value</em>, <em class="sig-param">opposite=False</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt'/'greater than'/'&gt;'. 'eq'/'equal'/'='</em><em>, </em><em>'lt'/'lesser than'/'&lt;'</em>) – the operator to filter the column by (larger than, equal or smaller than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
<p>Example usage: filt.number_filter(‘baseMean’,’gt’,57, inplace=False)         will return a Filter object in which all rows have a value greater than 57 in the column ‘baseMean’.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">pairplot</code><span class="sig-paren">(</span><em class="sig-param">sample_list: list = 'all'</em>, <em class="sig-param">log2: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Plot pairwise relationships in the dataset.         Can plot both single samples and average multiple replicates.         For more information see the documentation of seaborn.pairplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_list</strong> (<em>'all'</em><em>, </em><em>list</em><em>, or </em><em>nested list.</em>) – A list of the sample names and/or grouped sample names to be included in the pairplot.         All specified samples must be present in the CountFilter object.         To average multiple replicates of the same condition, they can be grouped in an inner list.         Example input:         [[‘SAMPLE1A’, ‘SAMPLE1B’, ‘SAMPLE1C’], [‘SAMPLE2A’, ‘SAMPLE2B’, ‘SAMPLE2C’],’SAMPLE3’ , ‘SAMPLE6’]</p></li>
<li><p><strong>log2</strong> (<em>bool</em>) – if True, the pairplot will be calculated with log2 of the dataframe, and not with the raw data.         If False (default), the pairplot will be calculated with the raw data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>An instance of seaborn.PairGrid.</p>
<div class="figure align-center" id="id2">
<a class="reference internal image-reference" href="_images/pairplot.png"><img alt="_images/pairplot.png" src="_images/pairplot.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of pairplot()</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">pca</code><span class="sig-paren">(</span><em class="sig-param">sample_names: list = 'all'</em>, <em class="sig-param">n_components=3</em>, <em class="sig-param">sample_grouping: list = None</em><span class="sig-paren">)</span></dt>
<dd><p>runs and plots a PCA for a given set of samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_names</strong> (<em>'all'</em><em> or </em><em>list.</em>) – the names of the relevant samples in a list.         Example input: [“1_REP_A”, “1_REP_B”, “1_REP_C”, “2_REP_A”, “2_REP_B”, “2_REP_C”, “2_REP_D”, “3_REP_A”]</p></li>
<li><p><strong>n_components</strong> – number of PCA components to return         :param sample_grouping: Optional. Indicates which samples are grouped together as replicates,         so they will be colored similarly in the PCA plot. A list of indices from 0 and up, that indicates the sample          grouping.          For example, if sample_names is:         [“1_REP_A”, “1_REP_B”, “1_REP_C”, “2_REP_A”, “2_REP_B”, “2_REP_C”, “2_REP_D”, “3_REP_A”],         then the sample_grouping will be:         [0, 0, 0, 1, 1, 1, 1, 2]</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A tuple whose first element is an sklearn.decomposition.pca object,         and second element is a list of matplotlib.axis objects.</p>
<div class="figure align-center" id="id3">
<a class="reference internal image-reference" href="_images/pca.png"><img alt="_images/pca.png" src="_images/pca.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of pca()</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">plot_expression</code><span class="sig-paren">(</span><em class="sig-param">features: list</em>, <em class="sig-param">sample_grouping: dict</em>, <em class="sig-param">count_unit: str = 'Reads per million'</em><span class="sig-paren">)</span></dt>
<dd><p>Plot the average expression and standard error of the specified features under the specified conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features</strong> (<em>str</em><em> or </em><em>list of strings</em>) – the feature/features to plot expression for.</p></li>
<li><p><strong>sample_grouping</strong> (<em>dict</em><em>, </em><em>with condition names as keys         and list of the sample numbers for each condition as a list</em>) – a dictionary of the conditions to plot expression for.         Each key should be a name of a conditions, and the value for each key is         a list of the numbers of columns to be used as samples of that condition.         For example, if the first 3 columns are replicates of the condition ‘condition 1’ and         the last 3 column are replicates of the condition ‘condition 2’, then sample_grouping should be:         {‘condition 1’:[0, 1, 2], ‘condition 2’:[3, 4, 5]}</p></li>
<li><p><strong>count_unit</strong> (<em>str</em><em>, </em><em>default 'Reads per million'</em>) – The unit of the count data. Will be displayed in the y axis.</p></li>
</ul>
</dd>
</dl>
<div class="figure align-center" id="id4">
<a class="reference internal image-reference" href="_images/plot_expression.png"><img alt="_images/plot_expression.png" src="_images/plot_expression.png" style="width: 403.6px; height: 294.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Example plot of plot_expression()</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Print the feature indices in the Filter object, separated by newline (‘</p>
</div></blockquote>
<p>‘).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param">alt_filename=None</em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> – If None, file name will be generated automatically according to the filtering methods used.         if string, will be the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">scatter_sample_vs_sample</code><span class="sig-paren">(</span><em class="sig-param">sample1: str</em>, <em class="sig-param">sample2: str</em>, <em class="sig-param">xlabel: str = None</em>, <em class="sig-param">ylabel: str = None</em>, <em class="sig-param">deseq_highlight=None</em><span class="sig-paren">)</span></dt>
<dd><p>Generate a scatter plot where every dot is a feature, the x value is log10 of reads         (counts, RPM, RPKM, TPM, etc) in sample1, the y value is log10 of reads in sample2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample1</strong> – str/list. Name of the first sample from the CountFilter object.         If sample1 is a list, they will be avarged as replicates.</p></li>
<li><p><strong>sample2</strong> – str/list. Name of the second sample from the CountFilter object.         If sample2 is a list, they will be averaged as replicates.</p></li>
<li><p><strong>xlabel</strong> – optional. If not specified, sample1 will be used as xlabel.</p></li>
<li><p><strong>ylabel</strong> – optional. If not specified, sample2 will be used as ylabel.</p></li>
<li><p><strong>deseq_highlight</strong> – DESeqFilter object or iterable of WBGene indices(optional).         If specified, the points in the scatter corresponding to the WBGene indices in deseq_highlight will be         highlighted in red.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a matplotlib axis object.</p>
<div class="figure align-center" id="id5">
<a class="reference internal image-reference" href="_images/rpm_vs_rpm.png"><img alt="_images/rpm_vs_rpm.png" src="_images/rpm_vs_rpm.png" style="width: 525.6px; height: 550.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of scatter_sample_vs_sample()</span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param">percentile: float</em>, <em class="sig-param">column: str</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         above and below the specified percentile in the spcfieid column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a list of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_reads</code><span class="sig-paren">(</span><em class="sig-param">threshold: float = 5</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the features in the current CountFilter object into two complementary, non-overlapping CountFilter         objects, based on the their maximum expression level. The first object will contain only highly-expressed          features (which have reads over the specified threshold in at least one sample). The second object will          contain only lowly-expressed features (which have reads below the specified threshold in all samples).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>threshold</strong> – A float. The minimal number of reads (counts, RPM, RPKM, TPM etc) a feature needs to have         in at least one sample in order to be         included in the “highly expressed” object and no the “lowly expressed” object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A tuple containing two CountFilter objects: the first has only highly-expressed features,         and the second has only lowly-expressed features.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_ref_table_attr</code><span class="sig-paren">(</span><em class="sig-param">attributes: tuple = None</em>, <em class="sig-param">ref: str = 'predefined'</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Big Table attributes.         Each object contains only features that match its indicated Big Table attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> – list of Big Table attributes to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A list of Filter objects, each containing only features that match one Big Table attribute; the Filter objects         appear in the list in the same order the Big Table attributes were given in.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">return_type: str = 'set'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param">n=5</em><span class="sig-paren">)</span></dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>returns the last n rows of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param">column: str</em>, <em class="sig-param">operator: str</em>, <em class="sig-param">value: str</em>, <em class="sig-param">opposite=False</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq'/'equal'/'='</em><em>, </em><em>'ct'/'contains'/'in'</em><em>, </em><em>'bw'/'begins with'</em><em>, </em><em>'ew'/'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
<p>Example usage: filt.text_filters(‘name’,’sw’,’pseudo’, inplace=False)         will return a Filter object in which all rows have a value that starts with ‘pseudo’ in the column ‘name’.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">violin_plot</code><span class="sig-paren">(</span><em class="sig-param">samples='all'</em><span class="sig-paren">)</span></dt>
<dd><p>Generates a violin plot of the specified samples in the CountFilter object.         Can plot both single samples and average multiple replicates.         It is recommended to use this function on normalized values and not on absolute read values.         Box inside the violin plot indicates 25% and 75% percentiles, and the white dot indicates the median.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>samples</strong> (<em>'all'</em><em> or </em><em>list.</em>) – A list of the sample names and/or grouped sample names to be plotted in the violin plot.         All specified samples must be present in the CountFilter object.         To average multiple replicates of the same condition, they can be grouped in an inner list.         Example input:         [[‘SAMPLE1A’, ‘SAMPLE1B’, ‘SAMPLE1C’], [‘SAMPLE2A’, ‘SAMPLE2B’, ‘SAMPLE2C’],’SAMPLE3’ , ‘SAMPLE6’]</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a seaborn violin object.</p>
<div class="figure align-center" id="id6">
<a class="reference internal image-reference" href="_images/violin.png"><img alt="_images/violin.png" src="_images/violin.png" style="width: 1152.0px; height: 556.1999999999999px;" /></a>
<p class="caption"><span class="caption-text">Example plot of violin_plot()</span><a class="headerlink" href="#id6" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">DESeqFilter</code><span class="sig-paren">(</span><em class="sig-param">fname: str</em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">rnalysis.filtering.Filter</span></code></a></p>
<p>A class that receives a DESeq output file and can filter it according to various characteristics.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the DESeq output file contents.         The DataFrame is modified upon usage of filter operations. .</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
</dl>
<dl class="method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param">percentiles: list = [0.01, 0.25, 0.5, 0.75, 0.99]</em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>Summary statistics of the dataset.
:rtype: Series or DataFrame</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em>, <em class="sig-param">inplace: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_abs_log2_fold_change</code><span class="sig-paren">(</span><em class="sig-param">abslog2fc: float = 1</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features whose absolute log2 fold change is below the indicated threshold.         For example: if log2fc is 2.0, all features whose log2 fold change is between 1 and -1 (went up less than         two-fold or went down less than two-fold) will be filtered out.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>abslog2fc</strong> – The threshold absolute log2 fold change for filtering out a feature. Float or int.         All features whose absolute log2 fold change is lower than log2fc will be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current DESeqFilter object. If False,         the function will return a new DESeqFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of DESeqFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param">biotype='protein_coding'</em>, <em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>str</em><em> or </em><em>list</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_by_ref_table_attr</code><span class="sig-paren">(</span><em class="sig-param">attributes: list = None</em>, <em class="sig-param">mode='union'</em>, <em class="sig-param">inclusion: bool = True</em>, <em class="sig-param">ref: str = 'predefined'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters features by inclusion in or exclusion from a Big Table attribute, or multiple Big Table attributes.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> – list of Big Table attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – ‘union’ or ‘intersection’. If ‘union’, filters out every feature that does not match at least one         of the indicated Big Table attributes. If ‘intersection’,         filters out every feature that does not match all of the indicated Big Table attributes.</p></li>
<li><p><strong>inclusion</strong> – boolean. If True (default), features will be filtered by inclusion in the specified criteria         (meaning features that  belong to the attributes are kept         and features that don’t belong to the attributes are filtered out).         If False, features will be filtered out by exclusion from the specified criteria         (meaning features that DON’T belong to the attributes         are kept, and features that do belong to the attributes are filtered out).</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_fold_change_direction</code><span class="sig-paren">(</span><em class="sig-param">direction: str = 'pos'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out features according to the direction in which they changed between the two conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>direction</strong> – ‘pos’ or ‘neg’. If ‘pos’, will keep only features that have positive log2foldchange.         If ‘neg’, will keep only features that have negative log2foldchange.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current DESeqFilter object. If False,         the function will return a new DESeqFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of DESeqFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param">percentile: float</em>, <em class="sig-param">column: str</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_significant</code><span class="sig-paren">(</span><em class="sig-param">alpha: float = 0.1</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all features which did not change significantly, according to the provided alpha.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alpha</strong> – the significance threshold to determine which genes will be filtered. between 0 and 1.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current DESeqFilter object. If False,         the function will return a new DESeqFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of DESeqFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param">col: str</em>, <em class="sig-param">n: int = 100</em>, <em class="sig-param">ascending: bool = True</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all features except the ‘n’ most significantly-changed features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col</strong> (<em>str</em>) – The column to filter by</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – sort ascending (keep lowest values) or descending (keep highest values).</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param">n=5</em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>returns the first n rows of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">index_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; set</dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         Warning: if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A set of WBGene names.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">index_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str</dt>
<dd><p>Returns a string of all feature indices in the current DataFrame separated by newline. </p>
<p>This includes all of the feature indices which were not filtered out by previously-used filter methods. Warning: if any duplicate features exist in the filter object (same index appears more than once), the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A string of WBGene indices separated by newlines (\n). For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em>, <em class="sig-param">inplace: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param">column: str</em>, <em class="sig-param">operator: str</em>, <em class="sig-param">value</em>, <em class="sig-param">opposite=False</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt'/'greater than'/'&gt;'. 'eq'/'equal'/'='</em><em>, </em><em>'lt'/'lesser than'/'&lt;'</em>) – the operator to filter the column by (larger than, equal or smaller than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
<p>Example usage: filt.number_filter(‘baseMean’,’gt’,57, inplace=False)         will return a Filter object in which all rows have a value greater than 57 in the column ‘baseMean’.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Print the feature indices in the Filter object, separated by newline (‘</p>
</div></blockquote>
<p>‘).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param">alt_filename=None</em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> – If None, file name will be generated automatically according to the filtering methods used.         if string, will be the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param">percentile: float</em>, <em class="sig-param">column: str</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         above and below the specified percentile in the spcfieid column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a list of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_ref_table_attr</code><span class="sig-paren">(</span><em class="sig-param">attributes: tuple = None</em>, <em class="sig-param">ref: str = 'predefined'</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Big Table attributes.         Each object contains only features that match its indicated Big Table attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> – list of Big Table attributes to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A list of Filter objects, each containing only features that match one Big Table attribute; the Filter objects         appear in the list in the same order the Big Table attributes were given in.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_fold_change_direction</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Splits the features in the current DESeqFilter object into two complementary, non-overlapping DESeqFilter         objects, based on the direction of their log2foldchange. The first object will contain only features with a         positive log2foldchange, the second object will contain only features with a negative log2foldchange.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>a tuple containing two DESeqFilter objects: the first has only features with positive log2 fold change,         and the other has only features with negative log2 fold change.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">return_type: str = 'set'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param">n=5</em><span class="sig-paren">)</span></dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>returns the last n rows of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param">column: str</em>, <em class="sig-param">operator: str</em>, <em class="sig-param">value: str</em>, <em class="sig-param">opposite=False</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq'/'equal'/'='</em><em>, </em><em>'ct'/'contains'/'in'</em><em>, </em><em>'bw'/'begins with'</em><em>, </em><em>'ew'/'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
<p>Example usage: filt.text_filters(‘name’,’sw’,’pseudo’, inplace=False)         will return a Filter object in which all rows have a value that starts with ‘pseudo’ in the column ‘name’.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">volcano_plot</code><span class="sig-paren">(</span><em class="sig-param">alpha: float = 0.1</em><span class="sig-paren">)</span></dt>
<dd><p>Plots a volcano plot (log2(fold change) vs -log10(adj. p-value)) of the DESeqFilter object.         Significantly upregulated features are colored in red,         and significantly downregulated features are colored in blue.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alpha</strong> (<em>float between 0 and 1</em>) – the significance threshold to color data points as significantly up/down-regulated.</p>
</dd>
</dl>
<div class="figure align-center" id="id7">
<a class="reference internal image-reference" href="_images/volcano.png"><img alt="_images/volcano.png" src="_images/volcano.png" style="width: 448.0px; height: 333.2px;" /></a>
<p class="caption"><span class="caption-text">Example plot of volcano_plot()</span><a class="headerlink" href="#id7" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">FoldChangeFilter</code><span class="sig-paren">(</span><em class="sig-param">fname: str</em>, <em class="sig-param">numerator_name: str</em>, <em class="sig-param">denominator_name: str</em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">rnalysis.filtering.Filter</span></code></a></p>
<p>A class that contains a single column, representing the gene-specific fold change between two conditions.
Warning: this class does not support ‘inf’ and ‘0’ values, and importing a file with such values could lead     to incorrect filtering and statistical analyses.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the fold change values.         The DataFrame is modified upon usage of filter operations. .</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
<dt>numerator: str</dt><dd><p>Name of the numerator used to calculate the fold change.</p>
</dd>
<dt>denominator: str</dt><dd><p>Name of the denominator used to calculate the fold change.</p>
</dd>
</dl>
<dl class="method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param">percentiles: list = [0.01, 0.25, 0.5, 0.75, 0.99]</em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>Summary statistics of the dataset.
:rtype: Series or DataFrame</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em>, <em class="sig-param">inplace: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_abs_log2_fold_change</code><span class="sig-paren">(</span><em class="sig-param">abslog2fc: float = 1</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features whose absolute log2 fold change is below the indicated threshold.         For example: if log2fc is 1.0, all features whose log2 fold change is between 1 and -1 (went up less than         two-fold or went down less than two-fold) will be filtered out.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>abslog2fc</strong> – The threshold absolute log2 fold change for filtering out a feature. Float or int.         All features whose absolute log2 fold change is lower than log2fc will be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current FoldChangeFilter object. If False,         the function will return a new FoldChangeFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of FoldChangeFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param">biotype='protein_coding'</em>, <em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>str</em><em> or </em><em>list</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_by_ref_table_attr</code><span class="sig-paren">(</span><em class="sig-param">attributes: list = None</em>, <em class="sig-param">mode='union'</em>, <em class="sig-param">inclusion: bool = True</em>, <em class="sig-param">ref: str = 'predefined'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters features by inclusion in or exclusion from a Big Table attribute, or multiple Big Table attributes.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> – list of Big Table attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – ‘union’ or ‘intersection’. If ‘union’, filters out every feature that does not match at least one         of the indicated Big Table attributes. If ‘intersection’,         filters out every feature that does not match all of the indicated Big Table attributes.</p></li>
<li><p><strong>inclusion</strong> – boolean. If True (default), features will be filtered by inclusion in the specified criteria         (meaning features that  belong to the attributes are kept         and features that don’t belong to the attributes are filtered out).         If False, features will be filtered out by exclusion from the specified criteria         (meaning features that DON’T belong to the attributes         are kept, and features that do belong to the attributes are filtered out).</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_fold_change_direction</code><span class="sig-paren">(</span><em class="sig-param">direction: str = 'pos'</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Filters out features according to the direction in which they changed between the two conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>direction</strong> – ‘pos’ or ‘neg’. If ‘pos’, will keep only features that have positive log2foldchange.         If ‘neg’, will keep only features that have negative log2foldchange.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current FoldChangeFilter object. If False,         the function will return a new FoldChangeFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of FoldChangeFilter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param">percentile: float</em>, <em class="sig-param">column: str</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param">col: str</em>, <em class="sig-param">n: int = 100</em>, <em class="sig-param">ascending: bool = True</em>, <em class="sig-param">opposite: bool = False</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Removes all features except the ‘n’ most significantly-changed features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>col</strong> (<em>str</em>) – The column to filter by</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – sort ascending (keep lowest values) or descending (keep highest values).</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param">n=5</em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>returns the first n rows of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">index_set</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; set</dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         Warning: if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A set of WBGene names.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">index_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span> &#x2192; str</dt>
<dd><p>Returns a string of all feature indices in the current DataFrame separated by newline. </p>
<p>This includes all of the feature indices which were not filtered out by previously-used filter methods. Warning: if any duplicate features exist in the filter object (same index appears more than once), the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>A string of WBGene indices separated by newlines (\n). For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em>, <em class="sig-param">inplace: bool = False</em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param">column: str</em>, <em class="sig-param">operator: str</em>, <em class="sig-param">value</em>, <em class="sig-param">opposite=False</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt'/'greater than'/'&gt;'. 'eq'/'equal'/'='</em><em>, </em><em>'lt'/'lesser than'/'&lt;'</em>) – the operator to filter the column by (larger than, equal or smaller than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
<p>Example usage: filt.number_filter(‘baseMean’,’gt’,57, inplace=False)         will return a Filter object in which all rows have a value greater than 57 in the column ‘baseMean’.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Print the feature indices in the Filter object, separated by newline (‘</p>
</div></blockquote>
<p>‘).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">randomization_test</code><span class="sig-paren">(</span><em class="sig-param">ref</em>, <em class="sig-param">alpha: float = 0.05</em>, <em class="sig-param">reps=10000</em>, <em class="sig-param">save_csv: bool = False</em>, <em class="sig-param">fname=None</em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Perform a randomization test to examine whether the fold change of a group of specific genomic features         is significantly different than the fold change of a background set of genomic features.</p>
<dl class="field-list simple">
<dt class="field-odd">type ref</dt>
<dd class="field-odd"><p>FoldChangeFilter</p>
</dd>
<dt class="field-even">param ref</dt>
<dd class="field-even"><p>A reference FoldChangeFilter object which contains the fold change for every reference gene.         Will be used to calculate the expected score and to perform randomizations.</p>
</dd>
<dt class="field-odd">type alpha</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param alpha</dt>
<dd class="field-even"><p>Indicates the threshold for significance (alpha).</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.         10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:         r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list simple">
<dt class="field-odd">rtype</dt>
<dd class="field-odd"><p>pandas DataFrame</p>
</dd>
<dt class="field-even">return</dt>
<dd class="field-even"><p></p></dd>
</dl>
<p>A Dataframe with the number of given genes, the observed fold change for the given group of genes,         the expected fold change for a group of genes of that size and the p value for the comparison.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param">alt_filename=None</em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> – If None, file name will be generated automatically according to the filtering methods used.         if string, will be the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param">percentile: float</em>, <em class="sig-param">column: str</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         above and below the specified percentile in the spcfieid column.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a list of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_by_ref_table_attr</code><span class="sig-paren">(</span><em class="sig-param">attributes: tuple = None</em>, <em class="sig-param">ref: str = 'predefined'</em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Big Table attributes.         Each object contains only features that match its indicated Big Table attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> – list of Big Table attributes to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A list of Filter objects, each containing only features that match one Big Table attribute; the Filter objects         appear in the list in the same order the Big Table attributes were given in.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">split_fold_change_direction</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Splits the features in the current FoldChangeFilter object into two complementary, non-overlapping         FoldChangeFilter objects, based on the direction of their log2foldchange.         The first object will contain only features with a positive log2foldchange,         the second object will contain only features with a negative log2foldchange.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>a tuple containing two FoldChangeFilter objects: the first has only features with positive log2 fold change,         and the other has only features with negative log2 fold change.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">return_type: str = 'set'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param">n=5</em><span class="sig-paren">)</span></dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>returns the last n rows of the Filter object.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param">column: str</em>, <em class="sig-param">operator: str</em>, <em class="sig-param">value: str</em>, <em class="sig-param">opposite=False</em>, <em class="sig-param">inplace=True</em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq'/'equal'/'='</em><em>, </em><em>'ct'/'contains'/'in'</em><em>, </em><em>'bw'/'begins with'</em><em>, </em><em>'ew'/'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
<p>Example usage: filt.text_filters(‘name’,’sw’,’pseudo’, inplace=False)         will return a Filter object in which all rows have a value that starts with ‘pseudo’ in the column ‘name’.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">return_type: str = 'set'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="rnalysis-enrichment-module">
<h2>rnalysis.enrichment module<a class="headerlink" href="#rnalysis-enrichment-module" title="Permalink to this headline">¶</a></h2>
<p>This module receives a set of feature names, and can perform various enrichment analyses on them, such as GO enrichment, tissue/phenotype enrichment, enrichment and depletion of Reference Table attributes, etc. Furthermore, set operations (union, intersect, difference, symmetric difference) can be performed on two FeatureSet objects. Results of enrichment analyses can be saved to .csv files.</p>
<dl class="class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.enrichment.</code><code class="sig-name descname">FeatureSet</code><span class="sig-paren">(</span><em class="sig-param">gene_set: set = None</em>, <em class="sig-param">set_name: str = ''</em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>receives a filtered gene set and preforms various enrichment analyses</p>
<dl class="method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param">ref: str = 'C:\\Users\\Guy_Teichman\\PycharmProjects\\RNAlysis\\rnalysis\\gene_names_and_biotype.csv'</em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the gene set and their count.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref</strong> – Name of the reference file used to determine biotype. Default is ce11 (included in the package).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Calculates the set difference of the WBGene indices from multiple FeatureSet objects         (the indices that appear in the first FeatureSet object but NOT in the other objects).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – The objects against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">enrich_randomization</code><span class="sig-paren">(</span><em class="sig-param">attributes: list = None</em>, <em class="sig-param">fdr: float = 0.05</em>, <em class="sig-param">reps=10000</em>, <em class="sig-param">biotype: str = 'protein_coding'</em>, <em class="sig-param">background_genes: set = None</em>, <em class="sig-param">ref_path: str = 'predefined'</em>, <em class="sig-param">save_csv: bool = False</em>, <em class="sig-param">fname=None</em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Calculates enrichment scores, p-values and q-values         for enrichment and depletion of selected attributes from a reference table.         P-values are calculated using a randomization test with the formula p = (successes + 1)/(repeats + 1).         P-values are corrected for multiple comparisons using         the Benjamini–Hochberg step-up procedure (original FDR method).         Enrichment/depletion is determined automatically by the calculated enrichment score:         if log2(enrichment score) is positive then enrichment is assumed,         and if log2(enrichment score) is negative then depletion is assumed.         In plots, for the clarity of display, complete depletion (linear enrichment = 0)         appears with the smallest value in the scale.</p>
<dl class="field-list simple">
<dt class="field-odd">type attributes</dt>
<dd class="field-odd"><p>str, int, iterable (list, tuple, set, etc) of str/int, or ‘all’.</p>
</dd>
<dt class="field-even">param attributes</dt>
<dd class="field-even"><p>An iterable of attribute names or attribute numbers         (according to their order in the reference table).         If ‘all’, all of the attributes in the reference table will be used.         If None, a manual input prompt will be raised.</p>
</dd>
<dt class="field-odd">type fdr</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param fdr</dt>
<dd class="field-even"><p>Indicates the FDR threshold for significance.</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.         10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">type ref_path</dt>
<dd class="field-odd"><p>‘predefined’ (default), str or pathlib.Path</p>
</dd>
<dt class="field-even">param ref_path</dt>
<dd class="field-even"><p>the path of the reference table file to be used as reference.</p>
</dd>
<dt class="field-odd">type biotype</dt>
<dd class="field-odd"><p>str specifying a specific biotype, list/set of strings each specifying a biotype, or ‘all’.         Default ‘protein_coding’.</p>
</dd>
<dt class="field-even">param biotype</dt>
<dd class="field-even"><p>the biotype you want your background genes to have. ‘all’ will include all biotypes,         ‘protein_coding’ will include only protein-coding genes in the reference, etc.         Cannot be specified together with ‘background_genes’.</p>
</dd>
<dt class="field-odd">type background_genes</dt>
<dd class="field-odd"><p>set of feature indices, filtering.Filter object, or enrichment.FeatureSet object</p>
</dd>
<dt class="field-even">param background_genes</dt>
<dd class="field-even"><p>a set of specific feature indices to be used as background genes.         Cannot be specified together with ‘biotype’.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:         r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list simple">
<dt class="field-odd">return</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<p>a pandas DataFrame with the indicated attribute names as rows/index, and the columns ‘log2_fold_enrichment’
and ‘pvalue’.</p>
<div class="figure align-center" id="id8">
<a class="reference internal image-reference" href="_images/bigtable_en.png"><img alt="_images/bigtable_en.png" src="_images/bigtable_en.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of enrich_randomization</span><a class="headerlink" href="#id8" title="Permalink to this image">¶</a></p>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">enrich_randomization_parallel</code><span class="sig-paren">(</span><em class="sig-param">attributes: list = None</em>, <em class="sig-param">fdr: float = 0.05</em>, <em class="sig-param">reps=10000</em>, <em class="sig-param">biotype: str = 'protein_coding'</em>, <em class="sig-param">background_genes: set = None</em>, <em class="sig-param">ref_path: str = 'predefined'</em>, <em class="sig-param">save_csv: bool = False</em>, <em class="sig-param">fname=None</em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><blockquote>
<div><p>Calculates enrichment scores, p-values and q-values         for enrichment and depletion of selected attributes from a reference table using parallel processing.         Parallel processing makes this function generally faster than FeatureSet.enrich_randomization.         To use it you must first start an ipcluster, using rnalysis.general.start_ipcluster().         P-values are calculated using a randomization test with the formula p = (successes + 1)/(repeats + 1).         P-values are corrected for multiple comparisons using         the Benjamini–Hochberg step-up procedure (original FDR method).         Enrichment/depletion is determined automatically by the calculated enrichment score:         if log2(enrichment score) is positive then enrichment is assumed,         and if log2(enrichment score) is negative then depletion is assumed.         In plots, for the clarity of display, complete depletion (linear enrichment = 0)         appears with the smallest value in the scale.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">type attributes</dt>
<dd class="field-odd"><p>str, int, iterable (list, tuple, set, etc) of str/int, or ‘all’</p>
</dd>
<dt class="field-even">param attributes</dt>
<dd class="field-even"><p>An iterable of attribute names or attribute numbers        (according to their order in the reference table).        If ‘all’, all of the attributes in the reference table will be used.        If None, a manual input prompt will be raised.</p>
</dd>
<dt class="field-odd">type fdr</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param fdr</dt>
<dd class="field-even"><p>Indicates the FDR threshold for significance.</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.        10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">param ref_path</dt>
<dd class="field-odd"><p>the path of the Big Table file to be used as reference.</p>
</dd>
<dt class="field-even">type biotype</dt>
<dd class="field-even"><p>str specifying a specific biotype, or ‘all’. Default ‘protein_coding’.</p>
</dd>
<dt class="field-odd">param biotype</dt>
<dd class="field-odd"><p>the biotype you want your background genes to have. ‘all’ will include all biotypes,        ‘protein_coding’ will include only protein-coding genes in the reference, etc.        Cannot be specified together with ‘background_genes’.</p>
</dd>
<dt class="field-even">type background_genes</dt>
<dd class="field-even"><p>set of feature indices, filtering.Filter object, or enrichment.FeatureSet object</p>
</dd>
<dt class="field-odd">param background_genes</dt>
<dd class="field-odd"><p>a set of specific feature indices to be used as background genes.        Cannot be specified together with ‘biotype’.</p>
</dd>
<dt class="field-even">type save_csv</dt>
<dd class="field-even"><p>bool, default False</p>
</dd>
<dt class="field-odd">param save_csv</dt>
<dd class="field-odd"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-even">type fname</dt>
<dd class="field-even"><p>str or pathlib.Path</p>
</dd>
<dt class="field-odd">param fname</dt>
<dd class="field-odd"><p>The full path and name of the file to which to save the results. For example:        r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list simple">
<dt class="field-odd">return</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<p>a pandas DataFrame with the indicated attribute names as rows/index, and the columns ‘log2_fold_enrichment’
and ‘pvalue’.</p>
<div class="figure align-center" id="id9">
<a class="reference internal image-reference" href="_images/bigtable_en.png"><img alt="_images/bigtable_en.png" src="_images/bigtable_en.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of big table enrichment</span><a class="headerlink" href="#id9" title="Permalink to this image">¶</a></p>
</div>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">go_enrichment</code><span class="sig-paren">(</span><em class="sig-param">mode: str = 'go'</em>, <em class="sig-param">alpha: float = 0.05</em>, <em class="sig-param">save_csv: bool = False</em>, <em class="sig-param">fname: str = None</em><span class="sig-paren">)</span></dt>
<dd><p>Analyzes GO, Tissue and/or Phenotype enrichment of the given group of features.         Uses the the Anatomy, Phenotype and Gene Ontology annotations for C. elegans.         Corrected p-values are calculated using hypergeometric statistics.         For more details see GitHub page of the developers: <a class="reference external" href="https://github.com/dangeles/TissueEnrichmentAnalysis">https://github.com/dangeles/TissueEnrichmentAnalysis</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> – the enrichment you wish to perform. ‘go’ for gene ontology enrichment,         ‘tissue’ for tissue enrichment, ‘phenotype’ for phenotype enrichment.</p></li>
<li><p><strong>alpha</strong> – float. Significance threshold. Default is 0.05</p></li>
<li><p><strong>save_csv</strong> – bool. False by default. If True, save the result to a csv.</p></li>
<li><p><strong>fname</strong> – Name and path in which to save the results. Must be filled if save_csv is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a DataFrame which contains the significant enrichmenet terms</p>
<div class="figure align-center" id="id10">
<a class="reference internal image-reference" href="_images/go_en.png"><img alt="_images/go_en.png" src="_images/go_en.png" style="width: 560.0px; height: 318.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Example plot of GO enrichment</span><a class="headerlink" href="#id10" title="Permalink to this image">¶</a></p>
</div>
<div class="figure align-center" id="id11">
<a class="reference internal image-reference" href="_images/tissue_en.png"><img alt="_images/tissue_en.png" src="_images/tissue_en.png" style="width: 560.0px; height: 318.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Example plot of Tissue enrichment</span><a class="headerlink" href="#id11" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Calculates the set intersection of the WBGene indices from multiple FeatureSet objects         (the indices that exist in ALL of the FeatureSet objects).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – The objects against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">save_txt</code><span class="sig-paren">(</span><em class="sig-param">fname</em><span class="sig-paren">)</span></dt>
<dd><p>Save the list of features in the FeatureSet object under the specified filename and path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fname</strong> (<em>str</em><em>, </em><em>pathlib.Path</em>) – full filename/path for the output file. Can include the ‘.txt’ suffix but doesn’t have to.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param">other</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>Calculates the set symmetric difference of the WBGene indices from two FeatureSet objects         (the indices that appear in EXACTLY ONE of the FeatureSet objects, and not both/neither).         A-symmetric difference-B is equivalent to (A-difference-B)-union-(B-difference-A).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – A second  object against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param">*others</em>, <em class="sig-param">inplace: bool = True</em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Calculates the set union of the WBGene indices from multipple FeatureSet objects         (the indices that exist in at least one of the FeatureSet objects).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – The objects against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="rnalysis-general-module">
<h2>rnalysis.general module<a class="headerlink" href="#rnalysis-general-module" title="Permalink to this headline">¶</a></h2>
<p>This module contains general-purpose functions, such as loading and saving files, removing table rows with no WBGene  index, identifying type of input variables, etc. This module is used mainly by other modules.</p>
<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">check_is_df</code><span class="sig-paren">(</span><em class="sig-param">inp</em><span class="sig-paren">)</span></dt>
<dd><p>checks whether an input file is a pandas DataFrame, a string that represent a path of a .csv file, a Path object     of a .csv file, or an invalid input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>inp</strong> – the input we wish to test</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if pandas DataFrame, False if a string/Path object that leads to a .csv file.    Raises ValueError otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">filter_low_rpm</code><span class="sig-paren">(</span><em class="sig-param">df: pandas.core.frame.DataFrame</em>, <em class="sig-param">threshold: float = 5</em><span class="sig-paren">)</span></dt>
<dd><p>remove all features which have less then ‘threshold’ reads per million in all conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas DataFrame to be filtered</p></li>
<li><p><strong>threshold</strong> – float. The minimal rpm a feature needs to have in at least one sample in order not to be filtered out.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>A filtered DataFrame</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">is_reference_table_defined</code><span class="sig-paren">(</span><em class="sig-param">settings_start_phrase='reference_table_path='</em><span class="sig-paren">)</span></dt>
<dd><p>Check whether a reference table path is defined.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>settings_start_phrase</strong> – The key phrase to look for in settings.ini (‘reference_table_path=’).</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>True if defined, False if not defined.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">load_csv</code><span class="sig-paren">(</span><em class="sig-param">filename: str</em>, <em class="sig-param">idx_col: int = None</em>, <em class="sig-param">drop_gene_names: bool = True</em>, <em class="sig-param">squeeze=False</em>, <em class="sig-param">comment: str = None</em><span class="sig-paren">)</span></dt>
<dd><p>loads a csv df into a pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – name of the csv file to be loaded</p></li>
<li><p><strong>idx_col</strong> (<em>int</em><em>, </em><em>default None</em>) – number of column to be used as index. default is None, meaning no column will be used as index.</p></li>
<li><p><strong>drop_gene_names</strong> (<em>bool</em><em>, </em><em>default True</em>) – bool. If True, tries to drop the ‘genes’ column from the DataFrame (useful for filtering).</p></li>
<li><p><strong>squeeze</strong> (<em>bool</em><em>, </em><em>default False</em>) – If the parsed data only contains one column then return a Series.</p></li>
<li><p><strong>comment</strong> (<em>str</em><em> (</em><em>optional</em><em>)</em>) – Indicates remainder of line should not be parsed.     If found at the beginning of a line, the line will be ignored altogether. This parameter must be a single character.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a pandas dataframe of the csv file</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">norm_reads_to_rpm</code><span class="sig-paren">(</span><em class="sig-param">htcount: pandas.core.frame.DataFrame</em>, <em class="sig-param">features: pandas.core.frame.DataFrame</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a dataframe of reads/counts and a dataframe of feature counts (ambiguous, no feature, not aligned, etc).     Divides each column in the read dataframe by (total reads + ambiguous + no feature)*10^-6 in order to normalize the     reads to RPM (reads per million).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>htcount</strong> – DataFrame of reads/counts</p></li>
<li><p><strong>features</strong> – the corresponding DataFrame of</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
<p>a DataFrame normalized to RPM</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">normalize_file</code><span class="sig-paren">(</span><em class="sig-param">htcount_filename: str</em>, <em class="sig-param">feature_filename: str</em><span class="sig-paren">)</span></dt>
<dd><p>receives an htcount filename and a feature filename, normalizes the htcount file to rpm, removes unindexed rows,     and saves the file with the suffix ‘_rpm’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>htcount_filename</strong> – str or pathlib.Path object. Name of the htcount file to be normalized</p></li>
<li><p><strong>feature_filename</strong> – str or pathlib.path object. Name of the corresponding feature file to be used for normalizing</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_gene_name_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains gene names (like ‘daf-2’ or ‘lin15B’).     Parses the string into a set of gene names.     The format of a gene name is a sequence consisting of the expression     ‘[a-z]{3,4}’, the character ‘-‘, and the expression ‘[A-Z,0-9]{1,4}’.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_sequence_name_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains sequence names (such as ‘Y55D5A.5’).     Parses the string into a set of WBGene indices.     The format of a sequence name is a sequence consisting of the expression ‘[A-Z,0-9]{5,6}’,     the character ‘.’, and a digit.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_wbgene_string</code><span class="sig-paren">(</span><em class="sig-param">string</em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains WBGene indices. Parses the string into a set of WBGene indices.     The format of a WBGene index is ‘WBGene’ and exactly 8 digits.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">read_reference_table_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Attempt to read the reference table path from settings.ini.     If the path was not previously defined, will prompt user to define it.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p></p>
</dd>
</dl>
<p>The path of the reference table.</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">remove_unindexed_rows</code><span class="sig-paren">(</span><em class="sig-param">df: pandas.core.frame.DataFrame</em><span class="sig-paren">)</span></dt>
<dd><p>removes rows which have no WBGene index.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>df</strong> – a DataFrame with WBGene indices</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a new DataFrame in which all rows with no WBGene index are removed</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">save_to_csv</code><span class="sig-paren">(</span><em class="sig-param">df: pandas.core.frame.DataFrame</em>, <em class="sig-param">filename: str</em>, <em class="sig-param">suffix: str = None</em>, <em class="sig-param">index: bool = True</em><span class="sig-paren">)</span></dt>
<dd><p>save a pandas DataFrame to csv.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas DataFrame to be saved</p></li>
<li><p><strong>filename</strong> – a string or pathlib.Path object stating the original name of the file</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>default None</em>) – A suffix to be added to the original name of the file. If None, no suffix will be added.</p></li>
<li><p><strong>index</strong> – if True, saves the DataFrame with the indices. If false, ignores the index.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">set_reference_table_path</code><span class="sig-paren">(</span><em class="sig-param">path: str = None</em><span class="sig-paren">)</span></dt>
<dd><p>Set the path for a user-defined reference table for filtering and enrichment analyses.     The path will be saved in a settings.ini file that will persist through sessions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>path</strong> (<em>str</em>) – The full path and name of the .csv reference file.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">start_ipcluster</code><span class="sig-paren">(</span><em class="sig-param">n_engines: int = 'default'</em><span class="sig-paren">)</span></dt>
<dd><p>Start an ipyparallel ipcluster in order to perform parallelized computation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_engines</strong> (<em>int</em><em> or </em><em>'default'</em>) – if ‘default’, will initiate the default amount of engines.     Otherwise, will initiate n_engines engines.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">start_parallel_session</code><span class="sig-paren">(</span><em class="sig-param">n_engines: int = 'default'</em><span class="sig-paren">)</span></dt>
<dd><p>Stop previous ipyparallel ipcluster and start a new one in order to perform parallelized computation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_engines</strong> (<em>int</em><em> or </em><em>'default'</em>) – if ‘default’, will initiate the default amount of engines.     Otherwise, will initiate n_engines engines.</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">stop_ipcluster</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Stop a previously started ipyparallel ipcluster.</p>
</dd></dl>

</div>
<div class="section" id="rnalysis-feature-name-to-wbgene-module">
<h2>rnalysis.feature_name_to_wbgene module<a class="headerlink" href="#rnalysis-feature-name-to-wbgene-module" title="Permalink to this headline">¶</a></h2>
<p>This module can append WBGene indices to feature names in a DESeq/HTCount .csv file, according to a reference table (such as the Big Table).</p>
<dl class="class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.feature_name_to_wbgene.</code><code class="sig-name descname">GeneNameTranslator</code><span class="sig-paren">(</span><em class="sig-param">target_filename</em>, <em class="sig-param">reference_filename: str = 'C:\Users\Guy_Teichman\PycharmProjects\RNAlysis\rnalysis\gene_names_and_biotype.csv'</em>, <em class="sig-param">ref_wbgene_col: str = 'gene'</em>, <em class="sig-param">ref_gene_col: str = 'gene ID'</em>, <em class="sig-param">ref_seq_col: str = 'Sequence ID'</em>, <em class="sig-param">ref_other_col: str = 'Other IDs'</em>, <em class="sig-param">target_gene_col: str = 'genes'</em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class that takes a file with gene names that does not have WBGene indices, and a reference file that includes
gene names and their corresponding WBGene indices, and adds WBGene indices to the original file.</p>
<dl class="method">
<dt>
<code class="sig-name descname">generate_target_dict</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Generate the feature name to WBGene dictionary,
and translate the imported pandas DataFrame feature names to WBGene.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">import_target</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>import the target filename into a pandas DataFrame</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">read_reference</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>load the reference file into a pandas DataFrame</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">rewrite_target</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>save the target file as a csv with the suffix “_WBGene”.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-name descname">run</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>load the specified file, translate its feature names to WBGenes, and save it as a csv with the “_WBGene” suffix.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.feature_name_to_wbgene.</code><code class="sig-name descname">name_to_wbgene_folder</code><span class="sig-paren">(</span><em class="sig-param">pth: str</em><span class="sig-paren">)</span></dt>
<dd><p>iterate over all files in a directory, translate their feature names to WBGene
and save them as csv in the same folder.
:param pth: path of the folder on which to iterate
:return:
a list of all GeneNameTranslator objects</p>
</dd></dl>

<dl class="function">
<dt>
<code class="sig-prename descclassname">rnalysis.feature_name_to_wbgene.</code><code class="sig-name descname">run_all</code><span class="sig-paren">(</span><em class="sig-param">path</em><span class="sig-paren">)</span></dt>
<dd><p>run the translator on a list of files</p>
</dd></dl>

</div>
</div>


           </div>

          </div>
          <footer>

    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">


        <a href="rnalysis.feature_name_to_wbgene.GeneNameTranslator.run.html" class="btn btn-neutral float-left" title="rnalysis.feature_name_to_wbgene.GeneNameTranslator.run" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>

    </div>


  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Guy Teichman

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>



  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>






</body>
</html>
