

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>rnalysis package &mdash; RNAlysis 1.3.4 documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/copybutton.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html">
          

          
            
            <img src="_static/logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is RNAlysis?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#main-features">Main Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#where-to-get-it">Where to get it</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
</ul>
<p class="caption"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary</a></li>
</ul>
<p class="caption"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>rnalysis package</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/rnalysis.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rnalysis-package">
<h1>rnalysis package<a class="headerlink" href="#rnalysis-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rnalysis-filtering-module">
<h2>rnalysis.filtering module<a class="headerlink" href="#rnalysis-filtering-module" title="Permalink to this headline">¶</a></h2>
<p>This module can filter, normalize, intersect and visualize tabular data such as read counts and differential expression data.
Any tabular data saved in a csv format can be imported. Use this module to perform various filtering operations on your data,  normalize your data, perform set operations (union, intersection, etc), run basic exploratory analyses and plots (such as PCA, clustergram, violin plots, scatter, etc), save the filtered data to your computer, and more.
When you save filtered/modified data, its new file name will include by default  all of the operations performed on it, in the order they were performed, to allow easy traceback of your analyses.</p>
<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">CountFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fname</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>pathlib.Path<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">drop_columns</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">rnalysis.filtering.Filter</span></code></a></p>
<p>A class that receives a count matrix and can filter it according to various characteristics.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the count matrix contents.         The DataFrame is modified upon usage of filter operations.</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
<dt>index_set: set</dt><dd><p>All of the indices in the current DataFrame (which were not removed by previously used filter methods)         as a set.</p>
</dd>
<dt>index_string: string</dt><dd><p>A string of all feature indices in the current DataFrame separated by newline.</p>
</dd>
<dt>triplicates: list</dt><dd><p>Returns a nested list of the column names in the CountFilter, grouped by alphabetical order into triplicates.         For example, if counts.columns is [‘A_rep1’,’A_rep2’,’A_rep3’,’B_rep1’,’B_rep2’,_B_rep3’], then         counts.triplicates will be  [[‘A_rep1’,’A_rep2’,’A_rep3’],[‘B_rep1’,’B_rep2’,_B_rep3’]]</p>
</dd>
</dl>
<dl class="py method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_format</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'short'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.
:type return_format: ‘short’ or ‘long’ (default ‘short’)
:param return_format: ‘short’ returns a short-form DataFrame, which states the biotypes         in the Filter object and their count. ‘long’ returns a long-form DataFrame,
which also provides descriptive statistics of each column per biotype.
:param ref: Name of the biotype reference table used to determine biotype. Default is ce11 (included in the package).
:rtype: pandas.DataFrame
:returns: a pandas DataFrame showing the number of values belonging to each biotype,         as well as additional descriptive statistics of format==’long’.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># short-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># long-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">return_format</span><span class="o">=</span><span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">               baseMean               ...           padj</span>
<span class="go">                  count         mean  ...            75%            max</span>
<span class="go">biotype                               ...</span>
<span class="go">protein_coding     26.0  1823.089609  ...   1.005060e-90   9.290000e-68</span>
<span class="go">pseudogene          1.0  2688.043701  ...   1.800000e-94   1.800000e-94</span>
<span class="go">unknown             1.0  2085.995094  ...  3.070000e-152  3.070000e-152</span>

<span class="go">[3 rows x 48 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">box_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">scatter</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'log10(RPM + 1)'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generates a box plot of the specified samples in the CountFilter object in log10 scale.         Can plot both single samples and average multiple replicates.         It is recommended to use this function on normalized values and not on absolute read values.         The box indicates 25% and 75% percentiles, and the white dot indicates the median.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>'all'</em><em> or </em><em>list.</em>) – A list of the sample names and/or grouped sample names to be plotted.         All specified samples must be present in the CountFilter object.         To average multiple replicates of the same condition, they can be grouped in an inner list.         Example input:         [[‘SAMPLE1A’, ‘SAMPLE1B’, ‘SAMPLE1C’], [‘SAMPLE2A’, ‘SAMPLE2B’, ‘SAMPLE2C’],’SAMPLE3’ , ‘SAMPLE6’]</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em> (</em><em>default 'Log10</em><em>(</em><em>RPM + 1</em><em>)</em><em>'</em><em>)</em>) – the label of the Y axis.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a seaborn box plot object.</p>
</dd>
</dl>
<div class="figure align-center" id="id1">
<a class="reference internal image-reference" href="???.png"><img alt="???.png" src="???.png" /></a>
<p class="caption"><span class="caption-text">Example plot of box_plot()</span><a class="headerlink" href="#id1" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">clustergram</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_names</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">metric</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'euclidean'</span></em>, <em class="sig-param"><span class="n">linkage</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'average'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Performs hierarchical clustering and plots a clustergram on the base-2 log of a given set of samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_names</strong> (<em>'all'</em><em> or </em><em>list.</em>) – the names of the relevant samples in a list.         Example input: [“condition1_rep1”, “condition1_rep2”, “condition1_rep3”,         “condition2_rep1”, “condition3_rep1”, “condition3_rep2”]</p></li>
<li><p><strong>metric</strong> (<em>'euclidean'</em><em>, </em><em>'hamming'</em><em>, </em><em>'correlation'</em><em>, or </em><em>any other         distance metric available in scipy.spatial.distance.pdist</em>) – the distance metric to use in the clustergram.         For all possible inputs and their meaning see scipy.spatial.distance.pdist documentation online.</p></li>
<li><p><strong>linkage</strong> (<em>'single'</em><em>, </em><em>'average'</em><em>, </em><em>'complete'</em><em>, </em><em>'weighted'</em><em>, </em><em>'centroid'</em><em>, </em><em>'median'</em><em> or </em><em>'ward'.</em>) – the linkage method to use in the clustergram.         For all possible inputs and their meaning see scipy.cluster.hierarchy.linkage documentation online.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A seaborn clustermap object.</p>
</dd>
</dl>
<div class="figure align-center" id="id2">
<a class="reference internal image-reference" href="_images/clustergram.png"><img alt="_images/clustergram.png" src="_images/clustergram.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of clustergram()</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">columns</code></dt>
<dd><p>list of column names</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentiles</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">0.01, 0.25, 0.5, 0.75, 0.99</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Summary statistics of the dataset.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Series or DataFrame</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">1%         0.000000      0.000000      0.000000      0.000000</span>
<span class="go">25%        6.000000      6.250000      1.250000      0.250000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">75%     2637.000000   2479.000000   6030.500000   4669.750000</span>
<span class="go">99%    15054.950000  12714.290000  21955.390000  15603.510000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># show the deciles (10%, 20%, 30%... 90%) of the columns</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">(</span><span class="n">percentiles</span><span class="o">=</span><span class="p">[</span><span class="n">decile</span><span class="o">/</span><span class="mi">10</span> <span class="k">for</span> <span class="n">decile</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">)])</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">10%        0.000000      0.200000      0.000000      0.000000</span>
<span class="go">20%        1.400000      3.200000      1.000000      0.000000</span>
<span class="go">30%       15.000000     15.700000      2.600000      1.000000</span>
<span class="go">40%       28.400000     26.800000     14.000000      9.000000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">60%       82.000000    106.800000     44.000000     33.000000</span>
<span class="go">70%      484.200000    395.500000    305.000000    302.500000</span>
<span class="go">80%     3398.600000   3172.600000   7981.400000   6213.000000</span>
<span class="go">90%     8722.100000   7941.800000  16449.500000  12129.900000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">df</code></dt>
<dd><p>pandas.DataFrame with the data</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00007063&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007071&#39;,</span>
<span class="go"> &#39;WBGene00007074&#39;, &#39;WBGene00007075&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007079&#39;,</span>
<span class="go"> &#39;WBGene00014997&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00044022&#39;,</span>
<span class="go"> &#39;WBGene00044951&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00077504&#39;}</span>
</pre></div>
</div>
<p># calculate difference and filter in-place
&gt;&gt;&gt; d.difference(counts, a_set, inplace=True)
Filtered 2 features, leaving 26 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">enhanced_box_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">scatter</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'log10(RPM + 1)'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generates an enhanced box plot of the specified samples in the CountFilter object in log10 scale.         Can plot both single samples and average multiple replicates.         It is recommended to use this function on normalized values and not on absolute read values.         The box indicates 25% and 75% percentiles, and the white dot indicates the median.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>'all'</em><em> or </em><em>list.</em>) – A list of the sample names and/or grouped sample names to be plotted.         All specified samples must be present in the CountFilter object.         To average multiple replicates of the same condition, they can be grouped in an inner list.         Example input:         [[‘SAMPLE1A’, ‘SAMPLE1B’, ‘SAMPLE1C’], [‘SAMPLE2A’, ‘SAMPLE2B’, ‘SAMPLE2C’],’SAMPLE3’ , ‘SAMPLE6’]</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em> (</em><em>default 'Log10</em><em>(</em><em>RPM + 1</em><em>)</em><em>'</em><em>)</em>) – the label of the Y axis.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a seaborn enhanced box plot object.</p>
</dd>
</dl>
<div class="figure align-center" id="id3">
<a class="reference internal image-reference" href="???.png"><img alt="???.png" src="???.png" /></a>
<p class="caption"><span class="caption-text">Example plot of enhanced_box_plot()</span><a class="headerlink" href="#id3" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype/biotypes.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>string</em><em> or </em><em>list of strings</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the biotype reference file used to determine biotypes.         Default is the path defined by the user in the settings.yaml file.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 9 features, leaving 13 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39; or &#39;pseudogene&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">([</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="s1">&#39;pseudogene&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 0 features, leaving 22 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'union'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters features according to user-defined attributes from an Attribute Reference Table.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).         To learn more about user-defined attributes and Attribute Reference Tables, read the user guide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>string</em><em> or </em><em>list of strings</em><em>,         </em><em>which are column titles in the user-defined Attribute Reference Table.</em>) – attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – If ‘union’, filters out every genomic feature that does not belong to one or more         of the indicated attributes. If ‘intersection’,         filters out every genomic feature that does not belong to ALL of the indicated attributes.</p></li>
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – filename/path of the attribute reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em><em> (</em><em>default False</em><em>)</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attribute &#39;attribute1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attributes &#39;attribute1&#39; OR &#39;attribute3&#39; (union)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 14 features, leaving 8 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to both attributes &#39;attribute1&#39; AND &#39;attribute3&#39; (intersection)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 19 features, leaving 3 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to either &#39;attribute1&#39;,&#39;attribute3&#39; or both</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 8 features, leaving 14 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to both &#39;attribute1&#39; AND &#39;attribute3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 3 features, leaving 19 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_by_row_sum</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">5</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Removes features/rows whose sum is belove ‘threshold’.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>threshold</strong> (<em>float</em>) – The minimal sum a row should have in order not to be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of CountFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">filter_by_row_sum</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span> <span class="c1"># remove all rows whose sum is &lt;5</span>
<span class="go">Filtered 4 features, leaving 18 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_low_reads</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">5</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>remove features which have less then ‘threshold’ reads all columns.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>threshold</strong> (<em>float</em>) – The minimal number of reads (counts, rpm, rpkm, tpm, etc) a feature should have         in at least one sample in order not to be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of CountFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">filter_low_reads</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span> <span class="c1"># remove all rows whose values in all columns are all &lt;5</span>
<span class="go">Filtered 6 features, leaving 16 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows whose value in the column &#39;log2FoldChange&#39; is below the 75th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows vulse value in the column &#39;log2FoldChange&#39; is above the 25th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns, then keep only the top ‘n’ rows.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>string</em><em> or </em><em>list of strings</em>) – Names of the column or columns to sort and then filter by.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bools</em><em> (</em><em>default True</em><em>)</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows with the highest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows which have the lowest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># and then the highest values in the column &#39;cond2&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="s1">&#39;cond2&#39;</span><span class="p">],</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">False</span><span class="p">])</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">fname</code></dt>
<dd><p>filename with full path</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">fold_change</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">numerator</span></em>, <em class="sig-param"><span class="n">denominator</span></em>, <em class="sig-param"><span class="n">numer_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'default'</span></em>, <em class="sig-param"><span class="n">denom_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'default'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Calculate the fold change between the numerator condition and the denominator condition,         and return it as a FoldChangeFilter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>numerator</strong> (<em>str</em><em>, or </em><em>list of strs</em>) – the CountFilter columns to be used as the numerator. If multiple arguments are given         in a list, they will be averaged.</p></li>
<li><p><strong>denominator</strong> (<em>str</em><em>, or </em><em>list of strs</em>) – the CountFilter columns to be used as the denominator. If multiple arguments are given         in a list, they will be averaged.</p></li>
<li><p><strong>numer_name</strong> (<em>str</em><em> or </em><em>'default'</em>) – name to give the numerator condition. If ‘default’, the name will be generarated         automatically from the names of numerator columns.</p></li>
<li><p><strong>denom_name</strong> (<em>str</em><em> or </em><em>'default'</em>) – name to give the denominator condition. If ‘default’, the name will be generarated         automatically from the names of denominator columns.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference internal" href="rnalysis.filtering.FoldChangeFilter.html#rnalysis.filtering.FoldChangeFilter" title="rnalysis.filtering.FoldChangeFilter">FoldChangeFilter</a></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A new instance of FoldChangeFilter</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted_fold_change.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate the fold change of mean(cond1_rep1,cond1_rep2)/mean(cond2_rep1,cond_2rep2)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">fold_change</span><span class="p">([</span><span class="s1">&#39;cond1_rep1&#39;</span><span class="p">,</span><span class="s1">&#39;cond1_rep2&#39;</span><span class="p">],[</span><span class="s1">&#39;cond2_rep1&#39;</span><span class="p">,</span><span class="s1">&#39;cond2_rep2&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">numerator</span>
<span class="go">&quot;Mean of [&#39;cond1_rep1&#39;, &#39;cond1_rep2&#39;]&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">denominator</span>
<span class="go">&quot;Mean of [&#39;cond2_rep1&#39;, &#39;cond2_rep2&#39;]&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
<span class="go">rnalysis.filtering.FoldChangeFilter</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">static </em><code class="sig-name descname">from_folder</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">folder_path</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">norm_to_rpm</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save_csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">counted_fname</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">uncounted_fname</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">input_format</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'.txt'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Iterates over HTSeq count .txt files in a given folder and combines them into a single CountFilter table.         Can also save the count data table and the uncounted data table to .csv files, and normalize the CountFilter         table to reads per million (RPM). Note that the saved data will always be count data, and not normalized data,         regardless if the CountFilter table was normalized or not.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>folder_path</strong> – str or pathlib.Path. Full path of the folder that contains individual htcount .txt files.</p></li>
<li><p><strong>norm_to_rpm</strong> – bool. If True, the CountFilter table will be automatically normalized to reads per         million (RPM). If False (defualt), the CountFilter object will not be normalized, and will instead contain         absolute count data (as in the original htcount .txt files).         Note that if save_csv is True, the saved .csv fill will contain ABSOLUTE COUNT DATA, as in the original         htcount .txt files, and NOT normalized data.</p></li>
<li><p><strong>save_csv</strong> – bool. If True, the joint DataFrame of count data and uncounted data will be saved         to two separate .csv files. The files will be saved in ‘folder_path’, and named according to the parameters         ‘counted_fname’ for the count data, and ‘uncounted_fname’ for the uncounted data (unaligned,         alignment not unique, etc).</p></li>
<li><p><strong>counted_fname</strong> – str. Name under which to save the combined count data table. Does not need to include         the ‘.csv’ suffix.</p></li>
<li><p><strong>uncounted_fname</strong> – counted_fname: str. Name under which to save the combined uncounted data.         Does not need to include the ‘.csv’ suffix.</p></li>
<li><p><strong>input_format</strong> – the file format of the input files. Default is ‘.txt’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>an CountFilter object containing the combined count data from all individual htcount .txt files in the         specified folder.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="o">.</span><span class="n">from_folder</span><span class="p">(</span><span class="s1">&#39;tests/test_count_from_folder&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="o">.</span><span class="n">from_folder</span><span class="p">(</span><span class="s1">&#39;tests/test_count_from_folder&#39;</span><span class="p">,</span> <span class="n">norm_to_rpm</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># This will also normalize the CountFilter to reads-per-million (RPM).</span>
<span class="go">Normalized the values of 10 features. Normalized inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="o">.</span><span class="n">from_folder</span><span class="p">(</span><span class="s1">&#39;tests/test_count_from_folder&#39;</span><span class="p">,</span> <span class="n">save_csv</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">counted_fname</span><span class="o">=</span><span class="s1">&#39;name_for_reads_csv_file&#39;</span><span class="p">,</span> <span class="n">uncounted_fname</span><span class="o">=</span><span class="s1">&#39;name_for_uncounted_reads_csv_file&#39;</span><span class="p">)</span> <span class="c1"># This will also save the counted reads and uncounted reads as separate .csv files</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the first n rows of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>
<span class="go">WBGene00000005  4028.154186        6.534112  ...  1.700000e-228  7.800000e-225</span>
<span class="go">WBGene00000006  1230.585240        7.157428  ...  2.070000e-216  7.590000e-213</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="c1"># return only the first 3 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>

<span class="go">[3 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_set</code></dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A set of WBGene names.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myset</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">myset</span><span class="p">)</span>
<span class="go">{&#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;,</span>
<span class="go">&#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00007077&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00007075&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_string</code></dt>
<dd><p>Returns a string of all feature indices in the current DataFrame, sorted alphabetically and separated by newline. </p>
<dl class="simple">
<dt>This includes all of the feature indices which were not filtered out by previously-used filter methods. </dt><dd><p>if any duplicate features exist in the filter object (same index appears more than once), </p>
</dd>
</dl>
<p>the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A string of WBGene indices separated by newlines (\n). </p>
</dd>
</dl>
<p>For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mystring</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_string</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mystring</span><span class="p">)</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate intersection and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;}</span>
</pre></div>
</div>
<p># calculate intersection and filter in-place
&gt;&gt;&gt; d.intersection(a_set, inplace=True)
Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">normalize_to_rpm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">special_counter_fname</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Normalizes the reads in the CountFilter to reads per million (RPM).         Uses a table of feature counts (ambiguous, no feature, not aligned, etc) from HTSeq’s output.         Divides each column in the CountFilter object by (total reads + ambiguous + no feature)*10^-6 .</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>special_counter_fname</strong> – the .csv file which contains feature information about the RNA library         (ambiguous, no feature, not aligned, etc).</p></li>
<li><p><strong>inplace</strong> – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If inplace is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s2">&quot;tests/counted.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">normalize_to_rpm</span><span class="p">(</span><span class="s2">&quot;tests/uncounted.csv&quot;</span><span class="p">)</span>
<span class="go">Normalized the values of 22 features. Normalized inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">normalize_with_scaling_factors</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">scaling_factor_fname</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>pathlib.Path<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Normalizes the reads in the CountFilter using pre-calculated scaling factors.         Receives a table of sample names and their corresponding size factors,         and divides each column in the CountFilter by the corresponding scaling factor.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>scaling_factor_fname</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – the .csv file which contains size factors for the different libraries.</p></li>
<li><p><strong>inplace</strong> – If True (default), filtering will be applied to the current CountFilter object. If False,         the function will return a new CountFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If inplace is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s2">&quot;tests/counted.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span><span class="o">.</span><span class="n">normalize_with_scaling_factors</span><span class="p">(</span><span class="s2">&quot;tests/scaling_factors.csv&quot;</span><span class="p">)</span>
<span class="go">Normalized the values of 22 features. Normalized inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt' / 'greater than' / '&gt;'</em><em>, </em><em>'eq' / 'equals' / '='</em><em>, </em><em>'lt' / 'lesser than' / '&lt;'</em>) – the operator to filter the column by (greater than, equal or lesser than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;gt&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;greater than&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;&gt;&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">pairplot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_list</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">log2</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Plot pairwise relationships in the dataset.         Can plot both single samples and average multiple replicates.         For more information see the documentation of seaborn.pairplot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_list</strong> (<em>'all'</em><em>, </em><em>list</em><em>, or </em><em>nested list.</em>) – A list of the sample names and/or grouped sample names to be included in the pairplot.         All specified samples must be present in the CountFilter object.         To average multiple replicates of the same condition, they can be grouped in an inner list.         Example input:         [[‘SAMPLE1A’, ‘SAMPLE1B’, ‘SAMPLE1C’], [‘SAMPLE2A’, ‘SAMPLE2B’, ‘SAMPLE2C’],’SAMPLE3’ , ‘SAMPLE6’]</p></li>
<li><p><strong>log2</strong> (<em>bool</em>) – if True, the pairplot will be calculated with log2 of the dataframe, and not with the raw data.         If False (default), the pairplot will be calculated with the raw data.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An instance of seaborn.PairGrid.</p>
</dd>
</dl>
<div class="figure align-center" id="id4">
<a class="reference internal image-reference" href="_images/pairplot.png"><img alt="_images/pairplot.png" src="_images/pairplot.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of pairplot()</span><a class="headerlink" href="#id4" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">pca</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_names</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">n_components</span><span class="o">=</span><span class="default_value">3</span></em>, <em class="sig-param"><span class="n">sample_grouping</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">labels</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>runs and plots a PCA for a given set of samples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_names</strong> (<em>'all'</em><em> or </em><em>list.</em>) – the names of the relevant samples in a list.         Example input: [“1_REP_A”, “1_REP_B”, “1_REP_C”, “2_REP_A”, “2_REP_B”, “2_REP_C”, “2_REP_D”, “3_REP_A”]</p></li>
<li><p><strong>n_components</strong> (<em>positive int</em><em> (</em><em>default 3</em><em>)</em>) – number of PCA components to return.</p></li>
<li><p><strong>sample_grouping</strong> (<em>list of positive integers</em><em>, </em><em>'triplicates'</em><em> or </em><em>None</em><em> (</em><em>default</em><em>)</em>) – Optional. Indicates which samples are grouped together as replicates,         so they will be colored similarly in the PCA plot. A list of indices from 1 and up, that indicates the sample          grouping.          For example, if sample_names is:         [“1_REP_A”, “1_REP_B”, “1_REP_C”, “2_REP_A”, “2_REP_B”, “2_REP_C”, “2_REP_D”, “3_REP_A”],         then the sample_grouping will be:         [1, 1, 1, 2, 2, 2, 2, 3].         If ‘triplicate’, then sample_groupins will automatically group samples into triplicates. For example:         [1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4].</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple whose first element is an sklearn.decomposition.pca object,         and second element is a list of matplotlib.axis objects.</p>
</dd>
</dl>
<div class="figure align-center" id="id5">
<a class="reference internal image-reference" href="_images/pca.png"><img alt="_images/pca.png" src="_images/pca.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of pca()</span><a class="headerlink" href="#id5" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">plot_expression</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">features</span><span class="p">:</span> <span class="n">list</span></em>, <em class="sig-param"><span class="n">sample_grouping</span><span class="p">:</span> <span class="n">dict</span></em>, <em class="sig-param"><span class="n">count_unit</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'Reads per million'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Plot the average expression and standard error of the specified features under the specified conditions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>features</strong> (<em>str</em><em> or </em><em>list of strings</em>) – the feature/features to plot expression for.</p></li>
<li><p><strong>sample_grouping</strong> (<em>dict</em><em>, </em><em>with condition names as keys         and list of the sample numbers</em><em> or </em><em>names for each condition as a list</em>) – a dictionary of the conditions to plot expression for.         Each key should be a name of a conditions, and the value for each key is         a list of the numbers of columns to be used as samples of that condition.         For example, if the first 3 columns are replicates of the condition ‘condition 1’ and         the last 3 column are replicates of the condition ‘condition 2’, then sample_grouping should be:         {‘condition 1’:[0, 1, 2], ‘condition 2’:[3, 4, 5]}</p></li>
<li><p><strong>count_unit</strong> (<em>str</em><em>, </em><em>default 'Reads per million'</em>) – The unit of the count data. Will be displayed in the y axis.</p></li>
</ul>
</dd>
</dl>
<div class="figure align-center" id="id6">
<a class="reference internal image-reference" href="_images/plot_expression.png"><img alt="_images/plot_expression.png" src="_images/plot_expression.png" style="width: 403.6px; height: 294.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Example plot of plot_expression()</span><a class="headerlink" href="#id6" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Print the feature indices in the Filter object, sorted alphabetically and separated by newline.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">print_features</span><span class="p">()</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alt_filename</span><span class="p">:</span> <span class="n">Union[None, str, pathlib.Path]</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> (<em>str</em><em>, </em><em>pathlib.Path</em><em>, or </em><em>None</em><em> (</em><em>default</em><em>)</em>) – If None, file name will be generated automatically         according to the filtering methods used.         If it’s a string, it will be used as the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">scatter_sample_vs_sample</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample1</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">sample2</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">xlabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">highlight</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate a scatter plot where every dot is a feature, the x value is log10 of reads         (counts, RPM, RPKM, TPM, etc) in sample1, the y value is log10 of reads in sample2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample1</strong> (<em>string</em><em> or </em><em>list of strings</em>) – Name of the first sample from the CountFilter object.         If sample1 is a list, they will be avarged as replicates.</p></li>
<li><p><strong>sample2</strong> (<em>string</em><em> or </em><em>list of strings</em>) – Name of the second sample from the CountFilter object.         If sample2 is a list, they will be averaged as replicates.</p></li>
<li><p><strong>xlabel</strong> (<em>str</em>) – optional. If not specified, sample1 will be used as xlabel.</p></li>
<li><p><strong>ylabel</strong> (<em>str</em>) – optional. If not specified, sample2 will be used as ylabel.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – optional. If not specified, a title will be generated automatically.</p></li>
<li><p><strong>highlight</strong> (<em>Filter object</em><em> or </em><em>iterable of strings</em>) – If specified, the points in the scatter corresponding to the names/features in ‘highlight’         will be highlighted in red.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a matplotlib axis object.</p>
</dd>
</dl>
<div class="figure align-center" id="id7">
<a class="reference internal image-reference" href="_images/rpm_vs_rpm.png"><img alt="_images/rpm_vs_rpm.png" src="_images/rpm_vs_rpm.png" style="width: 480.0px; height: 477.59999999999997px;" /></a>
<p class="caption"><span class="caption-text">Example plot of scatter_sample_vs_sample()</span><a class="headerlink" href="#id7" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">shape</code></dt>
<dd><p>(rows, columns)</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list of str</em>) – Names of the column or columns to sort by.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bool</em><em>, </em><em>default True</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, perform operation in-place.         Otherwise, returns a sorted copy of the Filter object without modifying the original.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None if inplace=True, a sorted Filter object otherwise.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00007063    633    451    365    388</span>
<span class="go">WBGene00007064     60     57     20     23</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00007066     55    266     46     39</span>
<span class="go">WBGene00007067     15     13      1      0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00077504      0      0      0      0</span>
<span class="go">WBGene00007069      0      2      1      0</span>
<span class="go">WBGene00077502      0      0      0      0</span>
<span class="go">WBGene00077503      1      4      2      0</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Attribute Reference Table attributes.         Each object contains only features that match its Attribute Reference Table attribute.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>list of strings</em>) – list of attribute names from the Attribute Reference Table to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of Filter objects, each containing only features that match one Attribute Reference Table attribute;         the Filter objects are returned in the same order the attributes were given in.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attribute1</span><span class="p">,</span><span class="n">attribute2</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         below and above the specified percentile in the spcfieid column.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">below</span><span class="p">,</span> <span class="n">above</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split_by_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Filtered 21 features, leaving 7 of the original 28 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_reads</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">threshold</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the features in the current CountFilter object into two complementary, non-overlapping CountFilter         objects, based on the their maximum expression level. The first object will contain only highly-expressed          features (which have reads over the specified threshold in at least one sample). The second object will          contain only lowly-expressed features (which have reads below the specified threshold in all samples).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>threshold</strong> – A float. The minimal number of reads (counts, RPM, RPKM, TPM etc) a feature needs to have         in at least one sample in order to be         included in the “highly expressed” object and no the “lowly expressed” object.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple containing two CountFilter objects: the first has only highly-expressed features,         and the second has only lowly-expressed features.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">low_expr</span><span class="p">,</span><span class="n">high_expr</span> <span class="o">=</span> <span class="n">c</span><span class="o">.</span><span class="n">split_by_reads</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
<span class="go">Filtered 6 features, leaving 16 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 16 features, leaving 6 of the original 22 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000024&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;,</span>
<span class="go">&#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000029&#39;, &#39;WBGene00000006&#39;,</span>
<span class="go">&#39;WBGene00007064&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;,</span>
<span class="go">&#39;WBGene00007074&#39;, &#39;WBGene00000025&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000027&#39;,</span>
<span class="go">&#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000012&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00077503&#39;,</span>
<span class="go">&#39;WBGene00000026&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000002&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;, &#39;WBGene00007077&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00043987&#39;,</span>
<span class="go">&#39;WBGene00007071&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame</dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>returns the last n rows of the Filter object.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span> <span class="c1"># returns the last 8 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000022   365.813048        6.101303  ...  2.740000e-97  2.400000e-94</span>
<span class="go">WBGene00000023  3168.566714        3.906719  ...  1.600000e-93  1.340000e-90</span>
<span class="go">WBGene00000024   221.925724        4.801676  ...  1.230000e-84  9.820000e-82</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[8 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq' / 'equals' / '='</em><em>, </em><em>'ct' / 'contains' / 'in'</em><em>, </em><em>'bw' / 'begins with'</em><em>, </em><em>'ew' / 'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/text_filters.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that have a value that starts with &#39;AC3&#39; in the column &#39;name&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">text_filters</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sw&#39;</span><span class="p">,</span><span class="s1">&#39;AC3&#39;</span><span class="p">)</span>
<span class="go">Filtered 17 features, leaving 5 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">triplicates</code></dt>
<dd><p>Returns a nested list of the column names in the CountFilter, grouped by alphabetical order into triplicates.         For example, if counts.columns is [‘A_rep1’,’A_rep2’,’A_rep3’,’B_rep1’,’B_rep2’,_B_rep3’], then         counts.triplicates will be  [[‘A_rep1’,’A_rep2’,’A_rep3’],[‘B_rep1’,’B_rep2’,_B_rep3’]]</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate union and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000012&#39;,</span>
<span class="go">&#39;WBGene00000024&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00007069&#39;,</span>
<span class="go">&#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000026&#39;, &#39;WBGene00000029&#39;,</span>
<span class="go">&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;,</span>
<span class="go">&#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00000025&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;,</span>
<span class="go">&#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000027&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">violin_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">ylabel</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'log10(RPM + 1)'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generates a violin plot of the specified samples in the CountFilter object in log10 scale.         Can plot both single samples and average multiple replicates.         It is recommended to use this function on normalized values and not on absolute read values.         Box inside the violin plot indicates 25% and 75% percentiles, and the white dot indicates the median.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>samples</strong> (<em>'all'</em><em> or </em><em>list.</em>) – A list of the sample names and/or grouped sample names to be plotted.         All specified samples must be present in the CountFilter object.         To average multiple replicates of the same condition, they can be grouped in an inner list.         Example input:         [[‘SAMPLE1A’, ‘SAMPLE1B’, ‘SAMPLE1C’], [‘SAMPLE2A’, ‘SAMPLE2B’, ‘SAMPLE2C’],’SAMPLE3’ , ‘SAMPLE6’]</p></li>
<li><p><strong>ylabel</strong> (<em>str</em><em> (</em><em>default 'Log10</em><em>(</em><em>RPM + 1</em><em>)</em><em>'</em><em>)</em>) – the label of the Y axis.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a seaborn violin object.</p>
</dd>
</dl>
<div class="figure align-center" id="id8">
<a class="reference internal image-reference" href="_images/violin.png"><img alt="_images/violin.png" src="_images/violin.png" style="width: 1152.0px; height: 556.1999999999999px;" /></a>
<p class="caption"><span class="caption-text">Example plot of violin_plot()</span><a class="headerlink" href="#id8" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">DESeqFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fname</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>pathlib.Path<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">drop_columns</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">rnalysis.filtering.Filter</span></code></a></p>
<p>A class that receives a DESeq output file and can filter it according to various characteristics.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the DESeq output file contents.         The DataFrame is modified upon usage of filter operations. .</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
<dt>index_set: set</dt><dd><p>All of the indices in the current DataFrame (which were not removed by previously used filter methods)         as a set.</p>
</dd>
<dt>index_string: string</dt><dd><p>A string of all feature indices in the current DataFrame separated by newline.</p>
</dd>
</dl>
<dl class="py method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_format</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'short'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.
:type return_format: ‘short’ or ‘long’ (default ‘short’)
:param return_format: ‘short’ returns a short-form DataFrame, which states the biotypes         in the Filter object and their count. ‘long’ returns a long-form DataFrame,
which also provides descriptive statistics of each column per biotype.
:param ref: Name of the biotype reference table used to determine biotype. Default is ce11 (included in the package).
:rtype: pandas.DataFrame
:returns: a pandas DataFrame showing the number of values belonging to each biotype,         as well as additional descriptive statistics of format==’long’.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># short-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># long-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">return_format</span><span class="o">=</span><span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">               baseMean               ...           padj</span>
<span class="go">                  count         mean  ...            75%            max</span>
<span class="go">biotype                               ...</span>
<span class="go">protein_coding     26.0  1823.089609  ...   1.005060e-90   9.290000e-68</span>
<span class="go">pseudogene          1.0  2688.043701  ...   1.800000e-94   1.800000e-94</span>
<span class="go">unknown             1.0  2085.995094  ...  3.070000e-152  3.070000e-152</span>

<span class="go">[3 rows x 48 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">columns</code></dt>
<dd><p>list of column names</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentiles</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">0.01, 0.25, 0.5, 0.75, 0.99</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Summary statistics of the dataset.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Series or DataFrame</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">1%         0.000000      0.000000      0.000000      0.000000</span>
<span class="go">25%        6.000000      6.250000      1.250000      0.250000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">75%     2637.000000   2479.000000   6030.500000   4669.750000</span>
<span class="go">99%    15054.950000  12714.290000  21955.390000  15603.510000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># show the deciles (10%, 20%, 30%... 90%) of the columns</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">(</span><span class="n">percentiles</span><span class="o">=</span><span class="p">[</span><span class="n">decile</span><span class="o">/</span><span class="mi">10</span> <span class="k">for</span> <span class="n">decile</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">)])</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">10%        0.000000      0.200000      0.000000      0.000000</span>
<span class="go">20%        1.400000      3.200000      1.000000      0.000000</span>
<span class="go">30%       15.000000     15.700000      2.600000      1.000000</span>
<span class="go">40%       28.400000     26.800000     14.000000      9.000000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">60%       82.000000    106.800000     44.000000     33.000000</span>
<span class="go">70%      484.200000    395.500000    305.000000    302.500000</span>
<span class="go">80%     3398.600000   3172.600000   7981.400000   6213.000000</span>
<span class="go">90%     8722.100000   7941.800000  16449.500000  12129.900000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">df</code></dt>
<dd><p>pandas.DataFrame with the data</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00007063&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007071&#39;,</span>
<span class="go"> &#39;WBGene00007074&#39;, &#39;WBGene00007075&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007079&#39;,</span>
<span class="go"> &#39;WBGene00014997&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00044022&#39;,</span>
<span class="go"> &#39;WBGene00044951&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00077504&#39;}</span>
</pre></div>
</div>
<p># calculate difference and filter in-place
&gt;&gt;&gt; d.difference(counts, a_set, inplace=True)
Filtered 2 features, leaving 26 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_abs_log2_fold_change</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">abslog2fc</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features whose absolute log2 fold change is below the indicated threshold.         For example: if log2fc is 2.0, all features whose log2 fold change is between 1 and -1 (went up less than         two-fold or went down less than two-fold) will be filtered out.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>abslog2fc</strong> – The threshold absolute log2 fold change for filtering out a feature. Float or int.         All features whose absolute log2 fold change is lower than log2fc will be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current DESeqFilter object. If False,         the function will return a new DESeqFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of DESeqFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/sample_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_abs_log2_fold_change</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="c1"># keep only rows whose log2(fold change) is &gt;=2 or &lt;=-2</span>
<span class="go">Filtered 1 features, leaving 28 of the original 29 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype/biotypes.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>string</em><em> or </em><em>list of strings</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the biotype reference file used to determine biotypes.         Default is the path defined by the user in the settings.yaml file.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 9 features, leaving 13 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39; or &#39;pseudogene&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">([</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="s1">&#39;pseudogene&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 0 features, leaving 22 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'union'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters features according to user-defined attributes from an Attribute Reference Table.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).         To learn more about user-defined attributes and Attribute Reference Tables, read the user guide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>string</em><em> or </em><em>list of strings</em><em>,         </em><em>which are column titles in the user-defined Attribute Reference Table.</em>) – attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – If ‘union’, filters out every genomic feature that does not belong to one or more         of the indicated attributes. If ‘intersection’,         filters out every genomic feature that does not belong to ALL of the indicated attributes.</p></li>
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – filename/path of the attribute reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em><em> (</em><em>default False</em><em>)</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attribute &#39;attribute1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attributes &#39;attribute1&#39; OR &#39;attribute3&#39; (union)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 14 features, leaving 8 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to both attributes &#39;attribute1&#39; AND &#39;attribute3&#39; (intersection)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 19 features, leaving 3 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to either &#39;attribute1&#39;,&#39;attribute3&#39; or both</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 8 features, leaving 14 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to both &#39;attribute1&#39; AND &#39;attribute3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 3 features, leaving 19 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_fold_change_direction</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">direction</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'pos'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out features according to the direction in which they changed between the two conditions.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>direction</strong> – ‘pos’ or ‘neg’. If ‘pos’, will keep only features that have positive log2foldchange.         If ‘neg’, will keep only features that have negative log2foldchange.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current DESeqFilter object. If False,         the function will return a new DESeqFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of DESeqFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/sample_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_fold_change_direction</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">)</span> <span class="c1"># keep only rows with a positive log2(fold change) value</span>
<span class="go">Filtered 3 features, leaving 26 of the original 29 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/sample_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_fold_change_direction</span><span class="p">(</span><span class="s1">&#39;neg&#39;</span><span class="p">)</span> <span class="c1"># keep only rows with a negative log2(fold change) value</span>
<span class="go">Filtered 27 features, leaving 2 of the original 29 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/sample_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_fold_change_direction</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span> <span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># keep only rows with a non-positive log2(fold change) value</span>
<span class="go">Filtered 26 features, leaving 3 of the original 29 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows whose value in the column &#39;log2FoldChange&#39; is below the 75th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows vulse value in the column &#39;log2FoldChange&#39; is above the 25th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_significant</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Removes all features which did not change significantly, according to the provided alpha.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>alpha</strong> – the significance threshold to determine which genes will be filtered. between 0 and 1.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current DESeqFilter object. If False,         the function will return a new DESeqFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of DESeqFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/sample_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_significant</span><span class="p">(</span><span class="mf">0.1</span><span class="p">)</span> <span class="c1"># keep only rows whose adjusted p-value is &lt;=0.1</span>
<span class="go">Filtered 4 features, leaving 25 of the original 29 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/sample_deseq.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_significant</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="c1"># keep only rows whose adjusted p-value is &gt;0.1</span>
<span class="go">Filtered 25 features, leaving 4 of the original 29 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns, then keep only the top ‘n’ rows.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>string</em><em> or </em><em>list of strings</em>) – Names of the column or columns to sort and then filter by.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bools</em><em> (</em><em>default True</em><em>)</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows with the highest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows which have the lowest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># and then the highest values in the column &#39;cond2&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="s1">&#39;cond2&#39;</span><span class="p">],</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">False</span><span class="p">])</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">fname</code></dt>
<dd><p>filename with full path</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the first n rows of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>
<span class="go">WBGene00000005  4028.154186        6.534112  ...  1.700000e-228  7.800000e-225</span>
<span class="go">WBGene00000006  1230.585240        7.157428  ...  2.070000e-216  7.590000e-213</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="c1"># return only the first 3 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>

<span class="go">[3 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_set</code></dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A set of WBGene names.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myset</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">myset</span><span class="p">)</span>
<span class="go">{&#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;,</span>
<span class="go">&#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00007077&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00007075&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_string</code></dt>
<dd><p>Returns a string of all feature indices in the current DataFrame, sorted alphabetically and separated by newline. </p>
<dl class="simple">
<dt>This includes all of the feature indices which were not filtered out by previously-used filter methods. </dt><dd><p>if any duplicate features exist in the filter object (same index appears more than once), </p>
</dd>
</dl>
<p>the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A string of WBGene indices separated by newlines (\n). </p>
</dd>
</dl>
<p>For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mystring</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_string</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mystring</span><span class="p">)</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate intersection and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;}</span>
</pre></div>
</div>
<p># calculate intersection and filter in-place
&gt;&gt;&gt; d.intersection(a_set, inplace=True)
Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt' / 'greater than' / '&gt;'</em><em>, </em><em>'eq' / 'equals' / '='</em><em>, </em><em>'lt' / 'lesser than' / '&lt;'</em>) – the operator to filter the column by (greater than, equal or lesser than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;gt&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;greater than&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;&gt;&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Print the feature indices in the Filter object, sorted alphabetically and separated by newline.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">print_features</span><span class="p">()</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alt_filename</span><span class="p">:</span> <span class="n">Union[None, str, pathlib.Path]</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> (<em>str</em><em>, </em><em>pathlib.Path</em><em>, or </em><em>None</em><em> (</em><em>default</em><em>)</em>) – If None, file name will be generated automatically         according to the filtering methods used.         If it’s a string, it will be used as the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">shape</code></dt>
<dd><p>(rows, columns)</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list of str</em>) – Names of the column or columns to sort by.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bool</em><em>, </em><em>default True</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, perform operation in-place.         Otherwise, returns a sorted copy of the Filter object without modifying the original.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None if inplace=True, a sorted Filter object otherwise.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00007063    633    451    365    388</span>
<span class="go">WBGene00007064     60     57     20     23</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00007066     55    266     46     39</span>
<span class="go">WBGene00007067     15     13      1      0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00077504      0      0      0      0</span>
<span class="go">WBGene00007069      0      2      1      0</span>
<span class="go">WBGene00077502      0      0      0      0</span>
<span class="go">WBGene00077503      1      4      2      0</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Attribute Reference Table attributes.         Each object contains only features that match its Attribute Reference Table attribute.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>list of strings</em>) – list of attribute names from the Attribute Reference Table to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of Filter objects, each containing only features that match one Attribute Reference Table attribute;         the Filter objects are returned in the same order the attributes were given in.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attribute1</span><span class="p">,</span><span class="n">attribute2</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         below and above the specified percentile in the spcfieid column.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">below</span><span class="p">,</span> <span class="n">above</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split_by_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Filtered 21 features, leaving 7 of the original 28 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_fold_change_direction</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Splits the features in the current DESeqFilter object into two complementary, non-overlapping DESeqFilter         objects, based on the direction of their log2foldchange. The first object will contain only features with a         positive log2foldchange, the second object will contain only features with a negative log2foldchange.</p>
<dl class="field-list">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a tuple containing two DESeqFilter objects: the first has only features with positive log2 fold change,         and the other has only features with negative log2 fold change.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos</span><span class="p">,</span> <span class="n">neg</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split_fold_change_direction</span><span class="p">()</span>
<span class="go">Filtered 2 features, leaving 26 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000024&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;,</span>
<span class="go">&#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000029&#39;, &#39;WBGene00000006&#39;,</span>
<span class="go">&#39;WBGene00007064&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;,</span>
<span class="go">&#39;WBGene00007074&#39;, &#39;WBGene00000025&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000027&#39;,</span>
<span class="go">&#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000012&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00077503&#39;,</span>
<span class="go">&#39;WBGene00000026&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000002&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;, &#39;WBGene00007077&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00043987&#39;,</span>
<span class="go">&#39;WBGene00007071&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame</dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>returns the last n rows of the Filter object.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span> <span class="c1"># returns the last 8 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000022   365.813048        6.101303  ...  2.740000e-97  2.400000e-94</span>
<span class="go">WBGene00000023  3168.566714        3.906719  ...  1.600000e-93  1.340000e-90</span>
<span class="go">WBGene00000024   221.925724        4.801676  ...  1.230000e-84  9.820000e-82</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[8 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq' / 'equals' / '='</em><em>, </em><em>'ct' / 'contains' / 'in'</em><em>, </em><em>'bw' / 'begins with'</em><em>, </em><em>'ew' / 'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/text_filters.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that have a value that starts with &#39;AC3&#39; in the column &#39;name&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">text_filters</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sw&#39;</span><span class="p">,</span><span class="s1">&#39;AC3&#39;</span><span class="p">)</span>
<span class="go">Filtered 17 features, leaving 5 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate union and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000012&#39;,</span>
<span class="go">&#39;WBGene00000024&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00007069&#39;,</span>
<span class="go">&#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000026&#39;, &#39;WBGene00000029&#39;,</span>
<span class="go">&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;,</span>
<span class="go">&#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00000025&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;,</span>
<span class="go">&#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000027&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">volcano_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.1</span></em><span class="sig-paren">)</span></dt>
<dd><p>Plots a volcano plot (log2(fold change) vs -log10(adj. p-value)) of the DESeqFilter object.         Significantly upregulated features are colored in red,         and significantly downregulated features are colored in blue.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alpha</strong> (<em>float between 0 and 1</em>) – the significance threshold to color data points as significantly up/down-regulated.</p>
</dd>
</dl>
<div class="figure align-center" id="id9">
<a class="reference internal image-reference" href="_images/volcano.png"><img alt="_images/volcano.png" src="_images/volcano.png" style="width: 448.0px; height: 333.2px;" /></a>
<p class="caption"><span class="caption-text">Example plot of volcano_plot()</span><a class="headerlink" href="#id9" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">Filter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fname</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>pathlib.Path<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">drop_columns</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>An all-purpose Filter.</p>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas DataFrame</dt><dd><p>A DataFrame that contains the DESeq output file contents.         The DataFrame is modified upon usage of filter operations. .</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
<dt>index_set: set</dt><dd><p>All of the indices in the current DataFrame (which were not removed by previously used filter methods)         as a set.</p>
</dd>
<dt>index_string: string</dt><dd><p>A string of all feature indices in the current DataFrame separated by newline.</p>
</dd>
</dl>
<dl class="py method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_format</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'short'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.
:type return_format: ‘short’ or ‘long’ (default ‘short’)
:param return_format: ‘short’ returns a short-form DataFrame, which states the biotypes         in the Filter object and their count. ‘long’ returns a long-form DataFrame,
which also provides descriptive statistics of each column per biotype.
:param ref: Name of the biotype reference table used to determine biotype. Default is ce11 (included in the package).
:rtype: pandas.DataFrame
:returns: a pandas DataFrame showing the number of values belonging to each biotype,         as well as additional descriptive statistics of format==’long’.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># short-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># long-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">return_format</span><span class="o">=</span><span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">               baseMean               ...           padj</span>
<span class="go">                  count         mean  ...            75%            max</span>
<span class="go">biotype                               ...</span>
<span class="go">protein_coding     26.0  1823.089609  ...   1.005060e-90   9.290000e-68</span>
<span class="go">pseudogene          1.0  2688.043701  ...   1.800000e-94   1.800000e-94</span>
<span class="go">unknown             1.0  2085.995094  ...  3.070000e-152  3.070000e-152</span>

<span class="go">[3 rows x 48 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">columns</code></dt>
<dd><p>list of column names</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentiles</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">0.01, 0.25, 0.5, 0.75, 0.99</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Summary statistics of the dataset.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Series or DataFrame</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">1%         0.000000      0.000000      0.000000      0.000000</span>
<span class="go">25%        6.000000      6.250000      1.250000      0.250000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">75%     2637.000000   2479.000000   6030.500000   4669.750000</span>
<span class="go">99%    15054.950000  12714.290000  21955.390000  15603.510000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># show the deciles (10%, 20%, 30%... 90%) of the columns</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">(</span><span class="n">percentiles</span><span class="o">=</span><span class="p">[</span><span class="n">decile</span><span class="o">/</span><span class="mi">10</span> <span class="k">for</span> <span class="n">decile</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">)])</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">10%        0.000000      0.200000      0.000000      0.000000</span>
<span class="go">20%        1.400000      3.200000      1.000000      0.000000</span>
<span class="go">30%       15.000000     15.700000      2.600000      1.000000</span>
<span class="go">40%       28.400000     26.800000     14.000000      9.000000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">60%       82.000000    106.800000     44.000000     33.000000</span>
<span class="go">70%      484.200000    395.500000    305.000000    302.500000</span>
<span class="go">80%     3398.600000   3172.600000   7981.400000   6213.000000</span>
<span class="go">90%     8722.100000   7941.800000  16449.500000  12129.900000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">df</code></dt>
<dd><p>pandas.DataFrame with the data</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00007063&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007071&#39;,</span>
<span class="go"> &#39;WBGene00007074&#39;, &#39;WBGene00007075&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007079&#39;,</span>
<span class="go"> &#39;WBGene00014997&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00044022&#39;,</span>
<span class="go"> &#39;WBGene00044951&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00077504&#39;}</span>
</pre></div>
</div>
<p># calculate difference and filter in-place
&gt;&gt;&gt; d.difference(counts, a_set, inplace=True)
Filtered 2 features, leaving 26 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype/biotypes.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>string</em><em> or </em><em>list of strings</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the biotype reference file used to determine biotypes.         Default is the path defined by the user in the settings.yaml file.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 9 features, leaving 13 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39; or &#39;pseudogene&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">([</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="s1">&#39;pseudogene&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 0 features, leaving 22 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'union'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters features according to user-defined attributes from an Attribute Reference Table.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).         To learn more about user-defined attributes and Attribute Reference Tables, read the user guide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>string</em><em> or </em><em>list of strings</em><em>,         </em><em>which are column titles in the user-defined Attribute Reference Table.</em>) – attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – If ‘union’, filters out every genomic feature that does not belong to one or more         of the indicated attributes. If ‘intersection’,         filters out every genomic feature that does not belong to ALL of the indicated attributes.</p></li>
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – filename/path of the attribute reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em><em> (</em><em>default False</em><em>)</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attribute &#39;attribute1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attributes &#39;attribute1&#39; OR &#39;attribute3&#39; (union)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 14 features, leaving 8 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to both attributes &#39;attribute1&#39; AND &#39;attribute3&#39; (intersection)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 19 features, leaving 3 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to either &#39;attribute1&#39;,&#39;attribute3&#39; or both</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 8 features, leaving 14 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to both &#39;attribute1&#39; AND &#39;attribute3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 3 features, leaving 19 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows whose value in the column &#39;log2FoldChange&#39; is below the 75th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows vulse value in the column &#39;log2FoldChange&#39; is above the 25th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns, then keep only the top ‘n’ rows.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>string</em><em> or </em><em>list of strings</em>) – Names of the column or columns to sort and then filter by.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bools</em><em> (</em><em>default True</em><em>)</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows with the highest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows which have the lowest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># and then the highest values in the column &#39;cond2&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="s1">&#39;cond2&#39;</span><span class="p">],</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">False</span><span class="p">])</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">fname</code></dt>
<dd><p>filename with full path</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the first n rows of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>
<span class="go">WBGene00000005  4028.154186        6.534112  ...  1.700000e-228  7.800000e-225</span>
<span class="go">WBGene00000006  1230.585240        7.157428  ...  2.070000e-216  7.590000e-213</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="c1"># return only the first 3 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>

<span class="go">[3 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_set</code></dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A set of WBGene names.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myset</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">myset</span><span class="p">)</span>
<span class="go">{&#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;,</span>
<span class="go">&#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00007077&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00007075&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_string</code></dt>
<dd><p>Returns a string of all feature indices in the current DataFrame, sorted alphabetically and separated by newline. </p>
<dl class="simple">
<dt>This includes all of the feature indices which were not filtered out by previously-used filter methods. </dt><dd><p>if any duplicate features exist in the filter object (same index appears more than once), </p>
</dd>
</dl>
<p>the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A string of WBGene indices separated by newlines (\n). </p>
</dd>
</dl>
<p>For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mystring</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_string</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mystring</span><span class="p">)</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate intersection and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;}</span>
</pre></div>
</div>
<p># calculate intersection and filter in-place
&gt;&gt;&gt; d.intersection(a_set, inplace=True)
Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt' / 'greater than' / '&gt;'</em><em>, </em><em>'eq' / 'equals' / '='</em><em>, </em><em>'lt' / 'lesser than' / '&lt;'</em>) – the operator to filter the column by (greater than, equal or lesser than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;gt&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;greater than&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;&gt;&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Print the feature indices in the Filter object, sorted alphabetically and separated by newline.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">print_features</span><span class="p">()</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alt_filename</span><span class="p">:</span> <span class="n">Union[None, str, pathlib.Path]</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> (<em>str</em><em>, </em><em>pathlib.Path</em><em>, or </em><em>None</em><em> (</em><em>default</em><em>)</em>) – If None, file name will be generated automatically         according to the filtering methods used.         If it’s a string, it will be used as the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">shape</code></dt>
<dd><p>(rows, columns)</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list of str</em>) – Names of the column or columns to sort by.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bool</em><em>, </em><em>default True</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, perform operation in-place.         Otherwise, returns a sorted copy of the Filter object without modifying the original.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None if inplace=True, a sorted Filter object otherwise.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00007063    633    451    365    388</span>
<span class="go">WBGene00007064     60     57     20     23</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00007066     55    266     46     39</span>
<span class="go">WBGene00007067     15     13      1      0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00077504      0      0      0      0</span>
<span class="go">WBGene00007069      0      2      1      0</span>
<span class="go">WBGene00077502      0      0      0      0</span>
<span class="go">WBGene00077503      1      4      2      0</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Attribute Reference Table attributes.         Each object contains only features that match its Attribute Reference Table attribute.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>list of strings</em>) – list of attribute names from the Attribute Reference Table to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of Filter objects, each containing only features that match one Attribute Reference Table attribute;         the Filter objects are returned in the same order the attributes were given in.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attribute1</span><span class="p">,</span><span class="n">attribute2</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         below and above the specified percentile in the spcfieid column.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">below</span><span class="p">,</span> <span class="n">above</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split_by_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Filtered 21 features, leaving 7 of the original 28 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000024&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;,</span>
<span class="go">&#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000029&#39;, &#39;WBGene00000006&#39;,</span>
<span class="go">&#39;WBGene00007064&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;,</span>
<span class="go">&#39;WBGene00007074&#39;, &#39;WBGene00000025&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000027&#39;,</span>
<span class="go">&#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000012&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00077503&#39;,</span>
<span class="go">&#39;WBGene00000026&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000002&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;, &#39;WBGene00007077&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00043987&#39;,</span>
<span class="go">&#39;WBGene00007071&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame</dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>returns the last n rows of the Filter object.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span> <span class="c1"># returns the last 8 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000022   365.813048        6.101303  ...  2.740000e-97  2.400000e-94</span>
<span class="go">WBGene00000023  3168.566714        3.906719  ...  1.600000e-93  1.340000e-90</span>
<span class="go">WBGene00000024   221.925724        4.801676  ...  1.230000e-84  9.820000e-82</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[8 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq' / 'equals' / '='</em><em>, </em><em>'ct' / 'contains' / 'in'</em><em>, </em><em>'bw' / 'begins with'</em><em>, </em><em>'ew' / 'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/text_filters.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that have a value that starts with &#39;AC3&#39; in the column &#39;name&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">text_filters</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sw&#39;</span><span class="p">,</span><span class="s1">&#39;AC3&#39;</span><span class="p">)</span>
<span class="go">Filtered 17 features, leaving 5 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate union and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000012&#39;,</span>
<span class="go">&#39;WBGene00000024&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00007069&#39;,</span>
<span class="go">&#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000026&#39;, &#39;WBGene00000029&#39;,</span>
<span class="go">&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;,</span>
<span class="go">&#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00000025&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;,</span>
<span class="go">&#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000027&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">FoldChangeFilter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fname</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>pathlib.Path<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">numerator_name</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">denominator_name</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><code class="xref py py-class docutils literal notranslate"><span class="pre">rnalysis.filtering.Filter</span></code></a></p>
<dl class="simple">
<dt>A class that contains a single column, representing the gene-specific fold change between two conditions. </dt><dd><p>this class does not support ‘inf’ and ‘0’ values, and importing a file with such values could lead     to incorrect filtering and statistical analyses.</p>
</dd>
</dl>
<p><strong>Attributes</strong></p>
<dl class="simple">
<dt>df: pandas Series</dt><dd><p>A Series that contains the fold change values.         The Series is modified upon usage of filter operations. .</p>
</dd>
<dt>shape: tuple (rows, columns)</dt><dd><p>The dimensions of df.</p>
</dd>
<dt>columns: list</dt><dd><p>The columns of df.</p>
</dd>
<dt>fname: pathlib.Path</dt><dd><p>The path and filename for the purpose of saving df as a csv file.         Updates automatically when filter operations are applied.</p>
</dd>
<dt>index_set: set</dt><dd><p>All of the indices in the current DataFrame (which were not removed by previously used filter methods)         as a set.</p>
</dd>
<dt>index_string: string</dt><dd><p>A string of all feature indices in the current DataFrame separated by newline.</p>
</dd>
<dt>numerator: str</dt><dd><p>Name of the numerator used to calculate the fold change.</p>
</dd>
<dt>denominator: str</dt><dd><p>Name of the denominator used to calculate the fold change.</p>
</dd>
</dl>
<dl class="py method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">return_format</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'short'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the Filter object and their count.
:type return_format: ‘short’ or ‘long’ (default ‘short’)
:param return_format: ‘short’ returns a short-form DataFrame, which states the biotypes         in the Filter object and their count. ‘long’ returns a long-form DataFrame,
which also provides descriptive statistics of each column per biotype.
:param ref: Name of the biotype reference table used to determine biotype. Default is ce11 (included in the package).
:rtype: pandas.DataFrame
:returns: a pandas DataFrame showing the number of values belonging to each biotype,         as well as additional descriptive statistics of format==’long’.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># short-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># long-form view</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">return_format</span><span class="o">=</span><span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">               baseMean               ...           padj</span>
<span class="go">                  count         mean  ...            75%            max</span>
<span class="go">biotype                               ...</span>
<span class="go">protein_coding     26.0  1823.089609  ...   1.005060e-90   9.290000e-68</span>
<span class="go">pseudogene          1.0  2688.043701  ...   1.800000e-94   1.800000e-94</span>
<span class="go">unknown             1.0  2085.995094  ...  3.070000e-152  3.070000e-152</span>

<span class="go">[3 rows x 48 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">columns</code></dt>
<dd></dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">denominator</code></dt>
<dd><p>name of the denominator</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">describe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentiles</span><span class="p">:</span> <span class="n">list</span> <span class="o">=</span> <span class="default_value">0.01, 0.25, 0.5, 0.75, 0.99</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate descriptive statistics that summarize the central tendency, dispersion and shape         of the dataset’s distribution, excluding NaN values.         For more information see the documentation of pandas.DataFrame.describe.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>percentiles</strong> (<em>list-like of numbers</em><em>, </em><em>optional</em>) – The percentiles to include in the output.         All should fall between 0 and 1.         The default is [.25, .5, .75], which returns the 25th, 50th, and 75th percentiles.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Summary statistics of the dataset.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Series or DataFrame</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">()</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">1%         0.000000      0.000000      0.000000      0.000000</span>
<span class="go">25%        6.000000      6.250000      1.250000      0.250000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">75%     2637.000000   2479.000000   6030.500000   4669.750000</span>
<span class="go">99%    15054.950000  12714.290000  21955.390000  15603.510000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="c1"># show the deciles (10%, 20%, 30%... 90%) of the columns</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">describe</span><span class="p">(</span><span class="n">percentiles</span><span class="o">=</span><span class="p">[</span><span class="n">decile</span><span class="o">/</span><span class="mi">10</span> <span class="k">for</span> <span class="n">decile</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">)])</span>
<span class="go">              cond1         cond2         cond3         cond4</span>
<span class="go">count     22.000000     22.000000     22.000000     22.000000</span>
<span class="go">mean    2515.590909   2209.227273   4230.227273   3099.818182</span>
<span class="go">std     4820.512674   4134.948493   7635.832664   5520.394522</span>
<span class="go">min        0.000000      0.000000      0.000000      0.000000</span>
<span class="go">10%        0.000000      0.200000      0.000000      0.000000</span>
<span class="go">20%        1.400000      3.200000      1.000000      0.000000</span>
<span class="go">30%       15.000000     15.700000      2.600000      1.000000</span>
<span class="go">40%       28.400000     26.800000     14.000000      9.000000</span>
<span class="go">50%       57.500000     52.500000     23.500000     21.000000</span>
<span class="go">60%       82.000000    106.800000     44.000000     33.000000</span>
<span class="go">70%      484.200000    395.500000    305.000000    302.500000</span>
<span class="go">80%     3398.600000   3172.600000   7981.400000   6213.000000</span>
<span class="go">90%     8722.100000   7941.800000  16449.500000  12129.900000</span>
<span class="go">max    15056.000000  12746.000000  22027.000000  15639.000000</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">df</code></dt>
<dd></dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in the first Filter object/set but NOT in the others.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist only in the first Filter object.         If ‘str’, returns a string of the WBGene indices that exist only in the first Filter object,         delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices         that exist only in the first Filter object/set (set difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00007063&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007071&#39;,</span>
<span class="go"> &#39;WBGene00007074&#39;, &#39;WBGene00007075&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007079&#39;,</span>
<span class="go"> &#39;WBGene00014997&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00044022&#39;,</span>
<span class="go"> &#39;WBGene00044951&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00077504&#39;}</span>
</pre></div>
</div>
<p># calculate difference and filter in-place
&gt;&gt;&gt; d.difference(counts, a_set, inplace=True)
Filtered 2 features, leaving 26 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_abs_log2_fold_change</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">abslog2fc</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">1</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features whose absolute log2 fold change is below the indicated threshold.         For example: if log2fc is 1.0, all features whose log2 fold change is between 1 and -1 (went up less than         two-fold or went down less than two-fold) will be filtered out.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>abslog2fc</strong> – The threshold absolute log2 fold change for filtering out a feature. Float or int.         All features whose absolute log2 fold change is lower than log2fc will be filtered out.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current FoldChangeFilter object. If False,         the function will return a new FoldChangeFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of FoldChangeFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/fc_1.csv&#39;</span><span class="p">,</span><span class="s1">&#39;numerator name&#39;</span><span class="p">,</span><span class="s1">&#39;denominator name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">filter_abs_log2_fold_change</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span> <span class="c1"># keep only rows whose log2(fold change) is &gt;=2 or &lt;=-2</span>
<span class="go">Filtered 18 features, leaving 4 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_biotype</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out all features that do not match the indicated biotype/biotypes.         Legal inputs: ‘protein_coding’,’pseudogene’,’piRNA’,’miRNA’,’ncRNA’,’lincRNA’,’rRNA’,’snRNA’,’snoRNA’.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>biotype</strong> (<em>string</em><em> or </em><em>list of strings</em>) – the biotypes which will not be filtered out.</p></li>
<li><p><strong>ref</strong> – Name of the biotype reference file used to determine biotypes.         Default is the path defined by the user in the settings.yaml file.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 9 features, leaving 13 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows whose biotype is &#39;protein_coding&#39; or &#39;pseudogene&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">([</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span><span class="s1">&#39;pseudogene&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 0 features, leaving 22 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'union'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters features according to user-defined attributes from an Attribute Reference Table.         When multiple attributes are given, filtering can be done in ‘union’ mode         (where features that belong to at least one attribute are not filtered out), or in ‘intersection’ mode         (where only features that belong to ALL attributes are not filtered out).         To learn more about user-defined attributes and Attribute Reference Tables, read the user guide.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>string</em><em> or </em><em>list of strings</em><em>,         </em><em>which are column titles in the user-defined Attribute Reference Table.</em>) – attributes to filter by.</p></li>
<li><p><strong>mode</strong> (<em>'union'</em><em> or </em><em>'intersection'.</em>) – If ‘union’, filters out every genomic feature that does not belong to one or more         of the indicated attributes. If ‘intersection’,         filters out every genomic feature that does not belong to ALL of the indicated attributes.</p></li>
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – filename/path of the attribute reference table to be used as reference.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em><em> (</em><em>default False</em><em>)</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attribute &#39;attribute1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to the attributes &#39;attribute1&#39; OR &#39;attribute3&#39; (union)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 14 features, leaving 8 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that belong to both attributes &#39;attribute1&#39; AND &#39;attribute3&#39; (intersection)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 19 features, leaving 3 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to either &#39;attribute1&#39;,&#39;attribute3&#39; or both</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 8 features, leaving 14 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that DON&#39;T belong to both &#39;attribute1&#39; AND &#39;attribute3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute3&#39;</span><span class="p">],</span><span class="n">mode</span><span class="o">=</span><span class="s1">&#39;intersection&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 3 features, leaving 19 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_fold_change_direction</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">direction</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'pos'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Filters out features according to the direction in which they changed between the two conditions.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>direction</strong> – ‘pos’ or ‘neg’. If ‘pos’, will keep only features that have positive log2foldchange.         If ‘neg’, will keep only features that have negative log2foldchange.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current FoldChangeFilter object. If False,         the function will return a new FoldChangeFilter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of FoldChangeFilter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/fc_1.csv&#39;</span><span class="p">,</span><span class="s1">&#39;numerator name&#39;</span><span class="p">,</span><span class="s1">&#39;denominator name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows with a positive log2(fold change) value</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">filter_fold_change_direction</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">)</span>
<span class="go">Filtered 10 features, leaving 12 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/fc_1.csv&#39;</span><span class="p">,</span><span class="s1">&#39;numerator name&#39;</span><span class="p">,</span><span class="s1">&#39;denominator name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span> <span class="c1"># keep only rows with a negative log2(fold change) value</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">filter_fold_change_direction</span><span class="p">(</span><span class="s1">&#39;neg&#39;</span><span class="p">)</span>
<span class="go">Filtered 14 features, leaving 8 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/fc_1.csv&#39;</span><span class="p">,</span><span class="s1">&#39;numerator name&#39;</span><span class="p">,</span><span class="s1">&#39;denominator name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows with a non-positive log2(fold change) value</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">filter_fold_change_direction</span><span class="p">(</span><span class="s1">&#39;pos&#39;</span><span class="p">,</span> <span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Removes all entries above the specified percentile in the specified column.         For example, if the column were ‘pvalue’ and the percentile was 0.5, then all features whose pvalue is above         the median pvalue will be filtered out.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If inplace is False, returns a new and filtered instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows whose value in the column &#39;log2FoldChange&#39; is below the 75th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the rows vulse value in the column &#39;log2FoldChange&#39; is above the 25th percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">filter_top_n</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">100</span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns, then keep only the top ‘n’ rows.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>string</em><em> or </em><em>list of strings</em>) – Names of the column or columns to sort and then filter by.</p></li>
<li><p><strong>n</strong> (<em>int</em>) – How many features to keep in the Filter object.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bools</em><em> (</em><em>default True</em><em>)</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of Filter.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows with the highest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only the 10 rows which have the lowest values in the columns &#39;cond1&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># and then the highest values in the column &#39;cond2&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="s1">&#39;cond2&#39;</span><span class="p">],</span><span class="n">n</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="p">[</span><span class="kc">True</span><span class="p">,</span><span class="kc">False</span><span class="p">])</span>
<span class="go">Filtered 12 features, leaving 10 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">fname</code></dt>
<dd></dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">head</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span></dt>
<dd><p>Return the first n rows of the Filter object. See pandas.DataFrame.head documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>returns the first n rows of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>
<span class="go">WBGene00000005  4028.154186        6.534112  ...  1.700000e-228  7.800000e-225</span>
<span class="go">WBGene00000006  1230.585240        7.157428  ...  2.070000e-216  7.590000e-213</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="c1"># return only the first 3 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>

<span class="go">[3 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_set</code></dt>
<dd><p>Returns all of the features in the current DataFrame (which were not removed by previously used filter methods)         as a set.         if any duplicate features exist in the filter object (same WBGene appears more than once),         the corresponding WBGene index will appear in the returned set ONLY ONCE.</p>
<dl class="field-list">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A set of WBGene names.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">myset</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">myset</span><span class="p">)</span>
<span class="go">{&#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;,</span>
<span class="go">&#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00007077&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00007075&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<em class="property">property </em><code class="sig-name descname">index_string</code></dt>
<dd><p>Returns a string of all feature indices in the current DataFrame, sorted alphabetically and separated by newline. </p>
<dl class="simple">
<dt>This includes all of the feature indices which were not filtered out by previously-used filter methods. </dt><dd><p>if any duplicate features exist in the filter object (same index appears more than once), </p>
</dd>
</dl>
<p>the corresponding index will appear in the returned string ONLY ONCE.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A string of WBGene indices separated by newlines (\n). </p>
</dd>
</dl>
<p>For example, “WBGene00000001\nWBGene00000003\nWBGene12345678”.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mystring</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">index_string</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">mystring</span><span class="p">)</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><p>Keep only the features that exist in ALL of the given Filter objects/sets.         Can be done inplace on the first Filter object, or return a set/string of features.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate intersection with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the intersecting WBGene indices. If ‘str’, returns a string of         the intersecting indices, delimited by a comma.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default False</em>) – If True, filtering will be applied to the current Filter object. If False (default),         the function will return a set/str that contains the intersecting indices.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>If inplace=False, returns a set/string of the WBGene indices that intersect between two Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate intersection and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;}</span>
</pre></div>
</div>
<p># calculate intersection and filter in-place
&gt;&gt;&gt; d.intersection(a_set, inplace=True)
Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">number_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a number filter (greater than, equal, lesser than) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'gt' / 'greater than' / '&gt;'</em><em>, </em><em>'eq' / 'equals' / '='</em><em>, </em><em>'lt' / 'lesser than' / '&lt;'</em>) – the operator to filter the column by (greater than, equal or lesser than)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;gt&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;greater than&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/test_deseq.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#keep only rows that have a value greater than 5900 in the column &#39;baseMean&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;&gt;&#39;</span><span class="p">,</span><span class="mi">5900</span><span class="p">)</span>
<span class="go">Filtered 26 features, leaving 2 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">numerator</code></dt>
<dd><p>name of the numerator</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">print_features</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Print the feature indices in the Filter object, sorted alphabetically and separated by newline.</p>
<dl class="field-list">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">print_features</span><span class="p">()</span>
<span class="go">WBGene00007063</span>
<span class="go">WBGene00007064</span>
<span class="go">WBGene00007066</span>
<span class="go">WBGene00007067</span>
<span class="go">WBGene00007069</span>
<span class="go">WBGene00007071</span>
<span class="go">WBGene00007074</span>
<span class="go">WBGene00007075</span>
<span class="go">WBGene00007076</span>
<span class="go">WBGene00007077</span>
<span class="go">WBGene00007078</span>
<span class="go">WBGene00007079</span>
<span class="go">WBGene00014997</span>
<span class="go">WBGene00043987</span>
<span class="go">WBGene00043988</span>
<span class="go">WBGene00043989</span>
<span class="go">WBGene00043990</span>
<span class="go">WBGene00044022</span>
<span class="go">WBGene00044951</span>
<span class="go">WBGene00077502</span>
<span class="go">WBGene00077503</span>
<span class="go">WBGene00077504</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">randomization_test</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ref</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">reps</span><span class="o">=</span><span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">save_csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">fname</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Perform a randomization test to examine whether the fold change of a group of specific genomic features         is significantly different than the fold change of a background set of genomic features.</p>
<dl class="field-list simple">
<dt class="field-odd">type ref</dt>
<dd class="field-odd"><p>FoldChangeFilter</p>
</dd>
<dt class="field-even">param ref</dt>
<dd class="field-even"><p>A reference FoldChangeFilter object which contains the fold change for every reference gene.         Will be used to calculate the expected score and to perform randomizations.</p>
</dd>
<dt class="field-odd">type alpha</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param alpha</dt>
<dd class="field-even"><p>Indicates the threshold for significance (alpha).</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.         10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:         r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list">
<dt class="field-odd">rtype</dt>
<dd class="field-odd"><p>pandas DataFrame</p>
</dd>
<dt class="field-even">return</dt>
<dd class="field-even"><p>A Dataframe with the number of given genes, the observed fold change for the given group of genes,         the expected fold change for a group of genes of that size and the p value for the comparison.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/fc_1.csv&#39;</span> <span class="p">,</span> <span class="s1">&#39;numerator&#39;</span> <span class="p">,</span> <span class="s1">&#39;denominator&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f_background</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter_biotype</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="c1">#keep only protein-coding genes as reference</span>
<span class="go">Filtered 9 features, leaving 13 of the original 22 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f_test</span> <span class="o">=</span> <span class="n">f_background</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 6 features, leaving 7 of the original 13 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rand_test_res</span> <span class="o">=</span> <span class="n">f_test</span><span class="o">.</span><span class="n">randomization_test</span><span class="p">(</span><span class="n">f_background</span><span class="p">)</span>
<span class="go">Calculating...</span>
<span class="go">   group size  observed fold change  ...      pval  significant</span>
<span class="go">0           7              2.806873  ...  0.360264        False</span>
</pre></div>
</div>
<p>[1 rows x 5 columns]</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">save_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">alt_filename</span><span class="p">:</span> <span class="n">Union[None, str, pathlib.Path]</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Saves the current filtered data to a .csv file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>alt_filename</strong> (<em>str</em><em>, </em><em>pathlib.Path</em><em>, or </em><em>None</em><em> (</em><em>default</em><em>)</em>) – If None, file name will be generated automatically         according to the filtering methods used.         If it’s a string, it will be used as the name of the saved file. Example input: ‘myfilename’</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">shape</code></dt>
<dd></dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">by</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ascending</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>bool<span class="p">, </span>List<span class="p">[</span>bool<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">na_position</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'last'</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Sort the rows by the values of specified column or columns.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>by</strong> (<em>str</em><em> or </em><em>list of str</em>) – Names of the column or columns to sort by.</p></li>
<li><p><strong>ascending</strong> (<em>bool</em><em> or </em><em>list of bool</em><em>, </em><em>default True</em>) – Sort ascending vs. descending. Specify list for multiple sort orders.         If this is a list of bools, it must have the same length as ‘by’.</p></li>
<li><p><strong>na_position</strong> (<em>'first'</em><em> or </em><em>'last'</em><em>, </em><em>default 'last'</em>) – If ‘first’, puts NaNs at the beginning; if ‘last’, puts NaNs at the end.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>default True</em>) – If True, perform operation in-place.         Otherwise, returns a sorted copy of the Filter object without modifying the original.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>None if inplace=True, a sorted Filter object otherwise.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00007063    633    451    365    388</span>
<span class="go">WBGene00007064     60     57     20     23</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00007066     55    266     46     39</span>
<span class="go">WBGene00007067     15     13      1      0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="s1">&#39;cond1&#39;</span><span class="p">,</span><span class="n">ascending</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">                cond1  cond2  cond3  cond4</span>
<span class="go">WBGene00044951      0      0      0      1</span>
<span class="go">WBGene00077504      0      0      0      0</span>
<span class="go">WBGene00007069      0      2      1      0</span>
<span class="go">WBGene00077502      0      0      0      0</span>
<span class="go">WBGene00077503      1      4      2      0</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_attribute</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">List<span class="p">[</span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into multiple Filter objects,         each corresponding to one of the specified Attribute Reference Table attributes.         Each object contains only features that match its Attribute Reference Table attribute.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attributes</strong> (<em>list of strings</em>) – list of attribute names from the Attribute Reference Table to filter by.</p></li>
<li><p><strong>ref</strong> – filename/path of the reference table to be used as reference.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of Filter objects, each containing only features that match one Attribute Reference Table attribute;         the Filter objects are returned in the same order the attributes were given in.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attribute1</span><span class="p">,</span><span class="n">attribute2</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_by_attribute</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span>
<span class="gp">... </span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 15 features, leaving 7 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_by_percentile</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">percentile</span><span class="p">:</span> <span class="n">float</span></em>, <em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em><span class="sig-paren">)</span></dt>
<dd><p>Splits the Filter object into two Filter objects:         below and above the specified percentile in the spcfieid column.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>percentile</strong> (<em>float between 0 and 1</em>) – The percentile that all features above it will be filtered out.</p></li>
<li><p><strong>column</strong> (<em>str</em>) – Name of the DataFrame column according to which the filtering will be performed.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a tuple of two Filter objects: the first contains all of the features below the specified percentile,         and the second contains all of the features above and equal to the specified percentile.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">below</span><span class="p">,</span> <span class="n">above</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">split_by_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Filtered 21 features, leaving 7 of the original 28 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">split_fold_change_direction</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Splits the features in the current FoldChangeFilter object into two complementary, non-overlapping         FoldChangeFilter objects, based on the direction of their log2(fold change).         The first object will contain only features with a positive log2(fold change),         the second object will contain only features with a negative log2(fold change).         Features with log2(fold change) = 0 will be ignored.</p>
<dl class="field-list">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a tuple containing two FoldChangeFilter objects: the first has only features with positive log2 fold change,         and the other has only features with negative log2 fold change.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/fc_1.csv&#39;</span><span class="p">,</span><span class="s1">&#39;numerator name&#39;</span><span class="p">,</span><span class="s1">&#39;denominator name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pos_log2fc</span><span class="p">,</span> <span class="n">neg_log2fc</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">split_fold_change_direction</span><span class="p">()</span>
<span class="go">Filtered 10 features, leaving 12 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 14 features, leaving 8 of the original 22 features. Filtering result saved to new object.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the WBGene indices that exist either in the first Filter object/set OR the second,         but NOT in both (set symmetric difference).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set.</em>) – a second Filter object/set to calculate symmetric difference with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the WBGene indices that exist in exactly one Filter object.         If ‘str’, returns a string of the WBGene indices that exist in exactly one Filter object, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that that exist t in exactly one Filter. (set symmetric difference).</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate difference and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000024&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;,</span>
<span class="go">&#39;WBGene00007069&#39;, &#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000029&#39;, &#39;WBGene00000006&#39;,</span>
<span class="go">&#39;WBGene00007064&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;,</span>
<span class="go">&#39;WBGene00007074&#39;, &#39;WBGene00000025&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000027&#39;,</span>
<span class="go">&#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000012&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00077503&#39;,</span>
<span class="go">&#39;WBGene00000026&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000002&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;, &#39;WBGene00007077&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00043987&#39;,</span>
<span class="go">&#39;WBGene00007071&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">tail</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">5</span></em><span class="sig-paren">)</span> &#x2192; pandas.core.frame.DataFrame</dt>
<dd><p>Return the last n rows of the Filter object. See pandas.DataFrame.tail documentation.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n</strong> (<em>positive int</em><em>, </em><em>default 5</em>) – Number of rows to show.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>returns the last n rows of the Filter object.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">()</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[5 rows x 6 columns]</span>
</pre></div>
</div>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span> <span class="c1"># returns the last 8 rows</span>
<span class="go">                   baseMean  log2FoldChange  ...        pvalue          padj</span>
<span class="go">WBGene00000022   365.813048        6.101303  ...  2.740000e-97  2.400000e-94</span>
<span class="go">WBGene00000023  3168.566714        3.906719  ...  1.600000e-93  1.340000e-90</span>
<span class="go">WBGene00000024   221.925724        4.801676  ...  1.230000e-84  9.820000e-82</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[8 rows x 6 columns]</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">text_filters</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">column</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">operator</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">value</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">opposite</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Applay a text filter (equals, contains, starts with, ends with) on a particular column in the Filter object.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>column</strong> (<em>str</em>) – name of the column to filter by</p></li>
<li><p><strong>operator</strong> (<em>str: 'eq' / 'equals' / '='</em><em>, </em><em>'ct' / 'contains' / 'in'</em><em>, </em><em>'bw' / 'begins with'</em><em>, </em><em>'ew' / 'ends with'</em>) – the operator to filter the column by (equals, contains, starts with, ends with)</p></li>
<li><p><strong>value</strong> (<em>number</em><em> (</em><em>int</em><em> or </em><em>float</em><em>)</em>) – the value to filter by</p></li>
<li><p><strong>opposite</strong> (<em>bool</em>) – If True, the output of the filtering will be the OPPOSITE of the specified         (instead of filtering out X, the function will filter out anything BUT X).         If False (default), the function will filter as expected.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), filtering will be applied to the current Filter object. If False,         the function will return a new Filter instance and the current instance will not be affected.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If ‘inplace’ is False, returns a new instance of the Filter object.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/text_filters.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># keep only rows that have a value that starts with &#39;AC3&#39; in the column &#39;name&#39;.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">filt</span><span class="o">.</span><span class="n">text_filters</span><span class="p">(</span><span class="s1">&#39;name&#39;</span><span class="p">,</span><span class="s1">&#39;sw&#39;</span><span class="p">,</span><span class="s1">&#39;AC3&#39;</span><span class="p">)</span>
<span class="go">Filtered 17 features, leaving 5 of the original 22 features. Filtered inplace.</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">return_type</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'set'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a set/string of the union of WBGene indices between multiple Filter objects         (the indices that exist in at least one of the Filter objects/sets).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter"><em>Filter</em></a><em> or </em><em>set objects.</em>) – Objects to calculate union with.</p></li>
<li><p><strong>return_type</strong> (<em>'set'</em><em> or </em><em>'str.</em>) – If ‘set’, returns a set of the union WBGene indices. If ‘str’, returns a string of         the union WBGene indices, delimited by a comma.</p></li>
</ul>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>set or str</p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>a set/string of the WBGene indices that exist in at least one of the Filter objects.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s1">&#39;tests/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># calculate union and return a set</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">counts</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000017&#39;, &#39;WBGene00000021&#39;, &#39;WBGene00044022&#39;, &#39;WBGene00077504&#39;, &#39;WBGene00000012&#39;,</span>
<span class="go">&#39;WBGene00000024&#39;, &#39;WBGene00007079&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000020&#39;, &#39;WBGene00000005&#39;, &#39;WBGene00007069&#39;,</span>
<span class="go">&#39;WBGene00007063&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00000026&#39;, &#39;WBGene00000029&#39;,</span>
<span class="go">&#39;WBGene00000002&#39;, &#39;WBGene00000003&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00044951&#39;,</span>
<span class="go">&#39;WBGene00000007&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00007066&#39;,</span>
<span class="go">&#39;WBGene00007076&#39;, &#39;WBGene00000013&#39;, &#39;WBGene00014997&#39;, &#39;WBGene00000023&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00007074&#39;,</span>
<span class="go">&#39;WBGene00000025&#39;, &#39;WBGene00000011&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00007071&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000018&#39;,</span>
<span class="go">&#39;WBGene00043989&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000016&#39;, &#39;WBGene00000027&#39;, &#39;WBGene00000028&#39;,</span>
<span class="go">&#39;WBGene00007075&#39;, &#39;WBGene00000022&#39;, &#39;WBGene00000009&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.filtering.</code><code class="sig-name descname">Pipeline</code><span class="sig-paren">(</span><em class="sig-param">filter_type: Union[Type[rnalysis.filtering.Filter], Type[rnalysis.filtering.DESeqFilter], Type[rnalysis.filtering.FoldChangeFilter], Type[rnalysis.filtering.CountFilter], str] = &lt;class 'rnalysis.filtering.Filter'&gt;</em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt>
<code class="sig-name descname">add_function</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">func</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>function<span class="p">, </span>str<span class="p">]</span></span></em>, <em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span></dt>
<dd><p>Add a function to the pipeline. Arguments can be stated with or without the correspoding keywords.         For example: Pipeline.add_function(‘sort’, ‘columnName’, ascending=False, na_position=’first’).         Pipelines support virtually all functions in the filtering module, including         filtering functions, normalization functions, splitting functions and plotting functions.         Do not include the ‘inplace’ argument when adding functions to a pipeline; instead,         include it when applying the pipeline to a Filter object using Pipeline.apply_to(). 
:param func: function to be added
:type func: function or name of function from the filtering module
:param args: unkeyworded arguments for the added function in their natural order. For example: 0.1, True
:param kwargs: keyworded arguments for the added function. For example: opposite=True</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">apply_to</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filter_object</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filter_object</strong> – </p></li>
<li><p><strong>inplace</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><dl class="field-list simple">
<dt class="field-odd">rtype</dt>
<dd class="field-odd"><p></p></dd>
</dl>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">filter_type</code></dt>
<dd><p>type of filter objects to which the Pipeline will be applied</p>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">functions</code></dt>
<dd><p>list of functions to perform</p>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">params</code></dt>
<dd><p>list of function parameters</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">remove_last_function</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Removes from the Pipeline the last function that was added to it. Removal is in-place.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="rnalysis-enrichment-module">
<h2>rnalysis.enrichment module<a class="headerlink" href="#rnalysis-enrichment-module" title="Permalink to this headline">¶</a></h2>
<p>This module can perform enrichment analyses on a given set of genomic features and visualize their intersections. These include gene ontology/tissue/phenotype enrichment, enrichment for user-defined attributes, set visualization ,etc. Results of enrichment analyses can be saved to .csv files.</p>
<dl class="py class">
<dt>
<em class="property">class </em><code class="sig-prename descclassname">rnalysis.enrichment.</code><code class="sig-name descname">FeatureSet</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">gene_set</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>List<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>Set<span class="p">[</span>str<span class="p">]</span><span class="p">, </span><a class="reference internal" href="rnalysis.filtering.Filter.html#rnalysis.filtering.Filter" title="rnalysis.filtering.Filter">rnalysis.filtering.Filter</a><span class="p">, </span><a class="reference internal" href="rnalysis.filtering.CountFilter.html#rnalysis.filtering.CountFilter" title="rnalysis.filtering.CountFilter">rnalysis.filtering.CountFilter</a><span class="p">, </span><a class="reference internal" href="rnalysis.filtering.DESeqFilter.html#rnalysis.filtering.DESeqFilter" title="rnalysis.filtering.DESeqFilter">rnalysis.filtering.DESeqFilter</a><span class="p">, </span><a class="reference internal" href="rnalysis.filtering.FoldChangeFilter.html#rnalysis.filtering.FoldChangeFilter" title="rnalysis.filtering.FoldChangeFilter">rnalysis.filtering.FoldChangeFilter</a><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">set_name</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">''</span></em><span class="sig-paren">)</span></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>receives a filtered gene set and preforms various enrichment analyses</p>
<dl class="py method">
<dt>
<code class="sig-name descname">biotypes</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Returns a DataFrame of the biotypes in the gene set and their count.</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – Path of the reference file used to determine biotype.         Default is the path predefined in the settings file.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">enrichment</span><span class="p">,</span> <span class="n">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="p">(</span><span class="s2">&quot;tests/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">biotypes</span><span class="p">(</span><span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Calculates the set difference of the indices from multiple FeatureSet objects         (the indices that appear in the first FeatureSet object but NOT in the other objects).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – The objects against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000006&#39;</span><span class="p">},</span> <span class="s1">&#39;set name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en2</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000004&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">en2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">en</span><span class="p">)</span>
<span class="go">FeatureSet: set name</span>
<span class="go">{&#39;WBGene00000006&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">enrich_hypergeometric</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>Iterable<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>str<span class="p">, </span>Iterable<span class="p">[</span>int<span class="p">]</span><span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">fdr</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">background_genes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">attr_ref_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">biotype_ref_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">save_csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">fname</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">return_fig</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Calculates enrichment scores, p-values and adjusted p-values         for enrichment and depletion of selected attributes from an Attribute Reference Table,         based on a hypergeometric test.         Background set is determined by either the input variable ‘background_genes’,         or by the input variable ‘biotype’ and a Biotype Reference Table.         P-values are calculated using a hypergeometric test:         Given M genes in the background set, n genes in the test set,         with N genes from the background set belonging to a specific attribute (or ‘success’)         and X genes from the test set belonging to that attribute.         If we were to randomly draw n genes from the background set (without replacement),         what is the probability of drawing X or more (in case of enrichment)/X or less (in case of depletion)         genes belonging to the given attribute?         P-values are corrected for multiple comparisons using         the Benjamini–Hochberg step-up procedure (original FDR method).         Enrichment/depletion is determined automatically by the calculated enrichment score:         if log2(enrichment score) is positive then enrichment is assumed,         and if log2(enrichment score) is negative then depletion is assumed.         In plots, for the clarity of display, complete depletion (linear enrichment = 0)         appears with the smallest value in the scale.</p>
<dl class="field-list simple">
<dt class="field-odd">type attributes</dt>
<dd class="field-odd"><p>str, int, iterable (list, tuple, set, etc) of str/int, or ‘all’.</p>
</dd>
<dt class="field-even">param attributes</dt>
<dd class="field-even"><p>An iterable of attribute names or attribute numbers         (according to their order in the Attribute Reference Table).         If ‘all’, all of the attributes in the Attribute Reference Table will be used.         If None, a manual input prompt will be raised.</p>
</dd>
<dt class="field-odd">type fdr</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param fdr</dt>
<dd class="field-even"><p>Indicates the FDR threshold for significance.</p>
</dd>
<dt class="field-odd">type attr_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param attr_ref_path</dt>
<dd class="field-even"><p>the path of the Attribute Reference Table from which user-defined attributes will be drawn.</p>
</dd>
<dt class="field-odd">type biotype_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param biotype_ref_path</dt>
<dd class="field-even"><p>the path of the Biotype Reference Table.         Will be used to generate background set if ‘biotype’ is specified.</p>
</dd>
<dt class="field-odd">type biotype</dt>
<dd class="field-odd"><p>str specifying a specific biotype, list/set of strings each specifying a biotype, or ‘all’.         Default ‘protein_coding’.</p>
</dd>
<dt class="field-even">param biotype</dt>
<dd class="field-even"><p>determines the background genes by their biotype. Requires specifying a Biotype Reference Table.         ‘all’ will include all genomic features in the reference table,         ‘protein_coding’ will include only protein-coding genes from the reference table, etc.         Cannot be specified together with ‘background_genes’.</p>
</dd>
<dt class="field-odd">type background_genes</dt>
<dd class="field-odd"><p>set of feature indices, filtering.Filter object, or enrichment.FeatureSet object</p>
</dd>
<dt class="field-even">param background_genes</dt>
<dd class="field-even"><p>a set of specific feature indices to be used as background genes.         Cannot be specified together with ‘biotype’.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:         r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list">
<dt class="field-odd">type return_fig</dt>
<dd class="field-odd"><p>bool (default False)</p>
</dd>
<dt class="field-even">param return_fig</dt>
<dd class="field-even"><p>if True, returns a matplotlib Figure object in addition to the results DataFrame.
:rtype: pd.DataFrame (default) or Tuple[pd.DataFrame, matplotlib.figure.Figure]
:return:         a pandas DataFrame with the indicated attribute names as rows/index, and the columns ‘log2_fold_enrichment’
and ‘pvalue’; and a matplotlib Figure, if ‘return_figure’ is set to True.</p>
<div class="figure align-center" id="id10">
<a class="reference internal image-reference" href="_images/enrichment_randomization.png"><img alt="_images/enrichment_randomization.png" src="_images/enrichment_randomization.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of enrich_hypergeometric()</span><a class="headerlink" href="#id10" title="Permalink to this image">¶</a></p>
</div>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">enrich_randomization</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>Iterable<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>str<span class="p">, </span>Iterable<span class="p">[</span>int<span class="p">]</span><span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">fdr</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">reps</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">background_genes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">attr_ref_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">biotype_ref_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">save_csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">fname</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">return_fig</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Calculates enrichment scores, p-values and adjusted p-values         for enrichment and depletion of selected attributes from an Attribute Reference Table.         Background set is determined by either the input variable ‘background_genes’,         or by the input variable ‘biotype’ and a Biotype Reference Table.         P-values are calculated using a randomization test with the formula p = (successes + 1)/(repeats + 1).         P-values are corrected for multiple comparisons using         the Benjamini–Hochberg step-up procedure (original FDR method).         Enrichment/depletion is determined automatically by the calculated enrichment score:         if log2(enrichment score) is positive then enrichment is assumed,         and if log2(enrichment score) is negative then depletion is assumed.         In plots, for the clarity of display, complete depletion (linear enrichment = 0)         appears with the smallest value in the scale.</p>
<dl class="field-list simple">
<dt class="field-odd">type attributes</dt>
<dd class="field-odd"><p>str, int, iterable (list, tuple, set, etc) of str/int, or ‘all’.</p>
</dd>
<dt class="field-even">param attributes</dt>
<dd class="field-even"><p>An iterable of attribute names or attribute numbers         (according to their order in the Attribute Reference Table).         If ‘all’, all of the attributes in the Attribute Reference Table will be used.         If None, a manual input prompt will be raised.</p>
</dd>
<dt class="field-odd">type fdr</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param fdr</dt>
<dd class="field-even"><p>Indicates the FDR threshold for significance.</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.         10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">type attr_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param attr_ref_path</dt>
<dd class="field-even"><p>the path of the Attribute Reference Table from which user-defined attributes will be drawn.</p>
</dd>
<dt class="field-odd">type biotype_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param biotype_ref_path</dt>
<dd class="field-even"><p>the path of the Biotype Reference Table.         Will be used to generate background set if ‘biotype’ is specified.</p>
</dd>
<dt class="field-odd">type biotype</dt>
<dd class="field-odd"><p>str specifying a specific biotype, list/set of strings each specifying a biotype, or ‘all’.         Default ‘protein_coding’.</p>
</dd>
<dt class="field-even">param biotype</dt>
<dd class="field-even"><p>determines the background genes by their biotype. Requires specifying a Biotype Reference Table.         ‘all’ will include all genomic features in the reference table,         ‘protein_coding’ will include only protein-coding genes from the reference table, etc.         Cannot be specified together with ‘background_genes’.</p>
</dd>
<dt class="field-odd">type background_genes</dt>
<dd class="field-odd"><p>set of feature indices, filtering.Filter object, or enrichment.FeatureSet object</p>
</dd>
<dt class="field-even">param background_genes</dt>
<dd class="field-even"><p>a set of specific feature indices to be used as background genes.         Cannot be specified together with ‘biotype’.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:         r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list">
<dt class="field-odd">type return_fig</dt>
<dd class="field-odd"><p>bool (default False)</p>
</dd>
<dt class="field-even">param return_fig</dt>
<dd class="field-even"><p>if True, returns a matplotlib Figure object in addition to the results DataFrame.
:rtype: pd.DataFrame (default) or Tuple[pd.DataFrame, matplotlib.figure.Figure]
:return: a pandas DataFrame with the indicated attribute names as rows/index, and the columns ‘log2_fold_enrichment’
and ‘pvalue’; and a matplotlib Figure, if ‘return_figure’ is set to True.</p>
<div class="figure align-center" id="id11">
<a class="reference internal image-reference" href="_images/enrichment_randomization.png"><img alt="_images/enrichment_randomization.png" src="_images/enrichment_randomization.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of enrich_randomization()</span><a class="headerlink" href="#id11" title="Permalink to this image">¶</a></p>
</div>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">enrich_randomization_parallel</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">attributes</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>Iterable<span class="p">[</span>str<span class="p">]</span><span class="p">, </span>str<span class="p">, </span>Iterable<span class="p">[</span>int<span class="p">]</span><span class="p">, </span>int<span class="p">]</span></span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">fdr</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">reps</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">10000</span></em>, <em class="sig-param"><span class="n">biotype</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'protein_coding'</span></em>, <em class="sig-param"><span class="n">background_genes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">attr_ref_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">biotype_ref_path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">save_csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">fname</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">return_fig</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><blockquote>
<div><p>Calculates enrichment scores, p-values and adjusted p-values         for enrichment and depletion of selected attributes from an Attribute Reference Table using parallel processing.         Background set is determined by either the input variable ‘background_genes’,         or by the input variable ‘biotype’ and a Biotype Reference Table.         Parallel processing makes this function generally faster than FeatureSet.enrich_randomization.         To use it you must first start a parallel session, using rnalysis.general.start_parallel_session().         P-values are calculated using a randomization test with the formula p = (successes + 1)/(repeats + 1).         P-values are corrected for multiple comparisons using         the Benjamini–Hochberg step-up procedure (original FDR method).         Enrichment/depletion is determined automatically by the calculated enrichment score:         if log2(enrichment score) is positive then enrichment is assumed,         and if log2(enrichment score) is negative then depletion is assumed.         In plots, for the clarity of display, complete depletion (linear enrichment = 0)         appears with the smallest value in the scale.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">type attributes</dt>
<dd class="field-odd"><p>str, int, iterable (list, tuple, set, etc) of str/int, or ‘all’</p>
</dd>
<dt class="field-even">param attributes</dt>
<dd class="field-even"><p>An iterable of attribute names or attribute numbers        (according to their order in the Attribute Reference Table).        If ‘all’, all of the attributes in the Attribute Reference Table will be used.        If None, a manual input prompt will be raised.</p>
</dd>
<dt class="field-odd">type fdr</dt>
<dd class="field-odd"><p>float between 0 and 1</p>
</dd>
<dt class="field-even">param fdr</dt>
<dd class="field-even"><p>Indicates the FDR threshold for significance.</p>
</dd>
<dt class="field-odd">type reps</dt>
<dd class="field-odd"><p>int larger than 0</p>
</dd>
<dt class="field-even">param reps</dt>
<dd class="field-even"><p>How many repetitions to run the randomization for.        10,000 is the default. Recommended 10,000 or higher.</p>
</dd>
<dt class="field-odd">type attr_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param attr_ref_path</dt>
<dd class="field-even"><p>the path of the Attribute Reference Table from which user-defined attributes will be drawn.</p>
</dd>
<dt class="field-odd">type biotype_ref_path</dt>
<dd class="field-odd"><p>str or pathlib.Path (default ‘predefined’)</p>
</dd>
<dt class="field-even">param biotype_ref_path</dt>
<dd class="field-even"><p>the path of the Biotype Reference Table.        Will be used to generate background set if ‘biotype’ is specified.</p>
</dd>
<dt class="field-odd">type biotype</dt>
<dd class="field-odd"><p>str specifying a specific biotype, or ‘all’. Default ‘protein_coding’.</p>
</dd>
<dt class="field-even">param biotype</dt>
<dd class="field-even"><p>determines the background genes by their biotype. Requires specifying a Biotype Reference Table.         ‘all’ will include all genomic features in the reference table,        ‘protein_coding’ will include only protein-coding genes from the reference table, etc.        Cannot be specified together with ‘background_genes’.</p>
</dd>
<dt class="field-odd">type background_genes</dt>
<dd class="field-odd"><p>set of feature indices, filtering.Filter object, or enrichment.FeatureSet object</p>
</dd>
<dt class="field-even">param background_genes</dt>
<dd class="field-even"><p>a set of specific feature indices to be used as background genes.        Cannot be specified together with ‘biotype’.</p>
</dd>
<dt class="field-odd">type save_csv</dt>
<dd class="field-odd"><p>bool, default False</p>
</dd>
<dt class="field-even">param save_csv</dt>
<dd class="field-even"><p>If True, will save the results to a .csv file, under the name specified in ‘fname’.</p>
</dd>
<dt class="field-odd">type fname</dt>
<dd class="field-odd"><p>str or pathlib.Path</p>
</dd>
<dt class="field-even">param fname</dt>
<dd class="field-even"><p>The full path and name of the file to which to save the results. For example:        r’C:dir</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>ile’. No ‘.csv’ suffix is required. If None (default), fname will be requested in a manual prompt.</dt><dd><dl class="field-list simple">
<dt class="field-odd">type return_fig</dt>
<dd class="field-odd"><p>bool (default False)</p>
</dd>
<dt class="field-even">param return_fig</dt>
<dd class="field-even"><p>if True, returns a matplotlib Figure object in addition to the results DataFrame.</p>
</dd>
<dt class="field-odd">rtype</dt>
<dd class="field-odd"><p>pd.DataFrame (default) or Tuple[pd.DataFrame, matplotlib.figure.Figure]</p>
</dd>
<dt class="field-even">return</dt>
<dd class="field-even"><p></p></dd>
</dl>
<p>a pandas DataFrame with the indicated attribute names as rows/index, and the columns ‘log2_fold_enrichment’
and ‘pvalue’; and a matplotlib Figure, if ‘return_figure’ is set to True.</p>
<div class="figure align-center" id="id12">
<a class="reference internal image-reference" href="_images/enrichment_randomization.png"><img alt="_images/enrichment_randomization.png" src="_images/enrichment_randomization.png" style="width: 768.0px; height: 370.8px;" /></a>
<p class="caption"><span class="caption-text">Example plot of enrich_randomization_parallel()</span><a class="headerlink" href="#id12" title="Permalink to this image">¶</a></p>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">gene_set</code></dt>
<dd><p>set of feature names/indices</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">go_enrichment</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mode</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">save_csv</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">fname</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Analyzes GO, Tissue and/or Phenotype enrichment for the given group of genomic features.         Uses the the Anatomy, Phenotype and Gene Ontology annotations for C. elegans.         Corrected p-values are calculated using hypergeometric statistics.         For more details see GitHub page of the developers: <a class="reference external" href="https://github.com/dangeles/TissueEnrichmentAnalysis">https://github.com/dangeles/TissueEnrichmentAnalysis</a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>mode</strong> (<em>'go'</em><em>, </em><em>'tissue'</em><em>, </em><em>'phenotype'</em><em> or </em><em>'all'</em><em> (</em><em>default 'all'</em><em>)</em>) – the enrichment you wish to perform. ‘go’ for gene ontology enrichment,         ‘tissue’ for tissue enrichment, ‘phenotype’ for phenotype enrichment, or ‘all’ for all three.</p></li>
<li><p><strong>alpha</strong> (<em>float between 0 and 1</em><em> (</em><em>default 0.05</em><em>)</em>) – Significance threshold.</p></li>
<li><p><strong>save_csv</strong> (<em>bool</em><em> (</em><em>default False</em><em>)</em>) – If True, save the result to a csv.</p></li>
<li><p><strong>fname</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – Name and path in which to save the results. Must be specified if save_csv is True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a DataFrame which contains the significant enrichmenet terms</p>
</dd>
</dl>
<div class="figure align-center" id="id13">
<a class="reference internal image-reference" href="_images/go_en.png"><img alt="_images/go_en.png" src="_images/go_en.png" style="width: 560.0px; height: 318.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Example plot of GO enrichment</span><a class="headerlink" href="#id13" title="Permalink to this image">¶</a></p>
</div>
<div class="figure align-center" id="id14">
<a class="reference internal image-reference" href="_images/tissue_en.png"><img alt="_images/tissue_en.png" src="_images/tissue_en.png" style="width: 560.0px; height: 318.40000000000003px;" /></a>
<p class="caption"><span class="caption-text">Example plot of Tissue enrichment</span><a class="headerlink" href="#id14" title="Permalink to this image">¶</a></p>
</div>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">intersection</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Calculates the set intersection of the indices from multiple FeatureSet objects         (the indices that exist in ALL of the FeatureSet objects).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – The objects against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000006&#39;</span><span class="p">},</span> <span class="s1">&#39;set name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en2</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000004&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">en2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">en</span><span class="p">)</span>
<span class="go">FeatureSet: set name</span>
<span class="go">{&#39;WBGene00000001&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">save_txt</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">fname</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>pathlib.Path<span class="p">]</span></span></em><span class="sig-paren">)</span></dt>
<dd><p>Save the list of features in the FeatureSet object under the specified filename and path.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fname</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – full filename/path for the output file. Can include the ‘.txt’ suffix but doesn’t have to.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt>
<code class="sig-name descname">set_name</code></dt>
<dd><p>name of the FeatureSet</p>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">symmetric_difference</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">other</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>Calculates the set symmetric difference of the indices from two FeatureSet objects         (the indices that appear in EXACTLY ONE of the FeatureSet objects, and not both/neither).         A-symmetric difference-B is equivalent to (A-difference-B)-union-(B-difference-A).</p>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>other</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – A second  object against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000006&#39;</span><span class="p">},</span> <span class="s1">&#39;set name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en2</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000004&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">symmetric_difference</span><span class="p">(</span><span class="n">en2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">en</span><span class="p">)</span>
<span class="go">FeatureSet: set name</span>
<span class="go">{&#39;WBGene00000002&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00000004&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt>
<code class="sig-name descname">union</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">others</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><blockquote>
<div><p>Calculates the set union of the indices from multipple FeatureSet objects         (the indices that exist in at least one of the FeatureSet objects).</p>
</div></blockquote>
<dl class="field-list">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>others</strong> (<a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet"><em>FeatureSet</em></a><em>, </em><em>set</em><em> or </em><em>str</em>) – The objects against which the current object will be compared.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em>) – If True (default), modifies the current instance of FeatureSet.         If False, returns a new instance of FeatureSet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>if inplace is False, returns a new instance of FeatureSet.</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000004&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000005&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000006&#39;</span><span class="p">},</span> <span class="s1">&#39;set name&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en2</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00000004&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">})</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">s</span><span class="p">,</span> <span class="n">en2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">en</span><span class="p">)</span>
<span class="go">FeatureSet: set name</span>
<span class="go">{&#39;WBGene00000003&#39;, &#39;WBGene00000004&#39;, &#39;WBGene00000001&#39;, &#39;WBGene00000002&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00000005&#39;}</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.enrichment.</code><code class="sig-name descname">upset_plot</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">objs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>str<span class="p">, </span><a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet">rnalysis.enrichment.FeatureSet</a><span class="p">, </span>Set<span class="p">[</span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">''</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate an UpSet plot of 2 or more sets, FeatureSets or attributes from the Attribute Reference Table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>objs</strong> (<em>a dictionary with 2</em><em> or </em><em>more entries</em><em>, </em><em>where the keys are the names of the sets</em><em>, </em><em>and the values are either     a FeatureSet</em><em>, </em><em>a python set of feature indices</em><em>, or </em><em>a name of a column in the Attribute Reference Table.     For example:     {'first set':{'gene1'</em><em>,</em><em>'gene2'</em><em>,</em><em>'gene3'}</em><em>, </em><em>'second set':'name_of_attribute_from_reference_table'}</em>) – the FeatureSets, python sets or user-defined attributes to plot.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – determines the title of the plot.</p></li>
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – the path of the Attribute Reference Table from which user-defined attributes will be drawn,     if such attributes are included in ‘objs’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>a dictionary of matplotlib axes, where the keys are ‘matrix’, ‘intersections’, ‘totals’, ‘shading’.</p>
<div class="figure align-center" id="id15">
<a class="reference internal image-reference" href="_images/upsetplot.png"><img alt="_images/upsetplot.png" src="_images/upsetplot.png" style="width: 552.3px; height: 340.2px;" /></a>
<p class="caption"><span class="caption-text">Example plot of upset_plot()</span><a class="headerlink" href="#id15" title="Permalink to this image">¶</a></p>
</div>
</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.enrichment.</code><code class="sig-name descname">venn_diagram</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">objs</span><span class="p">:</span> <span class="n">Dict<span class="p">[</span>str<span class="p">, </span>Union<span class="p">[</span>str<span class="p">, </span><a class="reference internal" href="rnalysis.enrichment.FeatureSet.html#rnalysis.enrichment.FeatureSet" title="rnalysis.enrichment.FeatureSet">rnalysis.enrichment.FeatureSet</a><span class="p">, </span>Set<span class="p">[</span>str<span class="p">]</span><span class="p">]</span><span class="p">]</span></span></em>, <em class="sig-param"><span class="n">title</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'default'</span></em>, <em class="sig-param"><span class="n">ref</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'predefined'</span></em>, <em class="sig-param"><span class="n">set_colors</span><span class="p">:</span> <span class="n">tuple</span> <span class="o">=</span> <span class="default_value">'r', 'g', 'b'</span></em>, <em class="sig-param"><span class="n">alpha</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">0.4</span></em>, <em class="sig-param"><span class="n">weighted</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">lines</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em>, <em class="sig-param"><span class="n">linecolor</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">'black'</span></em>, <em class="sig-param"><span class="n">linestyle</span><span class="o">=</span><span class="default_value">'solid'</span></em>, <em class="sig-param"><span class="n">linewidth</span><span class="o">=</span><span class="default_value">2.0</span></em>, <em class="sig-param"><span class="n">normalize_to</span><span class="p">:</span> <span class="n">float</span> <span class="o">=</span> <span class="default_value">1.0</span></em><span class="sig-paren">)</span></dt>
<dd><p>Generate a Venn diagram of 2 to 3 sets, FeatureSets or attributes from the Attribute Reference Table.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>objs</strong> (<em>a dictionary with 2-3 entries</em><em>, </em><em>where the keys are the names of the sets</em><em>, </em><em>and the values are either     a FeatureSet</em><em>, </em><em>a python set of feature indices</em><em>, or </em><em>a name of a column in the Attribute Reference Table.     For example:     {'first set':{'gene1'</em><em>,</em><em>'gene2'</em><em>,</em><em>'gene3'}</em><em>, </em><em>'second set':'name_of_attribute_from_reference_table'}</em>) – the FeatureSets, python sets or user-defined attributes to plot.</p></li>
<li><p><strong>set_colors</strong> (<em>tuple of matplotlib-format colors</em><em>, </em><em>the same size as 'objs'</em>) – determines the colors of the circles in the diagram.</p></li>
<li><p><strong>ref</strong> (<em>str</em><em> or </em><em>pathlib.Path</em><em> (</em><em>default 'predefined'</em><em>)</em>) – the path of the Attribute Reference Table from which user-defined attributes will be drawn,     if such attributes are included in ‘objs’.</p></li>
<li><p><strong>title</strong> (<em>str</em>) – determines the title of the plot.</p></li>
<li><p><strong>alpha</strong> (<em>a float between 0 and 1</em>) – determines the opacity of the circles.</p></li>
<li><p><strong>weighted</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – if True, the plot will be area-weighted.</p></li>
<li><p><strong>lines</strong> (<em>bool</em><em> (</em><em>default True</em><em>)</em>) – if True, adds an outline to the circles.</p></li>
<li><p><strong>linecolor</strong> (<em>matplotlib-format color</em><em> (</em><em>default 'black'</em><em>)</em>) – Determines the color of the circles’ outline.</p></li>
<li><p><strong>linestyle</strong> (<em>'solid'</em><em> or </em><em>'dashed'</em><em> (</em><em>default 'solid'</em><em>)</em>) – the style of the circles’ outline.</p></li>
<li><p><strong>linewidth</strong> (<em>float</em><em> (</em><em>default 2.0</em><em>)</em>) – the widdth of the circles’ outlines.</p></li>
<li><p><strong>normalize_to</strong> (<em>float</em><em> (</em><em>default 1.0</em><em>)</em>) – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>a tuple of a VennDiagram object; and a list of 2-3 Circle patches.</p>
<div class="figure align-center" id="id16">
<a class="reference internal image-reference" href="_images/venn.png"><img alt="_images/venn.png" src="_images/venn.png" style="width: 550.1999999999999px; height: 416.5px;" /></a>
<p class="caption"><span class="caption-text">Example plot of venn_diagram()</span><a class="headerlink" href="#id16" title="Permalink to this image">¶</a></p>
</div>
</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="rnalysis-general-module">
<h2>rnalysis.general module<a class="headerlink" href="#rnalysis-general-module" title="Permalink to this headline">¶</a></h2>
<p>This module contains general-purpose functions. Those include loading and saving files, reading and updating the settings file, identifying type of input variables, etc. This module is used mainly by other modules.</p>
<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">load_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">idx_col</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">drop_columns</span><span class="p">:</span> <span class="n">Union<span class="p">[</span>str<span class="p">, </span>List<span class="p">[</span>str<span class="p">]</span><span class="p">]</span></span> <span class="o">=</span> <span class="default_value">False</span></em>, <em class="sig-param"><span class="n">squeeze</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">comment</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>loads a csv df into a pandas dataframe.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>filename</strong> (<em>str</em><em> or </em><em>pathlib.Path</em>) – name of the csv file to be loaded</p></li>
<li><p><strong>idx_col</strong> (<em>int</em><em>, </em><em>default None</em>) – number of column to be used as index. default is None, meaning no column will be used as index.</p></li>
<li><p><strong>drop_columns</strong> (<em>str</em><em>, </em><em>list of str</em><em>, or </em><em>False</em><em> (</em><em>default False</em><em>)</em>) – if a string or list of strings are specified,     the columns of the same name/s will be dropped from the loaded DataFrame.</p></li>
<li><p><strong>squeeze</strong> (<em>bool</em><em>, </em><em>default False</em>) – If the parsed data only contains one column then return a Series.</p></li>
<li><p><strong>comment</strong> (<em>str</em><em> (</em><em>optional</em><em>)</em>) – Indicates remainder of line should not be parsed.     If found at the beginning of a line, the line will be ignored altogether. This parameter must be a single character.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>a pandas dataframe of the csv file</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_gene_name_string</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">string</span></em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains gene names (like ‘daf-2’ or ‘lin15B’).     Parses the string into a set of gene names.     The format of a gene name is a sequence consisting of the expression     ‘[a-z]{3,4}’, the character ‘-‘, and the expression ‘[A-Z,0-9]{1,4}’.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
<dl class="field-list simple">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">string</span> <span class="o">=</span> <span class="s1">&#39;saeg-2 lin-15B cyp-23A1lin-15A WBGene12345678</span><span class="si">%G</span><span class="s1">HF5H.3&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parsed</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">parse_gene_name_string</span><span class="p">(</span><span class="n">string</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span>
<span class="go">{&#39;saeg-2&#39;, &#39;lin-15B&#39;, &#39;cyp-23A1&#39;, &#39;lin-15A&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_sequence_name_string</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">string</span></em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains sequence names (such as ‘Y55D5A.5’).     Parses the string into a set of WBGene indices.     The format of a sequence name is a sequence consisting of the expression ‘[A-Z,0-9]{5,6}’,     the character ‘.’, and a digit.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
<dl class="field-list simple">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">string</span> <span class="o">=</span> <span class="s1">&#39;CELE_Y55D5A.5T23G5.6WBGene00000000 daf-16^^ZK662.4 &#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parsed</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">parse_sequence_name_string</span><span class="p">(</span><span class="n">string</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span>
<span class="go">{&#39;Y55D5A.5&#39;, &#39;T23G5.6&#39;, &#39;ZK662.4&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">parse_wbgene_string</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">string</span></em><span class="sig-paren">)</span></dt>
<dd><p>Receives a string that contains WBGene indices. Parses the string into a set of WBGene indices.     The format of a WBGene index is ‘WBGene’ and exactly 8 digits.
:type string: str
:param string: The string to be parsed. Can be any format of string.
:return:
a set of the WBGene indices that appear in the given string.</p>
<dl class="field-list simple">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">string</span> <span class="o">=</span>  <span class="s1">&#39;&#39;&#39;WBGene WBGenes WBGene12345678, WBGene98765432WBGene00000000&amp; the geneWBGene44444444daf-16A5gHB.5</span>
<span class="gp">... </span><span class="s1">WBGene55555555&#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">parsed</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">parse_wbgene_string</span><span class="p">(</span><span class="n">string</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">parsed</span><span class="p">)</span>
<span class="go">{&#39;WBGene12345678&#39;, &#39;WBGene44444444&#39;, &#39;WBGene98765432&#39;, &#39;WBGene55555555&#39;, &#39;WBGene00000000&#39;}</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">read_attr_ref_table_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Reads the Attribute Reference Table path from the settings file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the path of the Attribute Reference Table that is saved in the settings file.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_path</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">read_attr_ref_table_path</span><span class="p">()</span>
<span class="go">Attribute Reference Table used: my_attribute_reference_table_path</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">read_biotype_ref_table_path</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Reads the Biotype Reference Table path from the settings file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>the path of the Biotype Reference Table that is saved in the settings file.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_path</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">read_biotype_ref_table_path</span><span class="p">()</span>
<span class="go">Biotype Reference Table used: my_biotype_reference_table_path</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">reset_settings_file</code><span class="sig-paren">(</span><span class="sig-paren">)</span></dt>
<dd><p>Resets the local settings by deleting the local settings file. Warning: this action is irreversible!</p>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">save_to_csv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span><span class="p">:</span> <span class="n">pandas.core.frame.DataFrame</span></em>, <em class="sig-param"><span class="n">filename</span><span class="p">:</span> <span class="n">str</span></em>, <em class="sig-param"><span class="n">suffix</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em>, <em class="sig-param"><span class="n">index</span><span class="p">:</span> <span class="n">bool</span> <span class="o">=</span> <span class="default_value">True</span></em><span class="sig-paren">)</span></dt>
<dd><p>save a pandas DataFrame to csv.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> – pandas DataFrame to be saved</p></li>
<li><p><strong>filename</strong> – a string or pathlib.Path object stating the original name of the file</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>default None</em>) – A suffix to be added to the original name of the file. If None, no suffix will be added.</p></li>
<li><p><strong>index</strong> – if True, saves the DataFrame with the indices. If false, ignores the index.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">set_attr_ref_table_path</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Defines/updates the Attribute Reference Table path in the settings file.
:param path: the path you wish to set as the Attribute Reference Table path
:type path: str</p>
<dl class="field-list simple">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;my_attribute_reference_table_path&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">general</span><span class="o">.</span><span class="n">set_attr_ref_table_path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="go">Attribute Reference Table path set as: my_attribute_reference_table_path</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">set_biotype_ref_table_path</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">path</span><span class="p">:</span> <span class="n">str</span> <span class="o">=</span> <span class="default_value">None</span></em><span class="sig-paren">)</span></dt>
<dd><p>Defines/updates the Biotype Reference Table path in the settings file.
:param path: the path you wish to set as the Biotype Reference Table path
:type path: str</p>
<dl class="field-list simple">
<dt class="field-odd">Examples</dt>
<dd class="field-odd"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;my_biotype_reference_table_path&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">general</span><span class="o">.</span><span class="n">set_biotype_ref_table_path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="go">Biotype Reference Table path set as: my_biotype_reference_table_path</span>
</pre></div>
</div>
</dd></dl>

<dl class="py function">
<dt>
<code class="sig-prename descclassname">rnalysis.general.</code><code class="sig-name descname">start_parallel_session</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_engines</span><span class="p">:</span> <span class="n">int</span> <span class="o">=</span> <span class="default_value">'default'</span></em><span class="sig-paren">)</span></dt>
<dd><p>Stop previous ipyparallel ipcluster and start a new one in order to perform parallelized computation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_engines</strong> (<em>int</em><em> or </em><em>'default'</em>) – if ‘default’, will initiate the default amount of engines.     Otherwise, will initiate n_engines engines.</p>
</dd>
<dt class="field-even">Examples</dt>
<dd class="field-even"><p></p></dd>
</dl>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="kn">import</span> <span class="n">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">general</span><span class="o">.</span><span class="n">start_parallel_session</span><span class="p">()</span>
<span class="go">Starting parallel session...</span>
<span class="go">Parallel session started successfully</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Guy Teichman

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>