

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>RNAlysis user guide &mdash; RNAlysis 1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributing" href="contributing.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> RNAlysis
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">RNAlysis - <em>C. elegans</em> RNA sequencing data analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">RNAlysis user guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-filtering-module">RNAlysis filtering module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#working-with-filter-objects">Working with Filter objects</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#initialize-a-filter-object">Initialize a Filter object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#view-a-filter-object">View a Filter object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-operations">Filtering operations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-set-operations-on-multiple-filter-objects">Performing set operations on multiple Filter objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#saving-filter-results">Saving Filter results</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-a-reference-table-for-filter-operations">Using a reference table for filter operations</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-deseq2-output-files-with-filtering-deseqfilter">Filtering DESeq2 output files with filtering.DESeqFilter</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#loading-from-a-csv-file">Loading from a .csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-operations-unique-to-deseqfilter">Filtering operations unique to DESeqFilter</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-htseq-count-output-files-with-filtering-htcountfilter">Filtering HTSeq-count output files with filtering.HTCountFilter</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#generating-an-htcountfilter-object-from-a-folder-of-htseq-count-output-txt-files">Generating an HTCountFilter object from a folder of HTSeq-count output .txt files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#loading-from-a-pre-made-csv-file">Loading from a pre-made .csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-operations-unique-to-htcountfilter">Filtering operations unique to HTCountFilter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#normalizing-reads-with-htcountfilter">Normalizing reads with HTCountFilter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-visualization-and-clustering-analysis-with-htcountfilter">Data visualization and clustering analysis with HTCountFilter</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-fold-change-data-of-features-using-filtering-foldchangefilter">Filtering fold-change data of features using filtering.FoldChangeFilter</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#loading-fold-change-data-from-a-csv-file">Loading fold change data from a .csv file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generating-fold-change-data-from-an-existing-htcountfilter-object">Generating fold change data from an existing HTCountFilter object</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-enrichment-module">RNAlysis enrichment module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#working-with-enrichmentprocessing-objects">Working with EnrichmentProcessing objects</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#initialize-an-enrichmentprocessing-object">Initialize an EnrichmentProcessing object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#randomization-test-enrichment-analysis-for-user-defined-attributes">Randomization test enrichment analysis for user-defined attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-set-operations-on-multiple-enrichmentprocessing-objects">Performing set operations on multiple EnrichmentProcessing objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#saving-indices-from-enrichmentprocessing-to-a-txt-file">Saving indices from EnrichmentProcessing to a .txt file</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.feature_name_to_wbgene.html">rnalysis.feature_name_to_wbgene</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.html">rnalysis package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>RNAlysis user guide</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/rnalysis_user_guide.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="rnalysis-user-guide">
<h1>RNAlysis user guide<a class="headerlink" href="#rnalysis-user-guide" title="Permalink to this headline">¶</a></h1>
<div class="section" id="rnalysis-filtering-module">
<h2>RNAlysis filtering module<a class="headerlink" href="#rnalysis-filtering-module" title="Permalink to this headline">¶</a></h2>
<p>RNAlysis’s filtering module (rnalysis.filtering) is built to allow rapid and easy to understand filtering of various forms of RNA sequencing data. The module also contains specific methods for visualization and clustering of data.</p>
<p>The filtering module is built around Filter objects, which are containers for tabular sequencing data. You can use the different types of Filter objects to apply filtering operations to various types of tabular data. You will learn more about Filter objects in the next section.</p>
<div class="section" id="working-with-filter-objects">
<h3>Working with Filter objects<a class="headerlink" href="#working-with-filter-objects" title="Permalink to this headline">¶</a></h3>
<p>All Filter objects (DESeqFilter, HTCountFilter, FoldChangeFilter) work on the same principles,
and share many of the same functions and features. Each of them also has specific filtering, analysis and visualisation functions. In this section we will look into the general usage of Filter objects.</p>
<div class="section" id="initialize-a-filter-object">
<h4>Initialize a Filter object<a class="headerlink" href="#initialize-a-filter-object" title="Permalink to this headline">¶</a></h4>
<p>We will start by importing the filtering module:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="k">import</span> <span class="n">filtering</span>
</pre></div>
</div>
<p>We can now, for example, create a DESeqFilter object from a DESeq2 csv output file (see more details about DESeqFilter in sections below).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;D:\myfolder\my_deseq2_output.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="view-a-filter-object">
<h4>View a Filter object<a class="headerlink" href="#view-a-filter-object" title="Permalink to this headline">¶</a></h4>
<p>In order to view a glimpse of the file we imported we can use the ‘head’ and ‘tail’ functions.
By default ‘head’ will show the first 5 rows of the file, and ‘tail’ will show the last 5 rows,
but you can specify a specific number of lines to show.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
</pre></div>
</div>
<p>will give the following output:
#insert output here</p>
<p>We can also see the total number of rows an columns by accessing the ‘shape’ attribute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">shape</span>
</pre></div>
</div>
<p>which will return the following output:
#insert output here
meaning there are {} rows and {} columns in the file.</p>
</div>
<div class="section" id="filtering-operations">
<h4>Filtering operations<a class="headerlink" href="#filtering-operations" title="Permalink to this headline">¶</a></h4>
<p>Now we can start filtering the entries in the file according to parameters of our choosing.
Various filtering operations are applied directly to the Filter object. Those operations do not affect the original csv file, but its representation within the Filter object.
For example, we can the function ‘filter_percentile’ to remove all rows that are above the specified percentile (in our example, 75% percentile) in the specified column (in our example, ‘log2FoldChange’):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>If we now look at the shape of d, we will see that {} rows have been filtered out of the object, and we remain with {} rows.</p>
<p>By default, filtering operations on Filter objects are performed in-place, meaning the original object is modified. However, we can save the results into a new Filter object and leave the current object unaffected by passing the argument ‘inplace=False’ to any filtering function within RNAlysis. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d_filtered</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
<p>In this case, the object ‘d’ will remain unchanged, while ‘d_filtered’ will be a new Filter object which contains our filtered results. We can continue applying filters sequentially to the same Filter object, or using ‘inplace=False’ to create a new object at any point.</p>
<p>Another useful option is to perform an opposite filter. When we specify the parameter ‘opposite=True’ to any filtering function within RNAlysis, the filtering function will be performed in opposite. This means that all of the genomic features that were supposed to be filtered out are kept in the object, and the genomic features that were supposed to be kept in the object are filtered out.
For example, if we now wanted to remove the rows which are below the 25% percentile in the ‘log2FoldChange’ column, we will use the following code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Calling this function without the ‘opposite’ parameter would have removed all values except the bottom 25% of the ‘log2FoldChange’ column. When specifying ‘opposite’, we instead throw out the bottom 25% of the ‘log2FoldChange’ column and keep the rest.</p>
<p>There are many different filtering functions within the filtering module. Some of them are subtype-specific (such as ‘filter_low_reads’ for HTCountFilter objects and ‘filter_significant’ for DESeqFilter objects), while others can be applied to any Filter object. You can read more about the different functions and their usage in the project’s documentation.</p>
</div>
<div class="section" id="performing-set-operations-on-multiple-filter-objects">
<h4>Performing set operations on multiple Filter objects<a class="headerlink" href="#performing-set-operations-on-multiple-filter-objects" title="Permalink to this headline">¶</a></h4>
<p>In addition to using regular filters, it is also possible to use set operations such as union, intersection, difference and symmetric difference to combine the results of multiple Filter objects. Those set operations can be applied to any Filter object, as well as to python sets. The objects don’t have to be of the same subtype - you can, for example, look at the union of a DESeqFilter object, an HTCountFilter object and a python set:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;deseqfile.csv&#39;</span><span class="p">)</span>
<span class="n">h</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">HTCountFilter</span><span class="p">(</span><span class="s1">&#39;htseq_count_file.csv&#39;</span><span class="p">)</span>
<span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="n">union_result</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>When performing set operations, the return type will always be either a python set (default) or a string. This means you can use the output of the set operation as an input for yet another set operation. However, since the returned object is a set you cannot use Filter object functions such as ‘head’ and ‘save_csv’ on it, or apply filters to it directly.</p>
</div>
<div class="section" id="saving-filter-results">
<h4>Saving Filter results<a class="headerlink" href="#saving-filter-results" title="Permalink to this headline">¶</a></h4>
<p>At any point we can save the current result of our filtering to a new csv file, by using the ‘save_csv’ function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">save_csv</span><span class="p">()</span>
</pre></div>
</div>
<p>If no filename is specified, the file is given a name automatically based on the filtering operations performed on it, their order and their parameters.
We can view the current automatic filename by looking at the ‘fname’ attribute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">fname</span>
</pre></div>
</div>
<p>In this example, the automatic filename is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="s1">&#39;D:/myfolder/my_deseq2_output_below0.75percentile_below0.25percentileopposite.csv&#39;</span>
</pre></div>
</div>
<p>Alternatively, you can specify a filename:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span><span class="o">.</span><span class="n">save_csv</span><span class="p">(</span><span class="s1">&#39;alt_filename&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Instead of directly saving the results to a file, you can also get them as a set or string of genomic feature indices:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">set_output</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">features_set</span><span class="p">()</span>
<span class="n">str_output</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">features_string</span><span class="p">()</span>
</pre></div>
</div>
<p>Sets of genomic feature indices can be used later for enrichment analysis using the enrichment module (see below).</p>
</div>
<div class="section" id="using-a-reference-table-for-filter-operations">
<h4>Using a reference table for filter operations<a class="headerlink" href="#using-a-reference-table-for-filter-operations" title="Permalink to this headline">¶</a></h4>
<p>#set reference path
#read reference path
#what does reference table look like
#what do reference table attributes look like
#biotype table</p>
</div>
</div>
<div class="section" id="filtering-deseq2-output-files-with-filtering-deseqfilter">
<h3>Filtering DESeq2 output files with filtering.DESeqFilter<a class="headerlink" href="#filtering-deseq2-output-files-with-filtering-deseqfilter" title="Permalink to this headline">¶</a></h3>
<p>DESeqFilter objects are built to easily filter the output of R’s DESeq2 package. This package is meant to analyze differential expression of genomic features in sequencing data. You can read more about it here: {}
Like other filter objects, filtering operations on DESeqFilter are performed in-place by default,meaning the original object is modified.</p>
<p>In principle, any .csv file that contains differential expression analysis data with log2 fold change and adjusted p values can be used as input for DESeqFilter.
However, some DESeqFilter functions (such as ‘filter_significant’ and ‘filter_abs_log2_fold_change’) may only work on DESeq2 output files, and other unintended interactions may occur.</p>
<div class="section" id="loading-from-a-csv-file">
<h4>Loading from a .csv file<a class="headerlink" href="#loading-from-a-csv-file" title="Permalink to this headline">¶</a></h4>
<p>Loading a file into a DESeqFilter works as explained above for any Filter object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;my_file.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="filtering-operations-unique-to-deseqfilter">
<h4>Filtering operations unique to DESeqFilter<a class="headerlink" href="#filtering-operations-unique-to-deseqfilter" title="Permalink to this headline">¶</a></h4>
<p>There are a few filtering operations unique to DESeqFilter. Those include ‘filter_significant’, which removes statistically-insignificant rows according to a specified threshold; ‘filter_abs_log2_fold_change’, removes rows whose absolute value log2 fold change is below the specified threshold; ‘filter_fold_change_direction’ which removes either up-regulated (positive log2 fold change) or down-regulated (negative log2 fold change) rows; and ‘split_fold_change_direction’ which returns a DESeqFilter object with only up-regulated features and a DESeqFilter object with only down-regulated features.</p>
<p>The unique DESeqFilter filter operations expect specific column names (the column names automatically generated by DESeq2), and will not work with other column names:
‘log2FoldChange’,’pval’,’padj’.</p>
</div>
</div>
<div class="section" id="filtering-htseq-count-output-files-with-filtering-htcountfilter">
<h3>Filtering HTSeq-count output files with filtering.HTCountFilter<a class="headerlink" href="#filtering-htseq-count-output-files-with-filtering-htcountfilter" title="Permalink to this headline">¶</a></h3>
<p>You can read more about HTSeq-count here:
#link</p>
<p>In principle, any .csv file where the columns are different conditions/replicates and the rows include reads/normalized reads per genomic feature can be used as input for HTCountFilter. However, some HTCountFilter functions (such as ‘norm_reads_to_rpm’) will only work on HTSeq-count output files, and other unintended interactions may occur.</p>
<div class="section" id="generating-an-htcountfilter-object-from-a-folder-of-htseq-count-output-txt-files">
<h4>Generating an HTCountFilter object from a folder of HTSeq-count output .txt files<a class="headerlink" href="#generating-an-htcountfilter-object-from-a-folder-of-htseq-count-output-txt-files" title="Permalink to this headline">¶</a></h4>
<p>HTSeq-count receives as input an aligned SAM file. The native output of HTSeq-count is a text file with feature indices and read-per-genomic-feature, as well as information about reads that weren’t counted for any feature (alignment not unique, low alignment quality, ambiguous, unaligned, aligned to no feature). When running HTSeq-count on multiple SAM files (which could represent different conditions or replicates), the final output would be a directory of .txt files. RNAlysis can parse those .txt files into two .csv tables: in the first each row is a genomic feature and each column is a condition or replicate (a single .txt file), and in the second each row represents a category of reads not mapped to genomic features (alignment not unique, low alignment quality, etc). This is done with the ‘from_folder’ function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">h</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">HTCountFilter</span><span class="o">.</span><span class="n">from_folder</span><span class="p">(</span><span class="s1">&#39;my_folder_path&#39;</span><span class="p">,</span> <span class="n">save_reads_fname</span><span class="o">=</span><span class="s1">&#39;name_for_reads_csv_file&#39;</span><span class="p">,</span> <span class="n">save_not_counted_fname</span><span class="o">=</span><span class="s1">&#39;name_for_unmapped_reads_csv_file&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>By deault, ‘from_folder’ saves the generated tables as .csv files. However, you can avoid that by specifying ‘save_csv=False’.
It is also possible to automatically normalize the reads in the new HTCountFilter object to reads per million (RPM) using the unmapped reads data by specifying ‘norm_to_rpm=True’.</p>
</div>
<div class="section" id="loading-from-a-pre-made-csv-file">
<h4>Loading from a pre-made .csv file<a class="headerlink" href="#loading-from-a-pre-made-csv-file" title="Permalink to this headline">¶</a></h4>
<p>If you have previously generated a .csv file from HTSeq-count output files using RNAlysis, or have done so manually, you can directly load this .csv file into an HTCountFilter object as you would any other Filter object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">h</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">HTCountFilter</span><span class="p">(</span><span class="s1">&#39;my_csv_file.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="filtering-operations-unique-to-htcountfilter">
<h4>Filtering operations unique to HTCountFilter<a class="headerlink" href="#filtering-operations-unique-to-htcountfilter" title="Permalink to this headline">¶</a></h4>
<p>#filter low reads</p>
</div>
<div class="section" id="normalizing-reads-with-htcountfilter">
<h4>Normalizing reads with HTCountFilter<a class="headerlink" href="#normalizing-reads-with-htcountfilter" title="Permalink to this headline">¶</a></h4>
<p>HTCountFilter offers two methods for normalizing reads: reads per million (RPM) and DESeq2’s size factors. Data normalized in other methods (such as RPKM) can be used as input for HTCountFilter, but it cannot perform such normalization methods on its own.
#normalize to rpm
#normalize with size factors</p>
</div>
<div class="section" id="data-visualization-and-clustering-analysis-with-htcountfilter">
<h4>Data visualization and clustering analysis with HTCountFilter<a class="headerlink" href="#data-visualization-and-clustering-analysis-with-htcountfilter" title="Permalink to this headline">¶</a></h4>
<p>#pairplot
#clustermap
#pca</p>
</div>
</div>
<div class="section" id="filtering-fold-change-data-of-features-using-filtering-foldchangefilter">
<h3>Filtering fold-change data of features using filtering.FoldChangeFilter<a class="headerlink" href="#filtering-fold-change-data-of-features-using-filtering-foldchangefilter" title="Permalink to this headline">¶</a></h3>
<p>#fold change is calculated (1+reads)/(1+reads)
#0 and inf are not supported, and may lead to unintented results</p>
<div class="section" id="loading-fold-change-data-from-a-csv-file">
<h4>Loading fold change data from a .csv file<a class="headerlink" href="#loading-fold-change-data-from-a-csv-file" title="Permalink to this headline">¶</a></h4>
</div>
<div class="section" id="generating-fold-change-data-from-an-existing-htcountfilter-object">
<h4>Generating fold change data from an existing HTCountFilter object<a class="headerlink" href="#generating-fold-change-data-from-an-existing-htcountfilter-object" title="Permalink to this headline">¶</a></h4>
</div>
</div>
</div>
<div class="section" id="rnalysis-enrichment-module">
<h2>RNAlysis enrichment module<a class="headerlink" href="#rnalysis-enrichment-module" title="Permalink to this headline">¶</a></h2>
<p>RNAlysis’s enrichment module (rnalysis.enrichment) can be used to perform various enrichment analyses including gene ontology (GO) enrichment and enrichment for user-defined attributes. The module also includes basic set operations (union, intersection, difference, symmetric difference) between different sets of genomic features.</p>
<div class="section" id="working-with-enrichmentprocessing-objects">
<h3>Working with EnrichmentProcessing objects<a class="headerlink" href="#working-with-enrichmentprocessing-objects" title="Permalink to this headline">¶</a></h3>
<p>The enrichment module is built around EnrichmentProcessing objects, which are a container for a set of genomic features and their name (for example, ‘genes that are upregulated under hyperosmotic conditions’). All further anslyses of the set of features is done through the EnrichmentProcessing object.</p>
<div class="section" id="initialize-an-enrichmentprocessing-object">
<h4>Initialize an EnrichmentProcessing object<a class="headerlink" href="#initialize-an-enrichmentprocessing-object" title="Permalink to this headline">¶</a></h4>
<p>We will start by importing the enrichment module:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">rnalysis</span> <span class="k">import</span> <span class="n">enrichment</span>
</pre></div>
</div>
<p>An EnrichmentProcessing object can now be initialized by one of three methods.
The first method is to directly specify a python set of genomic feature indices:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">myset</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene0245200&#39;</span><span class="p">,</span><span class="s1">&#39; WBGene00402029&#39;</span><span class="p">}</span>
<span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">EnrichmentProcessing</span><span class="p">(</span><span class="n">myset</span><span class="p">,</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The second method is to extract a python set of genomic feature indices from an existing Filter object (see above for more information about Filter objects and the filtering module) using the function ‘features_set’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">h</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">HTCountFilter</span><span class="p">(</span><span class="s1">&#39;path_to_my_file.csv&#39;</span><span class="p">)</span>
<span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">EnrichmentProcessing</span><span class="p">(</span><span class="n">h</span><span class="o">.</span><span class="n">features_set</span><span class="p">(),</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The third method is not to specify a gene set at all:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">EnrichmentProcessing</span><span class="p">(</span><span class="n">set_name</span> <span class="o">=</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>At this point, you will be prompted to enter a string of feature indices seperated by newline. They will be automatically paresd into a python set.</p>
<p>EnrichmentProcessing objects have two attributes: gene_set, a python set containing genomic feature indices; and set_name, a string that describes the feature set (optional).</p>
</div>
<div class="section" id="randomization-test-enrichment-analysis-for-user-defined-attributes">
<h4>Randomization test enrichment analysis for user-defined attributes<a class="headerlink" href="#randomization-test-enrichment-analysis-for-user-defined-attributes" title="Permalink to this headline">¶</a></h4>
<p>#randomization
#parallel
#ref table
#background genes</p>
</div>
<div class="section" id="performing-set-operations-on-multiple-enrichmentprocessing-objects">
<h4>Performing set operations on multiple EnrichmentProcessing objects<a class="headerlink" href="#performing-set-operations-on-multiple-enrichmentprocessing-objects" title="Permalink to this headline">¶</a></h4>
<p>Similarly to Filter objects, it is possible to use set operations such as union, intersection, difference and symmetric difference to combine the feature sets of multiple EnrichmentProcessing objects. Those set operations can be applied to both EnrichmentProcessing objects and python sets. The objects don’t have to be of the same subtype - you can, for example, look at the union of an EnrichmentProcessing object and a python set:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">EnrichmentProcessing</span><span class="p">({</span><span class="s1">&#39;WBGene00003002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00004201&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00300139&#39;</span><span class="p">})</span>

<span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="n">union_result</span> <span class="o">=</span> <span class="n">en</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>When performing set operations, the return type will always be a python set. This means you can use the output of the set operation as an input for yet another set operation, or as input to a new EnrichmentProcessing object.</p>
</div>
<div class="section" id="saving-indices-from-enrichmentprocessing-to-a-txt-file">
<h4>Saving indices from EnrichmentProcessing to a .txt file<a class="headerlink" href="#saving-indices-from-enrichmentprocessing-to-a-txt-file" title="Permalink to this headline">¶</a></h4>
<p>It is possible to save the feature indices from an EnrichmentProcessing object to a .txt file, for use in online enrichment tools or simply to share the list of genomic features. This is done with the ‘save_txt’ function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">en</span><span class="o">.</span><span class="n">save_txt</span><span class="p">(</span><span class="s1">&#39;D:\path</span><span class="se">\f</span><span class="s1">ilename&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The feature indices will be saved to the text file in the specified path, separated by newline (‘n’).</p>
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributing" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, Guy Teichman

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>