<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>A-to-Z tutorial - example analysis &mdash; RNAlysis 3.9.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Frequently asked questions" href="faq.html" />
    <link rel="prev" title="RNAlysis quick-start video guide" href="quick_start.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">A-to-Z tutorial - example analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#intro-and-recommendations">Intro and recommendations</a></li>
<li class="toctree-l2"><a class="reference internal" href="#analysis-1-time-series-rna-sequencing">Analysis #1 - time series RNA sequencing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#the-dataset">The dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#start-rnalysis">Start <em>RNAlysis</em></a></li>
<li class="toctree-l3"><a class="reference internal" href="#data-preprocessing-and-exploratory-data-analysis">Data preprocessing and exploratory data analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#filter-out-lowly-expressed-genes">Filter out lowly-expressed genes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#examine-variance-in-our-data-with-principal-component-analysis">Examine variance in our data with Principal Component Analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#examine-similarity-between-developmental-stages">Examine similarity between developmental stages</a></li>
<li class="toctree-l4"><a class="reference internal" href="#compare-the-expression-of-specific-genes-over-the-developmental-stages">Compare the expression of specific genes over the developmental stages</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#clustering-analysis">Clustering analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#the-simple-approach-distance-based-clustering-with-k-medoids">The simple approach - distance-based clustering with K-Medoids</a></li>
<li class="toctree-l4"><a class="reference internal" href="#fine-tuning-our-approach-density-based-clustering-with-hdbscan">Fine-tuning our approach - density-based clustering with HDBSCAN</a></li>
<li class="toctree-l4"><a class="reference internal" href="#the-complex-approach-ensemble-based-clustering-with-clicom">The complex approach - ensemble-based clustering with CLICOM</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#enrichment-analysis">Enrichment analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#running-enrichment-analysis">Running enrichment analysis</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#create-analysis-report">Create analysis report</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#analysis-2-differential-expression-data">Analysis #2 - differential expression data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id6">The dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#quantify-fastq-files-and-differential-expression-analysis">Quantify FASTQ files and Differential Expression Analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#quality-control-with-fastqc">Quality control with <em>FastQC</em></a></li>
<li class="toctree-l4"><a class="reference internal" href="#trim-adapters-and-remove-low-quality-reads-with-cutadapt">Trim adapters and remove low-quality reads with <em>CutAdapt</em></a></li>
<li class="toctree-l4"><a class="reference internal" href="#quantify-gene-expression-with-kallisto">Quantify gene expression with <em>kallisto</em></a></li>
<li class="toctree-l4"><a class="reference internal" href="#run-differential-expression-analysis-with-deseq2">Run differential expression analysis with <em>DESeq2</em></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#data-filtering-and-visualization-with-customized-pipelines">Data filtering and visualization with customized Pipelines</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#create-a-customized-pipeline">Create a customized Pipeline</a></li>
<li class="toctree-l4"><a class="reference internal" href="#apply-the-pipeline-to-our-datasets">Apply the Pipeline to our datasets</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#visualizing-and-extracting-gene-set-interesctions">Visualizing and extracting gene set interesctions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#create-a-venn-diagram">Create a Venn diagram</a></li>
<li class="toctree-l4"><a class="reference internal" href="#extract-the-gene-subsets-we-are-interested-in">Extract the gene subsets we are interested in</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id9">Enrichment analysis</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#define-a-background-set">Define a background set</a></li>
<li class="toctree-l4"><a class="reference internal" href="#define-our-custom-enrichment-attributes">Define our custom enrichment attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#id11">Running enrichment analysis</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#final-words">Final words</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide.html">User guide - programmatic interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">A-to-Z tutorial - example analysis</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="a-to-z-tutorial-example-analysis">
<h1>A-to-Z tutorial - example analysis<a class="headerlink" href="#a-to-z-tutorial-example-analysis" title="Permalink to this heading"></a></h1>
<section id="intro-and-recommendations">
<h2>Intro and recommendations<a class="headerlink" href="#intro-and-recommendations" title="Permalink to this heading"></a></h2>
<p>In this A-to-Z tutorial, we are going to analyze two datasets step-by-step, with each analysis covering different aspects of using <em>RNAlysis</em>.</p>
<p>These analyses both focus on RNA sequencing data from nematodes, but the same principles can be applied to data from any organism.</p>
<p>If you want to see more resources and exercises demonstrating how to use <em>RNAlysis</em>, please follow this <a class="reference external" href="https://conwaymsserver.ucd.ie/guides/rnaseq_analysis/rnalysis.html">RNAlysis RNA-seq analysis guide</a> as well, which was kindly provided by <a class="reference external" href="https://www.mwilm-conway-ucd.de/">Prof. Matthias Wilm</a>.</p>
</section>
<section id="analysis-1-time-series-rna-sequencing">
<h2>Analysis #1 - time series RNA sequencing<a class="headerlink" href="#analysis-1-time-series-rna-sequencing" title="Permalink to this heading"></a></h2>
<section id="the-dataset">
<h3>The dataset<a class="headerlink" href="#the-dataset" title="Permalink to this heading"></a></h3>
<p>For the first analysis, we will use a publicly available dataset from <a class="reference external" href="https://downloads.wormbase.org/species/c_elegans/annotation/RNASeq_controls_FPKM/c_elegans.PRJNA13758.current.RNASeq_controls_FPKM.dat">WormBase</a>.
This dataset describes the median expression level of each gene over the developmental stages of <em>C. elegans</em> nematodes, averaged from RNA-sequencing short-read experiments that have been identified in the literature as being ‘controls’.</p>
<p>The original table linked above contains the mean expression, median expression, and number of samples for each developmental stage. Since we are only interested in the medians, before starting the analysis, I copied the Median columns to a new table, and gave the columns more readable names.
You can find the table I created <a class="reference external" href="https://raw.githubusercontent.com/GuyTeichman/RNAlysis/master/tests/test_files/elegans_developmental_stages.tsv">here</a>.</p>
</section>
<section id="start-rnalysis">
<h3>Start <em>RNAlysis</em><a class="headerlink" href="#start-rnalysis" title="Permalink to this heading"></a></h3>
<p>Let’s start by opening <em>RNAlysis</em> - you can learn how to do that by reading the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/installation.html">Installation page</a>.</p>
<p>When you open <em>RNAlysis</em> for the first time, you will be asked whether or not do you want to enable report generation.
<em>RNAlysis</em> allows you to generate automatic, interactive HTML reports of your RNA sequencing data analysis.
This feature is not mandatory, but is highly recommended. You can read more about it in this <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide_gui.html#rnalysis-interactive-analysis-reports">User Guide section</a>.
For this sample analysis, we will enable report generation.</p>
<a class="reference internal image-reference" href="_images/01a00_enable_report_gen.png"><img alt="Enable automatic report generation" src="_images/01a00_enable_report_gen.png" style="width: 600px;" /></a>
<p>We can now load the dataset: in the main window, click on the “Load” button and choose the table’s file from your computer.
We will then use the drop-down menu to change our table type from “Other” to “Count matrix”. This will allow us to later on use analysis methods that are dedicated to count matrix-style datasets.</p>
<a class="reference internal image-reference" href="_images/01a01_load_table.png"><img alt="Loading a table into *RNAlysis* - choose the table type" src="_images/01a01_load_table.png" style="width: 600px;" /></a>
<p>Once we picked a table type, a new option will appear, allowing us to specify whether our table was pre-normalized. We will set this option to “True”, since our dataset was already normalized to Fragments Per Kilobase Million (FPKM).
This is not a mandatory step, but if we don’t do it, <em>RNAlysis</em> will warn us whenever we run analyses that expect normalized values.</p>
<a class="reference internal image-reference" href="_images/01a02_load_table.png"><img alt="Loading a table into *RNAlysis* - set the table as pre-normalized" src="_images/01a02_load_table.png" style="width: 600px;" /></a>
<p>Finally, we can click the “start” button to actually open our table on <em>RNAlysis</em>.
The window will now display a preview of our table, as well as a short summary of our table’s content (table name, table type, number of rows and columns).</p>
<p>At the bottom of the window we can also see a log box - this is where <em>RNAlysis</em> will display notes and warnings regarding our analysis.</p>
<a class="reference internal image-reference" href="_images/01b01_view_table.png"><img alt="Preview the loaded table" src="_images/01b01_view_table.png" style="width: 600px;" /></a>
<p>If we want to see the entire table, we can click on the “View full table” button, and the entire table will appear in a new window:</p>
<a class="reference internal image-reference" href="_images/01b02_view_table.png"><img alt="View the table" src="_images/01b02_view_table.png" style="width: 600px;" /></a>
<p>At any point during the analysis, we can click the “Save table” button to save our table into a new file.</p>
</section>
<section id="data-preprocessing-and-exploratory-data-analysis">
<h3>Data preprocessing and exploratory data analysis<a class="headerlink" href="#data-preprocessing-and-exploratory-data-analysis" title="Permalink to this heading"></a></h3>
<p>We can now begin exploring our data! Let’s start with a pre-processing step - removing lowly-expressed genes.</p>
<section id="filter-out-lowly-expressed-genes">
<h4>Filter out lowly-expressed genes<a class="headerlink" href="#filter-out-lowly-expressed-genes" title="Permalink to this heading"></a></h4>
<p>We want to filter out the genes that have not been expressed or that have low expression across all samples.
Lowly-expressed genes can negatively affect our analysis downstream, since the % error in the expression of these genes is relatively high, and these genes are likely to add noise rather than useful signal to our analysis.</p>
<p>In this particular example, we also want to analyze a fairly small group of genes. This is because later down the line we will take our data through clustering analysis, and trying to cluster such a large group of genes could take your computer an extremely long time to finish.
Therefore, we are going to filter our table, so that we keep only genes with 50 or more normalized reads in at least 1 experimental condition.</p>
<p>To apply a filtering function to our table, click on the “Filter” tab, and select a function from the drop-down menu that opens:</p>
<a class="reference internal image-reference" href="_images/01c01_filter_low_reads.png"><img alt="Choose a filtering function" src="_images/01c01_filter_low_reads.png" style="width: 600px;" /></a>
<p><em>RNAlysis</em> contains a wide array of filtering functions - filtering genes by expression level, fold change direction and magnitude, statistical significance, removing rows with missing data, and many more.
We are going to select “Filter genes with low expression in all columns”. This function will filter our any gene whose expression level is lower than X in every single column. This means we only keep genes that are desirably expressed in at least one experimental condition.
If you are not sure what a function does, you can click on the blue question mark button next to the function’s name to read a short description, or go to the function’s help page by clicking on the blue link at the bottom of the main window.</p>
<p>Once we choose a function, the window will expand, and the filtering function’s parameters will be displayed below. We can modify these parameters to choose exactly how to filter our table.</p>
<p>In this example, we are going to set the parameter <cite>threshold</cite> to 50 FPKM. This means that any genes which have less than 50 FPKM in <strong>all</strong> of the table’s columns will be filtered out.
If you are not sure what a parameter does, you can hover your cursor over its name, or click on the blue question mark button next to the parameter’s name.</p>
<p>Every filtering function in <em>RNAlysis</em> supports two additional parameters: <cite>opposite</cite> and <cite>inplace</cite>.</p>
<p><cite>opposite</cite>, as the name indicates, allows you to apply the <strong>inverse</strong> of your filtering operation to your table. For example, in our case, instead of filtering out lowly-expressed genes, we will filter out everything <strong>but</strong> the lowly-expressed genes.</p>
<p><cite>inplace</cite> allows us to choose whether we want to apply the filtering operation in the same window (default), or to keep an unfiltered version of the table and apply the filtering operation in a new tab.
This can be conveniet if we want to try multiple filtering approaches in parallel, or if we need to use the unfiltered table later down the line.
Regardless of what we choose, the original file we loaded will not be changed by our filtering unless we expelicitly save our filtering results, and any filtering operation we apply can be undone with the click of a button.</p>
<p>Once we are happy with the parameters we chose, we can scroll down and click on the “Apply” button to apply the filtering function to our table.</p>
<a class="reference internal image-reference" href="_images/01c02_filter_low_reads.png"><img alt="Set filtering parameters" src="_images/01c02_filter_low_reads.png" style="width: 600px;" /></a>
<p>After clicking “Apply”, we can see that our action has been added to the <em>command history</em> panel on the right of the window. We can undo and redo any operation we apply inplace by clicking on a specific row in this panel.
Moreover, we can see that a short summary of our filtering action has printed to the log box at the bottom, and that the name and size of the table have been updated.</p>
<a class="reference internal image-reference" href="_images/01c03_filter_low_reads.png"><img alt="Filtering results" src="_images/01c03_filter_low_reads.png" style="width: 600px;" /></a>
</section>
<section id="examine-variance-in-our-data-with-principal-component-analysis">
<h4>Examine variance in our data with Principal Component Analysis<a class="headerlink" href="#examine-variance-in-our-data-with-principal-component-analysis" title="Permalink to this heading"></a></h4>
<p>The first analysis we will perform is Principal Component Analysis, or PCA.
Principal component analysis is a dimensionality-reduction method. Meaning, it can reduce the dimensionality of large data sets (e.g. the expression of thousands of genes), by transforming the expression data of these genes into a smaller dataset that still contains most of the information from the original dataset.</p>
<p>Since our dataset contains thousands of genes, it can be difficult to see how the different conditions vary in the expression of those genes.
To remedy that, PCA analysis can “rearrange” the axes of our data such that we can capture most of the variance of our dataset in very few dimensions (~2-10 dimensions).</p>
<p>You can read more about PCA <a class="reference external" href="https://builtin.com/data-science/step-step-explanation-principal-component-analysis">here</a>.</p>
<p>To run a PCA, click on the “Visualize” tab, and select “Principal Component Analysis” from the drop-down menu.</p>
<a class="reference internal image-reference" href="_images/01d01_pca.png"><img alt="Choose a visualization function function" src="_images/01d01_pca.png" style="width: 600px;" /></a>
<p>The ‘samples’ parameter allows you to choose which samples will be analyzed with PCA, and also lets you group these samples into sub-groups for the purpose of visualization (for example, group replicate data by experimental condition).
That way, each sub-group will be drawn with a different color on the final graph.
In our case we only have one column per condition, and we want to examine them all, so we don’t need to change this parameter.</p>
<p>By default, <em>RNAlysis</em> will apply a power-transform (Box-Cox) to the data before standardazing it and running PCA. This is the case for many functions in <em>RNAlysis</em>, since applying power transform minimizes undesirable characteristics of counts data, such as skeweness, mean-variance dependence, and extreme values.
However, this feature can always be disabled with the <cite>power_transform</cite> parameter.</p>
<p>Whether we apply a power transform to our data or not, <em>RNAlysis</em> will then standardize our data to neutralize any differences in absolute gene expression level, and then applies the Principal Component Analysis</p>
<p>Let’s apply the analysis and look at the output:</p>
<a class="reference internal image-reference" href="_images/01d02_pca.png"><img alt="Principal Component Analysis" src="_images/01d02_pca.png" style="width: 600px;" /></a>
<p><em>RNAlysis</em> generated three graphs, depicting all of the pair-wise combinations between Principal Components #1-3.
We can visualize less or more principal components by changing the value of the <cite>n_components</cite> parameter.</p>
<p>Usually the most relevant graph is the one depicting the first and second PCs, since they explain the largest amount of variance in the data.
However, when the % of variance explained by preceeding PCs is similar (perhaps PCs 2 and 3 explain about the same % of variance), some of the PCs may be more informative than others, so it’s worth keeping an eye out for those.</p>
<p>In our case, we can see that PCs 1-2 together explain ~75% of the variance in our dataset. Interestingly, the PCA shows a semi-circular pattern, ordered by the developmental stage of the worms.
My hypothesis would be that PC1 arranges the samples by their relative “germline” content - embryos are mostly gonads, adult nematodes contain a rather large quantity of germ cells, L4 larvae is the developmental stage where germline proliferation begins, and during the L1-L3 stages the relative germline content of the worms is relatively minimal.
PC2 appears to arrange the samples by their developmental stage, with embryos appearing at the top of the graph and adults at the bottom.</p>
</section>
<section id="examine-similarity-between-developmental-stages">
<h4>Examine similarity between developmental stages<a class="headerlink" href="#examine-similarity-between-developmental-stages" title="Permalink to this heading"></a></h4>
<p>Let’s now examine the distribution of gene expression across developmental stages with a <a class="reference external" href="https://pythonbasics.org/seaborn-pairplot/">Pair-plot</a>.
Pair-plots displays the pairwise relationships between samples, or experimental conditions, in our dataset, and also display a histogram of gene expression values within each sample/condition.</p>
<p>To generate a Pair-plot, select “Pair-plot” from the function drop-down menu.</p>
<p>The Pair-plot function allows you to group samples by experimental condition in order to average them together.</p>
<p>We also have the option, like with PCA, to choose whether or not to power transform our data.</p>
<p>Let’s click “Apply” and check out the result:</p>
<a class="reference internal image-reference" href="_images/01e02_pairplot.png"><img alt="Pair-plot" src="_images/01e02_pairplot.png" style="width: 600px;" /></a>
<p>We can spot some interesting trends from this plot: the embryo condition seems to be the most dissimilar from the rest of the conditions; sequential developmental stages seem to be fairly correlated with one another; and the earlier developmental stages appear to have a larger fraction of unexpressed genes compared to the later developmental stages.</p>
</section>
<section id="compare-the-expression-of-specific-genes-over-the-developmental-stages">
<h4>Compare the expression of specific genes over the developmental stages<a class="headerlink" href="#compare-the-expression-of-specific-genes-over-the-developmental-stages" title="Permalink to this heading"></a></h4>
<p>We already have a hypothesis about the expression of some genes over the developmental stages, which we would like to test.
For example, the gene <em>oma-1</em> (gene ID <strong>WBGene00003864</strong>) should be expressed almost exclusively in the two-cell stage of embryonic development, and we expect the gene <em>skn-1</em> (gene ID <strong>WBGene00004804</strong>) to show a fairly consistent expression level across development.
Let’s go to the “Visualize” tab to plot the expression of these genes over the developmental stages.</p>
<p>The <cite>features</cite> parameter will let us choose which genomic features we want to plot expression for. Since we can decide to add as many genes as we would like to this graph, <em>RNAlysis</em> will give us the option to choose how many genes to input.
To start, click on the “Set input” button next to the <cite>features</cite> parameter:</p>
<a class="reference internal image-reference" href="_images/01f01_plot_expression.png"><img alt="Choose genes to plot" src="_images/01f01_plot_expression.png" style="width: 600px;" /></a>
<p>In the pop-up window that opens, you can use the “Add field” and “Remove field” buttons to choose how many genes to plot the expression of.
Then, in each field, enter the gene ID of a gene you want to visualize. When you’re done, click on the “Done” button.</p>
<a class="reference internal image-reference" href="_images/01f02_plot_expression.png"><img alt="Choose genes to plot - part 2" src="_images/01f02_plot_expression.png" style="width: 600px;" /></a>
<p>Let’s also set the count unit to FPKM, and click “Apply” to create the plot:</p>
<a class="reference internal image-reference" href="_images/01f03_plot_expression.png"><img alt="Expression plot" src="_images/01f03_plot_expression.png" style="width: 600px;" /></a>
<p>Interestingly, we can see that the expression of <em>oma-1</em> is actually highest in the adult worms. This is possibly because the adult worm contains a large number of unlaid embryos, some of which are the two-cell stage.</p>
<p>Our data only has one sample per condition, so our bar plots do not convey any information about the variance in gene expression within each condition.
However, if we had multiple sample per condition, we could have grouped the samples into subgroup (like with the PCA and Pair-plot functions), and then each bar plot will also display a scatter of the expression values in each sub-group, and the standard error of the expression value.</p>
</section>
</section>
<section id="clustering-analysis">
<h3>Clustering analysis<a class="headerlink" href="#clustering-analysis" title="Permalink to this heading"></a></h3>
<p>We are interested in how different groups of genes change in expression level over the life cycle of the worm. We can use clustering analysis to group the genes in this dataset by their expression pattern over the developmental stages.
There is an abundance of approaches when it comes to clustering analysis of gene expression. To illustrate this point, we will cluster our data using three different types of clustering algorithms, arranged from the simplest to the most complex.
These algorithms are only a few representatives of the many clustering methods available in <em>RNAlysis</em>.</p>
<section id="the-simple-approach-distance-based-clustering-with-k-medoids">
<h4>The simple approach - distance-based clustering with K-Medoids<a class="headerlink" href="#the-simple-approach-distance-based-clustering-with-k-medoids" title="Permalink to this heading"></a></h4>
<p>K-Medoids clustering is a distance-based clustering method, where the algorithm attempts to divide all of our genes into K clusters (K is the number of clusters we are looking for), with each cluster being centered around one gene (Medoid).
K-Medoids clustering takes in two major parameters - the number of clusters we expect (K), and the distance metric by which we want to measure the similarity of expression between genes.</p>
<p>Specifying the exact number of clusters we expect is a bit challenging for us, since we aren’t actually sure how many biologically-meaningful clusters are there in our data.
Moreover, this number could depend on how fine-grained we want our analysis to be - we could reasonably divide our genes into a small number of more generalized clusters (such as “genes expressed more in the start of development” vs “genes expressed more near the end of development”), or we could further divide our genes into smaller groups based on their exact temporal expression pattern.</p>
<p>Fortunately, some methods were developed to suggest a good number of clusters for our dataset (a “good number of clusters” meaning that the genes in each clusters are most similar to each other and most different than genes in other clusters). Two of these methods, known as the Silhouette Method and the Gap Statistic, are available in <em>RNAlysis</em>.
We will use the Gap Statistic method to determine some good options for the number of clusters in our analysis.</p>
<p>To start, let’s click on the Clustering tab and choose K-Medoids clustering from the drop-down menu.
We can then set the value of the parameter <cite>n_clusters</cite> to ‘gap’, to indicate that we want to use the Gap Statistic to determine the number of clusters in this analysis:</p>
<a class="reference internal image-reference" href="_images/01g01_kmedoids.png"><img alt="K-Medoids clustering setup - choose the number of clusters using the Gap Statistic" src="_images/01g01_kmedoids.png" style="width: 600px;" /></a>
<p>Next, we can set the distance metric. Different distance metrics can be more or less effective on specific types of data.
<em>RNAlysis</em> offers a large array of distance metrics, about which you can read in the <em>RNAlysis</em> <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide_gui.html#data-clustering-with-countfilter">user guide</a>.
We will use a lesser-known distance metric called YR1, that was developed especially for time-series gene expression data and implemented in <em>RNAlysis</em>. You can read more about it in <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>:</p>
<a class="reference internal image-reference" href="_images/01g02_kmedoids.png"><img alt="K-Medoids clustering setup - choose the distance metric" src="_images/01g02_kmedoids.png" style="width: 600px;" /></a>
<p>We can now scroll all the way down, click the “Apply” button, and wait for the analysis to finish:</p>
<a class="reference internal image-reference" href="_images/01g03_kmedoids.png"><img alt="K-Medoids clustering - loading screen" src="_images/01g03_kmedoids.png" style="width: 600px;" /></a>
<p>Since K-Medoids is not a deterministic algorithm (it has randomized starting conditions every time you use it), the results you get will probably not be identical to those appearing in this tutorial, but they should be similar to what we observe here.
If you want to make sure your clustering results are 100% reproducible, you can set the <cite>random_seed</cite> parameter when setting up your clustering setup to a specific number.
Re-running the algorithm with an identical random seed will ensure that you get the exact same results every time.</p>
<p>One the clustering anslysis is finished, a few figures will open up. Let’s examine them one by one.
The first figure will show us the results of the Gap Statistic algorithm. The graph on the left will show us, for each value of <cite>n_clusters</cite> tested, the natural logarithm (ln) of within-cluster dispersion.
We expect this value to go down the more clusters there are (the more clusters there are, the fewer genes will be in each cluster - therefore the genes within each cluster will be more similar to each other), and therefore we show both the actual dispersion values for the clustering solutions we calculated (the blue line), and also dispersion values for clustering solutions on random data, drawn from the same distribution (the orange line).
On the graph to the right we can see the natural logarithm of the ratio between the observed and expected dispersion - this is the Gap Statistic.
We are looking for local ‘peaks’ in the graph -  values of <cite>n_clusters</cite> that have a larger Gap Statistic than their neighbors.
<em>RNAlysis</em> automatically picks the lowest value of <cite>n_clusters</cite> that fits this criterion, but also suggests other good values based on the results.</p>
<a class="reference internal image-reference" href="_images/01g04_kmedoids.png"><img alt="Gap Statistic results" src="_images/01g04_kmedoids.png" style="width: 600px;" /></a>
<p>In our case, <em>RNAlysis</em> recommended 3 clusters as the optimal number of clusters. This value might not be granular enough for the kind of analysis we want to run. Therefore, we will re-run the K-Medoids algorithm with the same parameters, but set the value of <cite>n_clusters</cite> to one of the next good values discovered in the Gap Statistic analysis - 11 clusters.</p>
<p>Next, we can look at the rest of the output of the K-Medoids clustering algorithm for <a href="#id2"><span class="problematic" id="id3">`</span></a>n_clusters`=11.
The first graph will show us the distributions of gene expression in each discovered cluster. Note that the expression values are regularized and power-transformed, since we are interested in grouping the different genes by their relative pattern of expression, and not by their absolute expression levels (highly/lowly-expressed genes).
The clusters are sorted by their size, from the biggest to the smallest cluster.
This type of graph can help us see the general expression pattern that characterizes each cluster. Moreover, it can help point out how internally similar our clusters are - indicating the quality of our clustering result.</p>
<a class="reference internal image-reference" href="_images/01g05_kmedoids.png"><img alt="K-Medoids clustering results" src="_images/01g05_kmedoids.png" style="width: 600px;" /></a>
<p>In this case, we can see that while some clusters seem very internally consistent, quite a few clusters seem to contain a significant number of ‘outlier’ genes.</p>
<p><em>RNAlysis</em> also generates a Principal Component Analysis projection of our gene expression data, marking the genes in each cluster with a different color.
This is another useful way to look at our clustering results - we would hope to see that the first two principal components explain a large degree of the variance in gene expression, and the genes in the same clusters will be grouped together in the graph.</p>
<a class="reference internal image-reference" href="_images/01g06_kmedoids.png"><img alt="K-Medoids clustering results - principal component analysis" src="_images/01g06_kmedoids.png" style="width: 600px;" /></a>
<p>Finally, the following window will open, prompting us to choose which output clusters we want to keep, and giving us the option to give these clusters a new name:</p>
<a class="reference internal image-reference" href="_images/01g07_kmedoids.png"><img alt="K-Medoids clustering results - choose which clusters to keep" src="_images/01g07_kmedoids.png" style="width: 600px;" /></a>
<p>For every cluster we choose to keep, <em>RNAlysis</em> will create a new copy of the original count matrix, which includes only the genes that belong to this specific cluster.
For now, we will choose to keep none of the clusters, so that we can try out other clustering approaches. Therefore, we click either OK or Cancel without selecting any clusters.</p>
</section>
<section id="fine-tuning-our-approach-density-based-clustering-with-hdbscan">
<h4>Fine-tuning our approach - density-based clustering with HDBSCAN<a class="headerlink" href="#fine-tuning-our-approach-density-based-clustering-with-hdbscan" title="Permalink to this heading"></a></h4>
<p>The next clustering approach we will use, HDBSCAN, belongs to a different category of clustering algorithms - density-based clustering.
HDBSCAN stands for Hierarchical Density-Based Spatial Clustering of Applications with Noise (see <a class="reference external" href="https://link.springer.com/chapter/10.1007/978-3-642-37456-2_14">the publication</a> for more details).
HDBSCAN offers multiple advantages over more traditional clustering methods:</p>
<ol class="arabic simple">
<li><p>HSBSCAN makes relatively few assumptions about the data - it assumes that the data contains noise, as well as some real clusters which we hope to discover.</p></li>
<li><p>Unlike most other clustering methods, HDBSCAN does not “force” every gene to belong to a cluster. Instead, it can classify genes as outliers, excluding them from the final clustering solution.</p></li>
<li><p>When using HDBSCAN, you don’t have to guess the number of clusters in the data. The main tuning parameter in HDBSCAN is <em>minimum cluster size</em> (<cite>min_cluster_size</cite>), which determines the smallest “interesting” cluster size we expect to find in the data.</p></li>
</ol>
<p>To run HDBSCAN, we need to pick a value for <cite>min_cluster_size</cite>.
Lower values of <cite>min_cluster_size</cite> will return a larger number of small clusters, revealing more fine-grained patterns in our gene expression data.
Higher values of <cite>min_cluster_size</cite> will return a smaller number of large clusters, revealing the most obvious and significant patterns in the data.
For our example, let’s pick a value of 75:</p>
<a class="reference internal image-reference" href="_images/01g11_hdbscan.png"><img alt="HDBSCAN clustering setup" src="_images/01g11_hdbscan.png" style="width: 600px;" /></a>
<p>We will, once again, use YR1 as the distance metric.</p>
<p>If we look at the clustering results, we can see that HDBSCAN ended up generating a much larger number of clusters than the previous method, and they look fairly internally consistent.</p>
<a class="reference internal image-reference" href="_images/01g12_hdbscan.png"><img alt="HDBSCAN clustering results" src="_images/01g12_hdbscan.png" style="width: 600px;" /></a>
<a class="reference internal image-reference" href="_images/01g13_hdbscan.png"><img alt="HDBSCAN clustering results - principal component analysis" src="_images/01g13_hdbscan.png" style="width: 600px;" /></a>
<p>Let’s now move on to the final clustering approach - ensemble-based clustering.</p>
</section>
<section id="the-complex-approach-ensemble-based-clustering-with-clicom">
<h4>The complex approach - ensemble-based clustering with CLICOM<a class="headerlink" href="#the-complex-approach-ensemble-based-clustering-with-clicom" title="Permalink to this heading"></a></h4>
<p>The last clustering approach we will use, CLICOM, is an emsemble-based clustering algorithm.
CLICOM (see <a class="reference external" href="https://doi.org/10.1016/j.eswa.2011.08.059">the publication</a> ) incorporates the results of multiple clustering solutions, which can come from different clustering algorithms with differing clustering parameters, and uses these clustering solutions to create a combined “concensus” clustering solution.
CLICOM offers multiple advantages over more traditional clustering methods:</p>
<ol class="arabic simple">
<li><p>The ensemble clustering approach allows you to combine the results of multiple clustering algorithms with multiple tuning parameters, potentially making up for the weaknesses of each individual clustering method, and only taking into account patterns that robustly appear in many clustering solutions.</p></li>
<li><p>When using CLICOM, you don’t have  to guess the final number of clusters in the data. The main tuning parameter in HDBSCAN is the <em>evidence threshold</em> (<cite>evidence_threshold</cite>).</p></li>
</ol>
<p><em>RNAlysis</em> offers a modified implementation of CLICOM. The modified version of the algorithm can, like the HDBSCAN algorithm, classify genes as outliers, excluding them from the final clustering solution.
Moreover, ths modified version of the algorithm can cluster each batch of biological/technical replicates in your data separately, which can reduce the influence of batch effect on clustering results, and increases the accuracy and robustness of your clustering results.</p>
<p>This modified version of CLICOM supports a few tuning parameters, in addition to the clustering solutions themselves:</p>
<ul class="simple">
<li><p><cite>evidence_threshold</cite>: how many clustering solutions (fraction between 0-1) have to agree about  two genes being clustered together in order for them to appear together in the final solution? A lower evidence threshold leads to fewer, large clusters, with fewer features being classified as outliers.</p></li>
<li><p><cite>cluster_unclustered_features</cite>: if set to True, CLICOM will force every gene to belong to a discovered cluster (like in the original implementation of the algorithm). Otherwise, genes can be classified as noise and remain unclustered (the modified algorithm).</p></li>
<li><p><cite>min_cluster_size</cite>: determines the minimal size of a cluster you would consider meaningful. Clusters smaller than this would be classified as noise and filtered out of the final result, or merged into other clusters (depending on the value of <cite>cluster_unclustered_features</cite>).</p></li>
<li><p><cite>replicates_grouping</cite>: allows you to group samples into technical/biological batches. The algorithm will then cluster each batch of samples separately, and use the CLICOM algorithm to find an ensemble clustering result from all of the separate clustering results.</p></li>
</ul>
<p>To start the analysis, let’s choose the CLICOM algorithm from the Clustering drop-down menu. A new window will open:</p>
<a class="reference internal image-reference" href="_images/01g21_clicom.png"><img alt="CLICOM clustering setup" src="_images/01g21_clicom.png" style="width: 600px;" /></a>
<p>On the left half of the window we can set the tuning parameters of the CLICOM algorithm. For our example, let’s set the evidence threshold to 0.5, and the minimum cluster size to 75.
Since our data doesn’t have replicates, we don’t need to tune the <cite>replicates_grouping</cite> parameter. If in the future you try to cluster data with biological replicates, keep this parameter in mind!</p>
<p>On the right half of the window we can add new clustering setups to our run of CLICOM. These clustering setups can be any of the clustering algorithms available in <em>RNAlysis</em>, and we can add as many as we want - including multiple clustering setups using the same algorithm.
To add a setup, pick an algorithm from the drop-down menu, set it’s parameters, and click the “Add Setup” button:</p>
<a class="reference internal image-reference" href="_images/01g22_clicom.png"><img alt="CLICOM clustering - add clustering setups" src="_images/01g22_clicom.png" style="width: 600px;" /></a>
<p>The setups you added will appear under “added setups” on the right, and you can delete a setup from that list if you want:</p>
<a class="reference internal image-reference" href="_images/01g23_clicom.png"><img alt="CLICOM clustering - examine added setups" src="_images/01g23_clicom.png" style="width: 600px;" /></a>
<p>Let’s add the two clustering setups we used earlier, plus a few more:</p>
<a class="reference internal image-reference" href="_images/01g24_clicom.png"><img alt="CLICOM clustering - multiple clustering setups" src="_images/01g24_clicom.png" style="width: 600px;" /></a>
<p>I chose to add, in addition to the two clustering setups from earlier, the following three clustering setups:</p>
<ul class="simple">
<li><p>K-Medoids clustering with 7 clusters and the Spearman distance metric</p></li>
<li><p>Hierarchical clustering 8 clusters, Euclidean distance metric and Ward linkage</p></li>
<li><p>HDBSCAN clustering with the Jackknife distance metric and minimal cluster size of 150</p></li>
</ul>
<p>Once we are happy with the clustering solutions and tuning parameters, we can click on the “Start CLICOM” button, and see progress reports in the output box on the main window of <em>RNAlysis</em>.</p>
<a class="reference internal image-reference" href="_images/01g25_clicom.png"><img alt="CLICOM clustering loading screen" src="_images/01g25_clicom.png" style="width: 600px;" /></a>
<p>Let’s look at the final result:</p>
<a class="reference internal image-reference" href="_images/01g26_clicom.png"><img alt="CLICOM clustering results" src="_images/01g26_clicom.png" style="width: 600px;" /></a>
<a class="reference internal image-reference" href="_images/01g27_clicom.png"><img alt="CLICOM clustering results - principal component analysis" src="_images/01g27_clicom.png" style="width: 600px;" /></a>
<p>Once again, since two of the clustering setups we chose are non-deterministic, the clusters you end up with might end up looking a bit differently from these.
Let’s choose a few good-looking clusters to keep, and give them a name that indicates their expression pattern:</p>
<a class="reference internal image-reference" href="_images/01g28_clicom.png"><img alt="The clusters we chose to keep" src="_images/01g28_clicom.png" style="width: 600px;" /></a>
<p>For this tutorial, I chose to keep clusters #1 (“down from embryo to L4”), #2 (“L4 peak”), and #9 (“Down from L1 to adult”).</p>
</section>
</section>
<section id="enrichment-analysis">
<h3>Enrichment analysis<a class="headerlink" href="#enrichment-analysis" title="Permalink to this heading"></a></h3>
<p>Now that we have extracted a few clusters of interest, we can try to characterize their biological significance using enrichment analysis. We will look at the most commonly-used enrichment analysis - Gene Ontology enrichment.</p>
<section id="running-enrichment-analysis">
<h4>Running enrichment analysis<a class="headerlink" href="#running-enrichment-analysis" title="Permalink to this heading"></a></h4>
<p>To open the Enrichent Analysis window, open the ‘Gene sets’ menu and click “Enrichment Analysis”:</p>
<a class="reference internal image-reference" href="_images/01h01_go_enrichment.png"><img alt="Pick 'Enrichment analysis' from the 'Gene sets' menu" src="_images/01h01_go_enrichment.png" style="width: 600px;" /></a>
<p>For basic enrichment analysis, we first need to choose our <em>enrichment set</em> (the gene set we are interested in - for example, “Down from L1 to adult”, cluster #9 we found earlier), and our background set (the reference genes we will be comparing our enrichment results to - for example, the genes in our original filtered table).
We can choose our sets from the two drop-down menus in the Enrichment window:</p>
<a class="reference internal image-reference" href="_images/01h02_go_enrichment.png"><img alt="Enrichment analysis setup - choose the enrichment and background sets" src="_images/01h02_go_enrichment.png" style="width: 600px;" /></a>
<p>Next, we can choose the dataset we want to draw annotations from. In our case, we will choose Gene Ontology (GO).
After picking the dataset, the window expanded to show us all of the parameters we can modify for our analysis:</p>
<a class="reference internal image-reference" href="_images/01h03_go_enrichment.png"><img alt="Enrichment analysis setup - choose the analysis type, organism, and gene ID type" src="_images/01h03_go_enrichment.png" style="width: 600px;" /></a>
<p>Let’s set the statistical test to ‘hypergeometric’, the organism to ‘Caenorhabditis elegans’ (matching our gene expression data), and the Gene ID Type to “WormBase” (matching the gene ID type of our original data table).</p>
<p>We will leave the rest of the settings on the default values, but keep in mind that you can customize the analysis to a significant degree: using different statistical tests (including a statistical test that doesn’t require a background gene set), using only specific types of GO annotations, propagating the annotations differently, etc’. You can read more about these options in the complete user guide.</p>
<p>Scroll to the bottom of the window and click on the “run” button to run the analysis:</p>
<a class="reference internal image-reference" href="_images/01h04_go_enrichment.png"><img alt="Enrichment analysis loading screen" src="_images/01h04_go_enrichment.png" style="width: 600px;" /></a>
<p>The enrichment window is going to minimize to allow you to read the log box on the main <em>RNAlysis</em> window, but you can enlarge it back if you want to look at the analysis parameters, or start a new analysis with the same parameters.
The log will update you on the number of annotations found, whether some genes were left entirely unannotated, how many gene IDs were successfully mapped to match the annotations, etc.</p>
<p>Once the analysis is done, we will be able to observe our results in multiple formats.</p>
<p>The first is a tabular format, showing all of the statistically significant GO terms we found (or the all of the tested GO terms, if we set the <cite>return_nonsignificant</cite> parameter to True).
The GO terms will be sorted by specificity, the most specific GO terms appearing at the top of the table.
The table also includes the statistics for each GO term (number of genes in the cluster, number of genes matching the GO term, expected number of genes to match the GO term, log2 fold change, p-value, and adjusted p-value).</p>
<a class="reference internal image-reference" href="_images/01h06_go_enrichment.png"><img alt="Enrichment analysis results - tabular format" src="_images/01h06_go_enrichment.png" style="width: 600px;" /></a>
<p>The second output format is an Ontology Graph, depicting the statistically-significant GO terms in each GO aspect (biological process, cellular component, and molecular function), as well as their ancestors in the ontology graph.
The hierarchy of GO terms in the graph matches the official Gene Ontology hierarchy.
The color of the terms on the graph indicates their log2 fold change, and the depth in the tree indicates the specificity of the term, with more specific GO terms being at the bottom.</p>
<a class="reference internal image-reference" href="_images/01h05_go_enrichment.png"><img alt="Enrichment analysis results - ontology graph" src="_images/01h05_go_enrichment.png" style="width: 800px;" /></a>
<p>The final output format is a bar plot depicting the log2 fold change values, as well as significance, of the 10 most specific GO terms that were found to be statistically significant in our analysis.</p>
<a class="reference internal image-reference" href="_images/01h07_go_enrichment.png"><img alt="Enrichment analysis results - bar plot" src="_images/01h07_go_enrichment.png" style="width: 600px;" /></a>
</section>
</section>
<section id="create-analysis-report">
<h3>Create analysis report<a class="headerlink" href="#create-analysis-report" title="Permalink to this heading"></a></h3>
<p>Once we are done analyzing the data, we can generate an interactive analysis report.
To do that, open the ‘File’ menu and click on “Create session report”:</p>
<a class="reference internal image-reference" href="_images/01_i01_generate_report.png"><img alt="Create interactive analysis report" src="_images/01_i01_generate_report.png" style="width: 600px;" /></a>
<p>Next, select a location on your computer to save the report to.
Once you do, <em>RNAlysis</em> will automatically generate a complete report of the analysis you ran up until this point.
The report should open on its own once is was generated.</p>
<p>Your report should look similar to this:</p>
<a class="reference internal image-reference" href="_images/01i04_view_report.png"><img alt="View interactive analysis report" src="_images/01i04_view_report.png" style="width: 600px;" /></a>
<p>The report includes all of the functions we applied throughout the analysis, and the tables and graphs we generated.
You can get a closer look at these elements by hovering your cursor on them:</p>
<a class="reference internal image-reference" href="_images/01_i05_hover.png"><img alt="Hover on a node to reveal more information" src="_images/01_i05_hover.png" style="width: 600px;" /></a>
<p>If you want to learn more about <em>RNAlysis</em> interactive reports, see this <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide_gui.html#rnalysis-interactive-analysis-reports">User Guide section</a>.</p>
<p>If you want, you can download <a class="reference external" href="https://raw.githubusercontent.com/GuyTeichman/RNAlysis/master/tests/test_files/sample_RNAlysis_report.zip">the report that I generated</a>.</p>
</section>
</section>
<section id="analysis-2-differential-expression-data">
<h2>Analysis #2 - differential expression data<a class="headerlink" href="#analysis-2-differential-expression-data" title="Permalink to this heading"></a></h2>
<section id="id6">
<h3>The dataset<a class="headerlink" href="#id6" title="Permalink to this heading"></a></h3>
<p>In the second analysis, we will analyze three publicly available datasets from GEO. Each individual dataset is an RNA sequencing experiment of <em>Caenorhabidtis elegans</em> nematodes, measuring the effect of a particular stress condition on gene expression.
Our goal in this analysis will be to examine the similarities between the different stress conditions, and to answer a specific question - how are Epigenetic genes and pathways affected by exposure to stress?</p>
<p>In this analysis, unlike the first analysis, we will not start from a pre-made counts table.
Instead, we are going to pre-process and quantify each dataset directly from the raw FASTQ files that are available on GEO.</p>
<p>The datasets we will use are:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>mRNA sequencing of worms exposed to osmotic stress: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107704">GSE107704</a>. This is a single-end unstranded RNA sequencing experiment.</dt><dd><ul class="simple">
<li><p>Control sample 1 - SRR6348197</p></li>
<li><p>Control sample 2 - SRR6348198</p></li>
<li><p>Control sample 3 - SRR6348199</p></li>
<li><p>Osmotic sample 1 - SRR6348200</p></li>
<li><p>Osmotic sample 2 - SRR6348201</p></li>
<li><p>Osmotic sample 3 - SRR6348202</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mRNA sequencing of worms exposed to starvation: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE124178">GSE124178</a>. This is a paired-end stranded RNA sequencing experiment.</dt><dd><ul class="simple">
<li><p>Control sample 1 - SRR8359132</p></li>
<li><p>Control sample 2 - SRR8359133</p></li>
<li><p>Control sample 3 - SRR8359142</p></li>
<li><p>Starvation sample 1 - SRR8359134</p></li>
<li><p>Starvation sample 2 - SRR8359135</p></li>
<li><p>Starvation sample 3 - SRR8359136</p></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>mRNA sequencing of worms exposed to heat shock: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE132838">GSE132838</a>. This is a paired-end stranded RNA sequencing experiment.</dt><dd><ul class="simple">
<li><p>Control sample 1 - SRR9310678</p></li>
<li><p>Control sample 2 - SRR9310679</p></li>
<li><p>Control sample 3 - SRR9310680</p></li>
<li><p>Heat sample 1 - SRR9310681</p></li>
<li><p>Heat sample 2 - SRR9310682</p></li>
<li><p>Heat sample 3 - SRR9310683</p></li>
</ul>
</dd>
</dl>
</li>
</ol>
</section>
<section id="quantify-fastq-files-and-differential-expression-analysis">
<h3>Quantify FASTQ files and Differential Expression Analysis<a class="headerlink" href="#quantify-fastq-files-and-differential-expression-analysis" title="Permalink to this heading"></a></h3>
<p>RNAlysis offers user the opportunity to start analysis right from raw FASTQ data.</p>
<p>Since our input data is raw FASTQ files, we will first have to pre-process them, quantify them, and then detect differentially expressed genes.
When using RNAlysis, you can start analyzing your data at any point in the analysis pipeline: raw FASTQ files, trimmed FASTQ files, quantified read counts, or differential expression tables.
This means you can combine <em>RNAlysis</em> with any other tools or pipelines you prefer.</p>
<p><em>RNAlysis</em> provides a graphical interface for <em>CutAdapt</em>, <em>kallisto</em>, and <em>DESeq2</em>, and we will use those tools in this tutorial analysis.</p>
<section id="quality-control-with-fastqc">
<h4>Quality control with <em>FastQC</em><a class="headerlink" href="#quality-control-with-fastqc" title="Permalink to this heading"></a></h4>
<p>The first step in analyzing RNA sequencing data is uaually quality control. This can help us get a general overview of our FASTQ files, and detect any potential issues with them.
One of the most popular tools for this purpose is <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/download.html#fastqc">FastQC</a>.
FastQC is an open-source tool, and it has both a friendly graphical interface and a programmatic API.
You can get it <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/download.html#fastqc">here</a>.</p>
</section>
<section id="trim-adapters-and-remove-low-quality-reads-with-cutadapt">
<h4>Trim adapters and remove low-quality reads with <em>CutAdapt</em><a class="headerlink" href="#trim-adapters-and-remove-low-quality-reads-with-cutadapt" title="Permalink to this heading"></a></h4>
<p>After doing a quality-control of our FASTQ files with FastQC, we can see that some of our samples had a small portion of reads that still contain adapter sequences.
Therefore, we will start our analysis by trimming the leftover adapters. We will also perform quality-trimming, removing bases with low quality scores from our reads.</p>
<p>The first dataset we will trim is the osmotic stress dataset, which happens to be single-end sequencing.
To begin, let’s open the “FASTQ” menu in <em>RNAlysis</em>, and under “Adapter trimming” choose “Single-end adapter trimming”:</p>
<a class="reference internal image-reference" href="_images/02a01_cutadapt.png"><img alt="FASTQ menu - adapter trimming" src="_images/02a01_cutadapt.png" style="width: 600px;" /></a>
<p>In the new window that opened, we can choose the folder that contains our raw FASTQ files, as well as the folder our trimmed files should be written to:</p>
<a class="reference internal image-reference" href="_images/02a02_cutadapt.png"><img alt="adapter trimming - set input and output folders" src="_images/02a02_cutadapt.png" style="width: 600px;" /></a>
<p>The input folder can contain any number of raw FASTQ files, and <em>RNAlysis</em> will trim them one by one and save the trimmed output files in the output folder.</p>
<p>Next, let’s set the adapter sequence we want to trim. The adapter used on these samples is the three-prime Illumina TruSeq adapter, with the sequence “AGATCGGAAGAGCACACGTCTGAACTCCAGTCA”.
Let’s click on the “Set input” button of the <cite>three_prime_adapters</cite> parameter:</p>
<a class="reference internal image-reference" href="_images/02a03_cutadapt.png"><img alt="adapter trimming - set input for adapter sequence" src="_images/02a03_cutadapt.png" style="width: 600px;" /></a>
<p>This will open a dialog box, where we can enter the sequence of our adapter:</p>
<a class="reference internal image-reference" href="_images/02a04_cutadapt.png"><img alt="adapter trimming - set adapter sequence" src="_images/02a04_cutadapt.png" style="width: 600px;" /></a>
<p>We will leave the quality-trimming parameters on their default values - trim bases below quality score 20 from the 3’ end of reads, trim flanking N bases from the reads, and filter out any read that ends up shorter than 10bp after trimming:</p>
<a class="reference internal image-reference" href="_images/02a05_cutadapt.png"><img alt="adapter trimming - quality trimming parameters" src="_images/02a05_cutadapt.png" style="width: 600px;" /></a>
<p>Let’s now scroll to the bottom of the window and set the <cite>discard_untrimmed_reads</cite> parameter to False, so that <em>CutAdapt</em> will not discard reads that were previously trimmed properly.
Usually we do want to discard any reads that didn’t contain adapter sequences, but in our case, most of the reads were already trimmed, so we don’t want to throw them away.</p>
<a class="reference internal image-reference" href="_images/02a06_cutadapt.png"><img alt="adapter trimming - do not discard untrimmed reads" src="_images/02a06_cutadapt.png" style="width: 600px;" /></a>
<p>Once we are happy with the trimming parameters, we can click on the “Start CutAdapt” button at the bottom of the window.
A loading screen will now appear, and the trimmed FASTQ files, as well as the trimming logs, will be saved to our output folder.</p>
<a class="reference internal image-reference" href="_images/02a07_cutadapt.png"><img alt="adapter trimming - loading screen" src="_images/02a07_cutadapt.png" style="width: 600px;" /></a>
<p>Let’s now apply the same trimming procedure to our paired end samples.
Open the Paired End adapter trimming from the FASTQ menu:</p>
<a class="reference internal image-reference" href="_images/02a08_cutadapt.png"><img alt="paired-end adapter trimming" src="_images/02a08_cutadapt.png" style="width: 600px;" /></a>
<p>First, we will set the output folder for our trimmed FASTQ files.
Then, we can set the adapter sequences: for read#1 we will use the TruSeq Read 1 adapter “AGATCGGAAGAGCACACGTCTGAACTCCAGTCA”, and for read#2 we will use the TruSeq Read 2 adapter “AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTG”.
We will leave the quality-trimming parameters the same as last time, and once again set the <cite>discard_untrimmed_reads</cite> parameter to False:</p>
<a class="reference internal image-reference" href="_images/02a09_cutadapt.png"><img alt="paired-end adapter trimming - quality trimming parameters" src="_images/02a09_cutadapt.png" style="width: 600px;" /></a>
<p>Finally, we can choose the files we want to trim. Since we are doing paired-end trimming, we have to make sure the files are properly paired.
Therefore, we will pick the read#1 and read#2 files separately by clicking on the “Add files” buttons:</p>
<a class="reference internal image-reference" href="_images/02a10_cutadapt.png"><img alt="paired-end adapter trimming - add files" src="_images/02a10_cutadapt.png" style="width: 600px;" /></a>
<p>After adding the files, we make sure the orders of the two file lists match, so that the files are properly paired:</p>
<a class="reference internal image-reference" href="_images/02a11_cutadapt.png"><img alt="paired-end adapter trimming - match-up file pairs" src="_images/02a11_cutadapt.png" style="width: 600px;" /></a>
<p>You can use the buttons at the bottom of each list to reorder the list, or remove some of the files you added:</p>
<a class="reference internal image-reference" href="_images/02a12_cutadapt.png"><img alt="paired-end adapter trimming - reorder the file lists" src="_images/02a12_cutadapt.png" style="width: 600px;" /></a>
<p>Once we are done setting up the trimming parameters, we can scroll down and click on the “Start CutAdapt” button.</p>
</section>
<section id="quantify-gene-expression-with-kallisto">
<h4>Quantify gene expression with <em>kallisto</em><a class="headerlink" href="#quantify-gene-expression-with-kallisto" title="Permalink to this heading"></a></h4>
<p>Now that out data has been pre-processed, we can proceed with quantifying the expression of each gene.
For this purpose we will use <a class="reference external" href="https://pachterlab.github.io/kallisto/about">kallisto</a> - a program for rapid quantification of transcript abundances.
<em>kallisto</em> uses a method called <strong>pseudoalignment</strong> to directly estimate the number of reads originating from each transcript in our target transcriptome, without actually aligning the reads to an exact physical position on the genome/transcriptome.
This means that alignment with <em>kallisto</em> is extremely quick, while providing results that are as good as traditional alignment methods.
Before proceeding with this step, make sure you have <a class="reference external" href="https://pachterlab.github.io/kallisto/download">installed kallisto</a> on your computer.</p>
<p>To run this analysis, in addition to our processed FASTQ files, we will need a transcriptome indexed by <em>kallisto</em>, and a matching GTF file describing the names of the transcripts and genes in the transcriptome.
<em>kallisto</em> provides pre-indexed transcriptomes and their matching GTF files for most common model organisms, which can be downloaded <a class="reference external" href="https://github.com/pachterlab/kallisto-transcriptome-indices/releases">here</a>.
Our data was sequenced from <em>Caenorhabditis elegans</em> nematodes, so we will download the <em>C. elegans</em> index and GTF files from the above link.
If your experiment requires a different transcriptome that is not available above, you can index any FASTA file through <em>RNAlysis</em>, by entering the “FASTQ” menu -&gt; “RNA sequencing quantification” -&gt; “Create kallisto index…”.</p>
<p>As we did earlier with adapter trimming, let’s begin quantification with our single-end osmotic stress samples. Open the “FASTQ” menu, and under “RNA sequencing quantification” select “Single-end RNA sequencing quantification…”:</p>
<a class="reference internal image-reference" href="_images/02b01_kallisto.png"><img alt="FASTQ menu - *kallisto* quantification" src="_images/02b01_kallisto.png" style="width: 600px;" /></a>
<p>Like before, in the new window that opened we can set the input folder containing our trimmed FASTQ files, and the output folder.
We should also set the path to our transcriptome index file, GTF file, and the folder in which you installed <em>kallisto</em> (unless you have also added it to your system’s PATH):</p>
<a class="reference internal image-reference" href="_images/02b02_kallisto.png"><img alt="single-end read quantification" src="_images/02b02_kallisto.png" style="width: 600px;" /></a>
<p>Since our data were sequenced with single-end reads, <em>kallisto</em> cannot estimate the size of the fragments in the sequencing run, so we will have to supply an estimate of the average fragment length and the standard deviaton. Let’s set them to 200 and 20 accordingly:</p>
<a class="reference internal image-reference" href="_images/02b03_kallisto.png"><img alt="single-end read quantification - fragment length" src="_images/02b03_kallisto.png" style="width: 600px;" /></a>
<p>We can optionally use the <cite>new_sample_names</cite> parameter to give our samples new, more readable names. For example:</p>
<a class="reference internal image-reference" href="_images/02b04_kallisto.png"><img alt="single-end read quantification - new sample names" src="_images/02b04_kallisto.png" style="width: 600px;" /></a>
<p>Once we are done setting up the quantification parameters, we can scroll down and click on the “Start kallisto quantify” button, then wait for the analysis to finish.
In the output folder, we can find the results of kallisto quantification for each of the individual FASTQ  files, each in its own sub-folder.
Alongside these files, we can find three CSV files (comma-separated values, can be opened with Microsoft Excel or with RNAlysis): a per-transcript count estimate table, a per-transcript TPM (transcripts per million) estimate table, and a per-gene scaled output table.</p>
<p>The per-gene scaled output table is generated using the <em>scaledTPM</em> method (scaling the TPM estimates up to the library size) as described by <a class="reference external" href="https://doi.org/10.12688/f1000research.7563.2">Soneson et al 2015</a> and used in the <a class="reference external" href="https://ycl6.gitbook.io/guide-to-rna-seq-analysis/differential-expression-analysis/tximport#scaling">tximport</a> R package.
This table format is considered un-normalized for library size, and can therefore be used directly by count-based statistical inference tools, such as DESeq2, for differential expression analysis later on.
<em>RNAlysis</em> will load this table automatically once the analysis is finished, and we can see it in a new tab of the main window:</p>
<a class="reference internal image-reference" href="_images/02b05_kallisto.png"><img alt="quantification output table - osmotic stress" src="_images/02b05_kallisto.png" style="width: 600px;" /></a>
<p>At the bottom of the window we can also see a log box - this is where <em>RNAlysis</em> will display notes and warnings regarding our analysis.</p>
<p>If you wanted to analyze differential expression for transcripts instead of genes, you could use the per-transcript count estimate table that is located in the output folder.</p>
<p>Let’s now apply the same quantification procedure to our paired-end starvation samples.
Open the Paired end quantification window from the FASTQ menu:</p>
<a class="reference internal image-reference" href="_images/02b06_kallisto.png"><img alt="paired-end read quantification" src="_images/02b06_kallisto.png" style="width: 600px;" /></a>
<p>Like before, we will set the path to our output folder, transcriptome index file, GTF file, and the folder in which you installed <em>kallisto</em> (unless you have also added it to your system’s PATH).
In addition, since these sequencing samples are Stranded, we will set the <cite>stranded</cite> parameter to “forward”:</p>
<a class="reference internal image-reference" href="_images/02b07_kallisto.png"><img alt="paired-end read quantification - set path parameters" src="_images/02b07_kallisto.png" style="width: 600px;" /></a>
<p>Next, we can choose the files we want to quantify.
Like we did when trimming paired-end adapters, we will pick the read#1 and read#2 files separately by clicking on the “Add files” buttons underneath each list, and then sort the files so the pairs match up:</p>
<a class="reference internal image-reference" href="_images/02b08_kallisto.png"><img alt="paired-end read quantification - add and sort files" src="_images/02b08_kallisto.png" style="width: 600px;" /></a>
<p>Finally, we can optionally use the <cite>new_sample_names</cite> parameter to give our sample pairs new, more readable names. For example:</p>
<a class="reference internal image-reference" href="_images/02b09_kallisto.png"><img alt="paired-end read quantification - new sample names" src="_images/02b09_kallisto.png" style="width: 600px;" /></a>
<p>We can now scroll to the bottom of the window and click on the “Start kallisto quantify” button, and wait for the analysis to finish.</p>
<p>Like with the single-end reads analysis, when the analysis is done, we will find the three summarized tables, as well as a subfolder with kallisto output for each pair of FASTQ files.</p>
<a class="reference internal image-reference" href="_images/02b10_kallisto.png"><img alt="paired-end quantification output table - starvation" src="_images/02b10_kallisto.png" style="width: 600px;" /></a>
<p>Last but not least, we can quantify our paired-end heat-shock data. The procedure follows the same principle of the starvation samples, with one difference - the heat shock reads are supplied in the reverse orientation, so we should set the <cite>stranded</cite> parameter to ‘reverse’.
This is what our results should look like:</p>
<a class="reference internal image-reference" href="_images/02b11_kallisto.png"><img alt="paired-end quantification output table - heat shock" src="_images/02b11_kallisto.png" style="width: 600px;" /></a>
<p>Let’s rename the tables to reflect the names of the experiments, and proceed to differential expression analysis.</p>
</section>
<section id="run-differential-expression-analysis-with-deseq2">
<h4>Run differential expression analysis with <em>DESeq2</em><a class="headerlink" href="#run-differential-expression-analysis-with-deseq2" title="Permalink to this heading"></a></h4>
<p>We can now start Differential Expression analysis.
For this purpose we will use <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a> - an R package for differential expression analysis.
Before proceeding with this step, make sure you have <a class="reference external" href="https://cran.r-project.org/bin/">installed R</a> on your computer.
You don’t have to install DESeq2 on your own - <em>RNAlysis</em> can install it for you, as long as you have installed the R language on your computer already.</p>
<p>To open the Differential Expression window, choose an <em>RNAlysis</em> tab with one of the scaled count tables, click on the “General” tab, and from the drop-down menu below select “Run DESeq2 differential expression”:</p>
<a class="reference internal image-reference" href="_images/02c01_deseq2.png"><img alt="Open the differential expression window" src="_images/02c01_deseq2.png" style="width: 600px;" /></a>
<p>The Differential Expression window should now open. On the left side of the window, set the path of your R installation (or keep it on ‘auto’ if you have previously added R to your computer’s PATH).</p>
<a class="reference internal image-reference" href="_images/02c02_deseq2.png"><img alt="Differential expression" src="_images/02c02_deseq2.png" style="width: 600px;" /></a>
<p>Next, we need to define a <strong>design matrix</strong> for each of our count tables.
The first column of the design matrix should contain the names of the samples in the count table.
Each other column should contain a variable to be added to the experimental design formula of the dataset. For example: experimental condition, genotype, or biological replicate/batch.
For example, the design matrix for our osmotic stress dataset would look like this:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 26%" />
<col style="width: 44%" />
<col style="width: 30%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>condition</p></th>
<th class="head"><p>batch</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Ctrl1</p></td>
<td><p>Ctrl</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-odd"><td><p>Ctrl2</p></td>
<td><p>Ctrl</p></td>
<td><p>B</p></td>
</tr>
<tr class="row-even"><td><p>Ctrl3</p></td>
<td><p>Ctrl</p></td>
<td><p>C</p></td>
</tr>
<tr class="row-odd"><td><p>Osm1</p></td>
<td><p>Osm</p></td>
<td><p>A</p></td>
</tr>
<tr class="row-even"><td><p>Osm2</p></td>
<td><p>Osm</p></td>
<td><p>B</p></td>
</tr>
<tr class="row-odd"><td><p>Osm3</p></td>
<td><p>Osm</p></td>
<td><p>C</p></td>
</tr>
</tbody>
</table>
<p>You can create your design matrix in a program like Microsoft Excel or Google Sheets, and then save it as a CSV or TSV file.</p>
<p>Once you have prepared your design matrix, choose that file from the DESeq2 window and click on the “Load design matrix” button:</p>
<a class="reference internal image-reference" href="_images/02c03_deseq2.png"><img alt="Differential expression - load sample table" src="_images/02c03_deseq2.png" style="width: 600px;" /></a>
<p>The right side of the window will now update, allowing you to choose which pairwise comparisons you want to run, based on your design matrix.
You can make as many pairwise comparisons as you want, each comparing two levels of one of the variables in the design matrix.
In our case, we are only interested in one comparison - osmotic stress VS control.
Note that the order of conditions in the comparison matters - the first condition will be the numerator in the comparison, and the second condition will be the denominator.</p>
<a class="reference internal image-reference" href="_images/02c04_deseq2.png"><img alt="Differential expression - choose pairwise comparisons" src="_images/02c04_deseq2.png" style="width: 600px;" /></a>
<p>After picking the comparisons you want to run, click on the “Start DESeq2”.</p>
<p>When the analysis ends, a dialog box will pop up, prompting you to choose which differential expression tables do you want to load into <em>RNAlysis</em>:</p>
<a class="reference internal image-reference" href="_images/02c05_deseq2.png"><img alt="Differential expression - choose tables to load" src="_images/02c05_deseq2.png" style="width: 600px;" /></a>
<p>After choosing to load the table, it will open in a new tab in <em>RNAlysis</em>:</p>
<a class="reference internal image-reference" href="_images/02c06_deseq2.png"><img alt="Differential expression - output table" src="_images/02c06_deseq2.png" style="width: 600px;" /></a>
<p>We can now repeat the same procedure with the other two count tables, and proceed with analyzing the differential expression tables.</p>
</section>
</section>
<section id="data-filtering-and-visualization-with-customized-pipelines">
<h3>Data filtering and visualization with customized Pipelines<a class="headerlink" href="#data-filtering-and-visualization-with-customized-pipelines" title="Permalink to this heading"></a></h3>
<p>Since we ran the differential expression analysis through <em>RNAlysis</em>, the differential expression tables were loaded into the program automatically.
Therefore, we can start analyzing the data straight away!</p>
<p>For each of the differential expression tables we generated, we would like to generate a volcano plot, to filter out genes which are not differentially expressed, and then split the differentially expressed genes into an ‘upregulated’ group and ‘downregulated’ group.</p>
<p>We could apply each of those operations to our data tables one by one, but that would take a long time, and there’s a decent chance we’ll make a mistake along the way.
Therefore, we will instead create a customized Pipeline containing all of those functions and their respective parameters, and then apply this Pipeline to all three tables at once.</p>
<section id="create-a-customized-pipeline">
<h4>Create a customized Pipeline<a class="headerlink" href="#create-a-customized-pipeline" title="Permalink to this heading"></a></h4>
<p>To create a Pipeline, let’s open the Pipelines menu and click on “New Pipeline”:</p>
<a class="reference internal image-reference" href="_images/02d01_pipeline.png"><img alt="Pipeline menu - new Pipeline" src="_images/02d01_pipeline.png" style="width: 600px;" /></a>
<p>In the new window that opened, we can name our Pipeline, and choose the type of table we want to apply the Pipeline to. Let’s select “differential expression” from the drop-down menu:</p>
<a class="reference internal image-reference" href="_images/02d02_pipeline.png"><img alt="Create Pipeline - choose table type" src="_images/02d02_pipeline.png" style="width: 600px;" /></a>
<p>After choosing a name and table type, we can click on the “Start” button.
The window will now update to show a preview of our new (empty) Pipeline:</p>
<a class="reference internal image-reference" href="_images/02d03_pipeline.png"><img alt="Create Pipeline - preview" src="_images/02d03_pipeline.png" style="width: 600px;" /></a>
<p>At this point, we can start adding functions to the Pipeline.
Adding functions to a Pipeline works very similarly to applying functions to tables, as we did in Analysis #1.
First, let’s click on the “Visualize” tab, and choose the “Volcano plot” function:</p>
<a class="reference internal image-reference" href="_images/02d04_pipeline.png"><img alt="Create Pipeline - Volcano plot" src="_images/02d04_pipeline.png" style="width: 600px;" /></a>
<p>A volcano plot can give us an overview of the results of a differential expression analysis - it will show us the distribution of adjusted p-values and log2 fold change values for our genes, and highlight the significantly up- and down-regulated genes.</p>
<p>Let’s click on the “Add to Pipeline” button:</p>
<a class="reference internal image-reference" href="_images/02d05_pipeline.png"><img alt="Create Pipeline - adding functions to Pipeline" src="_images/02d05_pipeline.png" style="width: 600px;" /></a>
<p>We can see that the Pipeline overview has updated to show the function we added to it, and this is displayed in the log textbox as well.</p>
<p>Let’s now add a filtering function to our Pipeline. Click on the “Filtering” tab and choose the “Filter by statistical significance” function from the drop-down menu:</p>
<a class="reference internal image-reference" href="_images/02d06_pipeline.png"><img alt="Create Pipeline - filter by statistical significance" src="_images/02d06_pipeline.png" style="width: 600px;" /></a>
<p>You may notice that the <cite>inplace</cite> parameter that is usually present for filtering functions is missing.
This is because when we apply the Pipeline later on, we can choose whether it will be applied inplace or not.</p>
<p>Finally, we will select the ‘Split by log2 fold-change direction’ function from the drop-down menu and add it to the Pipeline as well.
We can now click on the “Save Pipeline” button to save the Pipeline we created for later use. We can also click on the “Export Pipeline” button to export the Pipeline to a file, so that we can use it in future sessions, or share it with others.</p>
<a class="reference internal image-reference" href="_images/02d07_pipeline.png"><img alt="Create Pipeline - save and export" src="_images/02d07_pipeline.png" style="width: 600px;" /></a>
<p>We can now close the Pipeline window and resume our analysis.</p>
</section>
<section id="apply-the-pipeline-to-our-datasets">
<h4>Apply the Pipeline to our datasets<a class="headerlink" href="#apply-the-pipeline-to-our-datasets" title="Permalink to this heading"></a></h4>
<p>Now that the hard part is done, we can apply the Pipeline to our differential expression tables.
Open the “Pipelines” menu, and then under the “Apply Pipeline” menu, choose the Pipeline we created:</p>
<a class="reference internal image-reference" href="_images/02d08_pipeline.png"><img alt="Apply Pipeline" src="_images/02d08_pipeline.png" style="width: 600px;" /></a>
<p>Now, you will be prompted on whether you want to apply this Pipeline inplace or not. Let’s choose “no”, so that we keep a copy of our original tables.</p>
<a class="reference internal image-reference" href="_images/02d09_pipeline.png"><img alt="Apply Pipeline - not inplace" src="_images/02d09_pipeline.png" style="width: 300px;" /></a>
<p>Finally, you will be prompted to choose the tables to apply your Pipeline to. Let’s choose all three of our differential expression tables.</p>
<a class="reference internal image-reference" href="_images/02d10_pipeline.png"><img alt="Apply Pipeline - choose tables" src="_images/02d10_pipeline.png" style="width: 300px;" /></a>
<p>We can now examine the output of our Pipeline - three volcano plots, and six new tables - the significantly up/down-regulated genes from each differential expression table.</p>
<a class="reference internal image-reference" href="_images/02d11_pipeline.png"><img alt="Examine the Pipeline outputs" src="_images/02d11_pipeline.png" style="width: 600px;" /></a>
<a class="reference internal image-reference" href="_images/02d12_pipeline.png"><img alt="Examine the Pipeline outputs - volcano plot" src="_images/02d12_pipeline.png" style="width: 600px;" /></a>
<p>At this point, we could optionally rename our tables to make it easier to differentiate them later on.</p>
</section>
</section>
<section id="visualizing-and-extracting-gene-set-interesctions">
<h3>Visualizing and extracting gene set interesctions<a class="headerlink" href="#visualizing-and-extracting-gene-set-interesctions" title="Permalink to this heading"></a></h3>
<p>We want to know how much do the three stress conditions have in common. One easy way to do this is to visualize the intersections between the differentially expressed genes in the three stress conditions.</p>
<section id="create-a-venn-diagram">
<h4>Create a Venn diagram<a class="headerlink" href="#create-a-venn-diagram" title="Permalink to this heading"></a></h4>
<p>Open the “Gene sets” menu, and click on “Visualize Gene Sets…”:</p>
<a class="reference internal image-reference" href="_images/02e01_gene_sets.png"><img alt="Gene sets menu - Visualize Gene Sets" src="_images/02e01_gene_sets.png" style="width: 600px;" /></a>
<p>A new window will open.
On the left side of the window, we can choose which data tables/gene sets we want to visualize.
Let’s pick the three tables that contain significantly downregulated genes:</p>
<a class="reference internal image-reference" href="_images/02e02_gene_sets.png"><img alt="Visualize Gene Sets - select gene sets" src="_images/02e02_gene_sets.png" style="width: 600px;" /></a>
<p>Next, we can choose the type of graph we want to generate. <em>RNAlysis</em> supports Venn Diagrams for 2-3 gene sets, and <a class="reference external" href="https://doi.org/10.1109%2FTVCG.2014.2346248">UpSet plots</a> for any number of gene sets.
Let’s choose a Venn Diagram.
The window will now update and display various parameters to modify our graph, and a preview of the graph on the right:</p>
<a class="reference internal image-reference" href="_images/02e03_gene_sets.png"><img alt="Visualize Gene Sets - plot preview" src="_images/02e03_gene_sets.png" style="width: 600px;" /></a>
<p>We can change these plotting parameters to modify our graph - for example, to change the colors of the Venn circles, the title, or set whether or not our plot will be proportional to the size of the sets and their intersections.
Once we are happy, we can click on the “Generate graph” button to create a big version of our graph that we can export and share.</p>
<a class="reference internal image-reference" href="_images/02e04_gene_sets.png"><img alt="Intersection of significantly downregulated genes" src="_images/02e04_gene_sets.png" style="width: 600px;" /></a>
<p>From the looks of it, there is a rather large overlap between the three sets.
Let’s generate a similar graph for the tables containing the significantly upregulated genes:</p>
<a class="reference internal image-reference" href="_images/02e05_gene_sets.png"><img alt="Intersection of significantly upregulated genes" src="_images/02e05_gene_sets.png" style="width: 600px;" /></a>
</section>
<section id="extract-the-gene-subsets-we-are-interested-in">
<h4>Extract the gene subsets we are interested in<a class="headerlink" href="#extract-the-gene-subsets-we-are-interested-in" title="Permalink to this heading"></a></h4>
<p>Now that we have seen the intersection between the three sets, we want to get the actual list of genes that are significantly downreglated under all stress conditions.
To do that, let’s open the “Gene sets” menu once again, and this time click on “Set Operations…”:</p>
<a class="reference internal image-reference" href="_images/02f01_set_ops.png"><img alt="Gene sets menu - Set Operations" src="_images/02f01_set_ops.png" style="width: 600px;" /></a>
<p>A new window will open:</p>
<a class="reference internal image-reference" href="_images/02f02_set_ops.png"><img alt="Set Operations window" src="_images/02f02_set_ops.png" style="width: 600px;" /></a>
<p>On the left side of the window, we can choose which data tables/gene sets we want to intersect.
Let’s pick the three tables that contain significantly downregulated genes:</p>
<a class="reference internal image-reference" href="_images/02f03_set_ops.png"><img alt="Set Operations - select gene sets" src="_images/02f03_set_ops.png" style="width: 600px;" /></a>
<p>We will now see a simplified Venn Diagram depicting our three gene sets. We can now proceed to extract the subset we are interested in.
We can do this by choosing the “Intersection” set operation from the multiple choice list.</p>
<p>A drop-down menu will now appear, prompting us to pick the primary gene set in this operation.
Like filtering operations, some set operations can be applied ‘inplace’, filtering down an existing table instead of returning a new copy of the filtered gene set.
For this purpose, <em>RNAlysis</em> wants to know what is our “primary set” for this set operation - meaning, if we were to apply it inplace, which table should it be applied to?
Our choice will only matter if we apply this set operation inplace - but for our example we won’t be applying the operation inplace, so it doesn’t matter what we choose.</p>
<a class="reference internal image-reference" href="_images/02f04_set_ops.png"><img alt="Set Operations - intersection" src="_images/02f04_set_ops.png" style="width: 600px;" /></a>
<p>Once we pick a primary set, the subset representing the intersection between all three gene sets will now highlight.
<em>RNAlysis</em> supports a large number of set operations. You could try using a different set operation on the data - for example, using ‘majority-vote intersection’ to select all genes that were significantly downregulated in <em>at least</em> two stress conditions.</p>
<a class="reference internal image-reference" href="_images/02f05_set_ops.png"><img alt="Set Operations - operation preview" src="_images/02f05_set_ops.png" style="width: 600px;" /></a>
<p>Finally, let’s click on the “Apply” button to extract our gene set of interest. It opens in a new tab:</p>
<a class="reference internal image-reference" href="_images/02f06_set_ops.png"><img alt="Set Operations - gene set output" src="_images/02f06_set_ops.png" style="width: 600px;" /></a>
<p>Let’s rename it so we remember what it contains:</p>
<a class="reference internal image-reference" href="_images/02f07_set_ops.png"><img alt="Rename gene set" src="_images/02f07_set_ops.png" style="width: 600px;" /></a>
<p>Let’s now open the Set Operations window once more to intersect the other set of tables, containing the significantly upregulated genes.
This time, let’s use a different method to extract our subset of interest - simply click on the middle subset in the interactive Venn Diagram!</p>
<a class="reference internal image-reference" href="_images/02f08_set_ops.png"><img alt="Set Operations - select subset" src="_images/02f08_set_ops.png" style="width: 600px;" /></a>
<p>We can use this interactive graph to select exactly the subsets we are interested in, and click the “Apply” button to extract them.
After renaming the set, it should look like this:</p>
<a class="reference internal image-reference" href="_images/02f09_set_ops.png"><img alt="Set Operations - custom operation output" src="_images/02f09_set_ops.png" style="width: 600px;" /></a>
</section>
</section>
<section id="id9">
<h3>Enrichment analysis<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h3>
<p>We now have two lists of genes we want to analyze - genes that are significantly downregulated in all three stress conditions, and genes that are significantly upregulated in all three stress conditions.
Let’s now proceed to enrichment analysis!</p>
<section id="define-a-background-set">
<h4>Define a background set<a class="headerlink" href="#define-a-background-set" title="Permalink to this heading"></a></h4>
<p>In order to run enrichment analysis, we will need an appropriate background set.
One good example would be the set of all genes that are not lowly-expressed in at least one sequencing sample.
We can use the process we learned earlier to generate an appropriate background set:</p>
<ol class="arabic simple">
<li><p>Create a Pipeline for count tables, that will normalize our count tables and then filter out lowly-expressed genes (for example, normalize with Relative Log Expression, and filter out genes that don’t have at least 10 normalized reads in at least one sample)</p></li>
<li><p>Apply the Pipeline to the three count matrices we generated earlier</p></li>
<li><p>Use the Set Operation window to calculate the Union of the three filtered count matrices</p></li>
</ol>
<p>To keep this tutorial short, we will not go through these individual steps. If you are not sure how to filter count tables for lowly-expressed genes, read Analysis #1 above!</p>
<p>This is what your output should look like:</p>
<a class="reference internal image-reference" href="_images/02g01_enrichment.png"><img alt="Enrichment analysis - background set" src="_images/02g01_enrichment.png" style="width: 600px;" /></a>
<p>Now that we have a test set and a background set, we can finally approach enrichment analysis.</p>
</section>
<section id="define-our-custom-enrichment-attributes">
<h4>Define our custom enrichment attributes<a class="headerlink" href="#define-our-custom-enrichment-attributes" title="Permalink to this heading"></a></h4>
<p>Let’s assume that we want to specifically find out whether genes downregulated in all stress conditions are enriched for epigenetic genes.
Unfortunately, the most common sources for functional gene annotation, Gene Ontology (GO) and KEGG, do not contain a single, well-annotated category of epigenetic genes in <em>Caenorhabditis elegans</em> nematodes.
Therefore, to be able to test this hypothesis, we will have to source our own annotations.</p>
<p>Luckily, when using <em>RNAlysis</em>, you can define your own customized attributes and test enrichment for those attributes. These could be anything - from general categories like “epigenetic genes” and “newly-evolved genes” to highly specific categories like “genes whose codon usage deviates from the norm”.
The annotations can be either categorial (yes/no - whether a gene belongs to the category or not), or non-categorical (any numerical value - distance to nearest paralog in base-pairs, expression level in a certain tissue, etc).</p>
<p>A legal <strong>Attribute Reference Table</strong> for <em>RNAlysis</em> should follow the following format:
Every row in the table represents a gene/genomic feature.
The first column of the table should contain the gene names/IDs. These should be the same type of gene IDs you use in your data tables.
If the gene IDs in your data tables do not match those of your Attribute Reference Table, <em>RNAlysis</em> offers a function that can convert gene IDs for your data tables. You can it under the “general” tab.</p>
<p>Every other column in the Attribute Reference Table represents a user-defined attribute/category (in our case, “epigenetic genes”).
In each cell, set the value to NaN if the gene in this row is negative for the attribute (in our example - not an epigenetic gene), and any other value if the gene in this row is positive for the attirbute.
As an example, see this mock Attribute Reference Table:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 24%" />
<col style="width: 30%" />
<col style="width: 21%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>feature_indices</p></th>
<th class="head"><p>Epigenetic genes</p></th>
<th class="head"><p>Has paralogs</p></th>
<th class="head"><p>gut expression</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>WBGene0000001</p></td>
<td><p>1</p></td>
<td><p>NaN</p></td>
<td><p>13.7</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene0000002</p></td>
<td><p>NaN</p></td>
<td><p>1</p></td>
<td><p>241</p></td>
</tr>
<tr class="row-even"><td><p>WBGene0000003</p></td>
<td><p>NaN</p></td>
<td><p>1</p></td>
<td><p>3.6</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene0000004</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>NaN</p></td>
</tr>
<tr class="row-even"><td><p>WBGene0000005</p></td>
<td><p>1</p></td>
<td><p>NaN</p></td>
<td><p>21.5</p></td>
</tr>
</tbody>
</table>
<p>For our analysis, we curated a list of epigenetic genes based on gene lists from the following publications: <a class="reference external" href="https://doi.org/10.1016/j.cell.2016.02.057">Houri et al 2014</a>, <a class="reference external" href="https://doi.org/10.1371/journal.pntd.0001176">Dalzell et al 2011</a>, <a class="reference external" href="https://doi.org/10.1016/j.cell.2006.09.033">Yigit et al 2006</a>, and based on existing GO annotations.
You can find the pre-made curated Attribute Reference Table <a class="reference external" href="https://raw.githubusercontent.com/GuyTeichman/RNAlysis/development/tests/test_files/attribute_reference_table.csv">here</a>.</p>
</section>
<section id="id11">
<h4>Running enrichment analysis<a class="headerlink" href="#id11" title="Permalink to this heading"></a></h4>
<p>To open the Enrichent Analysis window, open the ‘Gene sets’ menu and click “Enrichment Analysis”:</p>
<a class="reference internal image-reference" href="_images/02h01_enrichment.png"><img alt="Pick 'Enrichment analysis' from the 'Gene sets' menu" src="_images/02h01_enrichment.png" style="width: 600px;" /></a>
<p>We can choose our test and background sets (as explained in Analysis #1) from the two drop-down menus in the Enrichment window:</p>
<a class="reference internal image-reference" href="_images/02h02_enrichment.png"><img alt="Enrichment analysis - choose gene sets" src="_images/02h02_enrichment.png" style="width: 600px;" /></a>
<p>Next, we can choose the dataset we want to draw annotations from. In our case, we will choose Custom Dataset Categorical Attributes.</p>
<a class="reference internal image-reference" href="_images/02h03_enrichment.png"><img alt="Enrichment analysis - Choose enrichment dataset" src="_images/02h03_enrichment.png" style="width: 600px;" /></a>
<p>After picking the dataset, the window expanded to show us all of the parameters we can modify for our analysis:</p>
<a class="reference internal image-reference" href="_images/02h04_enrichment.png"><img alt="Enrichment analysis - set enrichment parameters" src="_images/02h04_enrichment.png" style="width: 600px;" /></a>
<p>Let’s set the statistical test to ‘randomization’, and then set the name of our attributes and the path of our Attribute Reference Table:</p>
<a class="reference internal image-reference" href="_images/02h05_enrichment.png"><img alt="Enrichment analysis - set the attributes to enrich for" src="_images/02h05_enrichment.png" style="width: 600px;" /></a>
<p>Scroll to the bottom of the window and click on the “run” button to run the analysis:</p>
<a class="reference internal image-reference" href="_images/02h06_enrichment.png"><img alt="Enrichment analysis - loading screen" src="_images/02h06_enrichment.png" style="width: 600px;" /></a>
<p>The enrichment window is going to minimize to allow you to read the log box on the main <em>RNAlysis</em> window, but you can enlarge it back if you want to look at the analysis parameters, or start a new analysis with the same parameters.</p>
<p>if you examine the log, you will notice that <em>RNAlysis</em> issued a warning about some of the genes in our background set not being annotated.
<em>RNAlysis</em> will automatically detect and remove from your background set any genes that have no annotations associated with them, since the presence of those genes can make your enrichment results appear inflated.</p>
<p>If the number of genes removed from our background set was large, we may have cause for concern about the validity of your analysis.
However, since the percentage of genes removed is very small (123 out of 18924 genes), we can proceed as usual.</p>
<p>Once the analysis is done, we will be able to observe our results in two formats.</p>
<p>The first is a tabular format, showing all of the statistically significant attributes we found (or the all of the tests attributes, if we set the <cite>return_nonsignificant</cite> parameter to True).
The table also includes the statistics for each attribute (number of genes in the test set, number of genes matching the attribute, expected number of genes to match the attribute, log2 fold enrichment score, p-value, and adjusted p-value).</p>
<a class="reference internal image-reference" href="_images/02h07_enrichment.png"><img alt="Enrichment results - results table" src="_images/02h07_enrichment.png" style="width: 600px;" /></a>
<p>The final output format is a bar plot depicting the log2 fold enrichment scores, as well as significance, of the attributes we tested enrichment for.</p>
<a class="reference internal image-reference" href="_images/02h08_enrichment.png"><img alt="Enrichment results - bat plot" src="_images/02h08_enrichment.png" style="width: 600px;" /></a>
</section>
</section>
</section>
<section id="final-words">
<h2>Final words<a class="headerlink" href="#final-words" title="Permalink to this heading"></a></h2>
<p>This concludes our A-to-Z tutorial of <em>RNAlysis</em>. This tutorial could not possibly encompass every single feature in <em>RNAlysis</em>, but at this point you should be familiar enough with <em>RNAlysis</em> to start analyzing your own data.
If you want a deeper dive into all of the features available in <em>RNAlysis</em>, or want to learn how to write code with <em>RNAlysis</em>, check out the <a class="reference external" href="https://guyteichman.github.io/RNAlysis/build/user_guide_gui.html">user guide</a>.
If you still have questions about using <em>RNAlysis</em>, feel free to drop them in the <a class="reference external" href="https://github.com/GuyTeichman/RNAlysis/discussions">discussions area</a>.</p>
<p><em>RNAlysis</em> keeps updating all the time, so check out the documentation from time to time to find out what’s new.</p>
<p>We hope you found this tutorial helpful, and wish you luck with your future bioinformatic endeavors!</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="quick_start.html" class="btn btn-neutral float-left" title="RNAlysis quick-start video guide" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="faq.html" class="btn btn-neutral float-right" title="Frequently asked questions" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>