<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>User guide - programmatic interface &mdash; RNAlysis 3.9.2 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/sphinx_highlight.js"></script>
        <script src="_static/copybutton.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Glossary - programmatic interface" href="glossary.html" />
    <link rel="prev" title="User guide - graphical interface" href="user_guide_gui.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html">
            
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">About RNAlysis:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">What is <em>RNAlysis?</em></a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#what-can-i-do-with-rnalysis">What can I do with <em>RNAlysis</em>?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-install-it">How do I install it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#how-do-i-use-it">How do I use it?</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#dependencies">Dependencies</a></li>
<li class="toctree-l1"><a class="reference internal" href="readme.html#credits">Credits</a></li>
<li class="toctree-l1"><a class="reference external" href="https://github.com/GuyTeichman/RNAlysis">GitHub page</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">How can I contribute to <em>RNAlysis</em>?</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Starting with RNAlysis:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quick_start.html"><em>RNAlysis</em> quick-start video guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">A-to-Z tutorial - example analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_guide_gui.html">User guide - graphical interface</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">User guide - programmatic interface</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-fastq-module"><em>RNAlysis</em> fastq module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#working-with-the-fastq-module">Working with the fastq module</a></li>
<li class="toctree-l3"><a class="reference internal" href="#adapter-trimming-with-cutadapt">Adapter trimming with CutAdapt</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#single-end-adapter-trimming">Single-end adapter trimming</a></li>
<li class="toctree-l4"><a class="reference internal" href="#paired-end-adapter-trimming">Paired-end adapter trimming</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#rna-sequencing-quantification-with-kallisto">RNA sequencing quantification with kallisto</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#building-or-acquiring-index-files">Building or acquiring index files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#single-end-read-quantification">Single-end read quantification</a></li>
<li class="toctree-l4"><a class="reference internal" href="#paired-end-read-quantification">Paired-end read quantification</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#read-alignment-with-bowtie2">Read alignment with bowtie2</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#id2">Building or acquiring index files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#single-end-read-alignment">Single-end read alignment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#paired-end-read-alignment">Paired-end read alignment</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#feature-counting-with-featurecounts">Feature counting with featureCounts</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#single-end-read-feature-counting">Single-end read feature counting</a></li>
<li class="toctree-l4"><a class="reference internal" href="#paired-end-read-feature-counting">Paired-end read feature counting</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-filtering-module"><em>RNAlysis</em> filtering module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#working-with-filter-objects">Working with Filter objects</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#initialize-a-filter-object">Initialize a Filter object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#view-a-filter-object">View a Filter object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-operations">Filtering operations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-set-operations-on-multiple-filter-objects">Performing set operations on multiple Filter objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#saving-filter-results">Saving Filter results</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-an-attribute-reference-table-for-filter-operations">Using an Attribute Reference Table for filter operations</a></li>
<li class="toctree-l4"><a class="reference internal" href="#using-a-biotype-reference-table-for-filter-operations">Using a Biotype Reference Table for filter operations</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-deseq2-output-files-with-filtering-deseqfilter">Filtering DESeq2 output files with filtering.DESeqFilter</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#loading-from-a-csv-file">Loading from a <cite>csv</cite> file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-operations-unique-to-deseqfilter">Filtering operations unique to DESeqFilter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-visualization-and-exploratory-data-analysis-with-deseqfilter">Data visualization and exploratory data analysis with DESeqFilter</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-count-matrices-with-filtering-countfilter">Filtering count matrices with filtering.CountFilter</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#generating-an-countfilter-object-from-a-folder-of-htseq-count-output-txt-files">Generating an CountFilter object from a folder of HTSeq-count output .txt files</a></li>
<li class="toctree-l4"><a class="reference internal" href="#loading-from-a-pre-made-csv-file">Loading from a pre-made <cite>csv</cite> file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#filtering-operations-unique-to-countfilter">Filtering operations unique to CountFilter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#normalizing-reads-with-countfilter">Normalizing reads with CountFilter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-clustering-with-countfilter">Data clustering with CountFilter</a></li>
<li class="toctree-l4"><a class="reference internal" href="#data-visualization-and-exploratory-data-analysis-with-countfilter">Data visualization and exploratory data analysis with CountFilter</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#filtering-fold-change-data-of-features-using-filtering-foldchangefilter">Filtering fold-change data of features using filtering.FoldChangeFilter</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#loading-fold-change-data-from-a-csv-file">Loading fold change data from a <cite>csv</cite> file</a></li>
<li class="toctree-l4"><a class="reference internal" href="#generating-fold-change-data-from-an-existing-countfilter-object">Generating fold change data from an existing CountFilter object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-randomization-tests-on-a-foldchangefilter-object">Performing randomization tests on a FoldChangeFilter object</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#sequentially-applying-filtering-operations-using-pipelines">Sequentially applying filtering operations using Pipelines</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#creating-a-new-pipeline">Creating a new Pipeline</a></li>
<li class="toctree-l4"><a class="reference internal" href="#applying-pipelines-to-filter-objects">Applying Pipelines to Filter objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#importing-a-pipeline">Importing a Pipeline</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-enrichment-module"><em>RNAlysis</em> enrichment module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#working-with-featureset-objects">Working with FeatureSet objects</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#initialize-an-featureset-object">Initialize an FeatureSet object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#go-enrichment">GO Enrichment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#kegg-pathways-enrichment">KEGG Pathways enrichment</a></li>
<li class="toctree-l4"><a class="reference internal" href="#enrichment-analysis-for-user-defined-attributes">Enrichment analysis for user-defined attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-enrichment-analysis-for-non-categorical-user-defined-attributes">Performing enrichment analysis for non-categorical user-defined attributes</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-set-operations-and-visualisation-on-multiple-featureset-objects">Performing set operations and visualisation on multiple FeatureSet objects</a></li>
<li class="toctree-l4"><a class="reference internal" href="#saving-indices-from-featureset-to-a-txt-file">Saving indices from FeatureSet to a .txt file</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#working-with-rankedset-objects">Working with RankedSet objects</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#initialize-an-rankedset-object">Initialize an RankedSet object</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-single-set-go-enrichment-analysis-without-a-background-set">Performing single-set GO Enrichment analysis without a background set</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-single-set-kegg-enrichment-analysis-without-a-background-set">Performing single-set KEGG Enrichment analysis without a background set</a></li>
<li class="toctree-l4"><a class="reference internal" href="#performing-single-set-enrichment-analysis-for-user-defined-attributes-without-a-background-set">Performing single-set enrichment analysis for user-defined attributes without a background set</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#visualizing-sets-intersections-and-enrichment">Visualizing sets, intersections, and enrichment</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#plotting-results-of-enrichment-analysis">Plotting results of enrichment analysis</a></li>
<li class="toctree-l4"><a class="reference internal" href="#plotting-venn-diagrams-and-upset-plots">Plotting Venn Diagrams and UpSet Plots</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#rnalysis-general-module"><em>RNAlysis</em> general module</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#set-and-load-a-reference-table">Set and load a Reference Table</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#what-is-an-attribute-reference-table">What is an Attribute Reference Table?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#what-is-a-biotype-reference-table">What is a Biotype Reference Table?</a></li>
<li class="toctree-l4"><a class="reference internal" href="#set-a-reference-table-as-default">Set a Reference Table as default</a></li>
<li class="toctree-l4"><a class="reference internal" href="#load-the-default-attribute-reference-table-path">Load the default Attribute Reference Table path</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#parse-c-elegans-gene-names-wbgene-indices-and-sequence-names-using-regular-expressions">Parse <em>C. elegans</em> gene names, WBGene indices and sequence names using regular expressions</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="glossary.html">Glossary - programmatic interface</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Change history:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="history.html">History</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Modules:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.fastq.html">rnalysis.fastq</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.filtering.html">rnalysis.filtering</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.enrichment.html">rnalysis.enrichment</a></li>
<li class="toctree-l1"><a class="reference internal" href="rnalysis.general.html">rnalysis.general</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Indices and tables:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="genindex.html">Index</a></li>
<li class="toctree-l1"><a class="reference internal" href="api_reference.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">RNAlysis</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">User guide - programmatic interface</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/user_guide.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="user-guide-programmatic-interface">
<h1>User guide - programmatic interface<a class="headerlink" href="#user-guide-programmatic-interface" title="Permalink to this heading"></a></h1>
<section id="rnalysis-fastq-module">
<h2><em>RNAlysis</em> fastq module<a class="headerlink" href="#rnalysis-fastq-module" title="Permalink to this heading"></a></h2>
<p>The <em>fastq</em> module provides a unified programmatic interface to external tools that process FASTQ files.
Those currently include the <em>CutAdapt</em> adapter-trimming tool, the <em>kallisto</em> RNA-sequencing quantification tool,
the <em>bowtie2</em> alignment tool, and the <em>featureCounts</em> feature counting tool.</p>
<section id="working-with-the-fastq-module">
<h3>Working with the fastq module<a class="headerlink" href="#working-with-the-fastq-module" title="Permalink to this heading"></a></h3>
<p>The fastq module contains functions that interface with the different external tools.
Most of these tools require different sets of parameters for single-end vs paired-end reads, and therefore all of the functions in this module are split into single-end functions and paired-end functions.</p>
<dl>
<dt>We will start by importing the fastq module::</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">fastq</span>
</pre></div>
</div>
</dd>
</dl>
<p>You can now access the different functions of the fastq module, as explained in the sections below.</p>
</section>
<section id="adapter-trimming-with-cutadapt">
<h3>Adapter trimming with CutAdapt<a class="headerlink" href="#adapter-trimming-with-cutadapt" title="Permalink to this heading"></a></h3>
<p>Cutadapt is a program that finds and removes adapter sequences, primers, poly-A tails and other types of unwanted sequencea from high-throughput sequencing reads.
Cutadapt can also trim bases based on their quality scores, high incidence of N (‘any’) bases, and perform basic filtering of reads.</p>
<section id="single-end-adapter-trimming">
<h4>Single-end adapter trimming<a class="headerlink" href="#single-end-adapter-trimming" title="Permalink to this heading"></a></h4>
<dl>
<dt>If you want to trim single-end reads, use the <cite>trim_adapters_single_end</cite> functions::</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fastq_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/fastq/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adapter</span> <span class="o">=</span> <span class="s1">&#39;ATGGCGTTGA&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trim_adapters_single_end</span><span class="p">(</span><span class="n">fastq_folder</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">adapter</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<p>This function takes two mandatory parameters:
* <cite>fastq_folder</cite> - the folder containing the FASTQ files you want to trim
* <cite>output_folder</cite> - the folder to which you want to write the trimmed FASTQ files</p>
<p>The other parameters allow you to specify the adapter/adapters you wish to trim according to their type (3’ adapters, 5’ adapters, or any position adapters), and to apply additional trimming/filtering actions.
You can read more about those parameters in the <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/guide.html">Cutadapt user guide</a>.</p>
</section>
<section id="paired-end-adapter-trimming">
<h4>Paired-end adapter trimming<a class="headerlink" href="#paired-end-adapter-trimming" title="Permalink to this heading"></a></h4>
<dl>
<dt>If you want to trim paired-end reads, use the <cite>trim_adapters_paired_end</cite> functions::</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">read1_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;path/to/sample1read1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample2read1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample3read1.fastq&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read2_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;path/to/sample1read2.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample2read2.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample3read2.fastq&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adapter1</span> <span class="o">=</span> <span class="s1">&#39;ATGGCGT&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">adapter2</span> <span class="o">=</span> <span class="s1">&#39;CGTTATC&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trim_adapters_paired_end</span><span class="p">(</span><span class="n">read1_files</span><span class="p">,</span> <span class="n">read2_files</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">adapter1</span><span class="p">,</span> <span class="n">adapter2</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<p>This function takes three mandatory parameters:
* <cite>r1_files</cite> - list of paths the read#1 FASTQ files you want to trim
* <cite>r2_files</cite> - list of paths the read#2 FASTQ files you want to trim. The order of these should match the order of <cite>r1_files</cite>, so that they pair up.
* <cite>output_folder</cite> - the folder to which you want to write the trimmed FASTQ files</p>
<p>The other parameters allow you to specify the adapter/adapters you wish to trim according to their type (3’ adapters, 5’ adapters, or any position adapters), and to apply additional trimming/filtering actions.
Note that in paired-end mode, you can (and usually would) specify a different adapter for the read#1 files and the read#2 files.
You can read more about those parameters in the <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/guide.html">Cutadapt user guide</a>.</p>
</section>
</section>
<section id="rna-sequencing-quantification-with-kallisto">
<h3>RNA sequencing quantification with kallisto<a class="headerlink" href="#rna-sequencing-quantification-with-kallisto" title="Permalink to this heading"></a></h3>
<p><a class="reference external" href="https://pachterlab.github.io/kallisto/">kallisto</a> is a program for quantifying abundances of transcripts from bulk and single-cell RNA sequencing data.
Kallisto uses an approach known as <strong>pseudoalignment</strong> - identifying the transcripts from which the reads could have originated, instead of trying to pinpoint exactly how the sequences of the reads and transcripts align.
This allows kallisto to run incredibly quickly, but still produce accurate and valid results.
You can read more about it in the <a class="reference external" href="https://doi.org/10.1038/nbt.3519">kallisto publication</a>.</p>
<section id="building-or-acquiring-index-files">
<h4>Building or acquiring index files<a class="headerlink" href="#building-or-acquiring-index-files" title="Permalink to this heading"></a></h4>
<p>To quantify expression using <em>kallisto</em>, in addition to processed FASTQ files, you will need a transcriptome indexed by <em>kallisto</em>, and a matching GTF file describing the names of the transcripts and genes in the transcriptome.
kallisto provides pre-indexed transcriptomes and their matching GTF files for most common model organisms, which can be downloaded <a class="reference external" href="https://github.com/pachterlab/kallisto-transcriptome-indices/releases">here</a>.
If your experiment requires a different transcriptome that is not available above, you can index any FASTA file through <em>RNAlysis</em>, by using the function <cite>kallisto_create_index</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">transcriptome</span> <span class="o">=</span> <span class="s1">&#39;path/to/transcriptome.fasta&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fastq</span><span class="o">.</span><span class="n">kallisto_create_index</span><span class="p">(</span><span class="n">transcriptome</span><span class="p">)</span>
</pre></div>
</div>
<p>The function takes one mandatory parameter - <cite>transcriptome_fasts</cite>. This should point to the path of a fasta file containing your target transcriptome.</p>
<p>Three additional optional paremeters are supported:
* <cite>kallisto_installation_folder</cite> - path to the installation folder of <em>kallisto</em> (or ‘auto’, if you have already added it to your PATH).
* <cite>kmer_length</cite> - length of the k-mer used to construct the index. Must be an odd integer &lt;= 31. The default k-mer length is optimized for usual NGS read lengths (50-100bp) - if your reads are much shorter you may get better mapping results if you lower the k-mer length.
* <cite>make_unique</cite> - if True, replaces repeated target names with unique names.</p>
</section>
<section id="single-end-read-quantification">
<h4>Single-end read quantification<a class="headerlink" href="#single-end-read-quantification" title="Permalink to this heading"></a></h4>
<p>Once you have an indexed transcriptome and trimmed FASTQ files, you can proceed with quantification.
If your FASTQ files contain single-end reads, you can use the <cite>kallisto_quantify_single_end</cite> function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fastq_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/fastq/files&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_file</span> <span class="o">=</span> <span class="s1">&#39;path/to/index.idx&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">avg_frag_len</span> <span class="o">=</span> <span class="mi">200</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">stdev_frag_len</span> <span class="o">=</span> <span class="mi">40</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kallisto_quantify_single_end</span><span class="p">(</span><span class="n">fastq_folder</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">index_file</span><span class="p">,</span> <span class="n">avg_frag_len</span><span class="p">,</span> <span class="n">stdev_frag_len</span><span class="p">)</span>
</pre></div>
</div>
<p>The function takes 5 mandatory parameters:
* <cite>fastq_folder</cite> - the folder containing the FASTQ files you want to quantify
* <cite>output_folder</cite> - the folder to which you want to write the output of the quantification
* <cite>index_file</cite> - path to your index file
* <cite>average_fragment_length</cite> - estimated average length of fragments produced by your library preperation protocol
* <cite>stdev_fragment_length</cite> - estimated standard deviation of the length of fragments produced by your library preperation protocol</p>
<p>The FASTQ files will be individually quantified and saved in the output folder, each in its own sub-folder.
Alongside these files, three .csv files will be saved: a per-transcript count estimate table,
a per-transcript TPM estimate table, and a per-gene scaled output table.</p>
<p>The per-gene scaled output table is generated using the <em>scaledTPM</em> method
(scaling the TPM estimates up to the library size) as described by
<a class="reference external" href="https://doi.org/10.12688/f1000research.7563.2">Soneson et al 2015</a> and used in the
<a class="reference external" href="https://ycl6.gitbook.io/guide-to-rna-seq-analysis/differential-expression-analysis/tximport#scaling">tximport</a>
R package. This table format is considered un-normalized for library size,
and can therefore be used directly by count-based statistical inference tools such as DESeq2.
<em>RNAlysis</em> will return this table once the analysis is finished.</p>
</section>
<section id="paired-end-read-quantification">
<h4>Paired-end read quantification<a class="headerlink" href="#paired-end-read-quantification" title="Permalink to this heading"></a></h4>
<dl>
<dt>If your FASTQ files contain paired-end reads (meaning, two FASTQ files for each sample - a ‘read#1’ file and ‘read#2’ file), you can quantify them using the <cite>kallisto_quantify_paired_end</cite> function::</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">read1_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;path/to/sample1read1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample2read1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample3read1.fastq&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read2_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;path/to/sample1read2.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample2read2.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample3read2.fastq&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_file</span> <span class="o">=</span> <span class="s1">&#39;path/to/index.idx&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kallisto_quantify_paired_end</span><span class="p">(</span><span class="n">read1_files</span><span class="p">,</span> <span class="n">read2_files</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">index_file</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<p>The function takes 4 mandatory parameters:
* <cite>r1_files</cite> - list of paths the read#1 FASTQ files you want to quantify
* <cite>r2_files</cite> - list of paths the read#2 FASTQ files you want to quantify. The order of these should match the order of <cite>r1_files</cite>, so that they pair up.
* <cite>output_folder</cite> - the folder to which you want to write the output of the quantification
* <cite>index_file</cite> - path to your index file</p>
<p>The FASTQ file-pairs will be individually quantified and saved in the output folder, each in its own sub-folder.
Alongside these files, three .csv files will be saved: a per-transcript count estimate table,
a per-transcript TPM estimate table, and a per-gene scaled output table.
You can read more about these outputs in the section above.</p>
</section>
</section>
<section id="read-alignment-with-bowtie2">
<h3>Read alignment with bowtie2<a class="headerlink" href="#read-alignment-with-bowtie2" title="Permalink to this heading"></a></h3>
<p><a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2">Bowtie2</a> is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences.</p>
<section id="id2">
<h4>Building or acquiring index files<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h4>
<p>To align reads using bowtie2, in addition to processed FASTQ files, you will need a reference genome indexed by bowtie2.
Bowtie2 provides pre-indexed genome files for most common model organisms.
These can be downloaded from  the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie2 website</a> (from menu on the right).
If your experiment requires a different transcriptome that is not available above, you can index any FASTA file (or files) through <em>RNAlysis</em>, by using the function <cite>bowtie2_create_index</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">genome</span> <span class="o">=</span> <span class="s1">&#39;path/to/genome/file.fasta&#39;</span> <span class="c1"># can also be a list of paths to FASTA files</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bowtie2_create_index</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">)</span>
</pre></div>
</div>
<p>This function takes two mandatory parameters:
* <cite>genome_fastas</cite> - path to FASTA file/files containing the reference sequences you wish to align to
* <cite>output_folder</cite> - the folder to which you want to write the genome index files</p>
</section>
<section id="single-end-read-alignment">
<h4>Single-end read alignment<a class="headerlink" href="#single-end-read-alignment" title="Permalink to this heading"></a></h4>
<p>Once you have an index and trimmed FASTQ files, you can proceed with alignment.
If your FASTQ files contain single-end reads, you can align them using the <cite>bowtie2_align_single_end</cite> function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">fastq_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/fastq/files&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_file</span> <span class="o">=</span> <span class="s1">&#39;path/to/index.1.bt2&#39;</span> <span class="c1"># can be any of the index files</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bowtie2_align_single_end</span><span class="p">(</span><span class="n">fastq_folder</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">index_file</span><span class="p">)</span>
</pre></div>
</div>
<p>The function takes 3 mandatory parameters:
* <cite>fastq_folder</cite> - the folder containing the FASTQ files you want to align
* <cite>output_folder</cite> - the folder to which you want to write the output of the alignment (SAM files)
* <cite>index_file</cite> - path to your index file. This can be either the path to one of the multiple index files bowtie2 generates (doesn’t matter which), or a path to the prefix/name of the index files. The important thing is that all of the index files are in the same folder.</p>
<p>Bowtie2 supports many additional paramaters - you can read about them in the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie2 manual</a>.</p>
</section>
<section id="paired-end-read-alignment">
<h4>Paired-end read alignment<a class="headerlink" href="#paired-end-read-alignment" title="Permalink to this heading"></a></h4>
<p>Once you have an index and trimmed FASTQ files, you can proceed with alignment.
If your FASTQ files contain paired-end reads, you can align them using the <cite>bowtie2_align_paired_end</cite> function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">read1_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;path/to/sample1read1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample2read1.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample3read1.fastq&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">read2_files</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;path/to/sample1read2.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample2read2.fastq&#39;</span><span class="p">,</span> <span class="s1">&#39;path/to/sample3read2.fastq&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">index_file</span> <span class="o">=</span> <span class="s1">&#39;path/to/index.1.bt2&#39;</span> <span class="c1"># can be any of the index files</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">bowtie2_align_paired_end</span><span class="p">(</span><span class="n">read1_files</span><span class="p">,</span> <span class="n">read2_files</span><span class="p">,</span> <span class="n">output_folder</span><span class="p">,</span> <span class="n">index_file</span><span class="p">)</span>
</pre></div>
</div>
<p>The function takes 4 mandatory parameters:
* <cite>r1_files</cite> - list of paths the read#1 FASTQ files you want to align
* <cite>r2_files</cite> - list of paths the read#2 FASTQ files you want to align. The order of these should match the order of <cite>r1_files</cite>, so that they pair up.
* <cite>output_folder</cite> - the folder to which you want to write the output of the alignment (SAM files)
* <cite>index_file</cite> - path to your index file. This can be either the path to one of the multiple index files bowtie2 generates (doesn’t matter which), or a path to the prefix/name of the index files. The important thing is that all of the index files are in the same folder.</p>
<p>Bowtie2 supports many additional paramaters - you can read about them in the <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">bowtie2 manual</a>.</p>
</section>
</section>
<section id="feature-counting-with-featurecounts">
<h3>Feature counting with featureCounts<a class="headerlink" href="#feature-counting-with-featurecounts" title="Permalink to this heading"></a></h3>
<p><a class="reference external" href="https://subread.sourceforge.net/featureCounts.html">featureCounts</a>
is a highly efficient general-purpose read summarization program that counts mapped reads for genomic features such as genes, exons, promoter, gene bodies, genomic bins and chromosomal locations.
It can be used to count both RNA-seq and genomic DNA-seq reads.
featureCounts is part of the Subread package, which was also adapted into R as <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/Rsubread.html">RSubread</a>.</p>
<p>In order to use featureCounts through <em>RNAlysis</em>, you will need SAM/BAM files containing your aligned reads (such as those generated by <a class="reference external" href="https://bowtie-bio.sourceforge.net/bowtie2">bowtie2</a>), and a GTF annotation file that matches the genome you aligned your reads to.
Furthermore, you will need to define the folder that contains your SAM/BAM files, and the folder to which you want to write the outputs:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">input_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/input/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_folder</span> <span class="o">=</span> <span class="s1">&#39;path/to/output/folder&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gtf_file</span> <span class="o">=</span> <span class="s1">&#39;path/to/gtf_file.gtf&#39;</span>
</pre></div>
</div>
<p>Usually, you would want to assign reads to specific exons, and then sum up the exons into their corresponding genes, to get a final count of how many reads aligned to each gene.
If you want to count your reads differently, or your GTF file is non-standard, you can specify different values for the <cite>gtf_feature_type</cite> and <cite>gtf_attr_name</cite> parameters.</p>
<p>featureCounts supports many more options and parameters - you can read about them in the <em>RNAlysis</em> documentation and the <a class="reference external" href="https://rdrr.io/bioc/Rsubread/man/featureCounts.html">featureCounts documentation</a>.</p>
<section id="single-end-read-feature-counting">
<h4>Single-end read feature counting<a class="headerlink" href="#single-end-read-feature-counting" title="Permalink to this heading"></a></h4>
<dl>
<dt>To apply featureCounts to single-end reads, you can use the <cite>featurecounts_single_end</cite> function::</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">featurecounts_single_end</span><span class="p">(</span><span class="n">input_folder</span><span class="p">,</span><span class="n">output_folder</span><span class="p">,</span><span class="n">gtf_file</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
</section>
<section id="paired-end-read-feature-counting">
<h4>Paired-end read feature counting<a class="headerlink" href="#paired-end-read-feature-counting" title="Permalink to this heading"></a></h4>
<dl>
<dt>To apply featureCounts to paired-end reads, you can use the <cite>featurecounts_paired_end</cite> function::</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">featurecounts_paired_end</span><span class="p">(</span><span class="n">input_folder</span><span class="p">,</span><span class="n">output_folder</span><span class="p">,</span><span class="n">gtf_file</span><span class="p">)</span>
</pre></div>
</div>
</dd>
</dl>
<p>While the mandatory paramters of this function are identical to those of single-end read counting, paired-end feature counting supports a few additional optional parameters.
Those include minimum/maximum fragment lengths, and the option to exclude chimeric reads or pairs where only one read was aligned.</p>
</section>
</section>
</section>
<section id="rnalysis-filtering-module">
<h2><em>RNAlysis</em> filtering module<a class="headerlink" href="#rnalysis-filtering-module" title="Permalink to this heading"></a></h2>
<p>The filtering module (rnalysis.filtering) is built to allow rapid and easy-to-understand filtering of various forms of RNA sequencing data. The module also contains specific methods for visualization and clustering of data.</p>
<p>The filtering module is built around <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a>, which are containers for tabular sequencing data. You can use the different types of <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a> to apply filtering operations to various types of tabular data. You will learn more about <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a> in the next section.</p>
<section id="working-with-filter-objects">
<h3>Working with Filter objects<a class="headerlink" href="#working-with-filter-objects" title="Permalink to this heading"></a></h3>
<p>All <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a> (<a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a>, <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a>, <a class="reference internal" href="glossary.html#term-Filter"><span class="xref std std-term">Filter</span></a>, <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a>) work on the same principles,
and share many of the same functions and features. Each of them also has specific filtering, analysis and visualisation functions. In this section we will look into the general usage of <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a>.</p>
<section id="initialize-a-filter-object">
<h4>Initialize a Filter object<a class="headerlink" href="#initialize-a-filter-object" title="Permalink to this heading"></a></h4>
<p>We will start by importing the filtering module:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
</pre></div>
</div>
<p>We can now, for example, create a <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> object from a DESeq2 <cite>csv</cite> output file (see more details about <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> in sections below).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="view-a-filter-object">
<h4>View a Filter object<a class="headerlink" href="#view-a-filter-object" title="Permalink to this heading"></a></h4>
<p>In order to view a glimpse of the file we imported we can use the ‘head’ and ‘tail’ functions.
By default ‘head’ will show the first 5 rows of the file, and ‘tail’ will show the last 5 rows,
but you can specify a specific number of lines to show.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<span class="go">               baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000002  6820.755327        7.567762  ...   0.000000e+00   0.000000e+00</span>
<span class="go">WBGene00000003  3049.625670        9.138071  ...  4.660000e-302  4.280000e-298</span>
<span class="go">WBGene00000004  1432.911791        8.111737  ...  6.400000e-237  3.920000e-233</span>
<span class="go">WBGene00000005  4028.154186        6.534112  ...  1.700000e-228  7.800000e-225</span>
<span class="go">WBGene00000006  1230.585240        7.157428  ...  2.070000e-216  7.590000e-213</span>

<span class="go">[5 rows x 6 columns]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">8</span><span class="p">)</span>
<span class="go">               baseMean  log2FoldChange  ...         pvalue           padj</span>
<span class="go">WBGene00000022   365.813048        6.101303  ...  2.740000e-97  2.400000e-94</span>
<span class="go">WBGene00000023  3168.566714        3.906719  ...  1.600000e-93  1.340000e-90</span>
<span class="go">WBGene00000024   221.925724        4.801676  ...  1.230000e-84  9.820000e-82</span>
<span class="go">WBGene00000025  2236.185837        2.477374  ...  1.910000e-81  1.460000e-78</span>
<span class="go">WBGene00000026   343.648987       -4.037191  ...  2.320000e-75  1.700000e-72</span>
<span class="go">WBGene00000027   175.142856        6.352044  ...  1.580000e-74  1.120000e-71</span>
<span class="go">WBGene00000028   219.163200        3.913657  ...  3.420000e-72  2.320000e-69</span>
<span class="go">WBGene00000029  1066.242402       -2.811281  ...  1.420000e-70  9.290000e-68</span>

<span class="go">[8 rows x 6 columns]</span>
</pre></div>
</div>
<p>We can also see the total number of rows and columns by accessing the ‘shape’ attribute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(28, 6)</span>
</pre></div>
</div>
<p>meaning there are 28 rows and 6 columns in the file.</p>
</section>
<section id="filtering-operations">
<h4>Filtering operations<a class="headerlink" href="#filtering-operations" title="Permalink to this heading"></a></h4>
<p>Now we can start filtering the entries in the file according to parameters of our choosing.
Various filtering operations are applied directly to the <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a>. Those operations do not affect the original <cite>csv</cite> file, but its representation within the <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a>.
For example, we can the function ‘filter_percentile’ to remove all rows that are above the specified percentile (in our example, 75% percentile) in the specified column (in our example, ‘log2FoldChange’):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<p>If we now look at the shape of d, we will see that 5954 rows have been filtered out of the object, and we remain with 17781 rows.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(21, 6)</span>
</pre></div>
</div>
<p>By default, filtering operations on <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a> are performed in-place, meaning the original object is modified. However, we can save the results into a new <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a> and leave the current object unaffected by passing the argument ‘inplace=False’ to any filtering function within <em>RNAlysis</em>. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(28, 6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d_filtered</span> <span class="o">=</span> <span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d_filtered</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(21, 6)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(28, 6)</span>
</pre></div>
</div>
<p>In this case, the object ‘d’ remained unchanged, while ‘d_filtered’ became a new <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a> which contains our filtered results. We can continue applying filters sequentially to the same Filter object, or using ‘inplace=False’ to create a new object at any point.</p>
<p>Another useful option is to perform an opposite filter. When we specify the parameter ‘opposite=True’ to any filtering function within <em>RNAlysis</em>, the filtering function will be performed in opposite. This means that all of the genomic features that were supposed to be filtered out are kept in the object, and the genomic features that were supposed to be kept in the object are filtered out.
For example, if we now wanted to remove the rows which are below the 25% percentile in the ‘log2FoldChange’ column, we will use the following code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">,</span><span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<p>Calling this function without the ‘opposite’ parameter would have removed all values except the bottom 25% of the ‘log2FoldChange’ column. When specifying ‘opposite’, we instead throw out the bottom 25% of the ‘log2FoldChange’ column and keep the rest.</p>
<p>There are many different filtering functions within the filtering module. Some of them are subtype-specific (such as ‘filter_low_reads’ for <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> objects and ‘filter_significant’ for <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> objects), while others can be applied to any <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a>. You can read more about the different functions and their usage in the project’s documentation.</p>
</section>
<section id="performing-set-operations-on-multiple-filter-objects">
<h4>Performing set operations on multiple Filter objects<a class="headerlink" href="#performing-set-operations-on-multiple-filter-objects" title="Permalink to this heading"></a></h4>
<p>In addition to using regular filters, it is also possible to use set operations such as union, intersection, difference and symmetric difference to combine the results of multiple <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a>. Those set operations can be applied to any Filter object, as well as to python sets. The objects don’t have to be of the same subtype - you can, for example, look at the union of a <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> object, an <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object and a python set:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="n">counts</span><span class="p">,</span> <span class="n">a_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00007063&#39;, &#39;WBGene00007064&#39;, &#39;WBGene00007066&#39;, &#39;WBGene00007067&#39;, &#39;WBGene00007069&#39;, &#39;WBGene00007071&#39;,</span>
<span class="go"> &#39;WBGene00007074&#39;, &#39;WBGene00007075&#39;, &#39;WBGene00007076&#39;, &#39;WBGene00007077&#39;, &#39;WBGene00007078&#39;, &#39;WBGene00007079&#39;,</span>
<span class="go"> &#39;WBGene00014997&#39;, &#39;WBGene00043987&#39;, &#39;WBGene00043988&#39;, &#39;WBGene00043989&#39;, &#39;WBGene00043990&#39;, &#39;WBGene00044022&#39;,</span>
<span class="go"> &#39;WBGene00044951&#39;, &#39;WBGene00077502&#39;, &#39;WBGene00077503&#39;, &#39;WBGene00077504&#39;}</span>
</pre></div>
</div>
<p>When performing set operations, the return type can be either a python set (default) or a string. This means you can use the output of the set operation as an input for yet another set operation. However, since the returned object is a set you cannot use <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a> functions such as ‘head’ and ‘save_csv’ on it, or apply filters to it directly. Intersection and Difference in particular can be used in-place, which applies the filtering to the first <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a>.</p>
</section>
<section id="saving-filter-results">
<h4>Saving Filter results<a class="headerlink" href="#saving-filter-results" title="Permalink to this heading"></a></h4>
<p>At any point we can save the current result of our filtering to a new <cite>csv</cite> file, by using the ‘save_csv’ function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">save_csv</span><span class="p">()</span>
</pre></div>
</div>
<p>If no filename is specified, the file is given a name automatically based on the filtering operations performed on it, their order and their parameters.
We can view the current automatic filename by looking at the ‘fname’ attribute:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_percentile</span><span class="p">(</span><span class="mf">0.75</span><span class="p">,</span><span class="s1">&#39;log2FoldChange&#39;</span><span class="p">)</span>
<span class="go">Filtered 7 features, leaving 21 of the original 28 features. Filtered inplace.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">number_filters</span><span class="p">(</span><span class="s1">&#39;baseMean&#39;</span><span class="p">,</span><span class="s1">&#39;greater than&#39;</span><span class="p">,</span><span class="mi">500</span><span class="p">)</span>
<span class="go">Filtered 6 features, leaving 15 of the original 21 features. Filtered inplace.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">fname</span>
<span class="go">&#39;D:/myfolder/test_deseq_below0.75baseMeangt500.csv&#39;</span>
</pre></div>
</div>
<p>Alternatively, you can specify a filename:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">save_csv</span><span class="p">(</span><span class="s1">&#39;alternative_filename&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Instead of directly saving the results to a file, you can also get them as a set or string of genomic feature indices:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">index_set</span><span class="p">)</span>
<span class="go">{&#39;WBGene00000005&#39;, &#39;WBGene00000006&#39;, &#39;WBGene00000008&#39;, &#39;WBGene00000009&#39;, &#39;WBGene00000010&#39;, &#39;WBGene00000011&#39;,</span>
<span class="go"> &#39;WBGene00000012&#39;, &#39;WBGene00000014&#39;, &#39;WBGene00000015&#39;, &#39;WBGene00000017&#39;, &#39;WBGene00000019&#39;, &#39;WBGene00000021&#39;,</span>
<span class="go"> &#39;WBGene00000023&#39;, &#39;WBGene00000025&#39;, &#39;WBGene00000029&#39;}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">d</span><span class="o">.</span><span class="n">index_string</span><span class="p">)</span>
<span class="go">WBGene00000010</span>
<span class="go">WBGene00000012</span>
<span class="go">WBGene00000021</span>
<span class="go">WBGene00000023</span>
<span class="go">WBGene00000017</span>
<span class="go">WBGene00000015</span>
<span class="go">WBGene00000025</span>
<span class="go">WBGene00000008</span>
<span class="go">WBGene00000011</span>
<span class="go">WBGene00000014</span>
<span class="go">WBGene00000029</span>
<span class="go">WBGene00000006</span>
<span class="go">WBGene00000009</span>
<span class="go">WBGene00000005</span>
<span class="go">WBGene00000019</span>
</pre></div>
</div>
<p>Sets of genomic feature indices can be used later for enrichment analysis using the enrichment module (see below).</p>
</section>
<section id="using-an-attribute-reference-table-for-filter-operations">
<h4>Using an Attribute Reference Table for filter operations<a class="headerlink" href="#using-an-attribute-reference-table-for-filter-operations" title="Permalink to this heading"></a></h4>
<p>An <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> contains various user-defined attributes (such as ‘genes expressed in intestine’, ‘epigenetic genes’ or ‘genes that have paralogs’) and their value for each genomic feature.
You can read more about the <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> format and loading an <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> in the <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a> section.
Using the function Filter.filter_by_attribute(), you can filter your genomic features by one of the user-defined attributes in the Reference Table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 27 features, leaving 1 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
</section>
<section id="using-a-biotype-reference-table-for-filter-operations">
<h4>Using a Biotype Reference Table for filter operations<a class="headerlink" href="#using-a-biotype-reference-table-for-filter-operations" title="Permalink to this heading"></a></h4>
<p>A <a class="reference internal" href="glossary.html#term-Biotype-Reference-Table"><span class="xref std std-term">Biotype Reference Table</span></a> contains annotations of the biotype of each genomic features (‘protein_coding’, ‘piRNAs’, ‘lincRNAs’, ‘pseudogenes’, etc).
You can read more about the <a class="reference internal" href="glossary.html#term-Biotype-Reference-Table"><span class="xref std std-term">Biotype Reference Table</span></a> format and loading a <a class="reference internal" href="glossary.html#term-Biotype-Reference-Table"><span class="xref std std-term">Biotype Reference Table</span></a> in the <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a> section.
Using the function Filter.filter_biotype_from_ref_table(), you can filter your genomic features by their annotated biotype in the Biotype Reference Table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">filter_biotype_from_ref_table</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Filtered 2 features, leaving 26 of the original 28 features. Filtered inplace.</span>
</pre></div>
</div>
<p>You can also view the number of genomic features belonging to each biotype using the function Filter.biotypes_from_ref_table():</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes_from_ref_table</span><span class="p">()</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
</pre></div>
</div>
<p>Or view more elaborated descriptive statistics for eahc biotype by specifying return_format=’long’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span><span class="o">.</span><span class="n">biotypes_from_ref_table</span><span class="p">(</span><span class="n">return_format</span><span class="o">=</span><span class="s1">&#39;long&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>

<span class="go">               baseMean               ...           padj</span>
<span class="go">                  count         mean  ...            75%            max</span>
<span class="go">biotype                               ...</span>
<span class="go">protein_coding     26.0  1823.089609  ...   1.005060e-90   9.290000e-68</span>
<span class="go">pseudogene          1.0  2688.043701  ...   1.800000e-94   1.800000e-94</span>
<span class="go">unknown             1.0  2085.995094  ...  3.070000e-152  3.070000e-152</span>

<span class="go">[3 rows x 48 columns]</span>
</pre></div>
</div>
</section>
</section>
<section id="filtering-deseq2-output-files-with-filtering-deseqfilter">
<h3>Filtering DESeq2 output files with filtering.DESeqFilter<a class="headerlink" href="#filtering-deseq2-output-files-with-filtering-deseqfilter" title="Permalink to this heading"></a></h3>
<p><a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> objects are built to easily filter differential expression tables, such as those returned by the R package DESeq2.
Like other Filter Objects, filtering operations on <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> are performed in-place by default, meaning the original object is modified.</p>
<p>You can read more about DESeq2 here:
<a class="reference external" href="https://doi.org/doi:10.18129/B9.bioc.DESeq2">https://doi.org/doi:10.18129/B9.bioc.DESeq2</a></p>
<section id="loading-from-a-csv-file">
<h4>Loading from a <cite>csv</cite> file<a class="headerlink" href="#loading-from-a-csv-file" title="Permalink to this heading"></a></h4>
<p>Any <cite>csv</cite> file that contains differential expression analysis data with log2 fold change and adjusted p-values can be used as input for <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a>.
By default, <em>RNAlysis</em> assumes that log2 fold change values will be specified under a ‘log2FoldChange’ column, and adjusted p-values will be specified under a ‘padj’ column (as is the default in differential expression tables generated by DESeq2):</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 20%" />
<col style="width: 12%" />
<col style="width: 20%" />
<col style="width: 12%" />
<col style="width: 12%" />
<col style="width: 12%" />
<col style="width: 12%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>baseMean</p></th>
<th class="head"><p>log2FoldChange</p></th>
<th class="head"><p>lfcSE</p></th>
<th class="head"><p>stat</p></th>
<th class="head"><p>pvalue</p></th>
<th class="head"><p>padj</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>WBGene00000021</p></td>
<td><p>2688.044</p></td>
<td><p>3.329404</p></td>
<td><p>0.158938</p></td>
<td><p>20.94783</p></td>
<td><p>1.96E-97</p></td>
<td><p>1.80E-94</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000022</p></td>
<td><p>365.813</p></td>
<td><p>6.101303</p></td>
<td><p>0.291484</p></td>
<td><p>20.93189</p></td>
<td><p>2.74E-97</p></td>
<td><p>2.40E-94</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000023</p></td>
<td><p>3168.567</p></td>
<td><p>3.906719</p></td>
<td><p>0.190439</p></td>
<td><p>20.51433</p></td>
<td><p>1.60E-93</p></td>
<td><p>1.34E-90</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000024</p></td>
<td><p>221.9257</p></td>
<td><p>4.801676</p></td>
<td><p>0.246313</p></td>
<td><p>19.49419</p></td>
<td><p>1.23E-84</p></td>
<td><p>9.82E-82</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000025</p></td>
<td><p>2236.186</p></td>
<td><p>2.477374</p></td>
<td><p>0.129606</p></td>
<td><p>19.11463</p></td>
<td><p>1.91E-81</p></td>
<td><p>1.46E-78</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000026</p></td>
<td><p>343.649</p></td>
<td><p>-4.03719</p></td>
<td><p>0.219781</p></td>
<td><p>-18.3691</p></td>
<td><p>2.32E-75</p></td>
<td><p>1.70E-72</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000027</p></td>
<td><p>175.1429</p></td>
<td><p>6.352044</p></td>
<td><p>0.347777</p></td>
<td><p>18.26471</p></td>
<td><p>1.58E-74</p></td>
<td><p>1.12E-71</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000028</p></td>
<td><p>219.1632</p></td>
<td><p>3.913657</p></td>
<td><p>0.217802</p></td>
<td><p>17.96885</p></td>
<td><p>3.42E-72</p></td>
<td><p>2.32E-69</p></td>
</tr>
</tbody>
</table>
<p>Loading a file that follows this format into a <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> works similarly to other Filter objects:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>If your differential expression table does not follow this format, you can specify the exact names of the columns in your table that contain log2 fold change values and adjusted p-values:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/test_deseq.csv&quot;</span><span class="p">,</span> <span class="n">log2fc_col</span><span class="o">=</span><span class="s1">&#39;name of log2 fold change column&#39;</span><span class="p">,</span> <span class="n">padj_col</span><span class="o">=</span><span class="s1">&#39;name of adjusted p-value column&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Unique <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> functions (such as ‘filter_significant’ and ‘filter_abs_log2_fold_change’) will not execute properly if the log2 fold change column and adjusted p-value column are not defined correctly.</p>
</section>
<section id="filtering-operations-unique-to-deseqfilter">
<h4>Filtering operations unique to DESeqFilter<a class="headerlink" href="#filtering-operations-unique-to-deseqfilter" title="Permalink to this heading"></a></h4>
<p>There are a few filtering operations unique to DESeqFilter. Those include ‘filter_significant’, which removes statistically-insignificant rows according to a specified threshold; ‘filter_abs_log2_fold_change’, removes rows whose absolute value log2 fold change is below the specified threshold; ‘filter_fold_change_direction’ which removes either up-regulated (positive log2 fold change) or down-regulated (negative log2 fold change) rows; and ‘split_fold_change_direction’ which returns a <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> object with only up-regulated features and a <a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> object with only down-regulated features.</p>
</section>
<section id="data-visualization-and-exploratory-data-analysis-with-deseqfilter">
<h4>Data visualization and exploratory data analysis with DESeqFilter<a class="headerlink" href="#data-visualization-and-exploratory-data-analysis-with-deseqfilter" title="Permalink to this heading"></a></h4>
<p><a class="reference internal" href="glossary.html#term-DESeqFilter"><span class="xref std std-term">DESeqFilter</span></a> supports methods for visualization and exploratory analysis of differential expression data.</p>
<p>With DESeqFilter.volcano_plot, you can observe the direction, magnitude, and significance of differential expression within your data:</p>
<figure class="align-center" id="id47">
<a class="reference internal image-reference" href="_images/volcano.png"><img alt="_images/volcano.png" src="_images/volcano.png" style="width: 541.8px; height: 392.7px;" /></a>
<figcaption>
<p><span class="caption-text">Example output of DESeqFilter.volcano_plot()</span><a class="headerlink" href="#id47" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</section>
</section>
<section id="filtering-count-matrices-with-filtering-countfilter">
<h3>Filtering count matrices with filtering.CountFilter<a class="headerlink" href="#filtering-count-matrices-with-filtering-countfilter" title="Permalink to this heading"></a></h3>
<p><a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> objects are capable of visualizing, filtering, normalizing, and clustering of read count matrices (the output of feature-counting software such as HTSeq-count and featureCounts).
Data can be imported into a CountFilter objects either from a <cite>csv</cite> file, or directly from HTSeq-count output files as explained below.</p>
<p>You can read more about HTSeq-count here:
<a class="reference external" href="https://htseq.readthedocs.io/en/master/count.html">https://htseq.readthedocs.io/en/master/count.html</a></p>
<p>In principle, any <cite>csv</cite> file where the columns are different conditions/replicates and the rows include reads/normalized reads per genomic feature can be used as input for CountFilter. However, some <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> functions (such as ‘normalize_to_rpm_htseqcount’) will only work on HTSeq-count output files, and other unintended interactions may occur.</p>
<section id="generating-an-countfilter-object-from-a-folder-of-htseq-count-output-txt-files">
<span id="from-folder-ref"></span><h4>Generating an CountFilter object from a folder of HTSeq-count output .txt files<a class="headerlink" href="#generating-an-countfilter-object-from-a-folder-of-htseq-count-output-txt-files" title="Permalink to this heading"></a></h4>
<p>HTSeq-count receives as input an aligned SAM/BAM file. The native output of HTSeq-count is a text file with feature indices and read-per-genomic-feature, as well as information about reads that weren’t counted for any feature (alignment not unique, low alignment quality, ambiguous, unaligned, aligned to no feature).
An HTSeq-count output file would follow the following format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 83%" />
<col style="width: 17%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>WBGene00000001</p></td>
<td><p>376</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000002</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000003</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000004</p></td>
<td><p>18</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000005</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000006</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000007</p></td>
<td><p>6</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000008</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00000009</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00000010</p></td>
<td><p>177</p></td>
</tr>
<tr class="row-odd"><td><p>__no_feature</p></td>
<td><p>32</p></td>
</tr>
<tr class="row-even"><td><p>__ambiguous</p></td>
<td><p>12</p></td>
</tr>
<tr class="row-odd"><td><p>__too_low_aQual</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>__not_aligned</p></td>
<td><p>121</p></td>
</tr>
<tr class="row-odd"><td><p>__alignment_not_unique</p></td>
<td><p>100</p></td>
</tr>
</tbody>
</table>
<p>When running HTSeq-count on multiple SAM files (which could represent different conditions or replicates), the final output would be a directory of .txt files. <em>RNAlysis</em> can parse those .txt files into two <cite>csv</cite> tables: in the first each row is a genomic feature and each column is a condition or replicate (a single .txt file), and in the second each row represents a category of reads not mapped to genomic features (alignment not unique, low alignment quality, etc). This is done with the ‘from_folder_htseqcount’ function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="o">.</span><span class="n">from_folder_htseqcount</span><span class="p">(</span><span class="s1">&#39;tests/test_files/test_count_from_folder&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>By deault, ‘from_folder_htseqcount’ does not save the generated tables as <cite>csv</cite> files. However, you can choose to save them by specifying ‘save_csv=True’, and specifying their filenames in the arguments ‘counted_fname’ and ‘uncounted_fname’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="o">.</span><span class="n">from_folder_htseqcount</span><span class="p">(</span><span class="s1">&#39;tests/test_files/test_count_from_folder&#39;</span><span class="p">,</span> <span class="n">save_csv</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">counted_fname</span><span class="o">=</span><span class="s1">&#39;name_for_reads_csv_file&#39;</span><span class="p">,</span> <span class="n">uncounted_fname</span><span class="o">=</span><span class="s1">&#39;name_for_uncounted_reads_csv_file&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>It is also possible to automatically normalize the reads in the new <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object to reads per million (RPM) using the unmapped reads data by specifying ‘norm_to_rpm=True’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="o">.</span><span class="n">from_folder_htseqcount</span><span class="p">(</span><span class="s1">&#39;tests/test_files/test_count_from_folder&#39;</span><span class="p">,</span> <span class="n">norm_to_rpm</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="loading-from-a-pre-made-csv-file">
<h4>Loading from a pre-made <cite>csv</cite> file<a class="headerlink" href="#loading-from-a-pre-made-csv-file" title="Permalink to this heading"></a></h4>
<p>If you have previously generated a <cite>csv</cite> file from HTSeq-count output files using <em>RNAlysis</em>, or have done so manually, you can directly load this <cite>csv</cite> file into an <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object as you would any other Filter object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>A correct input to a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object would follow the following format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 36%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 16%" />
<col style="width: 16%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>cond1</p></th>
<th class="head"><p>cond2</p></th>
<th class="head"><p>cond3</p></th>
<th class="head"><p>cond4</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>WBGene00007063</p></td>
<td><p>633</p></td>
<td><p>451</p></td>
<td><p>365</p></td>
<td><p>388</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00007064</p></td>
<td><p>60</p></td>
<td><p>57</p></td>
<td><p>20</p></td>
<td><p>23</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00044951</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00007066</p></td>
<td><p>55</p></td>
<td><p>266</p></td>
<td><p>46</p></td>
<td><p>39</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00007067</p></td>
<td><p>15</p></td>
<td><p>13</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00007069</p></td>
<td><p>0</p></td>
<td><p>2</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00077502</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene00077503</p></td>
<td><p>1</p></td>
<td><p>4</p></td>
<td><p>2</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-even"><td><p>WBGene00077504</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
</section>
<section id="filtering-operations-unique-to-countfilter">
<h4>Filtering operations unique to CountFilter<a class="headerlink" href="#filtering-operations-unique-to-countfilter" title="Permalink to this heading"></a></h4>
<p>There are a few filtering operations unique to CountFilter. Those include ‘filter_low_reads’, which removes rows that have less than n reads in all columns.</p>
</section>
<section id="normalizing-reads-with-countfilter">
<h4>Normalizing reads with CountFilter<a class="headerlink" href="#normalizing-reads-with-countfilter" title="Permalink to this heading"></a></h4>
<p><a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> can normalize reads with either pre-made normalization methods <em>RNAlysis</em> supplies, or with user-defined scaling factors. Pre-normalized data can be used as input for CountFilter as well.</p>
<p><em>RNAlysis</em> supplies the following normalization methods:</p>
<ul class="simple">
<li><p>Relative Log Expression (RLE - ‘normalize_rle’), used by default by R’s DESeq2</p></li>
<li><p>Trimmed Mean of M-values (TMM - ‘normalize_tmm’), used by default by R’s edgeR</p></li>
<li><p>Quantile normalization, a generalization of Upper Quantile normalization (UQ - ‘normalize_quantile’), used by default by R’s Limma</p></li>
<li><p>Median of Ratios Normalization (MRN - ‘normalize_mrn’)</p></li>
<li><p>Reads Per Million (RPM - ‘normalize_to_rpm’)</p></li>
</ul>
<p>To normalize a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> with one of these functions, simply call the function with your preferred parameters, if there are any. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">normalize_rle</span><span class="p">()</span>
<span class="go">Normalized 22 features. Normalized inplace.</span>
</pre></div>
</div>
<p>To normalize a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> with user-generated scaling factors, we need a <cite>csv</cite> table with the scaling factor for each sample:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 25%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>sample1</p></th>
<th class="head"><p>sample2</p></th>
<th class="head"><p>sample3</p></th>
<th class="head"><p>sample4</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>0.96</p></td>
<td><p>1</p></td>
<td><p>0.78</p></td>
<td><p>1.23</p></td>
</tr>
</tbody>
</table>
<p>We would then supply the function with the path to the scaling factors file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">normalize_with_scaling_factors</span><span class="p">(</span><span class="s1">&#39;scaling_factors.csv&#39;</span><span class="p">)</span>
<span class="go">Normalized 22 features. Normalized inplace.</span>
</pre></div>
</div>
<p>The resulting <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object will be normalized with the scaling factors (dividing the value of each column by the value of the corresponding scaling factor).</p>
<p>To normalize a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> that originated from HTSeq-count to reads per million, we need a <cite>csv</cite> table with the special counters that appear in HTSeq-count output:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 40%" />
<col style="width: 15%" />
<col style="width: 15%" />
<col style="width: 15%" />
<col style="width: 15%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>sample1</p></th>
<th class="head"><p>sample2</p></th>
<th class="head"><p>sample3</p></th>
<th class="head"><p>sample4</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>__ambiguous</p></td>
<td><p>37</p></td>
<td><p>12</p></td>
<td><p>145</p></td>
<td><p>77</p></td>
</tr>
<tr class="row-odd"><td><p>__no_feature</p></td>
<td><p>9468</p></td>
<td><p>11354</p></td>
<td><p>14009</p></td>
<td><p>30287</p></td>
</tr>
<tr class="row-even"><td><p>__alignment_not_unique</p></td>
<td><p>108</p></td>
<td><p>290</p></td>
<td><p>557</p></td>
<td><p>893</p></td>
</tr>
<tr class="row-odd"><td><p>__too_low_aQual</p></td>
<td><p>0</p></td>
<td><p>5</p></td>
<td><p>12</p></td>
<td><p>9</p></td>
</tr>
<tr class="row-even"><td><p>__not_aligned</p></td>
<td><p>109853</p></td>
<td><p>277653</p></td>
<td><p>88653</p></td>
<td><p>96012</p></td>
</tr>
</tbody>
</table>
<p>Such a <cite>csv</cite> table is generated automatically when you create a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object from a folder of text files (CountFilter.from_folder_htseqcount(), see <a class="reference internal" href="user_guide_gui.html#from-folder-ref"><span class="std std-ref">Generating a count matrix from a folder of HTSeq-count output .txt files</span></a>).
We would then supply the normalization function with the path to the special counter file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">CountFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/counted.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">normalize_to_rpm_htseqcount</span><span class="p">(</span><span class="s2">&quot;tests/test_files/uncounted.csv&quot;</span><span class="p">)</span>
<span class="go">Normalized 22 features. Normalized inplace.</span>
</pre></div>
</div>
<p>The resulting <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object will be normalized to RPM with the formula (1,000,000 * reads in cell) / (sum of aligned reads + __no_feature + __ambiguous + __alignment_no_unique)</p>
</section>
<section id="data-clustering-with-countfilter">
<h4>Data clustering with CountFilter<a class="headerlink" href="#data-clustering-with-countfilter" title="Permalink to this heading"></a></h4>
<p><em>RNAlysis</em> supports a wide variety of clustering methods, which can group genomic features into clusters according to their similarity across different samples.</p>
<p>When clustering genomic features in a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object, the called function returns a tuple of <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> objects, with each object corresponding to one cluster of genomic features.</p>
<p>Expression plots of the resulting clusters can be generated in one of multiple styles:</p>
<blockquote>
<div><figure class="align-center" id="id48">
<a class="reference internal image-reference" href="_images/kmeans_all.png"><img alt="_images/kmeans_all.png" src="_images/kmeans_all.png" style="width: 560.0px; height: 360.0px;" /></a>
<figcaption>
<p><span class="caption-text">Example expression plot of clustering results with plot_style=’all’</span><a class="headerlink" href="#id48" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-center" id="id49">
<a class="reference internal image-reference" href="_images/kmeans_std_area.png"><img alt="_images/kmeans_std_area.png" src="_images/kmeans_std_area.png" style="width: 560.0px; height: 360.0px;" /></a>
<figcaption>
<p><span class="caption-text">Example expression plot of clustering results with plot_style=’std_area’</span><a class="headerlink" href="#id49" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-center" id="id50">
<a class="reference internal image-reference" href="_images/kmeans_std_bar.png"><img alt="_images/kmeans_std_bar.png" src="_images/kmeans_std_bar.png" style="width: 560.0px; height: 360.0px;" /></a>
<figcaption>
<p><span class="caption-text">Example expression plot of clustering results with plot_style=’std_bar’</span><a class="headerlink" href="#id50" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<figure class="align-center" id="id51">
<a class="reference internal image-reference" href="_images/clustering_PCA_clicom.png"><img alt="_images/clustering_PCA_clicom.png" src="_images/clustering_PCA_clicom.png" style="width: 380.8px; height: 360.0px;" /></a>
<figcaption>
<p><span class="caption-text">Example PCA plot of clustering results</span><a class="headerlink" href="#id51" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
<p>The expression plots can also by split into separate graphs, one for each discovered cluster.</p>
<p>All clustering methods in <em>RNAlysis</em> which require you to specify the expected number of clusters (such as K in K-Means clustering) allow multiple ways of specifying the number of clusters you want to find.
You can specify a single value:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">CountFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/counted.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">five_clusters</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_kmeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 7 features, leaving 15 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 21 features, leaving 1 of the original 22 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">five_clusters</span><span class="p">))</span>
<span class="go">5</span>
</pre></div>
</div>
<p>You can specify a list of values to be used, and each value will be calculated and returned separately:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">CountFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/counted.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">five_clusters</span><span class="p">,</span> <span class="n">two_clusters</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_kmeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">2</span><span class="p">])</span>
<span class="go">Filtered 21 features, leaving 1 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 7 features, leaving 15 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 4 features, leaving 18 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 18 features, leaving 4 of the original 22 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">five_clusters</span><span class="p">))</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">two_clusters</span><span class="p">))</span>
<span class="go">2</span>
</pre></div>
</div>
<p>Finally, you can use a model selection method to estimate the number of clusters in your dataset. <em>RNAlysis</em> supports both the Silhouette method and the Gap Statistic method:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">CountFilter</span><span class="p">(</span><span class="s2">&quot;tests/test_files/counted.csv&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">silhouette_clusters</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_kmeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="s1">&#39;silhouette&#39;</span><span class="p">)</span>
<span class="go">Estimating the optimal number of clusters using the Silhouette method in range 2:5...</span>
<span class="go">Using the Silhouette method, 4 was chosen as the best number of clusters (k).</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 6 features, leaving 16 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 20 features, leaving 2 of the original 22 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">silhouette_clusters</span><span class="p">))</span>
<span class="go">4</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gap_stat_clusters</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">split_kmeans</span><span class="p">(</span><span class="n">n_clusters</span><span class="o">=</span><span class="s1">&#39;gap&#39;</span><span class="p">)</span>
<span class="go">Estimating the optimal number of clusters using the Gap Statistic method in range 2:5...</span>
<span class="go">Using the Gap Statistic method, 2 was chosen as the best number of clusters (K).</span>
<span class="go">Filtered 4 features, leaving 18 of the original 22 features. Filtering result saved to new object.</span>
<span class="go">Filtered 18 features, leaving 4 of the original 22 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">gap_stat_clusters</span><span class="p">))</span>
<span class="go">2</span>
</pre></div>
</div>
<p>To help in evaluating the result of these model selection methods, <em>RNAlysis</em> will also plot a summary of their outcome:</p>
<a class="reference internal image-reference" href="_images/gap_statistic.png"><img alt="_images/gap_statistic.png" src="_images/gap_statistic.png" style="width: 60%;" /></a>
<a class="reference internal image-reference" href="_images/silhouette.png"><img alt="_images/silhouette.png" src="_images/silhouette.png" style="width: 30%;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
<section id="k-means-clustering">
<h5>K-Means clustering<a class="headerlink" href="#k-means-clustering" title="Permalink to this heading"></a></h5>
<p>K-means is a clustering method which partitions all of the data points into K clusters by minimizing the squared eucliean distance between points within each cluster.</p>
<p>The algorithm is initiated by picking a random starting point, and therefore the exact clustering results can change between runs.</p>
<p>The main advantage of K-means clustering is its simplicity - it contains one main tuning parameter (<em>K</em>, the expected number of clusters in the data).</p>
<a class="reference internal image-reference" href="_images/kmeans_all.png"><img alt="K-means clustering output figure" src="_images/kmeans_all.png" style="width: 450px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="k-medoids-clustering">
<h5>K-Medoids clustering<a class="headerlink" href="#k-medoids-clustering" title="Permalink to this heading"></a></h5>
<p>The K-medoids method is very similar to K-means. The main difference between the two is the way they define clusters and the distances between them:
K-medoids picks one data point as the ‘center’ (medoid) of each cluster.
In addition, K-medoids attempts to minimize the sum of dissimilarities within each cluster, instead of minimizing squared euclidean distance.</p>
<p>Due to these differences, the K-medoids algorithm supports the use of distance metrics other than eucliean distance through the <cite>metric</cite> parameter.</p>
<p>K-medoids clustering in <em>RNAlysis</em> supports the following distance metrics:</p>
<ul class="simple">
<li><p>eucliidean</p></li>
<li><p>cosine</p></li>
<li><p>pearson</p></li>
<li><p>spearman</p></li>
<li><p>manhattan (cityblock)</p></li>
<li><p>l1</p></li>
<li><p>l2</p></li>
<li><p>jackknife (see <a class="reference external" href="https://doi.org/10.1101%2Fgr.9.11.1106">Heyer, Kruglyak and Yooseph 1999</a>)</p></li>
<li><p>YS1 (see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>)</p></li>
<li><p>YR1 (see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>)</p></li>
<li><p>hammming</p></li>
<li><p>all other pairwise distance metrics supported by scikit-learn</p></li>
</ul>
<a class="reference internal image-reference" href="_images/kmedoids_all.png"><img alt="K-medoids clustering output figure" src="_images/kmedoids_all.png" style="width: 450px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="hierarchical-clustering">
<h5>Hierarchical clustering<a class="headerlink" href="#hierarchical-clustering" title="Permalink to this heading"></a></h5>
<p>Hierarchical clustering (or agglomerative clustering) is clustering method which aims to build a hierarchy of clusters.</p>
<p>In agglomerative hierarchical clustering, each data points starts in its own clusters.
The clustering algorithm then uses a distance metric (a measure of distance between pairs of data points)
and a linkage criterion
(determines the distance between sets of data points as a function of the pairwise distances between observations)
to group merge data points into clusters, and then further group those clusters into larger clusters based on their similarity.
Eventually, all of the observations are connected into a hierarchical tree.</p>
<p>We can decide to ‘cut’ the tree at any height in order to generate the final clustering solution.
This can be done by either specifying the estimated number of clusters like in K-means,
or by specifiying a distance threshold above which clusters will not be merged.</p>
<p>Hierarchical clustering in <em>RNAlysis</em> supports the following distance metrics:</p>
<ul class="simple">
<li><p>euclidean</p></li>
<li><p>cosine</p></li>
<li><p>pearson</p></li>
<li><p>spearman</p></li>
<li><p>manhattan (cityblock)</p></li>
<li><p>l1</p></li>
<li><p>l2</p></li>
<li><p>jackknife (see <a class="reference external" href="https://doi.org/10.1101%2Fgr.9.11.1106">Heyer, Kruglyak and Yooseph 1999</a>)</p></li>
<li><p>YS1 (see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>)</p></li>
<li><p>YR1 (see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>)</p></li>
</ul>
<a class="reference internal image-reference" href="_images/hierarchical_all.png"><img alt="Hierarchical clustering output figure" src="_images/hierarchical_all.png" style="width: 450px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="hdbscan-clustering">
<h5>HDBSCAN clustering<a class="headerlink" href="#hdbscan-clustering" title="Permalink to this heading"></a></h5>
<p>HDBSCAN stands for Hierarchical Density-Based Spatial Clustering of Applications with Noise (see <a class="reference external" href="https://link.springer.com/chapter/10.1007/978-3-642-37456-2_14">https://link.springer.com/chapter/10.1007/978-3-642-37456-2_14</a> ).
HDBSCAN offers multiple advantages over more traditional clustering methods:</p>
<ol class="arabic simple">
<li><p>HSBSCAN makes relatively few assumptions about the data - it assumes that the data contains noise, as well as some real clusters which we hope to discover.</p></li>
<li><p>Unlike most other clustering methods, HDBSCAN does not force every data point to belong to a cluster. Instead, it can classify data points as outliers, excluding them from the final clustering solution.</p></li>
<li><p>HDBSCAN does not require you to guess the number of clusters in the data. The main tuning parameter in HDBSCAN is <em>minimum cluster size</em> (<cite>min_cluster_size</cite>), which determines the smallest “interesting” cluster size we expect to find in the data.</p></li>
</ol>
<p>HDBSCAN supports additional tuning parameters, which you can read more about in the <a class="reference external" href="https://hdbscan.readthedocs.io/en/latest/parameter_selection.html">HDBSCAN documentation</a>:</p>
<p>HDBSCAN in <em>RNAlysis</em> supports the following distance metrics:</p>
<ul class="simple">
<li><p>eucliidean</p></li>
<li><p>cosine</p></li>
<li><p>pearson</p></li>
<li><p>spearman</p></li>
<li><p>manhattan (cityblock)</p></li>
<li><p>l1</p></li>
<li><p>l2</p></li>
<li><p>jackknife (see <a class="reference external" href="https://doi.org/10.1101%2Fgr.9.11.1106">Heyer, Kruglyak and Yooseph 1999</a>)</p></li>
<li><p>YS1 (see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>)</p></li>
<li><p>YR1 (see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>)</p></li>
<li><p>hammming</p></li>
<li><p>all other pairwise distance metrics elaborated in the <a class="reference external" href="https://hdbscan.readthedocs.io/en/latest/basic_hdbscan.html?#what-about-different-metrics">HDBSCAN documentation</a>.</p></li>
</ul>
<a class="reference internal image-reference" href="_images/hdbscan_all.png"><img alt="HDBSCAN output figure" src="_images/hdbscan_all.png" style="width: 450px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="clicom-clustering">
<h5>CLICOM clustering<a class="headerlink" href="#clicom-clustering" title="Permalink to this heading"></a></h5>
<p>CLICOM is an ensemble-based clustering algorithm (see <a class="reference external" href="https://doi.org/10.1016/j.eswa.2011.08.059">https://doi.org/10.1016/j.eswa.2011.08.059</a> ).
The CLICOM algorithm incorporates the results of multiple clustering solutions, which can come from different clustering algorithms with differing clustering parameters, and uses these clustering solutions to create a combined clustering solution.
CLICOM offers multiple advantages over more traditional clustering methods:</p>
<ol class="arabic simple">
<li><p>The ensemble clustering approach allows you to combine the results of multiple clustering algorithms with multiple tuning parameters, potentially making up for the weaknesses of each individual clustering method, and only taking into account patterns that robustly appear in many clustering solutions.</p></li>
<li><p>Unlike most other clustering methods, CLICOM does not have to force every data point to belong to a cluster. Instead, it can classify data points as outliers, excluding them from the final clustering solution.</p></li>
<li><p>CLICOM does not require you to guess the final number of clusters in the data. The main tuning parameter in HDBSCAN is the <em>evidence threshold</em> (<cite>evidence_threshold</cite>).</p></li>
</ol>
<p><em>RNAlysis</em> offers a modified implementation of CLICOM. This implementation of CLICOM supports a few tuning parameters, in addition to the clustering solutions themselves:
Moreover, ths modified version of the algorithm can cluster each batch of biological/technical replicates in your data separately, which can reduce the influence of batch effect on clustering results, and increases the accuracy and robustness of your clustering results.</p>
<ul class="simple">
<li><p><cite>evidence_threshold</cite>: a higher evidence threshold leads to fewer, large clusters, with fewer features being classified as outliers.</p></li>
<li><p><cite>cluster_unclustered_features</cite>: if True, CLICOM will force every feature to belong to a discovered cluster. Otherwise, features can be classified as noise and remain unclustered.</p></li>
<li><p><cite>min_cluster_size</cite>: determines the minimal size of a cluster you would consider meaningful. Clusters smaller than this would be classified as noise and filtered out of the final result, or merged into other clusters (depending on the value of <cite>cluster_unclustered_features</cite>).</p></li>
<li><p><cite>replicates_grouping</cite>: allows you to group samples into technical/biological batches. The algorithm will then cluster each batch of samples separately, and use the CLICOM algorithm to find an ensemble clustering result from all of the separate clustering results.</p></li>
</ul>
<a class="reference internal image-reference" href="_images/clicom_all.png"><img alt="CLICOM output figure" src="_images/clicom_all.png" style="width: 450px;" /></a>
<div class="line-block">
<div class="line"><br /></div>
</div>
</section>
<section id="specialized-clustering-distance-metrics">
<h5>Specialized clustering distance metrics<a class="headerlink" href="#specialized-clustering-distance-metrics" title="Permalink to this heading"></a></h5>
<p>In addition to the commonly-used distance metrics, such as euclidean distance and spearman correlation, <em>RNAlysis</em> offers a selection of distance metrics that were either developed especially for transcriptomics clustering, or found to work particularly well for transcriptomics clustering.
Those methods include:</p>
<p>1. jackknife distance - a modified Pearson dissimilarity coefficient.
Instead of measuring the linear correlation between expression levels of two genes, you measure the linear correlation coefficient N times (where N is the number of samples in the data), every time excluding a single sample from the correlation, and then taking the smallest correlation coefficient found.
The correlation score is then converted into a dissimilarity score.
This distance metric can detect linear correlation, like Pearson correlation, but is less sensitive to extreme values.
(see <a class="reference external" href="https://doi.org/10.1101%2Fgr.9.11.1106">Heyer, Kruglyak and Yooseph 1999</a>).
2. YR1 distance - a distance metric developed especially for time-series gene expression data.
This distance metric combines the Pearson dissimilarity, along with the positon of the minimal and maximal values of each sample, and the agreement of their slopes. These three values are combined into a single distance score.
This means that the YR1 metric captures more accurately the shape of the expression pattern of each gene, and ranks genes with similar expression patterns as more similar to one another.
(see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>).
3. YS1 distance - a distance metric developed especially for time-series gene expression data.
This distance metric combines the Spearman dissimilarity, along with the positon of the minimal and maximal values of each sample, and the agreement of their slopes. These three values are combined into a single distance score.
This means that the YS1 metric captures more accurately the shape of the expression pattern of each gene, and ranks genes with similar expression patterns as more similar to one another.
(see <a class="reference external" href="https://doi.org/10.1016/j.patrec.2007.09.015">Son and Baek 2007</a>).</p>
</section>
</section>
<section id="data-visualization-and-exploratory-data-analysis-with-countfilter">
<h4>Data visualization and exploratory data analysis with CountFilter<a class="headerlink" href="#data-visualization-and-exploratory-data-analysis-with-countfilter" title="Permalink to this heading"></a></h4>
<p><a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> includes multiple methods for visualization and exploratory analysis of count data.</p>
<p>With CountFilter.pairplot, you can get a quick overview of the distribution of counts within each sample, and the correlation between different samples:</p>
<figure class="align-center" id="id52">
<a class="reference internal image-reference" href="_images/pairplot.png"><img alt="_images/pairplot.png" src="_images/pairplot.png" style="width: 540.0px; height: 387.6px;" /></a>
<figcaption>
<p><span class="caption-text">Example output of CountFilter.pairplot()</span><a class="headerlink" href="#id52" title="Permalink to this image"></a></p>
</figcaption>
</figure>
<p>With CountFilter.pca, you can perform a principal component analysis and look for strong patterns in your dataset:</p>
<blockquote>
<div><figure class="align-center" id="id53">
<a class="reference internal image-reference" href="_images/pca.png"><img alt="_images/pca.png" src="_images/pca.png" style="width: 768.0px; height: 370.8px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of CountFilter.pca()</span><a class="headerlink" href="#id53" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
<p>With CountFilter.plot_expression, you can examine the average expression of specific genomic features under the specific conditions:</p>
<blockquote>
<div><figure class="align-center" id="id54">
<a class="reference internal image-reference" href="_images/plot_expression.png"><img alt="_images/plot_expression.png" src="_images/plot_expression.png" style="width: 384.0px; height: 288.0px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of CountFilter.plot_expression()</span><a class="headerlink" href="#id54" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
<p>With CountFilter.clustergram, you can cluster your samples according to specified distance and linkage metrics:</p>
<blockquote>
<div><figure class="align-center" id="id55">
<a class="reference internal image-reference" href="_images/clustergram.png"><img alt="_images/clustergram.png" src="_images/clustergram.png" style="width: 768.0px; height: 370.8px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of CountFilter.clustergram()</span><a class="headerlink" href="#id55" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
</section>
</section>
<section id="filtering-fold-change-data-of-features-using-filtering-foldchangefilter">
<h3>Filtering fold-change data of features using filtering.FoldChangeFilter<a class="headerlink" href="#filtering-fold-change-data-of-features-using-filtering-foldchangefilter" title="Permalink to this heading"></a></h3>
<p><a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> objects can perform filtering operations and randomization tests on fold change values between two conditions.</p>
<p>A <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> object can be calculated from a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object (you can read more about it in the <a class="reference internal" href="#fold-change-from-count-ref"><span class="std std-ref">Generating fold change data from an existing CountFilter object</span></a>), or imported from a <cite>csv</cite> file like other <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a>.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>by default, <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> assumes that fold change is calculated as (numerator_reads+1)/(denominator_reads+1), and does not support 0 and inf values. If you load a <cite>csv</cite> file which contains 0 and/or inf values into a <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> object, unintended results and interactions may occur.</p>
</div>
<p>Unlike other Filter object, the underlying data structure storing the values is a pandas Series and not a pandas DataFrame, and lacks the Columns attribute.</p>
<section id="loading-fold-change-data-from-a-csv-file">
<h4>Loading fold change data from a <cite>csv</cite> file<a class="headerlink" href="#loading-fold-change-data-from-a-csv-file" title="Permalink to this heading"></a></h4>
<p>Like with other objects from the Filter family, you can simply load a pre-existing or pre-calculated <cite>csv</cite> file into a <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> object. However, in addition to the file path you will also have to enter the name of the numerator condition and the name of the denominator condition:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/fc_1.csv&#39;</span><span class="p">,</span><span class="s1">&#39;name of numerator condition&#39;</span><span class="p">,</span> <span class="s1">&#39;name of denominator condition&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The names of the conditions are saved in the object attributes ‘numerator’ and ‘denominator’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">numerator</span>
<span class="go">&#39;name of numerator condition&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">denominator</span>
<span class="go">&#39;name of denominator condition&#39;</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>by default, <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> assumes that fold change is calculated as (mean_numerator_reads+1)/(mean_denominator_reads+1), and does not support 0 and inf values. If you load a <cite>csv</cite> file which contains 0 and/or inf values into a <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> object, unintended results and interactions may occur.</p>
</div>
</section>
<section id="generating-fold-change-data-from-an-existing-countfilter-object">
<span id="fold-change-from-count-ref"></span><h4>Generating fold change data from an existing CountFilter object<a class="headerlink" href="#generating-fold-change-data-from-an-existing-countfilter-object" title="Permalink to this heading"></a></h4>
<p>Alternatively, you can generate a <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> object from count data in a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object. We will start by loading a <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted_fold_change.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="glossary.html#term-CountFilter"><span class="xref std std-term">CountFilter</span></a> has the following columns:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span><span class="o">.</span><span class="n">columns</span>
<span class="go">[&#39;cond1_rep1&#39;, &#39;cond1_rep2&#39;, &#39;cond2_rep1&#39;, &#39;cond2_rep2&#39;, &#39;cond3_rep1&#39;, &#39;cond3_rep2&#39;]</span>
</pre></div>
</div>
<p>We will now calculate the fold change between the mean of condition1 and condition2. Fold change is calculated as (mean_numerator_reads+1)/(mean_denominator_reads+1). We will need to specify the numerator columns, the denominator columns, and the names of the numerator and denominator. Specifying names is optional - if no names are specified, they will be generator automatically from columns used as numerator and denominator. Since we have multiple replicates of each condition, we will specify all of them in a list:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">counts</span><span class="o">.</span><span class="n">fold_change</span><span class="p">([</span><span class="s1">&#39;cond1_rep1&#39;</span><span class="p">,</span><span class="s1">&#39;cond1_rep2&#39;</span><span class="p">],[</span><span class="s1">&#39;cond2_rep1&#39;</span><span class="p">,</span><span class="s1">&#39;cond2_rep2&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p>In this example we did not specify names for the numerator and denominator, and therefore they were generated automatically:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">numerator</span>
<span class="go">&quot;Mean of [&#39;cond1_rep1&#39;, &#39;cond1_rep2&#39;]&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f</span><span class="o">.</span><span class="n">denominator</span>
<span class="go">&quot;Mean of [&#39;cond2_rep1&#39;, &#39;cond2_rep2&#39;]&quot;</span>
</pre></div>
</div>
<p>We now have a <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> object that we can perform further filtering operations on.</p>
</section>
<section id="performing-randomization-tests-on-a-foldchangefilter-object">
<h4>Performing randomization tests on a FoldChangeFilter object<a class="headerlink" href="#performing-randomization-tests-on-a-foldchangefilter-object" title="Permalink to this heading"></a></h4>
<p>You can perform a randomization test to examine whether the fold change of a group of specific genomic features (for example, genes with a specific biological function) is significantly different than the fold change of a background set of genomic features.
To perform a randomization test you need two <a class="reference internal" href="glossary.html#term-FoldChangeFilter"><span class="xref std std-term">FoldChangeFilter</span></a> objects: one which contains the fold change values of all background genes, and another which contains the fold change values of your specific group of interest. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">FoldChangeFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/fc_1.csv&#39;</span> <span class="p">,</span> <span class="s1">&#39;numerator&#39;</span> <span class="p">,</span> <span class="s1">&#39;denominator&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f_background</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">filter_biotype_from_ref_table</span><span class="p">(</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/biotype_ref_table_for_tests.csv&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span> <span class="c1">#keep only protein-coding genes as reference</span>
<span class="go">Filtered 9 features, leaving 13 of the original 22 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">f_test</span> <span class="o">=</span> <span class="n">f_background</span><span class="o">.</span><span class="n">filter_by_attribute</span><span class="p">(</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/attr_ref_table_for_examples.csv&#39;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 6 features, leaving 7 of the original 13 features. Filtering result saved to new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">rand_test_res</span> <span class="o">=</span> <span class="n">f_test</span><span class="o">.</span><span class="n">randomization_test</span><span class="p">(</span><span class="n">f_background</span><span class="p">)</span>
<span class="go">Calculating...</span>
<span class="go">   group size  observed fold change  ...      pval  significant</span>
<span class="go">0           7              2.806873  ...  0.360264        False</span>

<span class="go">[1 rows x 5 columns]</span>
</pre></div>
</div>
<p>The output table would look like this:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 16%" />
<col style="width: 29%" />
<col style="width: 29%" />
<col style="width: 10%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>group size</p></th>
<th class="head"><p>observed fold change</p></th>
<th class="head"><p>expected fold change</p></th>
<th class="head"><p>pval</p></th>
<th class="head"><p>significant</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>7</p></td>
<td><p>2.806873</p></td>
<td><p>2.51828</p></td>
<td><p>0.36026</p></td>
<td><p>False</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="sequentially-applying-filtering-operations-using-pipelines">
<h3>Sequentially applying filtering operations using Pipelines<a class="headerlink" href="#sequentially-applying-filtering-operations-using-pipelines" title="Permalink to this heading"></a></h3>
<p><a class="reference internal" href="glossary.html#term-Pipeline"><span class="xref std std-term">Pipeline</span></a> objects allow you to group together multiple operations from the <em>filtering</em> module (such as filtering, splitting, normalizing, plotting or describing your data), and apply this group of operations to <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a> of your choice in a specific and consistent order.
Pipelines make your workflow easier to read and understand, help you avoid repetitive code, and makes your analyses more reproducible and less error-prone.</p>
<section id="creating-a-new-pipeline">
<h4>Creating a new Pipeline<a class="headerlink" href="#creating-a-new-pipeline" title="Permalink to this heading"></a></h4>
<p>To create a new empty <a class="reference internal" href="glossary.html#term-Pipeline"><span class="xref std std-term">Pipeline</span></a>, simply create a new Pipeline object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span> <span class="o">=</span> <span class="n">Pipeline</span><span class="p">()</span>
</pre></div>
</div>
<p>Because every <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a> has its own unique functions, a particular Pipeline can only contain functions of a specific Filter object type, and can only be applied to objects of that type.
By default, a new Pipeline’s <cite>filter_type</cite> is <a class="reference internal" href="glossary.html#term-Filter"><span class="xref std std-term">Filter</span></a>, and can only contain general functions from the <em>filtering</em> module that can apply to any Filter object.
If we wanted, for example, to create a Pipeline for DESeqFilter objects, we would have to specify the parameter <cite>filter_type</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">deseq_pipe</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">(</span><span class="s1">&#39;deseqfilter&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>One we have an empty <a class="reference internal" href="glossary.html#term-Pipeline"><span class="xref std std-term">Pipeline</span></a>, we can start adding functions to it.
We can do that either via the function’s name:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">(</span><span class="s1">&#39;DESeqFilter&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="s1">&#39;filter_significant&#39;</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.filter_significant()&#39; to the pipeline.</span>
</pre></div>
</div>
<p>or via the function itself:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">(</span><span class="s1">&#39;DESeqFilter&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="o">.</span><span class="n">filter_significant</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.filter_significant()&#39; to the pipeline.</span>
</pre></div>
</div>
<p>We can also specify the function’s arguments. We can specify both non-keyworded and keyworded arguments, just as we would if we called the function normally:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="n">filtering</span><span class="o">.</span><span class="n">Filter</span><span class="o">.</span><span class="n">filter_biotype_from_ref_table</span><span class="p">,</span> <span class="n">biotype</span><span class="o">=</span><span class="s1">&#39;protein_coding&#39;</span><span class="p">)</span>
<span class="go">Added function &#39;Filter.filter_biotype_from_ref_table(biotype=&#39;protein_coding&#39;)&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="s1">&#39;number_filters&#39;</span><span class="p">,</span> <span class="s1">&#39;column1&#39;</span><span class="p">,</span> <span class="s1">&#39;gt&#39;</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">opposite</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">Added function &#39;Filter.number_filters(&#39;column1&#39;, &#39;gt&#39;, value=5, opposite=True)&#39; to the pipeline.</span>
</pre></div>
</div>
<p>We can also view the functions currently in the Pipeline object, their arguments, and their order:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">pipe</span><span class="p">)</span>
<span class="go">Pipeline for Filter objects:</span>
<span class="go">    Filter.filter_biotype_from_ref_table(biotype=&#39;protein_coding&#39;)</span>
<span class="go">    Filter.number_filters(&#39;column1&#39;, &#39;gt&#39;, value=5, opposite=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">pipe</span><span class="p">))</span>
<span class="go">Pipeline(&#39;Filter&#39;): Filter.filter_biotype_from_ref_table(biotype=&#39;protein_coding&#39;)--&gt;Filter.number_filters(&#39;column1&#39;, &#39;gt&#39;, value=5, opposite=True)</span>
</pre></div>
</div>
<p>We can also remove functions from the Pipeline:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">remove_last_function</span><span class="p">()</span>
<span class="go">Removed function number_filters with parameters [&#39;column1&#39;, &#39;gt&#39;, value=5, opposite=True] from the pipeline.</span>
</pre></div>
</div>
<p>Now that we have a Pipeline with multiple functions, we can apply it to our Filter objects.</p>
</section>
<section id="applying-pipelines-to-filter-objects">
<h4>Applying Pipelines to Filter objects<a class="headerlink" href="#applying-pipelines-to-filter-objects" title="Permalink to this heading"></a></h4>
<p>Just like with other functions in the <em>filtering</em> module, the functions in a <a class="reference internal" href="glossary.html#term-Pipeline"><span class="xref std std-term">Pipeline</span></a> can be applied either inplace or returned as a new object.
You can determine that via the <cite>inplace</cite> argument of the function <cite>Pipeline.apply_to()</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># create the pipeline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">(</span><span class="s1">&#39;DESeqFilter&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="o">.</span><span class="n">filter_missing_values</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.filter_missing_values()&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="o">.</span><span class="n">filter_top_n</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;padj&#39;</span><span class="p">,</span> <span class="n">n</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.filter_top_n(by=&#39;padj&#39;, n=3)&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="s1">&#39;sort&#39;</span><span class="p">,</span> <span class="n">by</span><span class="o">=</span><span class="s1">&#39;baseMean&#39;</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.sort(by=&#39;baseMean&#39;)&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># load the Filter object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/test_files/test_deseq_with_nan.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># apply the Pipeline not-inplace</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d_filtered</span> <span class="o">=</span> <span class="n">pipe</span><span class="o">.</span><span class="n">apply_to</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Filtered 3 features, leaving 25 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Filtered 22 features, leaving 3 of the original 25 features. Filtering result saved to new object.</span>
<span class="go">Sorted 3 features. Sorting result saved to a new object.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># apply the Pipeline inplace</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">apply_to</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="go">Filtered 3 features, leaving 25 of the original 28 features. Filtered inplace.</span>
<span class="go">Filtered 22 features, leaving 3 of the original 25 features. Filtered inplace.</span>
<span class="go">Sorted 3 features. Sorted inplace.</span>
</pre></div>
</div>
<p>Note that only functions that can be applied inplace (such as filtering/normalizing) will be applied inplace.
If our pipeline contained other types of functions, they will not be applied inplace, and will instead be returned at the end of the Pipeline.</p>
<p>If we apply a Pipeline with functions that return additional outputs (such as Figures, DataFrames, etc), they will be returned in a dictionary alongside the Filter object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">filtering</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># create the pipeline</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">Pipeline</span><span class="p">(</span><span class="s1">&#39;DESeqFilter&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="s1">&#39;biotypes_from_ref_table&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/test_files/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.biotypes_from_ref_table(ref=&#39;tests/test_files/test_files/biotype_ref_table_for_tests.csv&#39;)&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="s1">&#39;filter_biotype_from_ref_table&#39;</span><span class="p">,</span> <span class="s1">&#39;protein_coding&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/test_files/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.filter_biotype_from_ref_table(&#39;protein_coding&#39;, ref=&#39;tests/test_files/test_files/biotype_ref_table_for_tests.csv&#39;)&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">pipe</span><span class="o">.</span><span class="n">add_function</span><span class="p">(</span><span class="s1">&#39;biotypes_from_ref_table&#39;</span><span class="p">,</span> <span class="n">ref</span><span class="o">=</span><span class="s1">&#39;tests/test_files/test_files/biotype_ref_table_for_tests.csv&#39;</span><span class="p">)</span>
<span class="go">Added function &#39;DESeqFilter.biotypes_from_ref_table(ref=&#39;tests/test_files/test_files/biotype_ref_table_for_tests.csv&#39;)&#39; to the pipeline.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># load the Filter object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">DESeqFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/test_files/test_deseq_with_nan.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># apply the Pipeline not-inplace</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">d_filtered</span><span class="p">,</span> <span class="n">output_dict</span> <span class="o">=</span> <span class="n">pipe</span><span class="o">.</span><span class="n">apply_to</span><span class="p">(</span><span class="n">d</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">Biotype Reference Table used: tests/test_files/test_files/biotype_ref_table_for_tests.csv</span>
<span class="go">Biotype Reference Table used: tests/test_files/test_files/biotype_ref_table_for_tests.csv</span>
<span class="go">Filtered 2 features, leaving 26 of the original 28 features. Filtering result saved to new object.</span>
<span class="go">Biotype Reference Table used: tests/test_files/test_files/biotype_ref_table_for_tests.csv</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">output_dict</span><span class="p">[</span><span class="s1">&#39;biotypes_1&#39;</span><span class="p">])</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="go">pseudogene         1</span>
<span class="go">unknown            1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="n">output_dict</span><span class="p">[</span><span class="s1">&#39;biotypes_2&#39;</span><span class="p">])</span>
<span class="go">                gene</span>
<span class="go">biotype</span>
<span class="go">protein_coding    26</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># apply the Pipeline inplace</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">output_dict_inplace</span> <span class="o">=</span> <span class="n">pipe</span><span class="o">.</span><span class="n">apply_to</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>
<span class="go">Biotype Reference Table used: tests/test_files/test_files/biotype_ref_table_for_tests.csv</span>
<span class="go">Biotype Reference Table used: tests/test_files/test_files/biotype_ref_table_for_tests.csv</span>
<span class="go">Filtered 2 features, leaving 26 of the original 28 features. Filtered inplace.</span>
<span class="go">Biotype Reference Table used: tests/test_files/test_files/biotype_ref_table_for_tests.csv</span>
</pre></div>
</div>
<p>When an output dictionary is returned, the keys in the dictionary will be the name of the function appended to the number of call made to this function in the Pipeline (in the example above, the first call to ‘biotypes_from_ref_table’ is under the key ‘biotypes_1’, and the second call to ‘biotypes_from_ref_table’ is under the key ‘biotypes_2’); and the values in the dictionary will be the returned values from those functions.
We can apply the same Pipeline to as many Filter objects as we want, as long as the type of the Filter object matches the Pipeline’s <cite>filter_type</cite>.</p>
</section>
<section id="importing-a-pipeline">
<h4>Importing a Pipeline<a class="headerlink" href="#importing-a-pipeline" title="Permalink to this heading"></a></h4>
<p>If you have previously exported a Pipeline, or you want to use a Pipeline that someone else exported, you can import Pipeline files into any <em>RNAlysis</em> session.
<em>RNAlysis</em> Pipelines are saved as YAML (.yaml) files. Those files contain the name of the Pipeline, the functions and parameters added to it, as well as some metadata such as the time it was exported.
To import a Pipeline into <em>RNAlysis</em>, use the Pipeline.import_pipeline() method.</p>
</section>
</section>
</section>
<section id="rnalysis-enrichment-module">
<h2><em>RNAlysis</em> enrichment module<a class="headerlink" href="#rnalysis-enrichment-module" title="Permalink to this heading"></a></h2>
<p>RNAlysis’s enrichment module (rnalysis.enrichment) can be used to perform various enrichment analyses including Gene Ontology (GO) enrichment and enrichment for user-defined attributes. The module also includes basic set operations (union, intersection, difference, symmetric difference) between different sets of genomic features.</p>
<section id="working-with-featureset-objects">
<h3>Working with FeatureSet objects<a class="headerlink" href="#working-with-featureset-objects" title="Permalink to this heading"></a></h3>
<p>The enrichment module is built around <a class="reference internal" href="glossary.html#term-FeatureSet"><span class="xref std std-term">FeatureSet</span></a> objects. A Featureset is a container for a set of gene/genomic feature IDs, and the set’s name (for example, ‘genes that are upregulated under hyperosmotic conditions’). All further anslyses of the set of features is done through the <a class="reference internal" href="glossary.html#term-FeatureSet"><span class="xref std std-term">FeatureSet</span></a> object.</p>
<section id="initialize-an-featureset-object">
<h4>Initialize an FeatureSet object<a class="headerlink" href="#initialize-an-featureset-object" title="Permalink to this heading"></a></h4>
<p>We will start by importing the enrichment module:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">enrichment</span>
</pre></div>
</div>
<p>A <a class="reference internal" href="glossary.html#term-FeatureSet"><span class="xref std std-term">FeatureSet</span></a> object can now be initialized by one of three methods.
The first method is to specify an existing Filter object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_filter_obj</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span> <span class="c1"># create a Filter object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">my_filter_obj</span><span class="p">,</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The second method is to directly specify a python set of genomic feature indices, or a python set generated from an existing <a class="reference internal" href="glossary.html#term-Filter-object"><span class="xref std std-term">Filter object</span></a> (see above for more information about <a class="reference internal" href="glossary.html#term-Filter-objects"><span class="xref std std-term">Filter objects</span></a> and the filtering module) using the function ‘index_set’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_python_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene0245200&#39;</span><span class="p">,</span><span class="s1">&#39; WBGene00402029&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">my_python_set</span><span class="p">,</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
<span class="go"># alternatively, using &#39;index_set&#39; on an existing Filter object:</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_other_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">my_filter_obj</span><span class="o">.</span><span class="n">index_set</span><span class="p">,</span><span class="s1">&#39; a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The third method is not to specify a gene set at all:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">set_name</span> <span class="o">=</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>At this point, you will be prompted to enter a string of feature indices seperated by newline. They will be automatically paresd into a python set.</p>
<p>FeatureSet objects have two attributes: gene_set, a python set containing genomic feature indices; and set_name, a string that describes the feature set (optional).</p>
</section>
<section id="go-enrichment">
<h4>GO Enrichment<a class="headerlink" href="#go-enrichment" title="Permalink to this heading"></a></h4>
<p>Using the <em>enrichment</em> module, you can perform enrichment analysis for Gene Ontology terms (GO enrichment).
You can read more about Gene Ontology on the <a class="reference external" href="http://geneontology.org/docs/ontology-documentation/?">Gene Ontology Consortium website</a>.</p>
<p>To perform GO Enrichment analysis, we will start by creating an FeatureSet object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;path_to_my_file.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">index_set</span><span class="p">,</span> <span class="s1">&#39;my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<section id="define-the-correct-organism-and-gene-id-type-for-your-dataset">
<h5>Define the correct <em>organism</em> and <em>gene ID type</em> for your dataset<a class="headerlink" href="#define-the-correct-organism-and-gene-id-type-for-your-dataset" title="Permalink to this heading"></a></h5>
<p>Since GO annotations refer to specific gene products, which can differ between different species, <em>RNAlysis</em> needs to know which organism your dataset refers to.
The organism can be specified as either the organism’s name, or the organism’s <em>NCBI Taxon ID</em> (for example: 6239 for <em>Caenorhabditis elegans</em>).</p>
<p>It is recommended to manually determine your organism’s <em>NCBI Taxon ID</em> to avoid mischaracterization of annotations.
However, if you are not sure, <em>RNAlysis</em> will attempt to automatically determine the correct <cite>organism</cite> by default, based on the gene IDs in your FeatureSet.</p>
<p>Furthermore, since different annotations use different gene ID types to annotate the same gene products (such as UniProtKB ID, Entrez Gene ID, or Wormbase WBGene), <em>RNAlysis</em> can translate gene IDs from one gene ID type to another.
In order to do that, you need to specify which gene ID type your dataset uses.</p>
</section>
<section id="define-the-background-set">
<h5>Define the background set<a class="headerlink" href="#define-the-background-set" title="Permalink to this heading"></a></h5>
<p>In enrichment analysis, we test whether our set of genomic features is enriched/depleted for a certain <em>GO Term</em>, in comparison to a more generalized set of genomic features that we determined as ‘background’.
This could be the set of all protein-coding genes, the set of all genomic features that show expression above a certain threshold, or any other set of background genes which you deem appropriate. Importantly, the background set must contain all of the genes in the enrichment set.</p>
<p>Enrichment analysis is usually performed on protein-coding genes. Therefore, by default, <em>RNAlysis</em> uses all of the protein-coding genes that have at least one GO Annotation as a background set.
If you don’t want to use the default setting, there are two methods of defining the background set:</p>
<p>The first method is to specify a biotype (such as ‘protein_coding’, ‘miRNA’ or ‘all’) under the parameter ‘biotype’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">go_enrichment</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, instead of using all of the protein-coding genes that have GO Annotations as background, we use every genomic feature with GO Annotations as background.
When specifying a biotype, the Biotype Reference Table that you specified is used to determine the biotype of each genomic feature.</p>
<p>The second method of defining the background set is to define a specific set of genomic features to be used as background:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_background_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;feature1&#39;</span><span class="p">,</span><span class="s1">&#39;feature2&#39;</span><span class="p">,</span><span class="s1">&#39;feature3&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">go_enrichment</span><span class="p">(</span><span class="n">background_genes</span><span class="o">=</span><span class="n">my_background_set</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, our background set consists of <em>feature1</em>, <em>feature2</em> and <em>feature3</em>.</p>
<p>It is not possible to specify both a biotype and a specific background set.</p>
<p>If some of the features in the background set or the enrichment set do no appear in the Reference Table, they will be ignored when calculating enrichment.</p>
</section>
<section id="choose-the-statistical-test-optional">
<h5>Choose the statistical test (optional)<a class="headerlink" href="#choose-the-statistical-test-optional" title="Permalink to this heading"></a></h5>
<p>Significance testing for GO enrichment analysis can be done using either the Hypergeometric Test, Fisher’s Exact Test, or a randomization test.</p>
<p>The hypergeometric test is defined as: Given <em>M</em> genes in the background set, <em>n</em> genes in the test set, with <em>N</em> genes from the background set belonging to a specific attribute (‘success’) and <em>X</em> genes from the test set belonging to that attribute.
If we were to randomly draw <em>n</em> genes from the background set (without replacement), what is the probability of drawing <em>X</em> or more (in case of enrichment)/<em>X</em> or less (in case of depletion) genes belonging to the given attribute?</p>
<p>The Fisher’s Exact test is similar in principle to the hypergeometric test, but is two-tailed by default, as opposed to the hypergeometric test which examines enrichment and depletion separately.</p>
<p>The randomization test is defined as: Given <em>M</em> genes in the background set, <em>n</em> genes in the test set, with <em>N</em> genes from the background set belonging to a specific attribute and <em>X</em> genes from the test set belonging to that attribute.
We performs the number of randomizations specified by the user (10,000 by default).
In each randomization we randomly draw a set of <em>n</em> genes from the background set (without replacement), and marks the randomization as a ‘success’ if the number of genes in the random set belonging to the attribute is &gt;= <em>X</em> (in case of enrichment) or &lt;= <em>X</em> (in case of depletion).
The p-values are calculated as <em>(number of sucesses + 1)/(number of repetitions + 1)</em>.
This is a positive-bias estimator of the exact p-value, which avoids exactly-zero p-values.
You can read more about the topic in the following publication: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/21044043">https://www.ncbi.nlm.nih.gov/pubmed/21044043</a></p>
<p>If you don’t specify which statistical test you want to use, the Fisher’s Exact Test will be used by default.</p>
<p>To choose the statistical test you want to use, utilize the <cite>statistical_test</cite> parameter, which accepts either ‘fisher’, ‘hypergeometric’, or ‘randomization’.
If you choose to use a randomization test, you can specify the number of randomization repititions to run using the <cite>randomization_reps</cite> parameter, and set the random seed using the <cite>random_seed</cite> parameter.</p>
</section>
<section id="filter-go-terms-by-go-aspects-optional">
<h5>Filter GO Terms by <em>GO aspects</em> (optional)<a class="headerlink" href="#filter-go-terms-by-go-aspects-optional" title="Permalink to this heading"></a></h5>
<p>Gene Ontology considers three discrete aspects by which gene products can be described:</p>
<ol class="arabic simple">
<li><p>Biological process - the general ‘biological objective’ the gene product contributes to (e.g. ‘locomotion’, ‘cell-cell signaling by wnt’)</p></li>
<li><p>Molecular function - the molecular process or activity carried out by the gene product (e.g. ‘antioxidant activity’, ‘ATP-dependent protein folding chaperone’)</p></li>
<li><p>Cellular component - the location of the gene product when it carries out its action (e.g. ‘P granule’, ‘mitochondrion’)</p></li>
</ol>
<p>Every GO term is exclusively associated with one of these <em>GO aspects</em>.
If you are interested in testing enrichment only for GO terms associated with a subset of these <em>GO aspects</em> you can specify which <em>GO aspects</em> to use through the <cite>aspects</cite> parameter.</p>
<p>If you don’t specify <em>GO aspects</em> to be included, <em>RNAlysis</em> will test enrichment for GO Terms from all <em>GO aspects</em> by default.</p>
</section>
<section id="filter-go-annotations-by-evidence-codes-optional">
<h5>Filter GO Annotations by Evidence Codes (optional)<a class="headerlink" href="#filter-go-annotations-by-evidence-codes-optional" title="Permalink to this heading"></a></h5>
<p>Every GO annotations is supported by an evidence code, which specifies what kind of evidence supports this annotation.
Evidence codes are divided into six categories:</p>
<ol class="arabic simple">
<li><p>experimental (there is evidence from an experiment directly supporting this annotation)</p></li>
<li><p>phylogenetic (annotations are derived from a phylogenetic model)</p></li>
<li><p>computational (annotations are based on in-silico analysis of gene sequence or other computational analysis)</p></li>
<li><p>author (annotations are based on the statement of the author in the cited reference)</p></li>
<li><p>curator (annotations are based on a curator’s judgement)</p></li>
<li><p>electronic (annotations are based on homology and/or sequence information, and were not manually reviewed)</p></li>
</ol>
<p>Each evidence category contains multiple evidence codes, each with its own definition.</p>
<p>You can choose to include only annotations with specific evidence codes, or to exclude annotations with specific annotation codes, using the <cite>evidence_types</cite> and <cite>excluded_evidence_types</cite> parameters.
You can specify either specific evidence codes (e.g. ‘IEA’, ‘IKR’), evidence categories (‘experimental’, ‘electronic’), or any combination of those.</p>
<p>If you don’t specify evidence types to be included/excluded, <em>RNAlysis</em> will use annotations with all evidence codes by default.</p>
<p>You can read more about GO evidence codes here:
<a class="reference external" href="http://geneontology.org/docs/guide-go-evidence-codes/">http://geneontology.org/docs/guide-go-evidence-codes/</a></p>
</section>
<section id="filter-go-annotations-by-database-optional">
<h5>Filter GO Annotations by database (optional)<a class="headerlink" href="#filter-go-annotations-by-database-optional" title="Permalink to this heading"></a></h5>
<p>GO annotations are curated by different databases, such as UniProt, WormBase, or The Arabidopsis Information Resource.
You can choose to include only annotations from specific databases, or to exclude annotations from specific databases, using the <cite>databases</cite> and <cite>excluded_databases</cite> parameters.</p>
<p>If you don’t specify databases to be included/excluded, <em>RNAlysis</em> will use annotations from all databases by default.</p>
</section>
<section id="filter-go-annotations-by-qualifiers-optional">
<h5>Filter GO Annotations by Qualifiers (optional)<a class="headerlink" href="#filter-go-annotations-by-qualifiers-optional" title="Permalink to this heading"></a></h5>
<p>Some GO annotations are modified by qualifiers. Each qualifier has a specific meaning within Gene Ontology:</p>
<p>1. the <em>NOT</em> qualifier - an explicit annotation that this particular gene product has been experimentally demonstrated <em>not</em> to be associated with the particular GO term.
Annotations with the <em>NOT</em> qualifier are usually ignored during enrichment analysis.
2. the <em>contributes_to</em> qualifier - indicates that this gene product facilitates, but does not directly carry out a function of a protein complex.
3. the <em>colocalizes_with</em> qualifier - indicates that this gene product associates with an organelle or complex.</p>
<p>You can choose to include only annotations with specific qualifiers, or to exclude annotations with a specific qualifier, using the <cite>qualifiers</cite> and <cite>excluded_qualifiers</cite> parameters.</p>
<p>If you don’t specify qualifiers to be included/excluded, <em>RNAlysis</em> will ignore annotations with the <em>NOT</em> qualifier by default, and use annotations with any other qualifiers (or no qualifiers at all).</p>
<p>You can read more about GO qualifiers here:
<a class="reference external" href="http://geneontology.org/docs/go-annotations/">http://geneontology.org/docs/go-annotations/</a></p>
</section>
<section id="choose-annotation-propagation-method-optional">
<h5>Choose annotation propagation method (optional)<a class="headerlink" href="#choose-annotation-propagation-method-optional" title="Permalink to this heading"></a></h5>
<p>Gene Ontology terms have a somewhat hierarchical relationship that is defined as a directed a-cyclic graph (DAG). This means that each GO term is a node in the graph, and that each node has defined parents that are less specific than itself, going up to the top of the graph.</p>
<p>For example:</p>
<blockquote>
<div><figure class="align-center" id="id56">
<a class="reference internal image-reference" href="/figures/http://geneontology.org/assets/hexose-biosynthetic-process.png"><img alt="/figures/http://geneontology.org/assets/hexose-biosynthetic-process.png" src="/figures/http://geneontology.org/assets/hexose-biosynthetic-process.png" /></a>
<figcaption>
<p><span class="caption-text">taken from the Gene Ontology Consortium site</span><a class="headerlink" href="#id56" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
<p>We can see that in this example the GO term ‘hexose biosynthetic process’ has two parents, one of which is the less specific term ‘hexose metabolic process’, and these relationships go all the way up to the GO term ‘metabolic process’.</p>
<p>Due to the relationships defined between GO terms, when a gene is annotated with a specific GO term, it makes sense that all of the less-specific parents of this GO term will also apply to said gene.
Therefore, when performing GO enrichment, we would usually ‘propagate’ every GO annotation to all of the GO terms upstream to it, all the way to the top of the GO graph.</p>
<p>Unfortunately, propagating GO annotations comes with some issues:
the defined relationship between GO terms introduces dependencies between neighboring GO terms, leading to correlation between enrichment results of different GO terms, and under-estimation of the False Discovery Rate of our analysis.
Moreover, since more specific GO terms by definition have less annotations than their less-specific parents, the most statistically significant enrichment results usually belong to the least-specific GO terms, which are not very biologically relevant.</p>
<p>To deal with this problem, several alternative propagation methods were developed to help de-correlate the GO graph structure and increase the specificity of our results without compromising accuracy.
You can read more about some suggested methods in the following publication:
<a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/16606683/">https://pubmed.ncbi.nlm.nih.gov/16606683/</a></p>
<p><em>RNAlysis</em> implements three of these propagation methods: <em>elim</em>, <em>weight</em>, and <em>all.m</em>.
You can decide which propagation method to use by specifying the <cite>propagation_method</cite> parameter: ‘no’ for no propagation of GO annotations, ‘classic’ for classic propagation of GO annotations, and ‘elim’/’weight’/’all.m’ for propagation using the <em>elim</em>/<em>weight</em>/<em>all.m</em> propagation algorithms.</p>
<p>If you don’t specify which propagation method to use in enrichment analysis, the <em>elim</em> method will be used by default.</p>
</section>
<section id="choose-plotting-parameters-optional">
<h5>Choose plotting parameters (optional)<a class="headerlink" href="#choose-plotting-parameters-optional" title="Permalink to this heading"></a></h5>
<p>After <em>RNAlysis</em> is done calculating the results of your enrichment analysis, it will automatically plot a summary of the enrichment results.
<em>RNAlysis</em> plots the results as a bar plot, with the Y axis showing log2 fold enrichment, and asterisks indicating whether this enrichment is statistically significant after correcting for multiple comparisons.</p>
<p>You can determine the orientation of the bar plot (horizontal or vertical) using the <cite>plot_horizontal</cite> parameter:</p>
<blockquote>
<div><figure class="align-center" id="id57">
<a class="reference internal image-reference" href="_images/plot_enrichment_results_go.png"><img alt="_images/plot_enrichment_results_go.png" src="_images/plot_enrichment_results_go.png" style="width: 560.0px; height: 406.0px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id15"><span class="problematic" id="id16">`</span></a>plot_horizontal`=True</span><a class="headerlink" href="#id57" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
<figure class="align-center" id="id58">
<a class="reference internal image-reference" href="_images/plot_enrichment_results_go_vertical.png"><img alt="_images/plot_enrichment_results_go_vertical.png" src="_images/plot_enrichment_results_go_vertical.png" style="width: 769.2px; height: 224.0px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id17"><span class="problematic" id="id18">`</span></a>plot_horizontal`=False</span><a class="headerlink" href="#id58" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
</div></blockquote>
<p>If you want to further customize this plot, you can request <em>RNAlysis</em> to return a Matplotlib Figure object of the barplot, by using the <cite>return_fig</cite> parameter.</p>
<p>If you don’t specify plotting parameters, <em>RNAlysis</em> will generate a horizontal bar plot by default, and will not return a Matplotlib Figure object of the bar plot.</p>
<p>In addition, <em>RNAlysis</em> can generate an ontology graph, depicting all of the statistically significant GO terms and their hierarchical relationships:</p>
<blockquote>
<div><figure class="align-center" id="id59">
<a class="reference internal image-reference" href="_images/ontology_graph.png"><img alt="_images/ontology_graph.png" src="_images/ontology_graph.png" style="width: 560.0px; height: 360.0px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id19"><span class="problematic" id="id20">`</span></a>plot_ontology_graph`=True</span><a class="headerlink" href="#id59" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
</div></blockquote>
<p>If you don’t want to generate this graph, you can set the parameter <cite>plot_ontology_graph</cite> to False.</p>
<p>Moreover, you can determine the file format of the generated graph (.pdf, .png, .svg, etc’), by setting the <cite>ontology_graph_format</cite> parameter.</p>
<p>If you don’t specify plotting parameters, <em>RNAlysis</em> will generate an ontology graph by default in a PDF format.</p>
</section>
<section id="enrichment-analysis-output">
<h5>Enrichment analysis output<a class="headerlink" href="#enrichment-analysis-output" title="Permalink to this heading"></a></h5>
<p>Running enrichment analysis will calculate enrichment for each of the GO terms, and return a pandas DataFrame in the following format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 12%" />
<col style="width: 16%" />
<col style="width: 13%" />
<col style="width: 4%" />
<col style="width: 6%" />
<col style="width: 20%" />
<col style="width: 9%" />
<col style="width: 9%" />
<col style="width: 12%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>name</p></th>
<th class="head"><p>samples</p></th>
<th class="head"><p>obs</p></th>
<th class="head"><p>exp</p></th>
<th class="head"><p>log2_fold_enrichment</p></th>
<th class="head"><p>pval</p></th>
<th class="head"><p>padj</p></th>
<th class="head"><p>significant</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="GO:0001556">GO:0001556</a></p></td>
<td><p>oocyte maturation</p></td>
<td><p>1327</p></td>
<td><p>451</p></td>
<td><p>319.52</p></td>
<td><p>0.49722119558</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="GO:0043186">GO:0043186</a></p></td>
<td><p>P granule</p></td>
<td><p>1327</p></td>
<td><p>89</p></td>
<td><p>244.87</p></td>
<td><p>-1.46013879322</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
</tbody>
</table>
<p>‘samples’ is the number of features that were used in the enrichment set. ‘obs’ is the observed number of features positive for the attribute in the enrichment set.
‘exp’ is the expected number of features positive for the attribute in the background set. ‘log2_fold_enrichment’ is log2 of the fold change ‘obs’/’exp’.</p>
</section>
</section>
<section id="kegg-pathways-enrichment">
<h4>KEGG Pathways enrichment<a class="headerlink" href="#kegg-pathways-enrichment" title="Permalink to this heading"></a></h4>
<p>Using the <em>enrichment</em> module, you can perform enrichment analysis for KEGG pathways.
You can read more about KEGG pathways on the <a class="reference external" href="https://www.genome.jp/kegg/pathway.html">KEGG website</a>.</p>
<p>To perform KEGG Enrichment analysis, we will start by creating an FeatureSet object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;path_to_my_file.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">index_set</span><span class="p">,</span> <span class="s1">&#39;my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<section id="id21">
<h5>Define the correct <em>organism</em> and <em>gene ID type</em> for your dataset<a class="headerlink" href="#id21" title="Permalink to this heading"></a></h5>
<p>Since KEGG annotations refer to specific gene products, which can differ between different species, <em>RNAlysis</em> needs to know which organism your dataset refers to.
The organism can be specified as either the organism’s name, or the organism’s <em>NCBI Taxon ID</em> (for example: 6239 for <em>Caenorhabditis elegans</em>).</p>
<p>It is recommended to manually determine your organism’s <em>NCBI Taxon ID</em> to avoid mischaracterization of annotations.
However, if you are not sure, <em>RNAlysis</em> will attempt to automatically determine the correct <cite>organism</cite> by default, based on the gene IDs in your FeatureSet.</p>
<p>Furthermore, since different annotations use different gene ID types to annotate the same gene products (such as UniProtKB ID, Entrez Gene ID, or Wormbase WBGene), <em>RNAlysis</em> can translate gene IDs from one gene ID type to another.
In order to do that, you need to specify which gene ID type your dataset uses.</p>
</section>
<section id="id22">
<h5>Define the background set<a class="headerlink" href="#id22" title="Permalink to this heading"></a></h5>
<p>In enrichment analysis, we test whether our set of genomic features is enriched/depleted for a certain KEGG pathway, in comparison to a more generalized set of genomic features that we determined as ‘background’.
This could be the set of all protein-coding genes, the set of all genomic features that show expression above a certain threshold, or any other set of background genes which you deem appropriate. Importantly, the background set must contain all of the genes in the enrichment set.</p>
<p>Enrichment analysis is usually performed on protein-coding genes. Therefore, by default, <em>RNAlysis</em> uses all of the protein-coding genes that have at least one KEGG annotation as a background set.
If you don’t want to use the default setting, there are two methods of defining the background set:</p>
<p>The first method is to specify a biotype (such as ‘protein_coding’, ‘miRNA’ or ‘all’) under the parameter ‘biotype’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">kegg_enrichment</span><span class="p">(</span><span class="n">biotype</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, instead of using all of the protein-coding genes that have GO Annotations as background, we use every genomic feature with GO Annotations as background.
When specifying a biotype, the Biotype Reference Table that you specified is used to determine the biotype of each genomic feature.</p>
<p>The second method of defining the background set is to define a specific set of genomic features to be used as background:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_background_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;feature1&#39;</span><span class="p">,</span><span class="s1">&#39;feature2&#39;</span><span class="p">,</span><span class="s1">&#39;feature3&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">kegg_enrichment</span><span class="p">(</span><span class="n">background_genes</span><span class="o">=</span><span class="n">my_background_set</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, our background set consists of <em>feature1</em>, <em>feature2</em> and <em>feature3</em>.</p>
<p>It is not possible to specify both a biotype and a specific background set.</p>
<p>If some of the features in the background set or the enrichment set do no appear in the Reference Table, they will be ignored when calculating enrichment.</p>
</section>
<section id="id23">
<h5>Choose the statistical test (optional)<a class="headerlink" href="#id23" title="Permalink to this heading"></a></h5>
<p>Significance testing for KEGG enrichment analysis can be done using either the Hypergeometric Test, Fisher’s Exact Test, or a randomization test.</p>
<p>The hypergeometric test is defined as: Given <em>M</em> genes in the background set, <em>n</em> genes in the test set, with <em>N</em> genes from the background set belonging to a specific attribute (‘success’) and <em>X</em> genes from the test set belonging to that attribute.
If we were to randomly draw <em>n</em> genes from the background set (without replacement), what is the probability of drawing <em>X</em> or more (in case of enrichment)/<em>X</em> or less (in case of depletion) genes belonging to the given attribute?</p>
<p>The Fisher’s Exact test is similar in principle to the hypergeometric test, but is two-tailed by default, as opposed to the hypergeometric test which examines enrichment and depletion separately.</p>
<p>The randomization test is defined as: Given <em>M</em> genes in the background set, <em>n</em> genes in the test set, with <em>N</em> genes from the background set belonging to a specific attribute and <em>X</em> genes from the test set belonging to that attribute.
We performs the number of randomizations specified by the user (10,000 by default).
In each randomization we randomly draw a set of <em>n</em> genes from the background set (without replacement), and marks the randomization as a ‘success’ if the number of genes in the random set belonging to the attribute is &gt;= <em>X</em> (in case of enrichment) or &lt;= <em>X</em> (in case of depletion).
The p-values are calculated as <em>(number of sucesses + 1)/(number of repetitions + 1)</em>.
This is a positive-bias estimator of the exact p-value, which avoids exactly-zero p-values.
You can read more about the topic in the following publication: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/21044043">https://www.ncbi.nlm.nih.gov/pubmed/21044043</a></p>
<p>If you don’t specify which statistical test you want to use, the Fisher’s Exact Test will be used by default.</p>
<p>To choose the statistical test you want to use, utilize the <cite>statistical_test</cite> parameter, which accepts either ‘fisher’, ‘hypergeometric’, or ‘randomization’.
If you choose to use a randomization test, you can specify the number of randomization repititions to run using the <cite>randomization_reps</cite> parameter, and set the random seed using the <cite>random_seed</cite> parameter.</p>
</section>
<section id="id24">
<h5>Choose plotting parameters (optional)<a class="headerlink" href="#id24" title="Permalink to this heading"></a></h5>
<p>After <em>RNAlysis</em> is done calculating the results of your enrichment analysis, it will automatically plot a summary of the enrichment results.
<em>RNAlysis</em> plots the results as a bar plot, with the Y axis showing log2 fold enrichment, and asterisks indicating whether this enrichment is statistically significant after correcting for multiple comparisons.</p>
<p>You can determine the orientation of the bar plot (horizontal or vertical) using the <cite>plot_horizontal</cite> parameter:</p>
<blockquote>
<div><figure class="align-center" id="id60">
<a class="reference internal image-reference" href="_images/plot_enrichment_results.png"><img alt="_images/plot_enrichment_results.png" src="_images/plot_enrichment_results.png" style="width: 224.0px; height: 260.8px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id25"><span class="problematic" id="id26">`</span></a>plot_horizontal`=True</span><a class="headerlink" href="#id60" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
<figure class="align-center" id="id61">
<a class="reference internal image-reference" href="_images/plot_enrichment_results_vertical.png"><img alt="_images/plot_enrichment_results_vertical.png" src="_images/plot_enrichment_results_vertical.png" style="width: 327.6px; height: 222.4px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id27"><span class="problematic" id="id28">`</span></a>plot_horizontal`=False</span><a class="headerlink" href="#id61" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
</div></blockquote>
<p>If you want to further customize this plot, you can request <em>RNAlysis</em> to return a Matplotlib Figure object of the barplot, by using the <cite>return_fig</cite> parameter.</p>
<p>If you don’t specify plotting parameters, <em>RNAlysis</em> will generate a horizontal bar plot by default, and will not return a Matplotlib Figure object of the bar plot.</p>
</section>
<section id="id29">
<h5>Enrichment analysis output<a class="headerlink" href="#id29" title="Permalink to this heading"></a></h5>
<p>Running enrichment analysis will calculate enrichment for each of the KEGG pathways, and return a pandas DataFrame in the following format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 7%" />
<col style="width: 41%" />
<col style="width: 9%" />
<col style="width: 3%" />
<col style="width: 4%" />
<col style="width: 14%" />
<col style="width: 6%" />
<col style="width: 6%" />
<col style="width: 8%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>KEGG ID</p></th>
<th class="head"><p>name</p></th>
<th class="head"><p>samples</p></th>
<th class="head"><p>obs</p></th>
<th class="head"><p>exp</p></th>
<th class="head"><p>log2_fold_enrichment</p></th>
<th class="head"><p>pval</p></th>
<th class="head"><p>padj</p></th>
<th class="head"><p>significant</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>cel00010</p></td>
<td><p>Glycolysis / Gluconeogenesis - Caenorhabditis elegans (nematode)</p></td>
<td><p>1327</p></td>
<td><p>451</p></td>
<td><p>319.52</p></td>
<td><p>0.49722119558</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>cel00030</p></td>
<td><p>Pentose phosphate pathway - Caenorhabditis elegans (nematode)</p></td>
<td><p>1327</p></td>
<td><p>89</p></td>
<td><p>244.87</p></td>
<td><p>-1.46013879322</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
</tbody>
</table>
<p>‘samples’ is the number of features that were used in the enrichment set. ‘obs’ is the observed number of features positive for the attribute in the enrichment set.
‘exp’ is the expected number of features positive for the attribute in the background set. ‘log2_fold_enrichment’ is log2 of the fold change ‘obs’/’exp’.</p>
</section>
</section>
<section id="enrichment-analysis-for-user-defined-attributes">
<h4>Enrichment analysis for user-defined attributes<a class="headerlink" href="#enrichment-analysis-for-user-defined-attributes" title="Permalink to this heading"></a></h4>
<p>Using the <em>enrichment</em> module, you can perform enrichment analysis for user-defined attributes (such as ‘genes expressed in intestine’, ‘epigenetic genes’, ‘genes that have paralogs’). The enrichment analysis can be performed using either the hypergeometric test or a randomization test.</p>
<p>Enrichment analysis for user-defined attributes is performed using FeatureSet.user_defined_enrichment. We will start by creating an FeatureSet object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">counts</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;path_to_my_file.csv&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">(</span><span class="n">counts</span><span class="o">.</span><span class="n">index_set</span><span class="p">,</span> <span class="s1">&#39;my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<section id="choose-which-user-defined-attributes-to-calculate-enrichment-for">
<h5>Choose which user-defined attributes to calculate enrichment for<a class="headerlink" href="#choose-which-user-defined-attributes-to-calculate-enrichment-for" title="Permalink to this heading"></a></h5>
<p>Our attributes should be defined in a Reference Table <cite>csv</cite> file. You can read more about Reference Tables and their format in the section <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a>.
Once we have a Reference Table, we can perform enrichment analysis for those attributes using the function FeatureSet.enrich_randomization.
If your Reference Tables are set to be the default Reference Tables (as explained in <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a>) you do not need to specify them when calling enrich_randomization. Otherwise, you need to specify your Reference Tables’ path.
The names of the attributes you want to calculate enrichment for can be specified as a list of names (for example, [‘attribute1’, ‘attribute2’]).</p>
</section>
<section id="id30">
<h5>Define the background set<a class="headerlink" href="#id30" title="Permalink to this heading"></a></h5>
<p>In enrichment analysis, we test whether our set of genomic features is enriched/depleted for a certain attribute, in comparison to a more generalized set of genomic features that we determined as ‘background’.
This could be the set of all protein-coding genes, the set of all genomic features that show expression above a certain threshold, or any other set of background genes which you deem appropriate. Importantly, the background set must contain all of the genes in the enrichment set.</p>
<p>Enrichment analysis is usually performed on protein-coding genes. Therefore, by default, <em>RNAlysis</em> uses all of the protein-coding genes that appear in the Attribute Reference Table as a background set.
If you don’t want to use the default setting, there are two methods of defining the background set:</p>
<p>The first method is to specify a biotype (such as ‘protein_coding’, ‘miRNA’ or ‘all’) under the parameter ‘biotype’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">en</span><span class="o">.</span><span class="n">user_defined_enrichment</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span> <span class="n">biotype</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, instead of using all of the protein-coding genes in the Attribute Reference Table as background, we use all of the genomic features in the Attribute Reference Table as background.
When specifying a biotype, the Biotype Reference Table that you specified is used to determine the biotype of each genomic feature.</p>
<p>The second method of defining the background set is to define a specific set of genomic features to be used as background:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_background_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;feature1&#39;</span><span class="p">,</span><span class="s1">&#39;feature2&#39;</span><span class="p">,</span><span class="s1">&#39;feature3&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">en</span><span class="o">.</span><span class="n">user_defined_enrichment</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span> <span class="n">background_genes</span><span class="o">=</span><span class="n">my_background_set</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, our background set consists of feature1, feature2 and feature3.</p>
<p>It is not possible to specify both a biotype and a specific background set.</p>
<p>If some of the features in the background set or the enrichment set do no appear in the Reference Table, they will be ignored when calculating enrichment.</p>
</section>
<section id="id31">
<h5>Choose the statistical test (optional)<a class="headerlink" href="#id31" title="Permalink to this heading"></a></h5>
<p>Significance testing for enrichment analysis can be done using either the Hypergeometric Test, Fisher’s Exact Test, or a randomization test.</p>
<p>The hypergeometric test is defined as: Given <em>M</em> genes in the background set, <em>n</em> genes in the test set, with <em>N</em> genes from the background set belonging to a specific attribute (‘success’) and <em>X</em> genes from the test set belonging to that attribute.
If we were to randomly draw <em>n</em> genes from the background set (without replacement), what is the probability of drawing <em>X</em> or more (in case of enrichment)/<em>X</em> or less (in case of depletion) genes belonging to the given attribute?</p>
<p>The Fisher’s Exact test is similar in principle to the hypergeometric test, but is two-tailed by default, as opposed to the hypergeometric test which examines enrichment and depletion separately.</p>
<p>The randomization test is defined as: Given <em>M</em> genes in the background set, <em>n</em> genes in the test set, with <em>N</em> genes from the background set belonging to a specific attribute and <em>X</em> genes from the test set belonging to that attribute.
We performs the number of randomizations specified by the user (10,000 by default).
In each randomization we randomly draw a set of <em>n</em> genes from the background set (without replacement), and marks the randomization as a ‘success’ if the number of genes in the random set belonging to the attribute is &gt;= <em>X</em> (in case of enrichment) or &lt;= <em>X</em> (in case of depletion).
The p-values are calculated as <em>(number of sucesses + 1)/(number of repetitions + 1)</em>.
This is a positive-bias estimator of the exact p-value, which avoids exactly-zero p-values.
You can read more about the topic in the following publication: <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/21044043">https://www.ncbi.nlm.nih.gov/pubmed/21044043</a></p>
<p>If you don’t specify which statistical test you want to use, the Fisher’s Exact Test will be used by default.</p>
<p>To choose the statistical test you want to use, utilize the <cite>statistical_test</cite> parameter, which accepts either ‘fisher’, ‘hypergeometric’, or ‘randomization’.
If you choose to use a randomization test, you may specify the number of randomization repititions to run using the <cite>randomization_reps</cite> parameter, and set the random seed using the <cite>random_seed</cite> parameter.</p>
</section>
<section id="id32">
<h5>Choose plotting parameters (optional)<a class="headerlink" href="#id32" title="Permalink to this heading"></a></h5>
<p>After performing enrichment analysis, <em>RNAlysis</em> will automatically plot a summary of your enrichment results as a bar plot of log-transformed enrichment scores.
You can determine the orientation of the bar plot (horizontal or vertical) using the <cite>plot_horizontal</cite> parameter:</p>
<blockquote>
<div><figure class="align-center" id="id62">
<a class="reference internal image-reference" href="_images/plot_enrichment_results.png"><img alt="_images/plot_enrichment_results.png" src="_images/plot_enrichment_results.png" style="width: 224.0px; height: 260.8px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id33"><span class="problematic" id="id34">`</span></a>plot_horizontal`=True</span><a class="headerlink" href="#id62" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
<figure class="align-center" id="id63">
<a class="reference internal image-reference" href="_images/plot_enrichment_results_vertical.png"><img alt="_images/plot_enrichment_results_vertical.png" src="_images/plot_enrichment_results_vertical.png" style="width: 327.6px; height: 222.4px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id35"><span class="problematic" id="id36">`</span></a>plot_horizontal`=False</span><a class="headerlink" href="#id63" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
</div></blockquote>
<p>If you want to further customize your plot, you can retreive the matplotlib Figure object of your plot using the <cite>return_fig</cite> parameter.
When it is set as ‘True’, <em>RNAlysis</em> will return the Figure object it generated in addition to the results table.</p>
</section>
<section id="id37">
<h5>Enrichment analysis output<a class="headerlink" href="#id37" title="Permalink to this heading"></a></h5>
<p>Running enrichment analysis will calculate enrichment for each of the specified attributes, and return a pandas DataFrame in the following format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 16%" />
<col style="width: 14%" />
<col style="width: 5%" />
<col style="width: 7%" />
<col style="width: 23%" />
<col style="width: 10%" />
<col style="width: 10%" />
<col style="width: 13%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>samples</p></th>
<th class="head"><p>obs</p></th>
<th class="head"><p>exp</p></th>
<th class="head"><p>log2_fold_enrichment</p></th>
<th class="head"><p>pval</p></th>
<th class="head"><p>padj</p></th>
<th class="head"><p>significant</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>attribute1</p></td>
<td><p>1327</p></td>
<td><p>451</p></td>
<td><p>319.52</p></td>
<td><p>0.49722119558</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>attribute2</p></td>
<td><p>1327</p></td>
<td><p>89</p></td>
<td><p>244.87</p></td>
<td><p>-1.46013879322</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
</tbody>
</table>
<p>‘samples’ is the number of features that were used in the enrichment set. ‘obs’ is the observed number of features positive for the attribute in the enrichment set.
‘exp’ is the expected number of features positive for the attribute in the background set. ‘log2_fold_enrichment’ is log2 of the fold change ‘obs’/’exp’.</p>
</section>
</section>
<section id="performing-enrichment-analysis-for-non-categorical-user-defined-attributes">
<h4>Performing enrichment analysis for non-categorical user-defined attributes<a class="headerlink" href="#performing-enrichment-analysis-for-non-categorical-user-defined-attributes" title="Permalink to this heading"></a></h4>
<p>Instead of peforming enrichment analysis for categorical attributes (“genes which are expressed exclusively in neurons”, “genes enriched in males”, “epigenetic gene-products”, etc), you can test whether your FeatureSet is enriched for a non-categorical attribute (“number of paralogs”, “gene length”, or any other numeric attribute) using the function <cite>FeatureSet.non_categorical_enrichment()</cite>.</p>
<section id="choose-which-user-defined-non-categorical-attributes-to-calculate-enrichment-for">
<h5>Choose which user-defined non-categorical attributes to calculate enrichment for<a class="headerlink" href="#choose-which-user-defined-non-categorical-attributes-to-calculate-enrichment-for" title="Permalink to this heading"></a></h5>
<p>The attributes should be defined in an Attribute Reference Table <cite>csv</cite> file. You can read more about Reference Tables and their format in the section <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a>.
Once we have an Attirubte Reference Table, we can perform enrichment analysis for those non-categorical attributes using the function <cite>FeatureSet.non_categorical_enrichment</cite>.
If your Reference Tables are set to be the default Reference Tables (as explained in <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a>) you do not need to specify them when calling non_categorical_enrichment. Otherwise, you need to specify your Reference Tables’ path.
The names of the attributes you want to calculate enrichment for can be specified as a list of names (for example, [‘attribute1’, ‘attribute2’]).</p>
<p>Note that the variables you use for non-categorical enrichment analysis must be non-categorical, and must be defined for every genomic feature in the background and test sets (meaning, no NaN values).</p>
</section>
<section id="id38">
<h5>Define the background set<a class="headerlink" href="#id38" title="Permalink to this heading"></a></h5>
<p>In enrichment analysis, we test whether our set of genomic features is enriched/depleted for a certain attribute, in comparison to a more generalized set of genomic features that we determined as ‘background’.
This could be the set of all protein-coding genes, the set of all genomic features that show expression above a certain threshold, or any other set of background genes which you deem appropriate. Importantly, the background set must contain all of the genes in the enrichment set.</p>
<p>Enrichment analysis is usually performed on protein-coding genes. Therefore, by default, <em>RNAlysis</em> uses all of the protein-coding genes that appear in the Attribute Reference Table as a background set.
If you don’t want to use the default setting, there are two methods of defining the background set:</p>
<p>The first method is to specify a biotype (such as ‘protein_coding’, ‘miRNA’ or ‘all’) under the parameter ‘biotype’:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">en</span><span class="o">.</span><span class="n">non_categorical_enrichment</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span> <span class="n">biotype</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, instead of using all of the protein-coding genes in the Attribute Reference Table as background, we use all of the genomic features in the Attribute Reference Table as background.
When specifying a biotype, the Biotype Reference Table that you specified is used to determine the biotype of each genomic feature.</p>
<p>The second method of defining the background set is to define a specific set of genomic features to be used as background:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_background_set</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;feature1&#39;</span><span class="p">,</span><span class="s1">&#39;feature2&#39;</span><span class="p">,</span><span class="s1">&#39;feature3&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">result</span> <span class="o">=</span> <span class="n">en</span><span class="o">.</span><span class="n">non_categorical_enrichment</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span><span class="s1">&#39;attribute2&#39;</span><span class="p">],</span> <span class="n">background_genes</span><span class="o">=</span><span class="n">my_background_set</span><span class="p">)</span>
</pre></div>
</div>
<p>In this example, our background set consists of feature1, feature2 and feature3.</p>
<p>It is not possible to specify both a biotype and a specific background set.</p>
<p>If some of the features in the background set or the enrichment set do no appear in the Reference Table, they will be ignored when calculating enrichment.</p>
</section>
<section id="id39">
<h5>Choose the statistical test (optional)<a class="headerlink" href="#id39" title="Permalink to this heading"></a></h5>
<p>When calculating enrichment for a non-categorical attribute, you can use either a parametric or non-parametric statistical test.</p>
<p>If the <cite>parametric_test</cite> parameter is True, <em>RNAlysis</em> will calculate enrichment using the <em>one-sample Student’s T-test</em>, comparing the value of the non-categorical attribute in your enrichment set to the mean value of the non-categorical attribute in the background set.</p>
<p>If the <cite>parametric_test</cite> parameter is False, <em>RNAlysis</em> will calculate enrichment using the non-parametric <em>one-sample Sign test</em>, comparing the value of the non-categorical attribute in your enrichment set to the median value of the non-categorical attribute in the background set.</p>
<p>The parametric T-test assumes that the values of your non-categorical attribute distribute normally. Therefore, if you are not sure your data meets this assumption, it is recommended to use the less-powerful non-parametric test.</p>
<p>If you don’t specify the statistical test, <em>RNAlysis</em> will automatically use a non-parametric test.</p>
</section>
<section id="id40">
<h5>Choose plotting parameters (optional)<a class="headerlink" href="#id40" title="Permalink to this heading"></a></h5>
<p>The results of non-categorical enrichment analysis will be plotted as a histogram, comparing the distribution of the non-categorical attribute within the enrichment set compared to the background set.
If you used a parametric test, the mean of each set will be marked on the histogram. Alternatively, if you used a non-parametric test, the median of each set will be marked on the graph.
The enrichment results of each attribute will be plotted in a different Figure.</p>
<p>To better fit the visualization of the results to your needs, you can specify two parameters:</p>
<p><cite>plot_log_scale</cite> can accept either <cite>True</cite> or <cite>False</cite>, and will determine whether the X-axis of the histogram will be plotting on a logarithmic or linear scale.</p>
<p><a href="#id41"><span class="problematic" id="id42">`</span></a>plot_style’ can accept either ‘overlap’ or ‘interleaved’, and will draw the histogram in one of the following two styles:</p>
<blockquote>
<div><figure class="align-center" id="id64">
<a class="reference internal image-reference" href="_images/hist_overlap.png"><img alt="_images/hist_overlap.png" src="_images/hist_overlap.png" style="width: 480.0px; height: 238.4px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id43"><span class="problematic" id="id44">`</span></a>plot_style`=’overlap’</span><a class="headerlink" href="#id64" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
<figure class="align-center" id="id65">
<a class="reference internal image-reference" href="_images/hist_interleaved.png"><img alt="_images/hist_interleaved.png" src="_images/hist_interleaved.png" style="width: 480.0px; height: 238.4px;" /></a>
<figcaption>
<p><span class="caption-text"><a href="#id45"><span class="problematic" id="id46">`</span></a>plot_style`=’interleaved’</span><a class="headerlink" href="#id65" title="Permalink to this image"></a></p>
<div class="legend">
</div>
</figcaption>
</figure>
</div></blockquote>
<p>If you don’t specify plotting parameters, <em>RNAlysis</em> will plot the histogram on a logarithmic scale using the ‘overlap’ style by default.</p>
<p>If you want to further customize your plot, you can retreive the matplotlib Figure object of your plot using the <cite>return_fig</cite> parameter.
When it is set as ‘True’, <em>RNAlysis</em> will return the Figure object it generated in addition to the results table.</p>
</section>
<section id="non-categorical-enrichment-analysis-output">
<h5>Non-Categorical Enrichment analysis output<a class="headerlink" href="#non-categorical-enrichment-analysis-output" title="Permalink to this heading"></a></h5>
<p>Running non-categorical enrichment analysis will calculate enrichment for each of the specified attributes, and return a pandas DataFrame in the following format:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 21%" />
<col style="width: 18%" />
<col style="width: 9%" />
<col style="width: 10%" />
<col style="width: 13%" />
<col style="width: 13%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"></th>
<th class="head"><p>samples</p></th>
<th class="head"><p>obs</p></th>
<th class="head"><p>exp</p></th>
<th class="head"><p>pval</p></th>
<th class="head"><p>padj</p></th>
<th class="head"><p>significant</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>attribute1</p></td>
<td><p>1327</p></td>
<td><p>451</p></td>
<td><p>319.52</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
<tr class="row-odd"><td><p>attribute2</p></td>
<td><p>1327</p></td>
<td><p>89</p></td>
<td><p>244.87</p></td>
<td><p>0.0000999</p></td>
<td><p>0.0000999</p></td>
<td><p>True</p></td>
</tr>
</tbody>
</table>
<p>‘samples’ is the number of features that were used in the enrichment set. ‘obs’ is the observed mean/median (depending if the statistical test was parametric or not) value of the non-categorical attribute in the enrichment set.
‘exp’ is the expected mean/median value of the non-categorical attribute in the background set.</p>
</section>
</section>
<section id="performing-set-operations-and-visualisation-on-multiple-featureset-objects">
<h4>Performing set operations and visualisation on multiple FeatureSet objects<a class="headerlink" href="#performing-set-operations-and-visualisation-on-multiple-featureset-objects" title="Permalink to this heading"></a></h4>
<p>Similarly to Filter objects, it is possible to use set operations such as union, intersection, difference and symmetric difference to combine the feature sets of multiple FeatureSet objects. Those set operations can be applied to both FeatureSet objects and python sets. The objects don’t have to be of the same subtype - you can, for example, look at the union of an FeatureSet object and a python set:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">en</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;WBGene00003002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00004201&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00300139&#39;</span><span class="p">})</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">s</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000002&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene00000003&#39;</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">union_result</span> <span class="o">=</span> <span class="n">en</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>
</div>
<p>When performing set operations, the return type will always be a python set. This means you can use the output of the set operation as an input for yet another set operation, or as input to a new FeatureSet object.</p>
<p>In addition, the enrichment module includes functions for visualisation of sets and overlap, such as enrichment.venn_diagram() and enrichment.upset_plot().
Both functions receive a similar input: a dictionary whose keys are the names of the sets, and values are either FeatureSet objects, Filter objects, sets of genomic feature names, or the name of an attribute from the <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> (you can read more about attributes in <a class="reference internal" href="user_guide_gui.html#reference-table-ref"><span class="std std-ref">Set and load a Reference Table</span></a>).
Venn diagrams are limited to 2-3 sets:</p>
<blockquote>
<div><figure class="align-center" id="id66">
<a class="reference internal image-reference" href="_images/venn.png"><img alt="_images/venn.png" src="_images/venn.png" style="width: 550.1999999999999px; height: 416.5px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of venn_diagram()</span><a class="headerlink" href="#id66" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
<p>While UpSet plots can include any number of sets:</p>
<blockquote>
<div><figure class="align-center" id="id67">
<a class="reference internal image-reference" href="_images/upsetplot.png"><img alt="_images/upsetplot.png" src="_images/upsetplot.png" style="width: 552.3px; height: 340.2px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of upset_plot()</span><a class="headerlink" href="#id67" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
</section>
<section id="saving-indices-from-featureset-to-a-txt-file">
<h4>Saving indices from FeatureSet to a .txt file<a class="headerlink" href="#saving-indices-from-featureset-to-a-txt-file" title="Permalink to this heading"></a></h4>
<p>It is possible to save the feature indices from an FeatureSet object to a .txt file, for use in online enrichment tools or simply to share the list of genomic features. This is done with the ‘save_txt’ function:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">en</span><span class="o">.</span><span class="n">save_txt</span><span class="p">(</span><span class="s1">&#39;D:\path</span><span class="se">\f</span><span class="s1">ilename&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>The feature indices will be saved to the text file in the specified path, separated by newline (’n’).</p>
</section>
</section>
<section id="working-with-rankedset-objects">
<h3>Working with RankedSet objects<a class="headerlink" href="#working-with-rankedset-objects" title="Permalink to this heading"></a></h3>
<p>The <a class="reference internal" href="glossary.html#term-RankedSet"><span class="xref std std-term">RankedSet</span></a> class represents a set of genomic features which are ranked by an inherent order. For example, genes could be ranked based on their mean expression level, their fold change between a pair of conditions, etc.
<a class="reference internal" href="glossary.html#term-RankedSet"><span class="xref std std-term">RankedSet</span></a> objects behave similarly to <a class="reference internal" href="glossary.html#term-FeatureSet"><span class="xref std std-term">FeatureSet</span></a> objects, but in addition they support <em>single-set enrichment analysis</em> - enrichment analysis without a background set.</p>
<p>The statistical test used to test significance in <em>single-set enrichment analysis</em> is the <em>generalized minimum hypergeometric test (XL-mHG)</em>, which measures the enrichment of the genomic features at the top and bottom of your ranked gene set compared to the rest of the gene set.
You can read more about the <em>minimum hypergeometric test</em> and its generalized version the <em>XL-mHG test</em> in the following publications:
<a class="reference external" href="https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.0030039">https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.0030039</a>
<a class="reference external" href="https://arxiv.org/abs/1507.07905">https://arxiv.org/abs/1507.07905</a></p>
<section id="initialize-an-rankedset-object">
<h4>Initialize an RankedSet object<a class="headerlink" href="#initialize-an-rankedset-object" title="Permalink to this heading"></a></h4>
<p>We will start by importing the enrichment module:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">enrichment</span>
</pre></div>
</div>
<p><a class="reference internal" href="glossary.html#term-RankedSet"><span class="xref std std-term">RankedSet</span></a> objects can be initialized by one of two methods.
The first method is to specify an existing Filter object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_filter_obj</span> <span class="o">=</span> <span class="n">filtering</span><span class="o">.</span><span class="n">CountFilter</span><span class="p">(</span><span class="s1">&#39;tests/test_files/counted.csv&#39;</span><span class="p">)</span> <span class="c1"># create a Filter object</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_ranked_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">RankedSet</span><span class="p">(</span><span class="n">my_filter_obj</span><span class="p">,</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>When using this method, the ranking of the genomic features will be determined by their current order within the <a class="reference internal" href="glossary.html#term-Filter"><span class="xref std std-term">Filter</span></a> object. You can use the ‘sort’ method of your Filter object to modify the current order of genomic features within the Filter object.</p>
<p>The second method is to directly specify a list, tuple, or numpy array of genomic feature names:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">my_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;WBGene00000001&#39;</span><span class="p">,</span><span class="s1">&#39;WBGene0245200&#39;</span><span class="p">,</span><span class="s1">&#39; WBGene00402029&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_ranked_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">RankedSet</span><span class="p">(</span><span class="n">my_list</span><span class="p">,</span> <span class="s1">&#39;a name for my set&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>RankedSet objects have three attributes: ranked_genes, a numpy array containing the ranked genomic features; gene_set, a python set containing the genomic features; and set_name, a string that describes the feature set (optional).</p>
</section>
<section id="performing-single-set-go-enrichment-analysis-without-a-background-set">
<h4>Performing single-set GO Enrichment analysis without a background set<a class="headerlink" href="#performing-single-set-go-enrichment-analysis-without-a-background-set" title="Permalink to this heading"></a></h4>
<p>Single-set GO Enrichment works much the same as normal Go Enrichment analysis, with two key differences:</p>
<p>First, when performing single-set GO Enrichment you do not define a background set to comapre your enrichment gene set against. Instead, you supply a <a class="reference internal" href="glossary.html#term-RankedSet"><span class="xref std std-term">RankedSet</span></a> that defines a meaningful ranking for your gene set.
Second, you cannot specify which statistical test to use, since the <em>XL-mHG</em> test has to be used.</p>
<p>An example for running single-set GO Enrichment would look like so:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ranked_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">RankedSet</span><span class="p">([</span><span class="s1">&#39;WBGene00000019&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000106&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000041&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000105&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go_en_result</span> <span class="o">=</span> <span class="n">ranked_set</span><span class="o">.</span><span class="n">single_set_go_enrichment</span><span class="p">(</span><span class="n">gene_id_type</span><span class="o">=</span><span class="s1">&#39;WormBase&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="performing-single-set-kegg-enrichment-analysis-without-a-background-set">
<h4>Performing single-set KEGG Enrichment analysis without a background set<a class="headerlink" href="#performing-single-set-kegg-enrichment-analysis-without-a-background-set" title="Permalink to this heading"></a></h4>
<p>Single-set KEGG Enrichment works much the same as normal KEGG Enrichment analysis, with two key differences:</p>
<p>First, when performing single-set KEGG Enrichment you do not define a background set to comapre your enrichment gene set against. Instead, you supply a <a class="reference internal" href="glossary.html#term-RankedSet"><span class="xref std std-term">RankedSet</span></a> that defines a meaningful ranking for your gene set.
Second, you cannot specify which statistical test to use, since the <em>XL-mHG</em> test has to be used.</p>
<p>An example for running single-set KEGG Enrichment would look like so:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ranked_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">RankedSet</span><span class="p">([</span><span class="s1">&#39;WBGene00000019&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000106&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000041&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000105&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">go_en_result</span> <span class="o">=</span> <span class="n">ranked_set</span><span class="o">.</span><span class="n">single_set_kegg_enrichment</span><span class="p">(</span><span class="n">gene_id_type</span><span class="o">=</span><span class="s1">&#39;WormBase&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="performing-single-set-enrichment-analysis-for-user-defined-attributes-without-a-background-set">
<h4>Performing single-set enrichment analysis for user-defined attributes without a background set<a class="headerlink" href="#performing-single-set-enrichment-analysis-for-user-defined-attributes-without-a-background-set" title="Permalink to this heading"></a></h4>
<p>Single-set enrichment analysis for user-defined attributes works much the same as normal enrichment analysis, with two key differences:</p>
<p>First, when performing single-set enrichment you do not define a background set to comapre your enrichment gene set against. Instead, you supply a <a class="reference internal" href="glossary.html#term-RankedSet"><span class="xref std std-term">RankedSet</span></a> that defines a meaningful ranking for your gene set.
Second, you cannot specify which statistical test to use, since the <em>XL-mHG</em> test has to be used.</p>
<p>An example for running single-set enrichment analysis would look like so:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ranked_set</span> <span class="o">=</span> <span class="n">enrichment</span><span class="o">.</span><span class="n">RankedSet</span><span class="p">([</span><span class="s1">&#39;WBGene00000019&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000106&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000041&#39;</span><span class="p">,</span> <span class="s1">&#39;WBGene00000105&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">en_result</span> <span class="o">=</span> <span class="n">ranked_set</span><span class="o">.</span><span class="n">single_set_enrichment</span><span class="p">([</span><span class="s1">&#39;attribute1&#39;</span><span class="p">,</span> <span class="s1">&#39;attribute3&#39;</span><span class="p">],</span> <span class="n">attr_ref_path</span><span class="o">=</span><span class="s1">&#39;tests/test_files/attr_ref_table_for_examples.csv&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
<section id="visualizing-sets-intersections-and-enrichment">
<h3>Visualizing sets, intersections, and enrichment<a class="headerlink" href="#visualizing-sets-intersections-and-enrichment" title="Permalink to this heading"></a></h3>
<section id="plotting-results-of-enrichment-analysis">
<h4>Plotting results of enrichment analysis<a class="headerlink" href="#plotting-results-of-enrichment-analysis" title="Permalink to this heading"></a></h4>
<p>If you want to plot existing enrichment results using <em>RNAlysis</em>, you can use the <cite>enrichment.plot_enrichment_results()</cite> function. It employs a similar API to the enrichment functions in the enrichment modules, but accepts pre-calculated DataFrames.</p>
</section>
<section id="plotting-venn-diagrams-and-upset-plots">
<h4>Plotting Venn Diagrams and UpSet Plots<a class="headerlink" href="#plotting-venn-diagrams-and-upset-plots" title="Permalink to this heading"></a></h4>
<p>To visualize set intersection using a Venn Diagram or UpSet plot, you first need to define the sets you want to visualize.
Those sets can be represented using a python dictionary, where each key represents the name of the set, and the corresponding value represents the set’s content (the genomic features that are part of the set).</p>
<p>There are three ways of specifying a set’s content:</p>
<ol class="arabic simple">
<li><p>A python set containing the names of the genomic features</p></li>
<li><p>A <cite>FeatureSet</cite> or <cite>RankedSet</cite> object containing the names of the genomic features</p></li>
<li><p>A column name from an Attribute Reference Table</p></li>
</ol>
<p>For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">enrichment</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">sets_to_visualize</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;First set&#39;</span><span class="p">:{</span><span class="s1">&#39;gene1&#39;</span><span class="p">,</span> <span class="s1">&#39;gene2&#39;</span><span class="p">,</span> <span class="s1">&#39;gene3&#39;</span><span class="p">},</span> <span class="s1">&#39;Second set&#39;</span><span class="p">:</span><span class="n">enrichment</span><span class="o">.</span><span class="n">FeatureSet</span><span class="p">({</span><span class="s1">&#39;gene2&#39;</span><span class="p">,</span><span class="s1">&#39;gene3&#39;</span><span class="p">,</span><span class="s1">&#39;gene4&#39;</span><span class="p">}),</span> <span class="s1">&#39;Third set&#39;</span><span class="p">:</span><span class="s1">&#39;attribute1&#39;</span><span class="p">}</span>
</pre></div>
</div>
<p>After defining the dictionary of sets, we can use it to plot set intersection via the <cite>enrichment.venn_diagram()</cite> or <cite>enrichment.upset_plot()</cite> functions.
Venn diagrams in <em>RNAlysis</em> are plotted with relatively accurate proportions and overlaps, and therefore only support up to 3 sets.</p>
<blockquote>
<div><figure class="align-center" id="id68">
<a class="reference internal image-reference" href="_images/venn.png"><img alt="_images/venn.png" src="_images/venn.png" style="width: 550.1999999999999px; height: 416.5px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of venn_diagram()</span><a class="headerlink" href="#id68" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
<p>UpSet plots support the visualization of much larger groups of sets. You can read more about UpSet plots here: <a class="reference external" href="https://upset.app/">https://upset.app/</a></p>
<blockquote>
<div><figure class="align-center" id="id69">
<a class="reference internal image-reference" href="_images/upsetplot.png"><img alt="_images/upsetplot.png" src="_images/upsetplot.png" style="width: 552.3px; height: 340.2px;" /></a>
<figcaption>
<p><span class="caption-text">Example plot of upset_plot()</span><a class="headerlink" href="#id69" title="Permalink to this image"></a></p>
</figcaption>
</figure>
</div></blockquote>
</section>
</section>
</section>
<section id="rnalysis-general-module">
<h2><em>RNAlysis</em> general module<a class="headerlink" href="#rnalysis-general-module" title="Permalink to this heading"></a></h2>
<p>RNAlysis’s general module (rnalysis.general) contains general functions that can be useful during analysis of RNA sequencing data, including regular expression parsers and setting the Reference Table path.</p>
<section id="set-and-load-a-reference-table">
<span id="reference-table-ref"></span><h3>Set and load a Reference Table<a class="headerlink" href="#set-and-load-a-reference-table" title="Permalink to this heading"></a></h3>
<section id="what-is-an-attribute-reference-table">
<h4>What is an Attribute Reference Table?<a class="headerlink" href="#what-is-an-attribute-reference-table" title="Permalink to this heading"></a></h4>
<p>You can perform enrichment analysis or filtering operations based on user-defined attributes (such as ‘genes expressed in intestine’, ‘epigenetic genes’, ‘genes that have paralogs’).
User-defined attributes should be defined in an <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> <cite>csv</cite> file. The format of the <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> is one row for each gene/genomic feature, and one column for each attribute. Features that are negative for the attribute (for example, genes that have no paralogs under the attribute ‘genes that have paralogs’) should have the value NaN specified for the attribute, and features that are positive for the attribute (for example, genes that have paralogs under the attribute ‘genes that have paralogs’) should have any value other than NaN. The value could be either a boolean value (in our example, ‘True’ or ‘1’ for genes that have paralogs), a number (in our example, the number of paralogs the gene has or the genomic distance to the nearest paralog), or any other value which is not NaN. See example for an <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> below:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 29%" />
<col style="width: 25%" />
<col style="width: 23%" />
<col style="width: 23%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>feature_indices</p></th>
<th class="head"><p>attribute1</p></th>
<th class="head"><p>attribute2</p></th>
<th class="head"><p>attribute3</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>WBGene0000001</p></td>
<td><p>1</p></td>
<td><p>NaN</p></td>
<td><p>13.7</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene0000002</p></td>
<td><p>NaN</p></td>
<td><p>1</p></td>
<td><p>241</p></td>
</tr>
<tr class="row-even"><td><p>WBGene0000003</p></td>
<td><p>NaN</p></td>
<td><p>1</p></td>
<td><p>3.6</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene0000004</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>NaN</p></td>
</tr>
<tr class="row-even"><td><p>WBGene0000005</p></td>
<td><p>1</p></td>
<td><p>NaN</p></td>
<td><p>21.5</p></td>
</tr>
</tbody>
</table>
</section>
<section id="what-is-a-biotype-reference-table">
<h4>What is a Biotype Reference Table?<a class="headerlink" href="#what-is-a-biotype-reference-table" title="Permalink to this heading"></a></h4>
<p>You can perform filtering operations or generate background-sets for enrichment analysis based on user-annotated biotypes (such as ‘protein_coding’, ‘pseudogene’, ‘piRNA’, etc).
User-annotated biotypes should be defined in a <a class="reference internal" href="glossary.html#term-Biotype-Reference-Table"><span class="xref std std-term">Biotype Reference Table</span></a> <cite>csv</cite> file. The format of the <a class="reference internal" href="glossary.html#term-Biotype-Reference-Table"><span class="xref std std-term">Biotype Reference Table</span></a> is one row for each gene/genomic feature, and a column titled ‘biotype’ (case insensitive). See example for a Biotype Reference Table below:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>feature_indices</p></th>
<th class="head"><p>biotype</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>WBGene0000001</p></td>
<td><p>protein_coding</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene0000002</p></td>
<td><p>protein_coding</p></td>
</tr>
<tr class="row-even"><td><p>WBGene0000003</p></td>
<td><p>pseudogene</p></td>
</tr>
<tr class="row-odd"><td><p>WBGene0000004</p></td>
<td><p>piRNA</p></td>
</tr>
<tr class="row-even"><td><p>WBGene0000005</p></td>
<td><p>lincRNA</p></td>
</tr>
</tbody>
</table>
</section>
<section id="set-a-reference-table-as-default">
<h4>Set a Reference Table as default<a class="headerlink" href="#set-a-reference-table-as-default" title="Permalink to this heading"></a></h4>
<p>Once we have an Attribute and/or Biotype Reference Table, we can set it to be the default reference table for all future uses of <em>RNAlysis</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;the_new_attribute_reference_table_path&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">general</span><span class="o">.</span><span class="n">set_attr_ref_table_path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="go">Attribute Reference Table path set as: the_new_attribute_reference_table_path</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">path</span><span class="o">=</span><span class="s2">&quot;the_new_biotype_reference_table_path&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">general</span><span class="o">.</span><span class="n">set_biotype_ref_table_path</span><span class="p">(</span><span class="n">path</span><span class="p">)</span>
<span class="go">Attribute Reference Table path set as: the_new_biotype_reference_table_path</span>
</pre></div>
</div>
<p>This will create a file called ‘settings.yaml’, which will store the full paths of your reference tables.
Whenever <em>RNAlysis</em> needs to use an Attribute/Biotype Reference Table and no other path is specified, <em>RNAlysis</em> will automatically use the path saved in the settings file.
The saved path can be changed any time using the general.set_attr_ref_table_path() and general.set_biotype_ref_table_path() functions.</p>
</section>
<section id="load-the-default-attribute-reference-table-path">
<h4>Load the default Attribute Reference Table path<a class="headerlink" href="#load-the-default-attribute-reference-table-path" title="Permalink to this heading"></a></h4>
<p>You can read the saved path from the settings file using the general.read_attr_ref_table_path() and general.read_biotype_ref_table_path() functions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">attr_table_path</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">read_attr_ref_table_path</span><span class="p">()</span>
<span class="go">Attribute Reference Table used: the_attribute_reference_table_path_that_was_saved_in_the_settings_file</span>

<span class="gp">&gt;&gt;&gt; </span><span class="n">biotype_table_path</span> <span class="o">=</span> <span class="n">general</span><span class="o">.</span><span class="n">read_biotype_ref_table_path</span><span class="p">()</span>
<span class="go">Biotype Reference Table used: the_biotype_reference_table_path_that_was_saved_in_the_settings_file</span>
</pre></div>
</div>
<p>If an <a class="reference internal" href="glossary.html#term-Attribute-Reference-Table"><span class="xref std std-term">Attribute Reference Table</span></a> path was not previously defined, you will be requested to define it when you run this function.</p>
</section>
</section>
<section id="parse-c-elegans-gene-names-wbgene-indices-and-sequence-names-using-regular-expressions">
<h3>Parse <em>C. elegans</em> gene names, WBGene indices and sequence names using regular expressions<a class="headerlink" href="#parse-c-elegans-gene-names-wbgene-indices-and-sequence-names-using-regular-expressions" title="Permalink to this heading"></a></h3>
<p>The general module includes functions which can parse <em>C. elegans</em> gene names (like <em>daf-2</em> or <em>lin-15B</em>), WBGene indices (like WBGene00023495) and sequence names (like Y55D5A.5 or T23G5.6).
For example, we could extract all WBGene indices from the following string:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="o">*</span><span class="n">RNAlysis</span><span class="o">*</span> <span class="kn">import</span> <span class="nn">general</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">my_string</span><span class="o">=</span><span class="s1">&#39;&#39;&#39;WBGene00000001 and WBGene00000002WBGene00000003</span>

<span class="go">        WBGene00000004g</span>
<span class="go">        &#39;&#39;&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="s1">general.parse_wbgene_string(my_string)</span>
<span class="go">{&#39;WBGene00000001&#39;,&#39;WBGene000000002&#39;,&#39;WBGene00000003&#39;,&#39;WBGene00000004&#39;}</span>
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="user_guide_gui.html" class="btn btn-neutral float-left" title="User guide - graphical interface" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="glossary.html" class="btn btn-neutral float-right" title="Glossary - programmatic interface" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Guy Teichman.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>